<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • rbioapi</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg"><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/rbioapi.html">Get Started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles-supported-services" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles &amp; Supported Services</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles-supported-services"><li><a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://moosa-r.com/" aria-label="Moosa Rezwani Portfolio Website"><span class="fa fa-solid fa-id-card"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="helper-functions">Helper functions<a class="anchor" aria-label="anchor" href="#helper-functions"></a></h2>

      <div class="section-desc"><p>Functions that assist you with the overall experience of rbioapi</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_options.html">rba_options()</a></code>

        </dt>
        <dd>Set rbioapi Global Options</dd>
      </dl><dl><dt>

          <code><a href="rba_connection_test.html">rba_connection_test()</a></code>

        </dt>
        <dd>Test if the Supported Services Are Responding</dd>
      </dl><dl><dt>

          <code><a href="rba_pages.html">rba_pages()</a></code>

        </dt>
        <dd>Get Multiple Pages of a Paginated Resource</dd>
      </dl></div><div class="section level2">
      <h2 id="enrichr-rba-enrichr-">Enrichr (rba_enrichr_*)<a class="anchor" aria-label="anchor" href="#enrichr-rba-enrichr-"></a></h2>

      <div class="section-desc"><p>Functions that connect you to <a href="https://maayanlab.cloud/Enrichr/" class="external-link">Enrichr</a> (Gene list enrichment analysis tool) API resources.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_enrichr.html">rba_enrichr()</a></code>

        </dt>
        <dd>A One-step Wrapper for Gene-list Enrichment Using Enrichr</dd>
      </dl><dl><dt>

          <code><a href="rba_enrichr_add_background.html">rba_enrichr_add_background()</a></code>

        </dt>
        <dd>Upload Background Gene-List to Enrichr</dd>
      </dl><dl><dt>

          <code><a href="rba_enrichr_add_list.html">rba_enrichr_add_list()</a></code>

        </dt>
        <dd>Upload Your Gene-List to Enrichr</dd>
      </dl><dl><dt>

          <code><a href="rba_enrichr_enrich.html">rba_enrichr_enrich()</a></code>

        </dt>
        <dd>Get Enrichr Enrichment Results</dd>
      </dl><dl><dt>

          <code><a href="rba_enrichr_gene_map.html">rba_enrichr_gene_map()</a></code>

        </dt>
        <dd>Find Enrichr Terms That Contain a Given Gene</dd>
      </dl><dl><dt>

          <code><a href="rba_enrichr_libs.html">rba_enrichr_libs()</a></code>

        </dt>
        <dd>Retrieve a List of available libraries from Enrichr</dd>
      </dl><dl><dt>

          <code><a href="rba_enrichr_view_list.html">rba_enrichr_view_list()</a></code>

        </dt>
        <dd>View an Uploaded Gene List</dd>
      </dl></div><div class="section level2">
      <h2 id="jaspar-rba-jaspar-">JASPAR (rba_jaspar_*)<a class="anchor" aria-label="anchor" href="#jaspar-rba-jaspar-"></a></h2>

      <div class="section-desc"><p>Functions that connect you to <a href="https://jaspar.elixir.no/" class="external-link">JASPAR</a> (Database of transcription factor binding profiles) API resources.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_jaspar_collections.html">rba_jaspar_collections()</a></code>

        </dt>
        <dd>List collections available in JASPAR</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_collections_matrices.html">rba_jaspar_collections_matrices()</a></code>

        </dt>
        <dd>List matrices available in a JASPAR collection</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_matrix.html">rba_jaspar_matrix()</a></code>

        </dt>
        <dd>Get a Position Frequency Matrices (PFM) with annotations</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_matrix_search.html">rba_jaspar_matrix_search()</a></code>

        </dt>
        <dd>Search matrix profiles available in JASPAR</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_matrix_versions.html">rba_jaspar_matrix_versions()</a></code>

        </dt>
        <dd>List matrix profile versions associated with a base ID</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_releases.html">rba_jaspar_releases()</a></code>

        </dt>
        <dd>Get information about JASPAR database releases</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_sites.html">rba_jaspar_sites()</a></code>

        </dt>
        <dd>Get binding sites of a matrix profile</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_species.html">rba_jaspar_species()</a></code>

        </dt>
        <dd>List available species in JASPAR</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_species_matrices.html">rba_jaspar_species_matrices()</a></code>

        </dt>
        <dd>List matrices available in JASPAR of a species</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_taxons.html">rba_jaspar_taxons()</a></code>

        </dt>
        <dd>List available taxonomic groups in JASPAR</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_taxons_matrices.html">rba_jaspar_taxons_matrices()</a></code>

        </dt>
        <dd>List matrices available in JASPAR of a taxonomic group</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_tffm.html">rba_jaspar_tffm()</a></code>

        </dt>
        <dd>Get a TF flexible models (TFFMs) information</dd>
      </dl><dl><dt>

          <code><a href="rba_jaspar_tffm_search.html">rba_jaspar_tffm_search()</a></code>

        </dt>
        <dd>Search TF flexible models (TFFMs) available in JASPAR</dd>
      </dl></div><div class="section level2">
      <h2 id="mieaa-rba-mieaa-">miEAA (rba_mieaa_*)<a class="anchor" aria-label="anchor" href="#mieaa-rba-mieaa-"></a></h2>

      <div class="section-desc"><p>Functions that connect you to <a href="https://ccb-compute2.cs.uni-saarland.de/mieaa2/" class="external-link">miEAA</a> (miRNA Enrichment Analysis and Annotation Tool) API resources.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_mieaa_cats.html">rba_mieaa_cats()</a></code>

        </dt>
        <dd>Get Supported Enrichment Categories for a Species and miRNA Type</dd>
      </dl><dl><dt>

          <code><a href="rba_mieaa_convert_type.html">rba_mieaa_convert_type()</a></code>

        </dt>
        <dd>Convert Between Mature and precursor miRNA Accession</dd>
      </dl><dl><dt>

          <code><a href="rba_mieaa_convert_version.html">rba_mieaa_convert_version()</a></code>

        </dt>
        <dd>Convert miRNA accession Between Different miRBase Versions</dd>
      </dl><dl><dt>

          <code><a href="rba_mieaa_enrich.html">rba_mieaa_enrich()</a></code>

        </dt>
        <dd>A One-step Wrapper for miRNA Enrichment Using miEAA</dd>
      </dl><dl><dt>

          <code><a href="rba_mieaa_enrich_results.html">rba_mieaa_enrich_results()</a></code>

        </dt>
        <dd>Retrieve Results of a finished Enrichment Analysis from miEAA</dd>
      </dl><dl><dt>

          <code><a href="rba_mieaa_enrich_status.html">rba_mieaa_enrich_status()</a></code>

        </dt>
        <dd>Check Status of a Submitted Enrichment Analysis in miEAA</dd>
      </dl><dl><dt>

          <code><a href="rba_mieaa_enrich_submit.html">rba_mieaa_enrich_submit()</a></code>

        </dt>
        <dd>Submit miEAA miRNA Enrichment Analysis Request</dd>
      </dl></div><div class="section level2">
      <h2 id="panther-rba-panther-">PANTHER (rba_panther_*)<a class="anchor" aria-label="anchor" href="#panther-rba-panther-"></a></h2>

      <div class="section-desc"><p>Functions that connect you to <a href="https://www.pantherdb.org/" class="external-link">PANTHER</a> (Protein Analysis THrough Evolutionary Relationships) API resources.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_panther_enrich.html">rba_panther_enrich()</a></code>

        </dt>
        <dd>PANTHER Over-Representation or Enrichment Analysis</dd>
      </dl><dl><dt>

          <code><a href="rba_panther_family.html">rba_panther_family()</a></code>

        </dt>
        <dd>Get PANTHER Families and Sub-Families</dd>
      </dl><dl><dt>

          <code><a href="rba_panther_homolog.html">rba_panther_homolog()</a></code>

        </dt>
        <dd>Search PANTHER for Homologs of Gene(s)</dd>
      </dl><dl><dt>

          <code><a href="rba_panther_info.html">rba_panther_info()</a></code>

        </dt>
        <dd>Get PANTHER database Information</dd>
      </dl><dl><dt>

          <code><a href="rba_panther_mapping.html">rba_panther_mapping()</a></code>

        </dt>
        <dd>Map A Gene-set to PANTHER Database</dd>
      </dl><dl><dt>

          <code><a href="rba_panther_ortholog.html">rba_panther_ortholog()</a></code>

        </dt>
        <dd>Search PANTHER for Orthologs of Gene(s)</dd>
      </dl><dl><dt>

          <code><a href="rba_panther_tree_grafter.html">rba_panther_tree_grafter()</a></code>

        </dt>
        <dd>PANTHER Tree Grafter Use this function to retrieve a PANTHER family's tree topology information with a node corresponding to your sequence grafted in the best location in that tree.</dd>
      </dl></div><div class="section level2">
      <h2 id="reactome-analysis-services-rba-reactome-">Reactome Analysis Services (rba_reactome_*)<a class="anchor" aria-label="anchor" href="#reactome-analysis-services-rba-reactome-"></a></h2>

      <div class="section-desc"><p>Functions that connect you to <a href="https://reactome.org/" class="external-link">Reactome</a> pathway database’s Analysis services API resources.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_analysis.html">rba_reactome_analysis()</a></code>

        </dt>
        <dd>Reactome Over-Representation or Expression Analysis</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_analysis_download.html">rba_reactome_analysis_download()</a></code>

        </dt>
        <dd>Download Different Reactome Analysis Results</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_analysis_import.html">rba_reactome_analysis_import()</a></code>

        </dt>
        <dd>Import Saved Analysis JSON to Reactome</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_analysis_mapping.html">rba_reactome_analysis_mapping()</a></code>

        </dt>
        <dd>Maps Molecule Identifiers</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_analysis_pdf.html">rba_reactome_analysis_pdf()</a></code>

        </dt>
        <dd>Generate PDF file with Reactome Analysis Results</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_analysis_species.html">rba_reactome_analysis_species()</a></code>

        </dt>
        <dd>Compare Human Pathways with with Other Species</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_analysis_token.html">rba_reactome_analysis_token()</a></code>

        </dt>
        <dd>Return the Results Associated with a Token</dd>
      </dl></div><div class="section level2">
      <h2 id="reactome-contents-services-rba-reactome-">Reactome Contents Services (rba_reactome_*)<a class="anchor" aria-label="anchor" href="#reactome-contents-services-rba-reactome-"></a></h2>

      <div class="section-desc"><p>Functions that connect you to <a href="https://reactome.org/" class="external-link">Reactome</a> pathway database’s Contents Services API resources.</p></div>


    </div><div class="section level2">

      <h3 id="database-info-queries">Database Info Queries<a class="anchor" aria-label="anchor" href="#database-info-queries"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_version.html">rba_reactome_version()</a></code>

        </dt>
        <dd>The version number of current database</dd>
      </dl></div><div class="section level2">

      <h3 id="disease-related-queries">Disease Related Queries<a class="anchor" aria-label="anchor" href="#disease-related-queries"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_diseases.html">rba_reactome_diseases()</a></code>

        </dt>
        <dd>Reactome Diseases</dd>
      </dl></div><div class="section level2">

      <h3 id="physical-entity-queries">Physical Entity Queries<a class="anchor" aria-label="anchor" href="#physical-entity-queries"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_complex_list.html">rba_reactome_complex_list()</a></code>

        </dt>
        <dd>Get Complexes That Include a Molecule</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_complex_subunits.html">rba_reactome_complex_subunits()</a></code>

        </dt>
        <dd>Get a Complex's Subunits</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_participant_of.html">rba_reactome_participant_of()</a></code>

        </dt>
        <dd>Get Larger Reactome Structures Which Include an Entity</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_entity_other_forms.html">rba_reactome_entity_other_forms()</a></code>

        </dt>
        <dd>Get Other forms of a Reactome Entity</dd>
      </dl></div><div class="section level2">

      <h3 id="queries-related-to-events">Queries Related to Events<a class="anchor" aria-label="anchor" href="#queries-related-to-events"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_event_ancestors.html">rba_reactome_event_ancestors()</a></code>

        </dt>
        <dd>Get Reactome Events Ancestors</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_event_hierarchy.html">rba_reactome_event_hierarchy()</a></code>

        </dt>
        <dd>Get Full Event Hierarchy of a Species</dd>
      </dl></div><div class="section level2">

      <h3 id="format-exporter">Format Exporter<a class="anchor" aria-label="anchor" href="#format-exporter"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_exporter_diagram.html">rba_reactome_exporter_diagram()</a></code>

        </dt>
        <dd>Get a Reactome Event Diagram</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_exporter_event.html">rba_reactome_exporter_event()</a></code>

        </dt>
        <dd>Exports A Reactome Event to SBGN or SBML</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_exporter_overview.html">rba_reactome_exporter_overview()</a></code>

        </dt>
        <dd>Get a Reactome Pathway Overview</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_exporter_reaction.html">rba_reactome_exporter_reaction()</a></code>

        </dt>
        <dd>Get a Reactome Reaction Event</dd>
      </dl></div><div class="section level2">

      <h3 id="molecule-interactors">Molecule Interactors<a class="anchor" aria-label="anchor" href="#molecule-interactors"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_interactors_psicquic.html">rba_reactome_interactors_psicquic()</a></code>

        </dt>
        <dd>The interface From Reactome to PSICQUIC</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_interactors_static.html">rba_reactome_interactors_static()</a></code>

        </dt>
        <dd>Get Static(IntAct) Interaction Information of a Protein</dd>
      </dl></div><div class="section level2">

      <h3 id="mapping-related-queries">Mapping Related Queries<a class="anchor" aria-label="anchor" href="#mapping-related-queries"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_mapping.html">rba_reactome_mapping()</a></code>

        </dt>
        <dd>Map External ID to Reactome Pathways/Reactions</dd>
      </dl></div><div class="section level2">

      <h3 id="orthology-related-queries">Orthology Related Queries<a class="anchor" aria-label="anchor" href="#orthology-related-queries"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_orthology.html">rba_reactome_orthology()</a></code>

        </dt>
        <dd>Get Orthologous (Computationally Inferred) Events</dd>
      </dl></div><div class="section level2">

      <h3 id="queries-related-to-participants">Queries Related to Participants<a class="anchor" aria-label="anchor" href="#queries-related-to-participants"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_participants.html">rba_reactome_participants()</a></code>

        </dt>
        <dd>Get Participants of a Reactome Event</dd>
      </dl></div><div class="section level2">

      <h3 id="pathway-related-queries">Pathway Related Queries<a class="anchor" aria-label="anchor" href="#pathway-related-queries"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_pathways_events.html">rba_reactome_pathways_events()</a></code>

        </dt>
        <dd>Get Events Contained in an Upstream Events</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_pathways_low.html">rba_reactome_pathways_low()</a></code>

        </dt>
        <dd>Get lower level pathways Containing a 'Physical Entity' or Event</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_pathways_top.html">rba_reactome_pathways_top()</a></code>

        </dt>
        <dd>Get Top Level Pathways in a Species</dd>
      </dl></div><div class="section level2">

      <h3 id="person-queries">Person Queries<a class="anchor" aria-label="anchor" href="#person-queries"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_people_id.html">rba_reactome_people_id()</a></code>

        </dt>
        <dd>A person by his identifiers</dd>
      </dl><dl><dt>

          <code><a href="rba_reactome_people_name.html">rba_reactome_people_name()</a></code>

        </dt>
        <dd>Get Persons Information by Name</dd>
      </dl></div><div class="section level2">

      <h3 id="common-data-retrieval">Common Data Retrieval<a class="anchor" aria-label="anchor" href="#common-data-retrieval"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_query.html">rba_reactome_query()</a></code>

        </dt>
        <dd>Query and Retrieve any Reactome knowledge-base Object</dd>
      </dl></div><div class="section level2">

      <h3 id="referenceentity-queries">ReferenceEntity Queries<a class="anchor" aria-label="anchor" href="#referenceentity-queries"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_xref.html">rba_reactome_xref()</a></code>

        </dt>
        <dd>Map Cross References IDs to Reactome ReferenceEntity</dd>
      </dl></div><div class="section level2">

      <h3 id="species-related-queries">Species Related Queries<a class="anchor" aria-label="anchor" href="#species-related-queries"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_reactome_species.html">rba_reactome_species()</a></code>

        </dt>
        <dd>Get Reactome Species</dd>
      </dl></div><div class="section level2">
      <h2 id="string-rba-string-">STRING (rba_string_*)<a class="anchor" aria-label="anchor" href="#string-rba-string-"></a></h2>

      <div class="section-desc"><p>Functions that connect you to <a href="https://string-db.org/" class="external-link">STRING</a> (Functional protein association networks) API resources.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_string_annotations.html">rba_string_annotations()</a></code>

        </dt>
        <dd>Retrieving Functional Annotation</dd>
      </dl><dl><dt>

          <code><a href="rba_string_enrichment.html">rba_string_enrichment()</a></code>

        </dt>
        <dd>Getting Functional Enrichment</dd>
      </dl><dl><dt>

          <code><a href="rba_string_enrichment_image.html">rba_string_enrichment_image()</a></code>

        </dt>
        <dd>Get STRING Enrichment Plot</dd>
      </dl><dl><dt>

          <code><a href="rba_string_enrichment_ppi.html">rba_string_enrichment_ppi()</a></code>

        </dt>
        <dd>Get Protein-Protein Interaction Enrichment</dd>
      </dl><dl><dt>

          <code><a href="rba_string_homology_inter.html">rba_string_homology_inter()</a></code>

        </dt>
        <dd>Get Similarity Scores Hits of Proteins in Different Species</dd>
      </dl><dl><dt>

          <code><a href="rba_string_homology_intra.html">rba_string_homology_intra()</a></code>

        </dt>
        <dd>Get Similarity Scores Hits of Proteins in a Species</dd>
      </dl><dl><dt>

          <code><a href="rba_string_interaction_partners.html">rba_string_interaction_partners()</a></code>

        </dt>
        <dd>Get All STRING Interaction Partners</dd>
      </dl><dl><dt>

          <code><a href="rba_string_interactions_network.html">rba_string_interactions_network()</a></code>

        </dt>
        <dd>Get STRING Network Interactions</dd>
      </dl><dl><dt>

          <code><a href="rba_string_map_ids.html">rba_string_map_ids()</a></code>

        </dt>
        <dd>Map a Set of Identifiers to STRING Identifiers</dd>
      </dl><dl><dt>

          <code><a href="rba_string_network_image.html">rba_string_network_image()</a></code>

        </dt>
        <dd>Get STRING Network Image</dd>
      </dl><dl><dt>

          <code><a href="rba_string_version.html">rba_string_version()</a></code>

        </dt>
        <dd>Get Current STRING Version</dd>
      </dl></div><div class="section level2">
      <h2 id="uniprot-rba-uniprot-">UniProt (rba_uniprot_*)<a class="anchor" aria-label="anchor" href="#uniprot-rba-uniprot-"></a></h2>

      <div class="section-desc"><p>Functions that connect you to <a href="https://www.uniprot.org/" class="external-link">UniProt</a> (Universal Protein Resource) API resources.</p></div>


    </div><div class="section level2">

      <h3 id="proteins-proteins">Proteins - Proteins<a class="anchor" aria-label="anchor" href="#proteins-proteins"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_uniprot_antigens.html">rba_uniprot_antigens()</a></code>

        </dt>
        <dd>Get Antigens by UniProt Accession</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_antigens_search.html">rba_uniprot_antigens_search()</a></code>

        </dt>
        <dd>Search Antigens in UniProt</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_epitope.html">rba_uniprot_epitope()</a></code>

        </dt>
        <dd>Retrieve Epitopes by Accession</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_epitope_search.html">rba_uniprot_epitope_search()</a></code>

        </dt>
        <dd>Search UniProt Epitopes</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_features.html">rba_uniprot_features()</a></code>

        </dt>
        <dd>Get UniProt protein sequence features by accession</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_features_search.html">rba_uniprot_features_search()</a></code>

        </dt>
        <dd>Search UniProt protein sequence features</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_mutagenesis.html">rba_uniprot_mutagenesis()</a></code>

        </dt>
        <dd>Get Mutagenesis by UniProt Accession</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_mutagenesis_search.html">rba_uniprot_mutagenesis_search()</a></code>

        </dt>
        <dd>Search Mutagenesis in UniProt</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteins.html">rba_uniprot_proteins()</a></code>

        </dt>
        <dd>Get UniProt entry by accession</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteins_crossref.html">rba_uniprot_proteins_crossref()</a></code>

        </dt>
        <dd>Get UniProt Entry by UniProt Cross-Reference Database and ID</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteins_search.html">rba_uniprot_proteins_search()</a></code>

        </dt>
        <dd>Search UniProt entries</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_rna_edit.html">rba_uniprot_rna_edit()</a></code>

        </dt>
        <dd>Retrieve Epitope by Accession</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_rna_edit_search.html">rba_uniprot_rna_edit_search()</a></code>

        </dt>
        <dd>Search RNA Editing in UniProt</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_variation.html">rba_uniprot_variation()</a></code>

        </dt>
        <dd>Get natural variants in UniProt by NIH-NCBI SNP database identifier</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_variation_search.html">rba_uniprot_variation_search()</a></code>

        </dt>
        <dd>Search UniProt Natural Variants</dd>
      </dl></div><div class="section level2">

      <h3 id="proteomics">Proteomics<a class="anchor" aria-label="anchor" href="#proteomics"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_uniprot_proteomics_hpp.html">rba_uniprot_proteomics_hpp()</a></code>

        </dt>
        <dd>Get HPP Proteomics data in UniProt</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteomics_hpp_search.html">rba_uniprot_proteomics_hpp_search()</a></code>

        </dt>
        <dd>Search HPP Proteomics data in UniProt</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteomics_non_ptm.html">rba_uniprot_proteomics_non_ptm()</a></code>

        </dt>
        <dd>Get Proteomics data in UniProt</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteomics_non_ptm_search.html">rba_uniprot_proteomics_non_ptm_search()</a></code>

        </dt>
        <dd>Search Proteomics data in UniProt</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteomics_ptm.html">rba_uniprot_proteomics_ptm()</a></code>

        </dt>
        <dd>Get Post-Translational Modification of UniProt Protein</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteomics_ptm_search.html">rba_uniprot_proteomics_ptm_search()</a></code>

        </dt>
        <dd>Search Post-Translational Modification Proteomics in UniProt</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteomics_species.html">rba_uniprot_proteomics_species()</a></code>

        </dt>
        <dd>Get UniProt Proteomics Metadata</dd>
      </dl></div><div class="section level2">

      <h3 id="proteomes">Proteomes<a class="anchor" aria-label="anchor" href="#proteomes"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_uniprot_proteomes.html">rba_uniprot_proteomes()</a></code>

        </dt>
        <dd>Get proteome by proteome/proteins UPID</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_proteomes_search.html">rba_uniprot_proteomes_search()</a></code>

        </dt>
        <dd>Search Proteomes in UniProt</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_genecentric.html">rba_uniprot_genecentric()</a></code>

        </dt>
        <dd>Get Gene-Centric proteins by UniProt Accession</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_genecentric_search.html">rba_uniprot_genecentric_search()</a></code>

        </dt>
        <dd>Search Gene-Centric Proteins</dd>
      </dl></div><div class="section level2">

      <h3 id="taxonomy">Taxonomy<a class="anchor" aria-label="anchor" href="#taxonomy"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_uniprot_taxonomy.html">rba_uniprot_taxonomy()</a></code>

        </dt>
        <dd>Get UniProt Taxonomy Nodes</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_taxonomy_lca.html">rba_uniprot_taxonomy_lca()</a></code>

        </dt>
        <dd>Get Lowest Common Ancestor (LCA) of Two Taxonomy Nodes</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_taxonomy_lineage.html">rba_uniprot_taxonomy_lineage()</a></code>

        </dt>
        <dd>Get Taxonomic Lineage</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_taxonomy_name.html">rba_uniprot_taxonomy_name()</a></code>

        </dt>
        <dd>Search UniProt Taxonomic Names</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_taxonomy_path.html">rba_uniprot_taxonomy_path()</a></code>

        </dt>
        <dd>Traverse UniProt Taxonomic Tree Path</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_taxonomy_relationship.html">rba_uniprot_taxonomy_relationship()</a></code>

        </dt>
        <dd>Get Shortest Path Between Two Taxonomy Nodes</dd>
      </dl></div><div class="section level2">

      <h3 id="coordinates">Coordinates<a class="anchor" aria-label="anchor" href="#coordinates"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_uniprot_coordinates.html">rba_uniprot_coordinates()</a></code>

        </dt>
        <dd>Get Genomic Coordinates of a Protein</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_coordinates_location.html">rba_uniprot_coordinates_location()</a></code>

        </dt>
        <dd>Search UniProt entries by taxonomy and genomic coordinates</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_coordinates_location_genome.html">rba_uniprot_coordinates_location_genome()</a></code>

        </dt>
        <dd>Get Genome coordinate by Gene Sequence position</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_coordinates_location_protein.html">rba_uniprot_coordinates_location_protein()</a></code>

        </dt>
        <dd>Get Genome coordinate by Protein Sequence position</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_coordinates_search.html">rba_uniprot_coordinates_search()</a></code>

        </dt>
        <dd>Search Genomic Coordinates of UniProt entries</dd>
      </dl></div><div class="section level2">

      <h3 id="uniparc">Uniparc<a class="anchor" aria-label="anchor" href="#uniparc"></a></h3>



    </div><div class="section level2">




      <dl><dt>

          <code><a href="rba_uniprot_uniparc.html">rba_uniprot_uniparc()</a></code>

        </dt>
        <dd>Get UniParc entry</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_uniparc_bestguess.html">rba_uniprot_uniparc_bestguess()</a></code>

        </dt>
        <dd>Get UniParc Longest Sequence for Entries</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_uniparc_search.html">rba_uniprot_uniparc_search()</a></code>

        </dt>
        <dd>Search UniParc Entries</dd>
      </dl><dl><dt>

          <code><a href="rba_uniprot_uniparc_sequence.html">rba_uniprot_uniparc_sequence()</a></code>

        </dt>
        <dd>Get UniParc Entries by Sequence</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://moosa-r.com/" class="external-link">Moosa Rezwani</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

