<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Using this function, you can retrieve highest Smith-Waterman bit scores
  among your input proteins and proteins in every other STRING species
  (e.g. the closest homologous protein of your input protein in other
  species).
  Bit Scores serve as similarity scores between protein sequence;
  And, according to STRING documentations, as a proxy for protein homology."><title>Get Similarity Scores Hits of Proteins in Different Species — rba_string_homology_inter • rbioapi</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get Similarity Scores Hits of Proteins in Different Species — rba_string_homology_inter"><meta property="og:description" content="Using this function, you can retrieve highest Smith-Waterman bit scores
  among your input proteins and proteins in every other STRING species
  (e.g. the closest homologous protein of your input protein in other
  species).
  Bit Scores serve as similarity scores between protein sequence;
  And, according to STRING documentations, as a proxy for protein homology."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-1HHD9Z0414"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-1HHD9Z0414');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/rbioapi.html">Get Started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-supported-services-amp-articles">Supported Services &amp; Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-supported-services-amp-articles">
    <a class="dropdown-item" href="../articles/rbioapi.html">1: rbioapi: Get Started</a>
    <a class="dropdown-item" href="../articles/rbioapi_enrichr.html">2.A: Enrichr &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_mieaa.html">2.B: miEEA &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_panther.html">2.C: PANTHER &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_reactome.html">2.D: Reactome &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_string.html">2.E: STRING &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_uniprot.html">2.F: UniProt &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_jaspar.html">2.G: JASPAR &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">3 Do with rbioapi: 1. Over-Representation (Enrichment) Analysis</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Get Similarity Scores Hits of Proteins in Different Species</h1>
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/HEAD/R/stringdb.R" class="external-link"><code>R/stringdb.R</code></a></small>
      <div class="d-none name"><code>rba_string_homology_inter.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Using this function, you can retrieve highest Smith-Waterman bit scores
  among your input proteins and proteins in every other STRING species
  (e.g. the closest homologous protein of your input protein in other
  species).
  Bit Scores serve as similarity scores between protein sequence;
  And, according to STRING documentations, as a proxy for protein homology.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">rba_string_homology_inter</span><span class="op">(</span><span class="va">ids</span>, species <span class="op">=</span> <span class="cn">NULL</span>, species_b <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>ids</dt>
<dd><p>Your protein ID(s). It is strongly recommended to supply
STRING IDs. See <code><a href="rba_string_map_ids.html">rba_string_map_ids</a></code> for more information.</p></dd>
<dt>species</dt>
<dd><p>Numeric: NCBI Taxonomy identifier of your input proteins;
Human Taxonomy ID is 9606. (Recommended, but optional if your input is
less than 100 IDs.)</p></dd>
<dt>species_b</dt>
<dd><p>(optional) Numeric: one or more NCBI Taxonomy identifiers
of species to limit the closets homologous proteins search.</p></dd>
<dt>...</dt>
<dd><p>rbioapi option(s). See <code><a href="rba_options.html">rba_options</a></code>'s
arguments manual for more information on available options.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A data frame with Your input proteins and it's closest homologous
  proteins among all other (or a defined) STRING species.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Note that this function will return the highest similarity score hits of
  your given protein(s) and their closets homologous proteins in other
  species. to retrieve similarity scores of different proteins within the
  same species see <code><a href="rba_string_homology_intra.html">rba_string_homology_intra</a></code>.
  <br> Similarity matrix is imported -by STRING- from:
  <a href="https://cube.univie.ac.at/resources/simap" class="external-link">Similarity Matrix of
  Proteins (SIMAP)</a></p>
    </div>
    <div class="section level2">
    <h2 id="corresponding-api-resources">Corresponding API Resources<a class="anchor" aria-label="anchor" href="#corresponding-api-resources"></a></h2>
    

<p>"POST https://string-db.org/api/[output-format]/homology_best?
 identifiers=[your_identifiers]"</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>Szklarczyk D, Gable AL, Lyon D, Junge A, Wyder S, Huerta-Cepas J,
  Simonovic M, Doncheva NT, Morris JH, Bork P, Jensen LJ, Mering CV.
  STRING v11: protein-protein association networks with increased coverage,
  supporting functional discovery in genome-wide experimental datasets.
  Nucleic Acids Res. 2019 Jan 8;47(D1):D607-D613. doi: 10.1093/nar/gky1131.
  PMID: 30476243; PMCID: PMC6323986.</p></li>
<li><p><a href="https://string-db.org/help/api/" class="external-link">STRING API Documentation</a></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="rba_string_map_ids.html">rba_string_map_ids</a>, <a href="rba_string_homology_intra.html">rba_string_homology_intra</a></code></p>
<p>Other "STRING": 
<code><a href="rba_string_annotations.html">rba_string_annotations</a>()</code>,
<code><a href="rba_string_enrichment_ppi.html">rba_string_enrichment_ppi</a>()</code>,
<code><a href="rba_string_enrichment.html">rba_string_enrichment</a>()</code>,
<code><a href="rba_string_homology_intra.html">rba_string_homology_intra</a>()</code>,
<code><a href="rba_string_interaction_partners.html">rba_string_interaction_partners</a>()</code>,
<code><a href="rba_string_interactions_network.html">rba_string_interactions_network</a>()</code>,
<code><a href="rba_string_map_ids.html">rba_string_map_ids</a>()</code>,
<code><a href="rba_string_network_image.html">rba_string_network_image</a>()</code>,
<code><a href="rba_string_version.html">rba_string_version</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co"># \donttest{</span>
<span class="fu">rba_string_homology_inter</span><span class="op">(</span>ids <span class="op">=</span> <span class="st">"p53"</span>,
    species <span class="op">=</span> <span class="fl">9606</span>,
    species_b <span class="op">=</span> <span class="fl">7070</span><span class="op">)</span>
<span class="co"># }</span>
<span class="co"># \donttest{</span>
<span class="fu">rba_string_homology_inter</span><span class="op">(</span>ids <span class="op">=</span> <span class="st">"ENSP00000269305"</span>, species <span class="op">=</span> <span class="fl">9606</span><span class="op">)</span>
<span class="co"># }</span>

</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

