<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>UniProt maintains 
  sequence annotations (features) that describe regions
  in the protein sequence. Using this function, you can search and
  retrieve UniProt proteins' sequence annotations (features). you may also
  refine your search query with variety of modifiers. — rba_uniprot_features_search • rbioapi</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="UniProt maintains 
  sequence annotations (features) that describe regions
  in the protein sequence. Using this function, you can search and
  retrieve UniProt proteins' sequence annotations (features). you may also
  refine your search query with variety of modifiers. — rba_uniprot_features_search" />
<meta property="og:description" content="Note that this is a search function. Thus, you are not required to fill
  every argument; You may use whatever combinations of arguments you see
  fit for your query.
   UniProt Entries are grouped in two sections:
Reviewed(Swiss-Prot): Manually annotated records with information
  extracted from literature and curator-evaluated computational analysis.
Unreviewed (TrEMBL): Computationally analyzed records that await
  full manual annotation.

" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-1HHD9Z0414"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-1HHD9Z0414');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rbioapi</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/rbioapi.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Supported Services
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/rbioapi.html">1: rbioapi: Get Started</a>
    </li>
    <li>
      <a href="../articles/rbioapi_enrichr.html">2.A: Enrichr &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_mieaa.html">2.B: miEEA &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_panther.html">2.C: PANTHER &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_reactome.html">2.D: Reactome &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_string.html">2.E: STRING &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_uniprot.html">2.F: UniProt &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_jaspar.html">2.G: JASPAR &amp; rbioapi</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>UniProt maintains <a href='https://www.uniprot.org/help/sequence_annotation'>
  sequence annotations (features)</a> that describe regions
  in the protein sequence. Using this function, you can search and
  retrieve UniProt proteins' sequence annotations (features). you may also
  refine your search query with variety of modifiers.</h1>
    <small class="dont-index">Source: <a href='https://github.com/moosa-r/rbioapi/blob/master/R/uniprot_proteins.R'><code>R/uniprot_proteins.R</code></a></small>
    <div class="hidden name"><code>rba_uniprot_features_search.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Note that this is a search function. Thus, you are not required to fill
  every argument; You may use whatever combinations of arguments you see
  fit for your query.
  <br /> UniProt Entries are grouped in two sections:</p><ol>
<li><p>Reviewed(Swiss-Prot): Manually annotated records with information
  extracted from literature and curator-evaluated computational analysis.</p></li>
<li><p>Unreviewed (TrEMBL): Computationally analyzed records that await
  full manual annotation.</p></li>
</ol>

    </div>

    <pre class="usage"><span class='fu'>rba_uniprot_features_search</span><span class='op'>(</span>
  accession <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene <span class='op'>=</span> <span class='cn'>NULL</span>,
  exact_gene <span class='op'>=</span> <span class='cn'>NULL</span>,
  protein <span class='op'>=</span> <span class='cn'>NULL</span>,
  reviewed <span class='op'>=</span> <span class='cn'>NULL</span>,
  organism <span class='op'>=</span> <span class='cn'>NULL</span>,
  taxid <span class='op'>=</span> <span class='cn'>NULL</span>,
  categories <span class='op'>=</span> <span class='cn'>NULL</span>,
  types <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>accession</th>
      <td><p><a href='https://www.uniprot.org/help/accession_numbers'>
UniProtKB primary or secondary accession</a>(s). You can provide up to 100
accession numbers.</p></td>
    </tr>
    <tr>
      <th>gene</th>
      <td><p><a href='https://www.uniprot.org/help/gene_name'>UniProt gene
name(s)</a>. You can provide up to 20 gene names. e.g. if you provide
"CD40", "CD40 ligand" will also be included.</p></td>
    </tr>
    <tr>
      <th>exact_gene</th>
      <td><p><a href='https://www.uniprot.org/help/gene_name'>UniProt
exact gene name(s)</a>. You can provide up to 20 exact gene names. e.g.
if you provide "CD40", "CD40 ligand" will not be included in the results.</p></td>
    </tr>
    <tr>
      <th>protein</th>
      <td><p><a href='https://www.uniprot.org/help/protein_names'>UniProt
protein name</a></p></td>
    </tr>
    <tr>
      <th>reviewed</th>
      <td><p>Logical: If TRUE, only return
"UniProtKB/Swiss-Prot" (reviewed) entries; If FALSE, only return TrEMBL
(un-reviewed) entries.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><a href='https://www.uniprot.org/taxonomy/'>Organism name</a>.</p></td>
    </tr>
    <tr>
      <th>taxid</th>
      <td><p>NIH-NCBI <a href='https://www.uniprot.org/taxonomy/'>Taxon ID</a>.
You can provide up to 20 taxon IDs.</p></td>
    </tr>
    <tr>
      <th>categories</th>
      <td><p><a href='https://www.uniprot.org/help/sequence_annotation'>
Sequence annotation (Features)</a> categories (subsection). accepted values
are: "MOLECULE_PROCESSING", "TOPOLOGY", "SEQUENCE_INFORMATION",
"STRUCTURAL", "DOMAINS_AND_SITES", "PTM", "VARIANTS" and/or "MUTAGENESIS".
You can provide up to 8 categories.</p></td>
    </tr>
    <tr>
      <th>types</th>
      <td><p><a href='https://www.uniprot.org/help/sequence_annotation'>
Sequence annotation (Features)</a> types. accepted values
are: "INIT_MET", "SIGNAL", "PROPEP", "TRANSIT", "CHAIN", "PEPTIDE",
"TOPO_DOM", "TRANSMEM", "DOMAIN", "REPEAT", "CA_BIND", "ZN_FING",
"DNA_BIND", "NP_BIND", "REGION", "COILED", "MOTIF", "COMPBIAS",
"ACT_SITE", "METAL", "BINDING", "SITE", "NON_STD", "MOD_RES", "LIPID",
"CARBOHYD", "DISULFID", "CROSSLNK", "VAR_SEQ", "VARIANT", "MUTAGEN",
"UNSURE", "CONFLICT", "NON_CONS", "NON_TER", "HELIX", "TURN", "STRAND"
and/or "INTRAMEM". You can provide up to 20 types.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>rbioapi option(s). Refer to <code><a href='rba_options.html'>rba_options</a></code>'s
arguments documentation for more information on available options.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>List where each element corresponds to one UniProt entity returned
  by your search query. The element itself is a sub-list containing all
  information that UniProt has about that entity.</p>
    <h2 class="hasAnchor" id="corresponding-api-resources"><a class="anchor" href="#corresponding-api-resources"></a>Corresponding API Resources</h2>

    

<p>"GET https://www.ebi.ac.uk/proteins/api/features"</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    
<ul>
<li><p>Andrew Nightingale, Ricardo Antunes, Emanuele Alpi, Borisas
  Bursteinas, Leonardo Gonzales, Wudong Liu, Jie Luo, Guoying Qi, Edd
  Turner, Maria Martin, The Proteins API: accessing key integrated protein
  and genome information, Nucleic Acids Research, Volume 45, Issue W1,
  3 July 2017, Pages W539–W544, https://doi.org/10.1093/nar/gkx237</p></li>
<li><p><a href='https://www.ebi.ac.uk/proteins/api/doc/'>Proteins API
  Documentation</a></p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other "UniProt - Features": 
<code><a href='rba_uniprot_features.html'>rba_uniprot_features</a>()</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># \donttest{</span>
<span class='fu'>rba_uniprot_features_search</span><span class='op'>(</span>accession <span class='op'>=</span> <span class='st'>"Q99616"</span><span class='op'>)</span>
<span class='co'># }</span>
<span class='co'># \donttest{</span>
<span class='fu'>rba_uniprot_features_search</span><span class='op'>(</span>gene <span class='op'>=</span> <span class='st'>"cd40"</span><span class='op'>)</span>
<span class='co'># }</span>
<span class='co'># \donttest{</span>
<span class='fu'>rba_uniprot_features_search</span><span class='op'>(</span>gene <span class='op'>=</span> <span class='st'>"cd40 ligand"</span><span class='op'>)</span>
<span class='co'># }</span>
<span class='co'># \donttest{</span>
<span class='fu'>rba_uniprot_features_search</span><span class='op'>(</span>gene <span class='op'>=</span> <span class='st'>"cd40"</span>,  reviewed <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
<span class='co'># }</span>
<span class='co'># \donttest{</span>
<span class='fu'>rba_uniprot_features_search</span><span class='op'>(</span>accession <span class='op'>=</span> <span class='st'>"Q99616"</span>,
    categories <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"MOLECULE_PROCESSING"</span>, <span class='st'>"TOPOLOGY"</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># }</span>
<span class='co'># \donttest{</span>
<span class='fu'>rba_uniprot_features_search</span><span class='op'>(</span>accession <span class='op'>=</span> <span class='st'>"Q99616"</span>, types <span class='op'>=</span> <span class='st'>"DISULFID"</span><span class='op'>)</span>
<span class='co'># }</span>

</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


