<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Using this function you can use PANTHER services to perform
  over-representation enrichment analysis. This statistical test will
  compare your input genes to a set of defined gene lists to determine
  if they are over/under-represented."><title>PANTHER Over-Representation Enrichment Analysis — rba_panther_enrich • rbioapi</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="PANTHER Over-Representation Enrichment Analysis — rba_panther_enrich"><meta property="og:description" content="Using this function you can use PANTHER services to perform
  over-representation enrichment analysis. This statistical test will
  compare your input genes to a set of defined gene lists to determine
  if they are over/under-represented."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/rbioapi.html">Get Started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-supported-services-amp-articles">Supported Services &amp; Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-supported-services-amp-articles">
    <a class="dropdown-item" href="../articles/rbioapi.html">1: rbioapi: Get Started</a>
    <a class="dropdown-item" href="../articles/rbioapi_enrichr.html">2.A: Enrichr &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_mieaa.html">2.B: miEEA &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_panther.html">2.C: PANTHER &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_reactome.html">2.D: Reactome &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_string.html">2.E: STRING &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_uniprot.html">2.F: UniProt &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_jaspar.html">2.G: JASPAR &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">3 Do with rbioapi: 1. Over-Representation (Enrichment) Analysis</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>PANTHER Over-Representation Enrichment Analysis</h1>
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/HEAD/R/panther.R" class="external-link"><code>R/panther.R</code></a></small>
      <div class="d-none name"><code>rba_panther_enrich.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Using this function you can use PANTHER services to perform
  over-representation enrichment analysis. This statistical test will
  compare your input genes to a set of defined gene lists to determine
  if they are over/under-represented.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">rba_panther_enrich</span><span class="op">(</span>
  <span class="va">genes</span>,
  <span class="va">organism</span>,
  <span class="va">annot_dataset</span>,
  test_type <span class="op">=</span> <span class="st">"FISHER"</span>,
  correction <span class="op">=</span> <span class="st">"FDR"</span>,
  cutoff <span class="op">=</span> <span class="cn">NULL</span>,
  ref_genes <span class="op">=</span> <span class="cn">NULL</span>,
  ref_organism <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>genes</dt>
<dd><p>Character vector of genes identifiers with maximum length of
10000. Can be any of: Ensemble gene ID, Ensemble protein ID, Ensemble
transcript ID, Entrez gene ID, gene symbol, NCBI GI, HGNC ID,
International protein index ID, NCBI UniGene ID, UniProt accession
and/or UniProt ID.</p></dd>
<dt>organism</dt>
<dd><p>(numeric) NCBI taxon ID. run <code><a href="rba_panther_info.html">rba_panther_info</a></code>
with argument 'what = "organisms"' to get a list of PANTHER's
supported organisms.</p></dd>
<dt>annot_dataset</dt>
<dd><p>A PANTHER dataset ID to test your input against it.
run <code><a href="rba_panther_info.html">rba_panther_info</a></code>with argument 'what = "datasets"' to
get a list of PANTHER's supported datasets.</p></dd>
<dt>test_type</dt>
<dd><p>statistical test type to calculate the p values. either
"FISHER" (default) or "BINOMIAL".</p></dd>
<dt>correction</dt>
<dd><p>p value correction method. either "FDR" (default),
"BONFERRONI" or "NONE".</p></dd>
<dt>cutoff</dt>
<dd><p>(Numeric) (Optional) a threshold to filter the results.
if correction is "FDR", the threshold will be applied to fdr column's
values; if otherwise, the threshold will be applied to p value column.</p></dd>
<dt>ref_genes</dt>
<dd><p>(Optional) A set of genes that will be used as the test's
background (reference/universe) gene set. If no value supplied, all of
the genes in specified organism will be used. maximum length and supported
IDs are the same as 'genes' argument.</p></dd>
<dt>ref_organism</dt>
<dd><p>(Optional) if 'ref_genes' is used, you can specify
the organisms which correspond to your supplied IDs in 'ref_genes'
argument. see 'organism' argument for supported values.</p></dd>
<dt>...</dt>
<dd><p>rbioapi option(s). See <code><a href="rba_options.html">rba_options</a></code>'s
arguments manual for more information on available options.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list with the prov</p>
    </div>
    <div class="section level2">
    <h2 id="corresponding-api-resources">Corresponding API Resources<a class="anchor" aria-label="anchor" href="#corresponding-api-resources"></a></h2>
    

<p>"POST http://www.pantherdb.org/services/oai/pantherdb/enrich/overrep"</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>Mi, H., Muruganujan, A., Ebert, D., Huang, X., &amp; Thomas, P. D.
  (2019). PANTHER version 14: more genomes, a new PANTHER GO-slim and
  improvements in enrichment analysis tools. Nucleic acids research, 47(D1),
  D419-D426.</p></li>
<li><p>Mi, H., Muruganujan, A., Huang, X., Ebert, D., Mills, C., Guo, X.,
  &amp; Thomas, P. D. (2019). Protocol Update for large-scale genome and gene
  function analysis with the PANTHER classification system (v. 14.0).
  Nature protocols, 14(3), 703-721.</p></li>
<li><p><a href="http://www.pantherdb.org/services/details.jsp" class="external-link">PANTHER
  Services Details</a></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other "PANTHER": 
<code><a href="rba_panther_family.html">rba_panther_family</a>()</code>,
<code><a href="rba_panther_homolog.html">rba_panther_homolog</a>()</code>,
<code><a href="rba_panther_info.html">rba_panther_info</a>()</code>,
<code><a href="rba_panther_mapping.html">rba_panther_mapping</a>()</code>,
<code><a href="rba_panther_ortholog.html">rba_panther_ortholog</a>()</code>,
<code><a href="rba_panther_tree_grafter.html">rba_panther_tree_grafter</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co"># \donttest{</span>
<span class="fu">rba_panther_enrich</span><span class="op">(</span>genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>,
        <span class="st">"CDK1"</span>, <span class="st">"KIF23"</span>, <span class="st">"PLK1"</span>, <span class="st">"RAC2"</span>, <span class="st">"RACGAP1"</span><span class="op">)</span>,
    organism <span class="op">=</span> <span class="fl">9606</span>, annot_dataset <span class="op">=</span> <span class="st">"GO:0008150"</span>,
    cutoff <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>
<span class="co"># }</span>

</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

