<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content='Using this function, you can perform Reactome Analysis In a convenient way.
  The Analysis Type will be chosen depending on your supplied
  input:
If you supply a vector or a single-columned table,
  "Over-Representation" analysis will be performed.
If you supply a multi-column table, with the first column being
  molecules identifiers and the rest being numeral expression values,
  "Expression" analysis will be performed.
See the details section for the accepted input types and format.'><title>Reactome Over-Representation or Expression Analysis — rba_reactome_analysis • rbioapi</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reactome Over-Representation or Expression Analysis — rba_reactome_analysis"><meta property="og:description" content='Using this function, you can perform Reactome Analysis In a convenient way.
  The Analysis Type will be chosen depending on your supplied
  input:
If you supply a vector or a single-columned table,
  "Over-Representation" analysis will be performed.
If you supply a multi-column table, with the first column being
  molecules identifiers and the rest being numeral expression values,
  "Expression" analysis will be performed.
See the details section for the accepted input types and format.'><meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/rbioapi.html">Get Started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles-amp-supported-services">Articles &amp; Supported Services</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles-amp-supported-services">
    <a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Reactome Over-Representation or Expression Analysis</h1>
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/HEAD/R/reactome_analysis.R" class="external-link"><code>R/reactome_analysis.R</code></a></small>
      <div class="d-none name"><code>rba_reactome_analysis.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Using this function, you can perform Reactome Analysis In a convenient way.
  The Analysis Type will be chosen depending on your supplied
  input:</p><ol><li><p>If you supply a vector or a single-columned table,
  "Over-Representation" analysis will be performed.</p></li>
<li><p>If you supply a multi-column table, with the first column being
  molecules identifiers and the rest being numeral expression values,
  "Expression" analysis will be performed.</p></li>
</ol><p>See the details section for the accepted input types and format.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">rba_reactome_analysis</span><span class="op">(</span>
  <span class="va">input</span>,
  input_format <span class="op">=</span> <span class="cn">NULL</span>,
  projection <span class="op">=</span> <span class="cn">FALSE</span>,
  interactors <span class="op">=</span> <span class="cn">FALSE</span>,
  species <span class="op">=</span> <span class="cn">NULL</span>,
  sort_by <span class="op">=</span> <span class="st">"ENTITIES_PVALUE"</span>,
  order <span class="op">=</span> <span class="st">"ASC"</span>,
  resource <span class="op">=</span> <span class="st">"TOTAL"</span>,
  p_value <span class="op">=</span> <span class="fl">1</span>,
  include_disease <span class="op">=</span> <span class="cn">TRUE</span>,
  min <span class="op">=</span> <span class="cn">NULL</span>,
  max <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>A vector, data frame, matrix or a local file path or URL
that points to your data. See "Details section" for more information of
how to organize and supply your input.</p></dd>
<dt>input_format</dt>
<dd><p>(Optional) This function will automatically identify
your supplied input's format. But in case of unexpected issues or if you
want to be explicit, set this argument to one of:</p><ul><li><p>"table": If you supplied a data frame or matrix as input.</p></li>
<li><p>"vector": If you supplied a simple vector (numeric or character) as
input.</p></li>
<li><p>"file": If you supplied a local file path pointing to a
correctly-formatted text file.</p></li>
<li><p>"url": If you supplied a URL pointing to a correctly-formatted
text file.</p></li>
</ul></dd>
<dt>projection</dt>
<dd><p>Logical (default = FALSE) Should non-human identifiers
be projected to their human equivalents? (using Reactome orthology data)</p></dd>
<dt>interactors</dt>
<dd><p>Logical (default = FALSE) Should IntAct interaction data
be used to increase the analysis background?</p></dd>
<dt>species</dt>
<dd><p>Numeric or Character: NCBI Taxonomy identifier (Human
is 9606), species name (e.g. "Homo sapiens") or Reactome DbId (e.g
Homo sapiens is 48887). See
 <code><a href="rba_reactome_species.html">rba_reactome_species</a></code> or
 <a href="https://reactome.org/content/schema/objects/Species/" class="external-link">Reactome
 Data Schema: Entries: Species</a>.</p></dd>
<dt>sort_by</dt>
<dd><p>Sort the result based on what column? available choices
are: "NAME", "TOTAL_ENTITIES", "TOTAL_INTERACTORS", "TOTAL_REACTIONS",
"FOUND_ENTITIES", "FOUND_INTERACTORS", "FOUND_REACTIONS", "ENTITIES_RATIO",
"ENTITIES_PVALUE", "ENTITIES_FDR" or "REACTIONS_RATIO"</p></dd>
<dt>order</dt>
<dd><p>Sort Order. Can be either "ASC" (default) or "DESC".</p></dd>
<dt>resource</dt>
<dd><p>Filter results based on the resource. Default is "TOTAL",
available choices are: "TOTAL", "UNIPROT", "ENSEMBL", "CHEBI", "IUPHAR",
"MIRBASE", "NCBI_PROTEIN", "EMBL", "COMPOUND", "ENTITIES_FDR" or
"PUBCHEM_COMPOUND".</p></dd>
<dt>p_value</dt>
<dd><p>Set a P value threshold. Only results with P value equal to
or less than your supplied threshold will be returned. (default = 1,
Meaning no P value filtering)</p></dd>
<dt>include_disease</dt>
<dd><p>Logical (default = TRUE) Should the disease pathways
be included in the results?</p></dd>
<dt>min</dt>
<dd><p>(numeric) Minimum number of entities that a pathways should have
to be included in the results.</p></dd>
<dt>max</dt>
<dd><p>(numeric) Maximum number of entities that a pathways should have
to be included in the results.</p></dd>
<dt>...</dt>
<dd><p>rbioapi option(s). See <code><a href="rba_options.html">rba_options</a></code>'s
arguments manual for more information on available options.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>List containing the results and information of your analysis. Note
  that you can use the token returned in the "summary" sub-list of the
  results (i.e. results$summary$token) to retrieve your results later or
  in other Reactome analysis functions.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>You can supply your table or vector input in numerous formats:</p><ol><li><p>A R object which can be data frame, matrix or a simple vector.</p></li>
<li><p>A path to a local text file in your device that contains the molecules
  data. (The file should be formatted correctly, see below.)</p></li>
<li><p>A URL pointing to a text file on the web that contains the molecules
  data. (The file should be formatted correctly, see below.</p></li>
</ol><p>If you supply a text file (as a local file path or URL), it should be
  in TSV (Tab-Separated Values) format; Column names should start with "#"
  character. Note that if you are providing the file for
  "Over-Representation" analysis (i.e. Single columned-data) this header
  line is optional and will be used as your 'Sample Name', otherwise it is
  required.
  <br> Also, form the "summary" element in the function's output, you can see
  how Reactome Interpreted your input and subsequently the type of analysis
  that has been performed.
  <br> There is no strict criteria about the type of your molecules Identifiers,
  Reactome will Map the IDs to it's internal database entities.
  Nevertheless, You can check if all your identifiers has been found in
  "identifiersNotFound" element in the function's output.
  <br> After Any Analysis, Reactome will associate a token to your analysis.
  It can be later used to in function that requires the token (e.g to retrieve
  the analysis results, download pdf).
  <br> Note that Reactome will store your token for only 7 days. You can
  download your full results with
  <code><a href="rba_reactome_analysis_download.html">rba_reactome_analysis_download</a></code>, and re-import it anytime to
  reactome (using <code><a href="rba_reactome_analysis_import.html">rba_reactome_analysis_import</a></code>) to generate
  a new token.</p>
    </div>
    <div class="section level2">
    <h2 id="corresponding-api-resources">Corresponding API Resources<a class="anchor" aria-label="anchor" href="#corresponding-api-resources"></a></h2>
    

<p>"POST https://reactome.org/AnalysisService/identifiers/form"
 <br> "POST https://reactome.org/AnalysisService/identifiers/url"
 <br> "POST https://reactome.org/AnalysisService/identifiers/form/projection"
 <br> "POST https://reactome.org/AnalysisService/identifiers/url/projection"</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>Marc Gillespie, Bijay Jassal, Ralf Stephan, Marija Milacic, Karen
  Rothfels, Andrea Senff-Ribeiro, Johannes Griss, Cristoffer Sevilla, Lisa
  Matthews, Chuqiao Gong, Chuan Deng, Thawfeek Varusai, Eliot Ragueneau,
  Yusra Haider, Bruce May, Veronica Shamovsky, Joel Weiser, Timothy Brunson,
  Nasim Sanati, Liam Beckman, Xiang Shao, Antonio Fabregat, Konstantinos
  Sidiropoulos, Julieth Murillo, Guilherme Viteri, Justin Cook, Solomon
  Shorser, Gary Bader, Emek Demir, Chris Sander, Robin Haw, Guanming Wu,
  Lincoln Stein, Henning Hermjakob, Peter D’Eustachio, The reactome
  pathway knowledgebase 2022, Nucleic Acids Research, 2021;, kab1028,
  https://doi.org/10.1093/nar/gkab1028</p></li>
<li><p>Griss J, Viteri G, Sidiropoulos K, Nguyen V, Fabregat A,
  Hermjakob H. ReactomeGSA - Efficient Multi-Omics Comparative Pathway
  Analysis. Mol Cell Proteomics. 2020 Sep 9. doi: 10.1074/mcp. PubMed
  PMID: 32907876.</p></li>
<li><p><a href="https://reactome.org/AnalysisService/" class="external-link">Reactome Analysis
  Services API Documentation</a></p></li>
<li><p><a href="https://reactome.org/cite" class="external-link">Citations note on Reactome website</a></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other "Reactome Analysis Service": 
<code><a href="rba_reactome_analysis_download.html">rba_reactome_analysis_download</a>()</code>,
<code><a href="rba_reactome_analysis_import.html">rba_reactome_analysis_import</a>()</code>,
<code><a href="rba_reactome_analysis_mapping.html">rba_reactome_analysis_mapping</a>()</code>,
<code><a href="rba_reactome_analysis_pdf.html">rba_reactome_analysis_pdf</a>()</code>,
<code><a href="rba_reactome_analysis_species.html">rba_reactome_analysis_species</a>()</code>,
<code><a href="rba_reactome_analysis_token.html">rba_reactome_analysis_token</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co"># \donttest{</span>
<span class="fu">rba_reactome_analysis</span><span class="op">(</span>input <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span><span class="op">)</span><span class="op">)</span>
<span class="co"># }</span>
<span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">rba_reactome_analysis</span><span class="op">(</span>input <span class="op">=</span> <span class="st">"c:/rbioapi/genes.txt"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">rba_reactome_analysis</span><span class="op">(</span>input <span class="op">=</span> <span class="st">"https://qazwsx.com/genes.txt"</span><span class="op">)</span>
<span class="op">}</span>

</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

