<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Using this function, you can search and retrieve
  Natural variant(s) that
  has been annotated in the protein's sequences. You may also refine your
  search with modifiers such as source type, disease etc. See
  &quot;Arguments section&quot; for more information."><title>Search UniProt Natural Variants — rba_uniprot_variation_search • rbioapi</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Search UniProt Natural Variants — rba_uniprot_variation_search"><meta property="og:description" content="Using this function, you can search and retrieve
  Natural variant(s) that
  has been annotated in the protein's sequences. You may also refine your
  search with modifiers such as source type, disease etc. See
  &quot;Arguments section&quot; for more information."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-1HHD9Z0414"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-1HHD9Z0414');
</script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/rbioapi.html">Get Started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-supported-services-amp-articles">Supported Services &amp; Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-supported-services-amp-articles">
    <a class="dropdown-item" href="../articles/rbioapi.html">1: rbioapi: Get Started</a>
    <a class="dropdown-item" href="../articles/rbioapi_enrichr.html">2.A: Enrichr &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_mieaa.html">2.B: miEEA &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_panther.html">2.C: PANTHER &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_reactome.html">2.D: Reactome &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_string.html">2.E: STRING &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_uniprot.html">2.F: UniProt &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_jaspar.html">2.G: JASPAR &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">3 Do with rbioapi: 1. Over-Representation (Enrichment) Analysis</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Search UniProt Natural Variants</h1>
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/HEAD/R/uniprot_proteins.R" class="external-link"><code>R/uniprot_proteins.R</code></a></small>
      <div class="d-none name"><code>rba_uniprot_variation_search.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Using this function, you can search and retrieve
  <a href="https://www.uniprot.org/help/variant" class="external-link">Natural variant(s)</a> that
  has been annotated in the protein's sequences. You may also refine your
  search with modifiers such as source type, disease etc. See
  "Arguments section" for more information.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">rba_uniprot_variation_search</span><span class="op">(</span>
  accession <span class="op">=</span> <span class="cn">NULL</span>,
  source_type <span class="op">=</span> <span class="cn">NULL</span>,
  consequence_type <span class="op">=</span> <span class="cn">NULL</span>,
  wild_type <span class="op">=</span> <span class="cn">NULL</span>,
  alternative_sequence <span class="op">=</span> <span class="cn">NULL</span>,
  location <span class="op">=</span> <span class="cn">NULL</span>,
  disease <span class="op">=</span> <span class="cn">NULL</span>,
  omim <span class="op">=</span> <span class="cn">NULL</span>,
  evidence <span class="op">=</span> <span class="cn">NULL</span>,
  taxid <span class="op">=</span> <span class="cn">NULL</span>,
  db_type <span class="op">=</span> <span class="cn">NULL</span>,
  db_id <span class="op">=</span> <span class="cn">NULL</span>,
  save_peff <span class="op">=</span> <span class="cn">FALSE</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>accession</dt>
<dd><p><a href="https://www.uniprot.org/help/accession_numbers" class="external-link">
UniProtKB primary or secondary accession</a>(s). You can supply up to 100
accession numbers.</p></dd>
<dt>source_type</dt>
<dd><p>Variation's source type. You can choose up to two of:
"UniProt", "large scale study" and/or "mixed".</p></dd>
<dt>consequence_type</dt>
<dd><p>Variation's consequence type. You can choose up to
two of: "missense", "stop gained" or "stop lost".</p></dd>
<dt>wild_type</dt>
<dd><p>Wild type amino acid. Accepted values are IUPAC
single-letter amino acid (e.g. D for    Aspartic acid) and "*" for stop
codon. You can supply up to 20 values.</p></dd>
<dt>alternative_sequence</dt>
<dd><p>Alternative amino acid. Accepted values are IUPAC
single-letter amino acid (e.g. D for    Aspartic acid) and "*" for stop codon
and "-" for deletion. You can supply up to 20 values.</p></dd>
<dt>location</dt>
<dd><p>A valid amino acid range (e.g. 10-25) within the sequence
range where the variation occurs. You can supply up to 20 values.</p></dd>
<dt>disease</dt>
<dd><p><a href="https://www.uniprot.org/diseases/" class="external-link">Human disease</a>
that are associated with a sequence variation. Accepted values are
disease name (e.g. Alzheimer disease 18), partial disease name
(Alzheimer) and/or disease acronym (e.g. AD). You can supply up to
20 values.</p></dd>
<dt>omim</dt>
<dd><p><a href="https://www.ncbi.nlm.nih.gov/omim" class="external-link">OMIM</a> ID that is
associated with a variation. You can supply up to 20 values.</p></dd>
<dt>evidence</dt>
<dd><p>Pubmed ID of the variation's
<a href="https://www.uniprot.org/citations/" class="external-link">citation</a> You can supply up
to 20 values.</p></dd>
<dt>taxid</dt>
<dd><p>NIH-NCBI <a href="https://www.uniprot.org/taxonomy/" class="external-link">Taxon ID</a>.
You can supply up to 20 taxon IDs.</p></dd>
<dt>db_type</dt>
<dd><p>cross-reference database of the variation. You can supply
up to two of the following:</p><ul><li><p>"dbSNP": <a href="https://www.ncbi.nlm.nih.gov/snp/" class="external-link">NIH-NCBI dbSNP
database</a>.</p></li>
<li><p>"cosmic curate": <a href="https://cancer.sanger.ac.uk/cosmic/" class="external-link">COSMIC
(the Catalogue of Somatic Mutations in Cancer)</a></p></li>
<li><p>"ClinVar": <a href="https://www.ncbi.nlm.nih.gov/clinvar/" class="external-link">NIH-NCBI
ClinVar</a></p></li>
</ul></dd>
<dt>db_id</dt>
<dd><p>The variation ID in a Cross-reference (external) database.
You can supply up to 20 values.</p></dd>
<dt>save_peff</dt>
<dd><p>Logical or Character:</p><ul><li><p>FALSE: (default) Do not save PEFF file, just return as a list object.</p></li>
<li><p>TRUE: Save as PEFF file to an automatically-generated path.</p></li>
<li><p>Character string: A valid file path to save the PEFF file.</p></li>
</ul></dd>
<dt>...</dt>
<dd><p>rbioapi option(s). See <code><a href="rba_options.html">rba_options</a></code>'s
arguments manual for more information on available options.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>List where each element corresponds to one UniProt entity returned
  by your search query. The element itself is a sub-list containing all
  information that UniProt has about that Variation.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Note that this is a search function. Thus, you are not required to fill
  every argument; You may use whatever combinations of arguments you see
  fit for your query.</p>
    </div>
    <div class="section level2">
    <h2 id="corresponding-api-resources">Corresponding API Resources<a class="anchor" aria-label="anchor" href="#corresponding-api-resources"></a></h2>
    

<p>"GET https://www.ebi.ac.uk/proteins/api/variation"</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    
<ul><li><p>Andrew Nightingale, Ricardo Antunes, Emanuele Alpi, Borisas
  Bursteinas, Leonardo Gonzales, Wudong Liu, Jie Luo, Guoying Qi, Edd
  Turner, Maria Martin, The Proteins API: accessing key integrated protein
  and genome information, Nucleic Acids Research, Volume 45, Issue W1,
  3 July 2017, Pages W539–W544, https://doi.org/10.1093/nar/gkx237</p></li>
<li><p><a href="https://www.ebi.ac.uk/proteins/api/doc/" class="external-link">Proteins API
  Documentation</a></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other "UniProt - Variation": 
<code><a href="rba_uniprot_variation.html">rba_uniprot_variation</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co"># \donttest{</span>
<span class="fu">rba_uniprot_variation_search</span><span class="op">(</span>accession <span class="op">=</span> <span class="st">"P05067"</span><span class="op">)</span>
<span class="co"># }</span>
<span class="co"># \donttest{</span>
<span class="fu">rba_uniprot_variation_search</span><span class="op">(</span>disease <span class="op">=</span> <span class="st">"alzheimer disease, 18"</span><span class="op">)</span>
<span class="co"># }</span>
<span class="co"># \donttest{</span>
<span class="fu">rba_uniprot_variation_search</span><span class="op">(</span>disease <span class="op">=</span> <span class="st">"alzheimer"</span>,
    wild_type <span class="op">=</span> <span class="st">"A"</span>, alternative_sequence <span class="op">=</span> <span class="st">"T"</span><span class="op">)</span>
<span class="co"># }</span>

</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

