<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>2.D: PANTHER &amp; rbioapi • rbioapi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2.D: PANTHER &amp; rbioapi">
<meta name="description" content="Connect to PANTHER in R with rbioapi package.
">
<meta property="og:description" content="Connect to PANTHER in R with rbioapi package.
">
<meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/rbioapi.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles-supported-services" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles &amp; Supported Services</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles-supported-services">
<li><a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://moosa-r.com/" aria-label="Moosa Rezwani Portfolio Website"><span class="fa fa-solid fa-id-card"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="rbioapi_panther_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="rbioapi_panther_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="rbioapi_panther_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_panther_files/datatables-binding-0.33/datatables.js"></script><link href="rbioapi_panther_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_panther_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_panther_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="rbioapi_panther_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="rbioapi_panther_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>2.D: PANTHER &amp; rbioapi</h1>
                        <h4 data-toc-skip class="author">Moosa
Rezwani</h4>
            
            <h4 data-toc-skip class="date">2024-12-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/master/vignettes/rbioapi_panther.Rmd" class="external-link"><code>vignettes/rbioapi_panther.Rmd</code></a></small>
      <div class="d-none name"><code>rbioapi_panther.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2 heading2">
<h2 class="heading2" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Directly quoting the the paper published by <a href="https://www.pantherdb.org" title="Protein Analysis THrough Evolutionary Relationships (PANTHER)" class="external-link">PANTHER</a>
(Protein Analysis THrough Evolutionary Relationships) authors:</p>
<blockquote>
<p>The PANTHER classification system (<a href="https://www.pantherdb.org" class="external-link uri">https://www.pantherdb.org</a>) a comprehensive system that
combines genomes, gene function , pathways and statistical analysis
tools to enable to analyze large-scale genome-wide experimental data.
The system (PANTHER v.14.0) covers 131 complete genomes organized gene
families and subfamilies; evolutionary relationships between are
represented in phylogenetic trees, multiple sequence and statistical
models (hidden Markov models (HMMs)). The families and subfamilies are
annotated with Gene Ontology (GO) terms, sequences are assigned to
PANTHER pathways. A suite of tools has built to allow users to browse
and query gene functions and analyze-scale experimental data with a
number of statistical tests. is widely used by bench scientists,
bioinformaticians, computer and systems biologists.</p>
<p>(source: Mi, Huaiyu, et al. “Protocol Update for large-scale genome
and gene function analysis with the PANTHER classification system (v.
14.0).” <em>Nature protocols</em> 14.3 (2019): 703-721)</p>
</blockquote>
<p>The available tools in PANTHER’s <strong>RESTful API
services</strong> can be divided into 3 broad categories: Mapping genes,
retrieving information, and research tools. Herein, we provide a very
short introduction; you can always check functions’ manuals for detailed
guides and examples.</p>
<hr>
</div>
<div class="section level2 heading2">
<h2 class="heading2" id="map-genes">Map genes<a class="anchor" aria-label="anchor" href="#map-genes"></a>
</h2>
<ul>
<li><p><code><a href="../reference/rba_panther_mapping.html">rba_panther_mapping()</a></code>: map your gene-set to PANTHER
database and retrieve attributes and annotations associated with your
genes</p></li>
<li><p><code><a href="../reference/rba_panther_ortholog.html">rba_panther_ortholog()</a></code>: Retrieve Orthologs of your
genes</p></li>
<li><p><code><a href="../reference/rba_panther_homolog.html">rba_panther_homolog()</a></code>: Retrieve Homologs of your
genes</p></li>
</ul>
<hr>
</div>
<div class="section level2 heading2">
<h2 class="heading2" id="get-information">Get information<a class="anchor" aria-label="anchor" href="#get-information"></a>
</h2>
<ul>
<li><p><code>rba_panther_info</code>(): Retrieve a list of PANTHER’s
supported organisms, datasets, families, or pathways</p></li>
<li><p><code>rba_panther_family</code>(): Retrieve Orthologs, MSA, or
Tree topology of a given PANTHER family.</p></li>
</ul>
<hr>
</div>
<div class="section level2 heading2">
<h2 class="heading2" id="research-tools">Research tools<a class="anchor" aria-label="anchor" href="#research-tools"></a>
</h2>
<div class="section level3 heading3">
<h3 class="heading3" id="gene-list-analysis">Gene List Analysis<a class="anchor" aria-label="anchor" href="#gene-list-analysis"></a>
</h3>
<p><code><a href="../reference/rba_panther_enrich.html">rba_panther_enrich()</a></code> is an equivalent to <a href="https://www.pantherdb.org/index.jsp" title="PANTHER Gene List Analysis" class="external-link">Gene List analysis tool’s
webpage.</a>. Here is a usage example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1 We get the available annotation datasets in PANTHER (we need to select one of them to submit an enrichment request)</span></span>
<span><span class="va">annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_panther_info.html">rba_panther_info</a></span><span class="op">(</span>what <span class="op">=</span> <span class="st">"datasets"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Retrieving available annotation datasets.</span></span>
<span><span class="co"># Note that you should enter the "id" of the datasets, not its label (e.g. entering "biological_process" is incorrect, you should rather enter "GO:0008150").</span></span>
<span><span class="co">## 2 We create a variable with our genes' IDs</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>,<span class="st">"CDK1"</span>,<span class="st">"KIF23"</span>,<span class="st">"PLK1"</span>,</span>
<span>           <span class="st">"RAC2"</span>,<span class="st">"RACGAP1"</span>,<span class="st">"RHOA"</span>,<span class="st">"RHOB"</span>, <span class="st">"PHF14"</span>, <span class="st">"RBM3"</span>, <span class="st">"MSL1"</span><span class="op">)</span></span>
<span><span class="co">## 3 Now we can submit the enrichment request.</span></span>
<span><span class="va">enriched</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_panther_enrich.html">rba_panther_enrich</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>                               organism <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>                               annot_dataset <span class="op">=</span> <span class="st">"ANNOT_TYPE_ID_PANTHER_PATHWAY"</span>,</span>
<span>                               cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="co">#&gt; Performing over-representation enrichment analysis of 15 input genes of organism 9606 against ANNOT_TYPE_ID_PANTHER_PATHWAY datasets.</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14"],[4,3,4,3,3,2,3,7,2,2,2,3,2,2],[75.17808219178082,164.64,28.58333333333333,49.59036144578313,46.77272727272727,124.7272727272727,28.19178082191781,0.5335555555555556,78.39999999999999,66.92682926829268,54.88,15.70992366412214,32.28235294117647,22.30894308943089],[3.08873635002442e-05,5.708570624979337e-05,0.0004932036099763925,0.001111655684318716,0.00105982213006343,0.003028809791268619,0.00341709084592112,0.003423709368731017,0.005172594475969506,0.006399569898821666,0.008658346912849519,0.01104906209473271,0.02103876340142172,0.04040751310457476],[0.05320699708454811,0.01822157434402332,0.1399416909620991,0.06049562682215744,0.0641399416909621,0.01603498542274052,0.1064139941690962,13.11953352769679,0.02551020408163265,0.02988338192419825,0.03644314868804665,0.1909620991253644,0.0619533527696793,0.08965014577259475],[73,25,192,83,88,22,146,18000,35,41,50,262,85,123],[1.930460218765263e-07,7.135713281224172e-07,9.247567687057361e-06,2.77913921079679e-05,3.311944156448219e-05,0.0001135803671725732,0.000149497724509049,0.0001711854684365508,0.0002909584392732847,0.0003999731186763542,0.0005952613502584045,0.0008286796571049536,0.001709399526365515,0.003535657396650291],["+","+","+","+","+","+","+","-","+","+","+","+","+","+"],["P04393","P00008","P00034","P00016","P00059","P00007","P00029",null,"P00009","P05918","P04398","P00031","P00053","P00027"],["Ras Pathway","Axon guidance mediated by Slit/Robo","Integrin signalling pathway","Cytoskeletal regulation by Rho GTPase","p53 pathway","Axon guidance mediated by semaphorins","Huntington disease","UNCLASSIFIED","Axon guidance mediated by netrin","p38 MAPK pathway","p53 pathway feedback loops 2","Inflammation mediated by chemokine and cytokine signaling pathway","T cell activation","Heterotrimeric G-protein signaling pathway-Gq alpha and Go alpha mediated pathway"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>number_in_list<\/th>\n      <th>fold_enrichment<\/th>\n      <th>fdr<\/th>\n      <th>expected<\/th>\n      <th>number_in_reference<\/th>\n      <th>pValue<\/th>\n      <th>plus_minus<\/th>\n      <th>term.id<\/th>\n      <th>term.label<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"number_in_list","targets":1},{"name":"fold_enrichment","targets":2},{"name":"fdr","targets":3},{"name":"expected","targets":4},{"name":"number_in_reference","targets":5},{"name":"pValue","targets":6},{"name":"plus_minus","targets":7},{"name":"term.id","targets":8},{"name":"term.label","targets":9}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p><strong>Please Note:</strong> Other services supported by rbioapi
also provide Over-representation analysis tools. Please see the vignette
article <a href="rbioapi_do_enrich.html">Do with rbioapi:
Over-Representation (Enrichment) Analysis in R</a> (<a href="https://rbioapi.moosa-r.com/articles/rbioapi_do_enrich.html">link
to the documentation site</a>) for an in-depth review.</p>
</div>
<div class="section level3 heading3">
<h3 class="heading3" id="tree-grafter">Tree grafter<a class="anchor" aria-label="anchor" href="#tree-grafter"></a>
</h3>
<p><code><a href="../reference/rba_panther_tree_grafter.html">rba_panther_tree_grafter()</a></code> is an equivalent to the “<a href="https://www.pantherdb.org/tools/sequenceSearchForm.jsp" class="external-link">Graft
sequence into PANTHER library of trees</a>” tool.</p>
<hr>
</div>
</div>
<div class="section level2 heading2">
<h2 class="heading2" id="citations">How to Cite?<a class="anchor" aria-label="anchor" href="#citations"></a>
</h2>
<p>To cite PANTHER (Please see <a href="https://www.pantherdb.org/publications.jsp#HowToCitePANTHER" class="external-link uri">https://www.pantherdb.org/publications.jsp#HowToCitePANTHER</a>):</p>
<ul>
<li>Huaiyu Mi, Dustin Ebert, Anushya Muruganujan, Caitlin Mills,
Laurent-Philippe Albou, Tremayne Mushayamaha, Paul D Thomas, PANTHER
version 16: a revised family classification, tree-based classification
tool, enhancer regions and extensive API, <em>Nucleic Acids
Research</em>, Volume 49, Issue D1, 8 January 2021, Pages D394–D403, <a href="https://doi.org/10.1093/nar/gkaa1106" class="external-link uri">https://doi.org/10.1093/nar/gkaa1106</a>
</li>
</ul>
<p>To cite rbioapi:</p>
<ul>
<li>Moosa Rezwani, Ali Akbar Pourfathollah, Farshid Noorbakhsh, rbioapi:
user-friendly R interface to biologic web services’ API, Bioinformatics,
Volume 38, Issue 10, 15 May 2022, Pages 2952–2953, <a href="https://doi.org/10.1093/bioinformatics/btac172" class="external-link uri">https://doi.org/10.1093/bioinformatics/btac172</a>
</li>
</ul>
<hr>
</div>
<div class="section level2 heading2">
<h2 class="heading2" id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_panther.html" title="2.C: PANTHER &amp; rbioapi">This article in rbioapi
documentation site</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html#section-enrichr-rba-enrichr-" title="rbioapi reference">Functions references in rbioapi
documentation site</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi
vignette index</a></p></li>
</ul>
<hr>
</div>
<div class="section level2 heading2">
<h2 class="heading2" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] rbioapi_0.8.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] httr_1.4.7        cli_3.6.3         knitr_1.49        rlang_1.1.4      </span></span>
<span><span class="co">#&gt;  [5] xfun_0.49         textshaping_0.4.1 jsonlite_1.8.9    DT_0.33          </span></span>
<span><span class="co">#&gt;  [9] htmltools_0.5.8.1 ragg_1.3.3        sass_0.4.9        rmarkdown_2.29   </span></span>
<span><span class="co">#&gt; [13] crosstalk_1.2.1   evaluate_1.0.1    jquerylib_0.1.4   fastmap_1.2.0    </span></span>
<span><span class="co">#&gt; [17] yaml_2.3.10       lifecycle_1.0.4   compiler_4.4.2    fs_1.6.5         </span></span>
<span><span class="co">#&gt; [21] htmlwidgets_1.6.4 systemfonts_1.1.0 digest_0.6.37     R6_2.5.1         </span></span>
<span><span class="co">#&gt; [25] curl_6.0.1        magrittr_2.0.3    bslib_0.8.0       tools_4.4.2      </span></span>
<span><span class="co">#&gt; [29] pkgdown_2.1.1     cachem_1.1.0      desc_1.4.3</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://moosa-r.com/" class="external-link">Moosa Rezwani</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
