<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rbioapi">
<title>2.A: Enrichr &amp; rbioapi • rbioapi</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2.A: Enrichr &amp; rbioapi">
<meta property="og:description" content="rbioapi">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/rbioapi.html">Get Started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-supported-services-amp-articles">Supported Services &amp; Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-supported-services-amp-articles">
    <a class="dropdown-item" href="../articles/rbioapi.html">1: rbioapi: Get Started</a>
    <a class="dropdown-item" href="../articles/rbioapi_enrichr.html">2.A: Enrichr &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_mieaa.html">2.B: miEEA &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_panther.html">2.C: PANTHER &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_reactome.html">2.D: Reactome &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_string.html">2.E: STRING &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_uniprot.html">2.F: UniProt &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_jaspar.html">2.G: JASPAR &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">3 Do with rbioapi: 1. Over-Representation (Enrichment) Analysis</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="rbioapi_enrichr_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="rbioapi_enrichr_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="rbioapi_enrichr_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_enrichr_files/datatables-binding-0.20/datatables.js"></script><link href="rbioapi_enrichr_files/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_enrichr_files/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_enrichr_files/dt-core-1.11.3/js/jquery.dataTables.min.js"></script><link href="rbioapi_enrichr_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="rbioapi_enrichr_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>2.A: Enrichr &amp; rbioapi</h1>
                        <h4 data-toc-skip class="author">Moosa Rezwani</h4>
            
            <h4 data-toc-skip class="date">2022-01-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/HEAD/vignettes/rbioapi_enrichr.Rmd" class="external-link"><code>vignettes/rbioapi_enrichr.Rmd</code></a></small>
      <div class="d-none name"><code>rbioapi_enrichr.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><a href="https://maayanlab.cloud/Enrichr" class="external-link">Enrichr</a> is a gene-set enrichment analysis tool developed in the <a href="https://icahn.mssm.edu/research/labs/maayan-laboratory" class="external-link">Ma’ayan lab</a>.</p>
</div>
<div class="section level2">
<h2 id="gene-set-library-concept-in-enrichr">Gene set library concept in Enrichr<a class="anchor" aria-label="anchor" href="#gene-set-library-concept-in-enrichr"></a>
</h2>
<p>Directly quoting from Enrichr’s help page;</p>
<blockquote>
<p>A <em>gene set library</em> is a set of related gene sets or enrichment terms. Each enrichment term in Enrichr’s results pages is organized by its gene set library. These libraries have been constructed from many sources such as published studies and major biological and biomedical online databases. Others have been created for and only available through Enrichr. For example, the ChEA 2015 library is a set of functional terms representing transcription factors profiled by ChIP-seq in mammalian cells. Each term is associated with a collection of putative targets inferred from the peaks identified in each ChIP-seq study.</p>
<p>(source: <a href="https://maayanlab.cloud/Enrichr/help#background" class="external-link uri">https://maayanlab.cloud/Enrichr/help#background</a>)</p>
</blockquote>
<p>To get a list of available Enrichr libraries, use:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">enrichr_libs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_libs.html">rba_enrichr_libs</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>In the returned data frame, you can find the names of available Enrichr libraries in “libraryName” column:</p>
<div id="htmlwidget-476e97b4cb9b3697abc5" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-476e97b4cb9b3697abc5">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192"],[13362,27884,6002,47172,47107,21493,1295,3185,2854,15057,4128,34061,7504,16399,12753,23726,32740,13373,19270,13236,14264,3096,22288,4533,10231,2741,5655,10406,10493,11251,8695,1759,2178,851,10061,11250,15406,17711,17576,15797,12232,13572,6454,3723,7588,7682,7324,8469,13121,26382,29065,280,13877,15852,4320,4271,10496,1678,756,3800,2541,1918,5863,6768,25651,19129,23939,23561,23877,15886,24350,3102,31132,30832,48230,5613,9559,9448,16725,19249,15090,16129,15309,15103,15022,15676,15854,15015,3788,3357,12668,12638,8973,7010,5966,15562,17850,17660,1348,934,2541,2041,5209,49238,2243,19586,22440,8184,18329,15755,10271,10427,10601,13822,8002,10089,13247,21809,23601,20883,19612,25983,19500,14893,17598,5902,12486,1073,19513,14433,8655,11459,19741,27360,13072,13464,13787,13929,16964,17258,10352,10471,12419,19378,6201,4558,3264,7802,8551,12444,9000,7744,6204,13420,14148,9813,1397,9116,17464,394,11851,8189,18704,5605,5718,14156,16979,4383,54974,12118,12361,9763,8078,7173,5833,14937,11497,11936,9767,14167,17851,16853,6654,1683,20414,26076,26338,25381,25409,11980,31158,30006],[275,1284,77,1370,509,3713,18,73,34,300,48,641,155,247,57,127,85,258,388,82,58,31,4368,37,158,5,342,715,200,100,100,25,89,15,106,166,300,300,300,176,343,301,301,47,35,78,172,122,305,1811,2123,9,304,912,129,128,201,21,12,48,39,39,51,47,807,1594,293,307,302,9,299,25,298,302,1429,36,73,63,1443,1443,282,292,308,318,290,310,279,321,159,153,300,300,64,87,51,887,300,300,19,13,39,42,300,1550,19,545,505,24,161,28,22,38,25,21,143,45,49,2316,2395,299,299,299,137,128,1208,109,299,100,117,36,61,39,270,802,26,45,200,200,200,200,58,76,491,37,45,54,22,92,98,23,20,363,387,32,122,49,13,22,63,73,586,421,100,39,31,40,295,146,483,448,124,139,102,43,100,33,80,34,33,80,102,360,136,10,112,250,250,250,250,250,805,815],["Genome_Browser_PWMs","TRANSFAC_and_JASPAR_PWMs","Transcription_Factor_PPIs","ChEA_2013","Drug_Perturbations_from_GEO_2014","ENCODE_TF_ChIP-seq_2014","BioCarta_2013","Reactome_2013","WikiPathways_2013","Disease_Signatures_from_GEO_up_2014","KEGG_2013","TF-LOF_Expression_from_GEO","TargetScan_microRNA","PPI_Hub_Proteins","GO_Molecular_Function_2015","GeneSigDB","Chromosome_Location","Human_Gene_Atlas","Mouse_Gene_Atlas","GO_Cellular_Component_2015","GO_Biological_Process_2015","Human_Phenotype_Ontology","Epigenomics_Roadmap_HM_ChIP-seq","KEA_2013","NURSA_Human_Endogenous_Complexome","CORUM","SILAC_Phosphoproteomics","MGI_Mammalian_Phenotype_Level_3","MGI_Mammalian_Phenotype_Level_4","Old_CMAP_up","Old_CMAP_down","OMIM_Disease","OMIM_Expanded","VirusMINT","MSigDB_Computational","MSigDB_Oncogenic_Signatures","Disease_Signatures_from_GEO_down_2014","Virus_Perturbations_from_GEO_up","Virus_Perturbations_from_GEO_down","Cancer_Cell_Line_Encyclopedia","NCI-60_Cancer_Cell_Lines","Tissue_Protein_Expression_from_ProteomicsDB","Tissue_Protein_Expression_from_Human_Proteome_Map","HMDB_Metabolites","Pfam_InterPro_Domains","GO_Biological_Process_2013","GO_Cellular_Component_2013","GO_Molecular_Function_2013","Allen_Brain_Atlas_up","ENCODE_TF_ChIP-seq_2015","ENCODE_Histone_Modifications_2015","Phosphatase_Substrates_from_DEPOD","Allen_Brain_Atlas_down","ENCODE_Histone_Modifications_2013","Achilles_fitness_increase","Achilles_fitness_decrease","MGI_Mammalian_Phenotype_2013","BioCarta_2015","HumanCyc_2015","KEGG_2015","NCI-Nature_2015","Panther_2015","WikiPathways_2015","Reactome_2015","ESCAPE","HomoloGene","Disease_Perturbations_from_GEO_down","Disease_Perturbations_from_GEO_up","Drug_Perturbations_from_GEO_down","Genes_Associated_with_NIH_Grants","Drug_Perturbations_from_GEO_up","KEA_2015","Gene_Perturbations_from_GEO_up","Gene_Perturbations_from_GEO_down","ChEA_2015","dbGaP","LINCS_L1000_Chem_Pert_up","LINCS_L1000_Chem_Pert_down","GTEx_Tissue_Expression_Down","GTEx_Tissue_Expression_Up","Ligand_Perturbations_from_GEO_down","Aging_Perturbations_from_GEO_down","Aging_Perturbations_from_GEO_up","Ligand_Perturbations_from_GEO_up","MCF7_Perturbations_from_GEO_down","MCF7_Perturbations_from_GEO_up","Microbe_Perturbations_from_GEO_down","Microbe_Perturbations_from_GEO_up","LINCS_L1000_Ligand_Perturbations_down","LINCS_L1000_Ligand_Perturbations_up","L1000_Kinase_and_GPCR_Perturbations_down","L1000_Kinase_and_GPCR_Perturbations_up","Reactome_2016","KEGG_2016","WikiPathways_2016","ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X","Kinase_Perturbations_from_GEO_down","Kinase_Perturbations_from_GEO_up","BioCarta_2016","HumanCyc_2016","NCI-Nature_2016","Panther_2016","DrugMatrix","ChEA_2016","huMAP","Jensen_TISSUES","RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO","MGI_Mammalian_Phenotype_2017","Jensen_COMPARTMENTS","Jensen_DISEASES","BioPlex_2017","GO_Cellular_Component_2017","GO_Molecular_Function_2017","GO_Biological_Process_2017","GO_Cellular_Component_2017b","GO_Molecular_Function_2017b","GO_Biological_Process_2017b","ARCHS4_Tissues","ARCHS4_Cell-lines","ARCHS4_IDG_Coexp","ARCHS4_Kinases_Coexp","ARCHS4_TFs_Coexp","SysMyo_Muscle_Gene_Sets","miRTarBase_2017","TargetScan_microRNA_2017","Enrichr_Libraries_Most_Popular_Genes","Enrichr_Submissions_TF-Gene_Coocurrence","Data_Acquisition_Method_Most_Popular_Genes","DSigDB","GO_Biological_Process_2018","GO_Cellular_Component_2018","GO_Molecular_Function_2018","TF_Perturbations_Followed_by_Expression","Chromosome_Location_hg19","NIH_Funded_PIs_2017_Human_GeneRIF","NIH_Funded_PIs_2017_Human_AutoRIF","Rare_Diseases_AutoRIF_ARCHS4_Predictions","Rare_Diseases_GeneRIF_ARCHS4_Predictions","NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions","NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions","Rare_Diseases_GeneRIF_Gene_Lists","Rare_Diseases_AutoRIF_Gene_Lists","SubCell_BarCode","GWAS_Catalog_2019","WikiPathways_2019_Human","WikiPathways_2019_Mouse","TRRUST_Transcription_Factors_2019","KEGG_2019_Human","KEGG_2019_Mouse","InterPro_Domains_2019","Pfam_Domains_2019","DepMap_WG_CRISPR_Screens_Broad_CellLines_2019","DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019","MGI_Mammalian_Phenotype_Level_4_2019","UK_Biobank_GWAS_v1","BioPlanet_2019","ClinVar_2019","PheWeb_2019","DisGeNET","HMS_LINCS_KinomeScan","CCLE_Proteomics_2020","ProteomicsDB_2020","lncHUB_lncRNA_Co-Expression","Virus-Host_PPI_P-HIPSTer_2020","Elsevier_Pathway_Collection","Table_Mining_of_CRISPR_Studies","COVID-19_Related_Gene_Sets","MSigDB_Hallmark_2020","Enrichr_Users_Contributed_Lists_2020","TG_GATES_2020","Allen_Brain_Atlas_10x_scRNA_2021","Descartes_Cell_Types_and_Tissue_2021","KEGG_2021_Human","WikiPathway_2021_Human","HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression","GO_Biological_Process_2021","GO_Cellular_Component_2021","GO_Molecular_Function_2021","MGI_Mammalian_Phenotype_Level_4_2021","CellMarker_Augmented_2021","Orphanet_Augmented_2021","COVID-19_Related_Gene_Sets_2021","PanglaoDB_Augmented_2021","Azimuth_Cell_Types_2021","PhenGenI_Association_2021","RNAseq_Automatic_GEO_Signatures_Human_Down","RNAseq_Automatic_GEO_Signatures_Human_Up","RNAseq_Automatic_GEO_Signatures_Mouse_Down","RNAseq_Automatic_GEO_Signatures_Mouse_Up","GTEx_Aging_Signatures_2021","HDSigDB_Human_2021","HDSigDB_Mouse_2021"],["http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/","http://jaspar.genereg.net/html/DOWNLOAD/","","http://amp.pharm.mssm.edu/lib/cheadownload.jsp","http://www.ncbi.nlm.nih.gov/geo/","http://genome.ucsc.edu/ENCODE/downloads.html","https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://www.reactome.org/download/index.html","http://www.wikipathways.org/index.php/Download_Pathways","http://www.ncbi.nlm.nih.gov/geo/","http://www.kegg.jp/kegg/download/","http://www.ncbi.nlm.nih.gov/geo/","http://www.targetscan.org/cgi-bin/targetscan/data_download.cgi?db=vert_61","http://amp.pharm.mssm.edu/X2K","http://www.geneontology.org/GO.downloads.annotations.shtml","https://pubmed.ncbi.nlm.nih.gov/22110038/","http://software.broadinstitute.org/gsea/msigdb/index.jsp","http://biogps.org/downloads/","http://biogps.org/downloads/","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.human-phenotype-ontology.org/","http://www.roadmapepigenomics.org/","http://amp.pharm.mssm.edu/lib/keacommandline.jsp","https://www.nursa.org/nursa/index.jsf","http://mips.helmholtz-muenchen.de/genre/proj/corum/","http://amp.pharm.mssm.edu/lib/keacommandline.jsp","http://www.informatics.jax.org/","http://www.informatics.jax.org/","http://www.broadinstitute.org/cmap/","http://www.broadinstitute.org/cmap/","http://www.omim.org/downloads","http://www.omim.org/downloads","http://mint.bio.uniroma2.it/download.html","http://www.broadinstitute.org/gsea/msigdb/collections.jsp","http://www.broadinstitute.org/gsea/msigdb/collections.jsp","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://portals.broadinstitute.org/ccle/home\n","http://biogps.org/downloads/","https://www.proteomicsdb.org/","http://www.humanproteomemap.org/index.php","http://www.hmdb.ca/downloads","ftp://ftp.ebi.ac.uk/pub/databases/interpro/","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.brain-map.org/","http://genome.ucsc.edu/ENCODE/downloads.html","http://genome.ucsc.edu/ENCODE/downloads.html","http://www.koehn.embl.de/depod/","http://www.brain-map.org/","http://genome.ucsc.edu/ENCODE/downloads.html","http://www.broadinstitute.org/achilles","http://www.broadinstitute.org/achilles","http://www.informatics.jax.org/","https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://humancyc.org/","http://www.kegg.jp/kegg/download/","http://pid.nci.nih.gov/","http://www.pantherdb.org/","http://www.wikipathways.org/index.php/Download_Pathways","http://www.reactome.org/download/index.html","http://www.maayanlab.net/ESCAPE/","http://www.ncbi.nlm.nih.gov/homologene","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://grants.nih.gov/grants/oer.htm\n","http://www.ncbi.nlm.nih.gov/geo/","http://amp.pharm.mssm.edu/Enrichr","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://amp.pharm.mssm.edu/Enrichr","http://www.ncbi.nlm.nih.gov/gap","https://clue.io/","https://clue.io/","http://www.gtexportal.org/","http://www.gtexportal.org/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://clue.io/","https://clue.io/","https://clue.io/","https://clue.io/","http://www.reactome.org/download/index.html","http://www.kegg.jp/kegg/download/","http://www.wikipathways.org/index.php/Download_Pathways","","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://humancyc.org/","http://pid.nci.nih.gov/","http://www.pantherdb.org/pathway/","https://ntp.niehs.nih.gov/drugmatrix/","http://amp.pharm.mssm.edu/Enrichr","http://proteincomplexes.org/","http://tissues.jensenlab.org/","http://www.ncbi.nlm.nih.gov/geo/","http://www.informatics.jax.org/","http://compartments.jensenlab.org/","http://diseases.jensenlab.org/","http://bioplex.hms.harvard.edu/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://sys-myo.rhcloud.com/","http://mirtarbase.mbc.nctu.edu.tw/","http://www.targetscan.org/","http://amp.pharm.mssm.edu/Enrichr","http://amp.pharm.mssm.edu/Enrichr","http://amp.pharm.mssm.edu/Enrichr","http://tanlab.ucdenver.edu/DSigDB/DSigDBv1.0/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.ncbi.nlm.nih.gov/geo/","http://hgdownload.cse.ucsc.edu/downloads.html","https://www.ncbi.nlm.nih.gov/pubmed/","https://www.ncbi.nlm.nih.gov/pubmed/","https://amp.pharm.mssm.edu/geneshot/","https://www.ncbi.nlm.nih.gov/gene/about-generif","https://www.ncbi.nlm.nih.gov/pubmed/","https://www.ncbi.nlm.nih.gov/pubmed/","https://www.ncbi.nlm.nih.gov/gene/about-generif","https://amp.pharm.mssm.edu/geneshot/","http://www.subcellbarcode.org/","https://www.ebi.ac.uk/gwas","https://www.wikipathways.org/","https://www.wikipathways.org/","https://www.grnpedia.org/trrust/","https://www.kegg.jp/","https://www.kegg.jp/","https://www.ebi.ac.uk/interpro/","https://pfam.xfam.org/","https://depmap.org/","https://depmap.org/","http://www.informatics.jax.org/","https://www.ukbiobank.ac.uk/tag/gwas/","https://tripod.nih.gov/bioplanet/","https://www.ncbi.nlm.nih.gov/clinvar/","http://pheweb.sph.umich.edu/","https://www.disgenet.org","http://lincs.hms.harvard.edu/kinomescan/","https://portals.broadinstitute.org/ccle","https://www.proteomicsdb.org/","https://amp.pharm.mssm.edu/lnchub/","http://phipster.org/","http://www.transgene.ru/disease-pathways/","","https://amp.pharm.mssm.edu/covid19","https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp ","https://maayanlab.cloud/Enrichr","https://toxico.nibiohn.go.jp/english/","https://portal.brain-map.org/","https://descartes.brotmanbaty.org/bbi/human-gene-expression-during-development/","https://www.kegg.jp/","https://www.wikipathways.org/","https://hubmapconsortium.github.io/ccf-asct-reporter/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.informatics.jax.org/","http://biocc.hrbmu.edu.cn/CellMarker/","http://www.orphadata.org/","https://maayanlab.cloud/covid19/","https://panglaodb.se/","https://azimuth.hubmapconsortium.org/","https://www.ncbi.nlm.nih.gov/gap/phegeni","https://maayanlab.cloud/archs4/","https://maayanlab.cloud/archs4/","https://maayanlab.cloud/archs4/","https://maayanlab.cloud/archs4/","https://gtexportal.org/","https://www.hdinhd.org/","https://www.hdinhd.org/"],[615,326,290,353,701,498,249,78,199,142,200,269,222,385,1136,2139,386,84,96,641,5192,1779,383,474,1796,1658,84,71,476,6100,6100,90,187,85,858,189,142,323,323,967,93,207,30,3906,311,941,205,402,2192,816,412,59,2192,109,216,216,476,239,125,179,209,104,404,1389,315,12,839,839,906,32876,906,428,2460,2460,395,345,33132,33132,2918,2918,261,286,286,261,401,401,312,312,96,96,3644,3644,1530,293,437,104,285,285,237,152,209,112,7876,645,995,1842,1302,5231,2283,1811,3915,636,972,3166,816,3271,10125,108,125,352,498,1724,1135,3240,683,121,1722,12,4026,5103,446,1151,1958,36,5687,12558,3725,2244,12558,5684,2244,3725,104,1737,472,176,571,308,303,1071,608,558,325,5261,857,1510,182,1161,9828,148,378,913,3729,6715,1721,802,205,50,1482,1190,766,172,320,622,344,6036,511,1274,4601,1097,3774,478,178,341,950,4269,4269,4216,4216,270,2564,2579],["ea115789fcbf12797fd692cec6df0ab4dbc79c6a","7d42eb43a64a4e3b20d721fc7148f685b53b6b30","849f222220618e2599d925b6b51868cf1dab3763","7ebe772afb55b63b41b79dd8d06ea0fdd9fa2630","ad270a6876534b7cb063e004289dcd4d3164f342","497787ebc418d308045efb63b8586f10c526af51","4a293326037a5229aedb1ad7b2867283573d8bcd","b343994a1b68483b0122b08650201c9b313d5c66","5c307674c8b97e098f8399c92f451c0ff21cbf68","248c4ed8ea28352795190214713c86a39fd7afab","eb26f55d3904cb0ea471998b6a932a9bf65d8e50","","f4029bf6a62c91ab29401348e51df23b8c44c90f","69c0cfe07d86f230a7ef01b365abcc7f6e52f138","f531ac2b6acdf7587a54b79b465a5f4aab8f00f9","6d655e0aa3408a7accb3311fbda9b108681a8486","8dab0f96078977223646ff63eb6187e0813f1433","0741451470203d7c40a06274442f25f74b345c9c","31191bfadded5f96983f93b2a113cf2110ff5ddb","e1d004d5797cbd2363ef54b1c3b361adb68795c6","bf120b6e11242b1a64c80910d8e89f87e618e235","17a138b0b70aa0e143fe63c14f82afb70bc3ed0a","e1bc8a398e9b21f9675fb11bef18087eda21b1bf","462045609440fa1e628a75716b81a1baa5bd9145","7d3566b12ebc23dd23d9ca9bb97650f826377b16","d047f6ead7831b00566d5da7a3b027ed9196e104","54dcd9438b33301deb219866e162b0f9da7e63a0","c3bfc90796cfca8f60cba830642a728e23a53565","0b09a9a1aa0af4fc7ea22d34a9ae644d45864bd6","9041f90cccbc18479138330228b336265e09021c","ebc0d905b3b3142f936d400c5f2a4ff926c81c37","cb2b92578a91e023d0498a334923ee84add34eca","27eca242904d8e12a38cf8881395bc50d57a03e1","5abad1fc36216222b0420cadcd9be805a0dda63e","e4cdcc7e259788fdf9b25586cce3403255637064","c76f5319c33c4833c71db86a30d7e33cd63ff8cf","aabdf7017ae55ae75a004270924bcd336653b986","45268b7fc680d05dd9a29743c2f2b2840a7620bf","5f531580ccd168ee4acc18b02c6bdf8200e19d08","eb38dbc3fb20adafa9d6f9f0b0e36f378e75284f","75c81676d8d6d99d262c9660edc024b78cfb07c9","","49351dc989f9e6ca97c55f8aca7778aa3bfb84b9","1905132115d22e4119bce543bdacaab074edb363","e2b4912cfb799b70d87977808c54501544e4cdc9","5216d1ade194ffa5a6c00f105e2b1899f64f45fe","fd1332a42395e0bc1dba82868b39be7983a48cc5","7e3e99e5aae02437f80b0697b197113ce3209ab0","3804715a63a308570e47aa1a7877f01147ca6202","56b6adb4dc8a2f540357ef992d6cd93dfa2907e5","55b56cd8cf2ff04b26a09b9f92904008b82f3a6f","d40701e21092b999f4720d1d2b644dd0257b6259","ea67371adec290599ddf484ced2658cfae259304","c209ae527bc8e98e4ccd27a668d36cd2c80b35b4","98366496a75f163164106e72439fb2bf2f77de4e","83a710c1ff67fd6b8af0d80fa6148c40dbd9bc64","a4c6e217a81a4a58ff5a1c9fc102b70beab298e9","70e4eb538daa7688691acfe5d9c3c19022be832b","711f0c02b23f5e02a01207174943cfeee9d3ea9c","e80d25c56de53c704791ddfdc6ab5eec28ae7243","47edfc012bcbb368a10b717d8dca103f7814b5a4","ab824aeeff0712bab61f372e43aebb870d1677a9","1f7eea2f339f37856522c1f1c70ec74c7b25325f","36e541bee015eddb8d53827579549e30fe7a3286","a7acc741440264717ff77751a7e5fed723307835","663b665b75a804ef98add689f838b68e612f0d2a","0f412e0802d76efa0374504c2c9f5e0624ff7f09","9ddc3902fb01fb9eaf1a2a7c2ff3acacbb48d37e","068623a05ecef3e4a5e0b4f8db64bb8faa3c897f","76fc5ec6735130e287e62bae6770a3c5ee068645","c9c2155b5ac81ac496854fa61ba566dcae06cc80","18a081774e6e0aaf60b1a4be7fd20afcf9e08399","53dedc29ce3100930d68e506f941ef59de05dc6b","499882af09c62dd6da545c15cb51c1dc5e234f78","712eb7b6edab04658df153605ec6079fa89fb5c7","010f1267055b1a1cb036e560ea525911c007a666","5e678b3debe8d8ea95187d0cd35c914017af5eb3","fedbf5e221f45ee60ebd944f92569b5eda7f2330","74b818bd299a9c42c1750ffe43616aa9f7929f02","103738763d89cae894bec9f145ac28167a90e611","1eb3c0426140340527155fd0ef67029db2a72191","cd95fe1b505ba6f28cd722cfba50fdea979d3b4c","74c4f0a0447777005b2a5c00c9882a56dfc62d7c","31baa39da2931ddd5f7aedf2d0bbba77d2ba7b46","555f68aef0a29a67b614a0d7e20b6303df9069c6","1bc2ba607f1ff0dda44e2a15f32a2c04767da18c","9e613dba78ef7e60676b13493a9dc49ccd3c8b3f","d0c3e2a68e8c611c669098df2c87b530cec3e132","957846cb05ef31fc8514120516b73cc65af7980e","3bd494146c98d8189898a947f5ef5710f1b7c4b2","1ccc5bce553e0c2279f8e3f4ddcfbabcf566623b","b54a0d4ba525eac4055c7314ca9d9312adcb220c","1f54638e8f45075fb79489f0e0ef906594cb0678","43f56da7540195ba3c94eb6e34c522a699b36da9","340be98b444cad50bb974df69018fd598e23e5e1","5426f7747965c23ef32cff46fabf906e2cd76bfa","bb9682d78b8fc43be842455e076166fcd02cefc3","78618915009cac3a0663d6f99d359e39a31b6660","13d9ab18921d5314a5b2b366f6142b78ab0ff6aa","d6a502ef9b4c789ed5e73ca5a8de372796e5c72a","3c1e1f7d1a651d9aaa198e73704030716fc09431","ca5f6abf7f75d9baae03396e84d07300bf1fd051","255c3db820d612f34310f22a6985dad50e9fe1fe","af271913344aa08e6a755af1d433ef15768d749a","249247d2f686d3eb4b9e4eb976c51159fac80a89","e8879ab9534794721614d78fe2883e9e564d7759","f0752e4d7f5198f86446678966b260c530d19d78","0705e59bff98deda6e9cbe00cfcdd871c85e7d04","56ec68c32d4e83edc2ee83bea0e9f6a3829b2279","3045dff8181367c1421627bb8e4c5a32c6d67f98","b8620b1a9d0d271d1a2747d8cfc63589dba39991","8fed21d22dfcc3015c05b31d942fdfc851cc8e04","b4018906e0a8b4e81a1b1afc51e0a2e7655403eb","d9da4dba4a3eb84d4a28a3835c06dfbbe5811f92","ecf39c41fa5bc7deb625a2b5761a708676e9db7c","8d8340361dd36a458f1f0a401f1a3141de1f3200","6404c38bffc2b3732de4e3fbe417b5043009fe34","4126374338235650ab158ba2c61cd2e2383b70df","5496ef9c9ae9429184d0b9485c23ba468ee522a8","ce60be284fdd5a9fc6240a355421a9e12b1ee84a","6721c5ed97b7772e4a19fdc3f797110df0164b75","8a468c3ae29fa68724f744cbef018f4f3b61c5ab","","6b7c7fe2a97b19aecbfba12d8644af6875ad99c4","79d13fb03d2fa6403f9be45c90eeda0f6822e269","e9b7d8ee237d0a690bd79d970a23a9fa849901ed","be2ca8ef5a8c8e17d7e7bd290e7cbfe0951396c0","17ce5192b9eba7d109b6d228772ea8ab222e01ef","287476538ab98337dbe727b3985a436feb6d192a","b5b77681c46ac58cd050e60bcd4ad5041a9ab0a9","e9ebe46188efacbe1056d82987ff1c70218fa7ae","79ff80ae9a69dd00796e52569e41422466fa0bee","34d08a4878c19584aaf180377f2ea96faa6a6eb1","fdab39c467ba6b0fb0288df1176d7dfddd7196d5","859b100fac3ca774ad84450b1fbb65a78fcc6b12","fc5bf033b932cf173633e783fc8c6228114211f8","375ff8cdd64275a916fa24707a67968a910329bb","0f7fb7f347534779ecc6c87498e96b5460a8d652","f77de51aaf0979dd6f56381cf67ba399b4640d28","25fa899b715cd6a9137f6656499f89cd25144029","0fb9ac92dbe52024661c088f71a1134f00567a8b","ee3adbac2da389959410260b280e7df1fd3730df","b50bb9480d8a77103fb75b331fd9dd927246939a","fef3864bcb5dd9e60cee27357eff30226116c49b","b0c9e9ebb9014f14561e896008087725a2db24b7","e7750958da20f585c8b6d5bc4451a5a4305514ba","5f8cf93e193d2bcefa5a37ccdf0eefac576861b0","3477bc578c4ea5d851dcb934fe2a41e9fd789bb4","187eb44b2d6fa154ebf628eba1f18537f64e797c","18dd5ec520fdf589a93d6a7911289c205e1ddf22","a6325ed264f9ac9e6518796076c46a1d885cca7a","0b08b32b20854ac8a738458728a9ea50c2e04800","b7c4ead26d0eb64f1697c030d31682b581c8bb56","f1bed632e89ebc054da44236c4815cdce03ef5ee","958fb52e6215626673a5acf6e9289a1b84d11b4a","e110851dfc763d30946f2abedcc2cd571ac357a0","0a95303f8059bec08836ecfe02ce3da951150547","6a7c7321b6b72c5285b722f7902d26a2611117cb","3c261626478ce9e6bf2c7f0a8014c5e901d43dc0","47ba06cdc92469ac79400fc57acd84ba343ba616","7094b097ae2301a1d6a5bd856a193b084cca993d","8c87c8346167bac2ba68195a32458aba9b1acfd1","45b597d7efa5693b7e4172b09c0ed2dda3305582","a592eed13e8e9496aedbab63003b965574e46a65","9196c760e3bcae9c9de1e3f87ad81f96bde24325","ad580f3864fa8ff69eaca11f6d2e7f9b86378d08","72b0346849570f66a77a6856722601e711596cb4","6952efda94663d4bd8db09bf6eeb4e67d21ef58c","8dc362703b38b30ac3b68b6401a9b20a58e7d3ef","9e32560437b11b4628b00ccf3d584360f7f7daee","46f8235cb585829331799a71aec3f7c082170219","","","","","","","","","","","","","","","","","","","","",""],[1,1,1,7,7,7,7,7,7,7,7,1,7,2,7,4,7,5,5,7,7,3,1,7,2,2,2,7,7,4,4,4,4,4,4,4,7,4,4,5,5,7,5,6,6,7,7,7,5,1,1,2,5,7,4,4,7,7,7,7,7,7,7,7,5,6,8,8,8,6,8,2,8,8,7,4,4,4,5,5,8,8,8,8,8,8,8,8,4,4,2,2,2,7,7,1,2,2,2,2,2,2,4,1,2,3,8,7,3,3,2,7,7,7,7,7,7,5,5,4,2,1,8,1,1,6,1,6,4,7,7,7,1,6,6,6,4,4,6,6,4,4,2,4,7,7,1,7,7,6,6,4,4,7,4,2,4,4,4,2,5,5,1,2,2,6,7,2,6,4,5,5,2,2,5,3,3,3,3,5,4,4,5,5,4,8,8,8,8,4,4,4]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>geneCoverage<\/th>\n      <th>genesPerTerm<\/th>\n      <th>libraryName<\/th>\n      <th>link<\/th>\n      <th>numTerms<\/th>\n      <th>appyter<\/th>\n      <th>categoryId<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[1,2,5,7]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div class="section level2">
<h2 id="enrichment-analysis-using-enrichr">Enrichment analysis using Enrichr<a class="anchor" aria-label="anchor" href="#enrichment-analysis-using-enrichr"></a>
</h2>
<p>To enrich your gene-set with Enrichr using rbioapi, you can take two approaches. we will begin with the simple one.</p>
<div class="section level3">
<h3 id="approach-1-using-the-wrapper-function">Approach 1: Using the Wrapper function<a class="anchor" aria-label="anchor" href="#approach-1-using-the-wrapper-function"></a>
</h3>
<p>Just fill the arguments of <code>rba_enrichr</code> according to the function’s manual; Simply supply your gene-set as a character vector and select the libraries.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 1 We create a variable with our genes' NCBI IDs</span>
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>,<span class="st">"CDK1"</span>,<span class="st">"KIF23"</span>,<span class="st">"PLK1"</span>,
           <span class="st">"RAC2"</span>,<span class="st">"RACGAP1"</span>,<span class="st">"RHOA"</span>,<span class="st">"RHOB"</span>, <span class="st">"PHF14"</span>, <span class="st">"RBM3"</span>, <span class="st">"MSL1"</span><span class="op">)</span>

<span class="co"># 2.a Enrich your genes using "MSigDB_Hallmark_2020" library</span>
<span class="va">enrichr_msig_hallmark</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr.html">rba_enrichr</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span>,
                                     gene_set_library <span class="op">=</span> <span class="st">"MSigDB_Hallmark_2020"</span><span class="op">)</span>
<span class="co"># 2.b Maybe you want to enrich using every library that contains the word "msig":</span>
<span class="va">enrichr_msig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr.html">rba_enrichr</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span>,
                            gene_set_library <span class="op">=</span> <span class="st">"msig"</span>,
                            regex_library_name <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co"># 2.c Or maybe you want to enrich using every library available at Enrichr:</span>
<span class="co"># enrichr_all &lt;- rba_enrichr(gene_list = genes,</span>
<span class="co">#                            gene_set_library = "all")</span></code></pre></div>
<p>Note that when only one Enrichr library is selected, a data frame with enrichment analysis result will be returned:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">enrichr_msig_hallmark</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    18 obs. of  9 variables:</span>
<span class="co">#&gt;  $ Term                : chr  "Mitotic Spindle" "G2-M Checkpoint" "E2F Targets" "Apoptosis" ...</span>
<span class="co">#&gt;  $ Overlap             : chr  "5/199" "4/200" "4/200" "3/161" ...</span>
<span class="co">#&gt;  $ P.value             : num  2.57e-07 1.22e-05 1.22e-05 2.17e-04 2.74e-03 ...</span>
<span class="co">#&gt;  $ Adjusted.P.value    : num  4.62e-06 7.29e-05 7.29e-05 9.76e-04 9.87e-03 ...</span>
<span class="co">#&gt;  $ Old.P.value         : int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ Old.Adjusted.P.value: int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ Odds.Ratio          : num  51 36.7 36.7 31.4 29.7 ...</span>
<span class="co">#&gt;  $ Combined.Score      : num  774 416 416 265 175 ...</span>
<span class="co">#&gt;  $ Genes               : chr  "CDC42;RACGAP1;PLK1;CDK1;KIF23" "RACGAP1;PLK1;CDK1;KIF23" "RACGAP1;PLK1;CDK1;BRCA1" "CDK2;BRCA1;RHOB" ...</span></code></pre></div>
<p>But when multiple libraries have been selected, the function’s output will be a list where each element is a data frame corresponding to one of the selected libraries:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">enrichr_msig</span>, <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ MSigDB_Computational       :'data.frame': 195 obs. of  9 variables:</span>
<span class="co">#&gt;  $ MSigDB_Oncogenic_Signatures:'data.frame': 26 obs. of  9 variables:</span>
<span class="co">#&gt;  $ MSigDB_Hallmark_2020       :'data.frame': 18 obs. of  9 variables:</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="approach-2-going-step-by-step">Approach 2: Going step-by-step<a class="anchor" aria-label="anchor" href="#approach-2-going-step-by-step"></a>
</h3>
<p>As you can see in <code><a href="../reference/rba_enrichr.html">rba_enrichr()</a></code>’s name, it is a wrapper function. It basically executes the following sequence of functions:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 1 Get a list of available Enrichr libraries</span>
<span class="va">libs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_libs.html">rba_enrichr_libs</a></span><span class="op">(</span>store_in_options <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># 2 Submit your gene-set to enrichr</span>
<span class="va">list_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_add_list.html">rba_enrichr_add_list</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span><span class="op">)</span>

<span class="co"># 3 Perform Enrichment analysis with your uploaded gene-set</span>
<span class="va">enriched</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_enrich.html">rba_enrichr_enrich</a></span><span class="op">(</span>user_list_id <span class="op">=</span> <span class="va">list_id</span><span class="op">$</span><span class="va">userListId</span>,
                               gene_set_library <span class="op">=</span> <span class="st">"Table_Mining_of_CRISPR_Studies"</span><span class="op">)</span>

<span class="co">## As always, use str() to see what you have:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">enriched</span>, <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    46 obs. of  9 variables:</span>
<span class="co">#&gt;  $ Term                : chr  "Reaction polymerase chain quantitative time, PMC6627898 (Table1)" "Anticancer corresponding targets potential non, PMC5981615 (Table1)" "Names alphabetical order corresponding listed, PMC6783930 (Table1)" "Supplemental, PMC6879830 (Tablesupplemental)" ...</span>
<span class="co">#&gt;  $ Overlap             : chr  "4/6" "3/25" "3/31" "10/2992" ...</span>
<span class="co">#&gt;  $ P.value             : num  3.07e-12 7.77e-07 1.51e-06 8.05e-06 6.26e-05 ...</span>
<span class="co">#&gt;  $ Adjusted.P.value    : num  1.41e-10 1.79e-05 2.32e-05 9.26e-05 4.80e-04 ...</span>
<span class="co">#&gt;  $ Old.P.value         : int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ Old.Adjusted.P.value: int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ Odds.Ratio          : num  3633.3 226.9 178.2 11.4 219.5 ...</span>
<span class="co">#&gt;  $ Combined.Score      : num  96318 3191 2388 134 2124 ...</span>
<span class="co">#&gt;  $ Genes               : chr  "CDC42;CDK2;CDK1;RHOA" "PLK1;CDK2;CDK1" "CDK2;CDK1;BRCA1" "CDC42;RBM3;RACGAP1;PLK1;CDK2;CDK1;PHF14;KIF23;BRCA1;MSL1" ...</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="see-also-in-functions-manuals">See also in Functions’ manuals<a class="anchor" aria-label="anchor" href="#see-also-in-functions-manuals"></a>
</h2>
<p>Some rbioapi Enrichr functions were not covered in this vignette, be sure to check their manuals:</p>
<ul>
<li><p><code><a href="../reference/rba_enrichr_gene_map.html">rba_enrichr_gene_map()</a></code></p></li>
<li><p><code><a href="../reference/rba_enrichr_view_list.html">rba_enrichr_view_list()</a></code></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_enrichr.html" title="https://rbioapi.moosa-r.com/articles/rbioapi_enrichr.html">This article in rbioapi documentation site</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html#section-enrichr-rba-enrichr-" title="rbioapi reference">Functions references in rbioapi documentation site</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi vignette index</a></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.7</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] rbioapi_0.7.5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] knitr_1.37        magrittr_2.0.1    R6_2.5.1          ragg_1.1.3       </span>
<span class="co">#&gt;  [5] rlang_0.4.12      fastmap_1.1.0     httr_1.4.2        stringr_1.4.0    </span>
<span class="co">#&gt;  [9] tools_4.1.2       DT_0.20           xfun_0.29         jquerylib_0.1.4  </span>
<span class="co">#&gt; [13] crosstalk_1.2.0   htmltools_0.5.2   systemfonts_1.0.2 yaml_2.2.1       </span>
<span class="co">#&gt; [17] digest_0.6.29     rprojroot_2.0.2   pkgdown_2.0.2     crayon_1.4.2     </span>
<span class="co">#&gt; [21] textshaping_0.3.6 purrr_0.3.4       htmlwidgets_1.5.4 sass_0.4.0       </span>
<span class="co">#&gt; [25] fs_1.5.2          curl_4.3.2        memoise_2.0.1     cachem_1.0.6     </span>
<span class="co">#&gt; [29] evaluate_0.14     rmarkdown_2.11    stringi_1.7.6     compiler_4.1.2   </span>
<span class="co">#&gt; [33] bslib_0.3.1       desc_1.4.0        jsonlite_1.7.3</span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
