<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Connect to Enrichr in R with rbioapi package. 
">
<title>2.A: Enrichr &amp; rbioapi • rbioapi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2.A: Enrichr &amp; rbioapi">
<meta property="og:description" content="Connect to Enrichr in R with rbioapi package. 
">
<meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.9</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/rbioapi.html">Get Started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles-amp-supported-services">Articles &amp; Supported Services</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles-amp-supported-services">
    <a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://moosa-r.com/" aria-label="Moosa Rezwani Portfolio Website">
    <span class="fa fa-solid fa-id-card"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="rbioapi_enrichr_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="rbioapi_enrichr_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_enrichr_files/datatables-binding-0.29/datatables.js"></script><link href="rbioapi_enrichr_files/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_enrichr_files/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_enrichr_files/dt-core-1.13.4/js/jquery.dataTables.min.js"></script><link href="rbioapi_enrichr_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="rbioapi_enrichr_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>2.A: Enrichr &amp; rbioapi</h1>
                        <h4 data-toc-skip class="author">Moosa
Rezwani</h4>
            
            <h4 data-toc-skip class="date">2023-09-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/HEAD/vignettes/rbioapi_enrichr.Rmd" class="external-link"><code>vignettes/rbioapi_enrichr.Rmd</code></a></small>
      <div class="d-none name"><code>rbioapi_enrichr.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 class="heading2" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><a href="https://maayanlab.cloud/Enrichr/" class="external-link">Enrichr</a> is a popular
gene-set enrichment analysis tool developed in the Ma’ayan Lab.</p>
<hr>
</div>
<div class="section level2">
<h2 class="heading2" id="gene-set-library-concept-in-enrichr">Gene set library concept in Enrichr<a class="anchor" aria-label="anchor" href="#gene-set-library-concept-in-enrichr"></a>
</h2>
<p>Directly quoting from Enrichr’s help page:</p>
<blockquote>
<p>A <em>gene set library</em> is a set of related gene sets or
enrichment terms […] These libraries have been constructed from many
sources such as published studies and major biological and biomedical
online databases. Others have been created for and only available
through Enrichr.</p>
<p>(source: <a href="https://maayanlab.cloud/Enrichr/help#background" class="external-link uri">https://maayanlab.cloud/Enrichr/help#background</a>)</p>
</blockquote>
<p>To get a list of the available libraries in Enrichr, use:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichr_libs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_libs.html">rba_enrichr_libs</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>In the returned data frame, you can find the names of available
Enrichr libraries in “libraryName” column. As you will see in the
following sections, you can use these names to request an enrichment
analysis based on the selected library or libraries.</p>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-2ea3d31c4c5fd5e6fa16" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-2ea3d31c4c5fd5e6fa16">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218"],[13362,27884,6002,47172,47107,21493,1295,3185,2854,15057,4128,34061,7504,16399,12753,23726,32740,13373,19270,13236,14264,3096,22288,4533,10231,2741,5655,10406,10493,11251,8695,1759,2178,851,10061,11250,15406,17711,17576,15797,12232,13572,6454,3723,7588,7682,7324,8469,13121,26382,29065,280,13877,15852,4320,4271,10496,1678,756,3800,2541,1918,5863,6768,25651,19129,23939,23561,23877,15886,24350,3102,31132,30832,48230,5613,9559,9448,16725,19249,15090,16129,15309,15103,15022,15676,15854,15015,3788,3357,12668,12638,8973,7010,5966,15562,17850,17660,1348,934,2541,2041,5209,49238,2243,19586,22440,8184,18329,15755,10271,10427,10601,13822,8002,10089,13247,21809,23601,20883,19612,25983,19500,14893,17598,5902,12486,1073,19513,14433,8655,11459,19741,27360,13072,13464,13787,13929,16964,17258,10352,10471,12419,19378,6201,4558,3264,7802,8551,12444,9000,7744,6204,13420,14148,9813,1397,9116,17464,394,11851,8189,18704,5605,5718,14156,16979,4383,54974,12118,12361,9763,8078,7173,5833,14937,11497,11936,9767,14167,17851,16853,6654,1683,20414,26076,26338,25381,25409,11980,31158,30006,13370,13697,2183,12765,1509,18365,13525,9525,9440,3857,10489,1198,1882,1552,6713,936,8220,9021,8076,14698,10972,12126,13662,18290,12081,12853],[275,1284,77,1370,509,3713,18,73,34,300,48,641,155,247,57,127,85,258,388,82,58,31,4368,37,158,5,342,715,200,100,100,25,89,15,106,166,300,300,300,176,343,301,301,47,35,78,172,122,305,1811,2123,9,304,912,129,128,201,21,12,48,39,39,51,47,807,1594,293,307,302,9,299,25,298,302,1429,36,73,63,1443,1443,282,292,308,318,290,310,279,321,159,153,300,300,64,87,51,887,300,300,19,13,39,42,300,1550,19,545,505,24,161,28,22,38,25,21,143,45,49,2316,2395,299,299,299,137,128,1208,109,299,100,117,36,61,39,270,802,26,45,200,200,200,200,58,76,491,37,45,54,22,92,98,23,20,363,387,32,122,49,13,22,63,73,586,421,100,39,31,40,295,146,483,448,124,139,102,43,100,33,80,34,33,80,102,360,136,10,112,250,250,250,250,250,805,815,103,343,18,13,100,1214,175,245,245,80,61,23,47,16,68,15,146,793,96,33,85,38,12,34,50,485],["Genome_Browser_PWMs","TRANSFAC_and_JASPAR_PWMs","Transcription_Factor_PPIs","ChEA_2013","Drug_Perturbations_from_GEO_2014","ENCODE_TF_ChIP-seq_2014","BioCarta_2013","Reactome_2013","WikiPathways_2013","Disease_Signatures_from_GEO_up_2014","KEGG_2013","TF-LOF_Expression_from_GEO","TargetScan_microRNA","PPI_Hub_Proteins","GO_Molecular_Function_2015","GeneSigDB","Chromosome_Location","Human_Gene_Atlas","Mouse_Gene_Atlas","GO_Cellular_Component_2015","GO_Biological_Process_2015","Human_Phenotype_Ontology","Epigenomics_Roadmap_HM_ChIP-seq","KEA_2013","NURSA_Human_Endogenous_Complexome","CORUM","SILAC_Phosphoproteomics","MGI_Mammalian_Phenotype_Level_3","MGI_Mammalian_Phenotype_Level_4","Old_CMAP_up","Old_CMAP_down","OMIM_Disease","OMIM_Expanded","VirusMINT","MSigDB_Computational","MSigDB_Oncogenic_Signatures","Disease_Signatures_from_GEO_down_2014","Virus_Perturbations_from_GEO_up","Virus_Perturbations_from_GEO_down","Cancer_Cell_Line_Encyclopedia","NCI-60_Cancer_Cell_Lines","Tissue_Protein_Expression_from_ProteomicsDB","Tissue_Protein_Expression_from_Human_Proteome_Map","HMDB_Metabolites","Pfam_InterPro_Domains","GO_Biological_Process_2013","GO_Cellular_Component_2013","GO_Molecular_Function_2013","Allen_Brain_Atlas_up","ENCODE_TF_ChIP-seq_2015","ENCODE_Histone_Modifications_2015","Phosphatase_Substrates_from_DEPOD","Allen_Brain_Atlas_down","ENCODE_Histone_Modifications_2013","Achilles_fitness_increase","Achilles_fitness_decrease","MGI_Mammalian_Phenotype_2013","BioCarta_2015","HumanCyc_2015","KEGG_2015","NCI-Nature_2015","Panther_2015","WikiPathways_2015","Reactome_2015","ESCAPE","HomoloGene","Disease_Perturbations_from_GEO_down","Disease_Perturbations_from_GEO_up","Drug_Perturbations_from_GEO_down","Genes_Associated_with_NIH_Grants","Drug_Perturbations_from_GEO_up","KEA_2015","Gene_Perturbations_from_GEO_up","Gene_Perturbations_from_GEO_down","ChEA_2015","dbGaP","LINCS_L1000_Chem_Pert_up","LINCS_L1000_Chem_Pert_down","GTEx_Tissue_Expression_Down","GTEx_Tissue_Expression_Up","Ligand_Perturbations_from_GEO_down","Aging_Perturbations_from_GEO_down","Aging_Perturbations_from_GEO_up","Ligand_Perturbations_from_GEO_up","MCF7_Perturbations_from_GEO_down","MCF7_Perturbations_from_GEO_up","Microbe_Perturbations_from_GEO_down","Microbe_Perturbations_from_GEO_up","LINCS_L1000_Ligand_Perturbations_down","LINCS_L1000_Ligand_Perturbations_up","L1000_Kinase_and_GPCR_Perturbations_down","L1000_Kinase_and_GPCR_Perturbations_up","Reactome_2016","KEGG_2016","WikiPathways_2016","ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X","Kinase_Perturbations_from_GEO_down","Kinase_Perturbations_from_GEO_up","BioCarta_2016","HumanCyc_2016","NCI-Nature_2016","Panther_2016","DrugMatrix","ChEA_2016","huMAP","Jensen_TISSUES","RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO","MGI_Mammalian_Phenotype_2017","Jensen_COMPARTMENTS","Jensen_DISEASES","BioPlex_2017","GO_Cellular_Component_2017","GO_Molecular_Function_2017","GO_Biological_Process_2017","GO_Cellular_Component_2017b","GO_Molecular_Function_2017b","GO_Biological_Process_2017b","ARCHS4_Tissues","ARCHS4_Cell-lines","ARCHS4_IDG_Coexp","ARCHS4_Kinases_Coexp","ARCHS4_TFs_Coexp","SysMyo_Muscle_Gene_Sets","miRTarBase_2017","TargetScan_microRNA_2017","Enrichr_Libraries_Most_Popular_Genes","Enrichr_Submissions_TF-Gene_Coocurrence","Data_Acquisition_Method_Most_Popular_Genes","DSigDB","GO_Biological_Process_2018","GO_Cellular_Component_2018","GO_Molecular_Function_2018","TF_Perturbations_Followed_by_Expression","Chromosome_Location_hg19","NIH_Funded_PIs_2017_Human_GeneRIF","NIH_Funded_PIs_2017_Human_AutoRIF","Rare_Diseases_AutoRIF_ARCHS4_Predictions","Rare_Diseases_GeneRIF_ARCHS4_Predictions","NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions","NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions","Rare_Diseases_GeneRIF_Gene_Lists","Rare_Diseases_AutoRIF_Gene_Lists","SubCell_BarCode","GWAS_Catalog_2019","WikiPathways_2019_Human","WikiPathways_2019_Mouse","TRRUST_Transcription_Factors_2019","KEGG_2019_Human","KEGG_2019_Mouse","InterPro_Domains_2019","Pfam_Domains_2019","DepMap_WG_CRISPR_Screens_Broad_CellLines_2019","DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019","MGI_Mammalian_Phenotype_Level_4_2019","UK_Biobank_GWAS_v1","BioPlanet_2019","ClinVar_2019","PheWeb_2019","DisGeNET","HMS_LINCS_KinomeScan","CCLE_Proteomics_2020","ProteomicsDB_2020","lncHUB_lncRNA_Co-Expression","Virus-Host_PPI_P-HIPSTer_2020","Elsevier_Pathway_Collection","Table_Mining_of_CRISPR_Studies","COVID-19_Related_Gene_Sets","MSigDB_Hallmark_2020","Enrichr_Users_Contributed_Lists_2020","TG_GATES_2020","Allen_Brain_Atlas_10x_scRNA_2021","Descartes_Cell_Types_and_Tissue_2021","KEGG_2021_Human","WikiPathway_2021_Human","HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression","GO_Biological_Process_2021","GO_Cellular_Component_2021","GO_Molecular_Function_2021","MGI_Mammalian_Phenotype_Level_4_2021","CellMarker_Augmented_2021","Orphanet_Augmented_2021","COVID-19_Related_Gene_Sets_2021","PanglaoDB_Augmented_2021","Azimuth_Cell_Types_2021","PhenGenI_Association_2021","RNAseq_Automatic_GEO_Signatures_Human_Down","RNAseq_Automatic_GEO_Signatures_Human_Up","RNAseq_Automatic_GEO_Signatures_Mouse_Down","RNAseq_Automatic_GEO_Signatures_Mouse_Up","GTEx_Aging_Signatures_2021","HDSigDB_Human_2021","HDSigDB_Mouse_2021","HuBMAP_ASCTplusB_augmented_2022","FANTOM6_lncRNA_KD_DEGs","MAGMA_Drugs_and_Diseases","PFOCR_Pathways","Tabula_Sapiens","ChEA_2022","Diabetes_Perturbations_GEO_2022","LINCS_L1000_Chem_Pert_Consensus_Sigs","LINCS_L1000_CRISPR_KO_Consensus_Sigs","Tabula_Muris","Reactome_2022","SynGO_2022","GlyGen_Glycosylated_Proteins_2022","IDG_Drug_Targets_2022","KOMP2_Mouse_Phenotypes_2022","Metabolomics_Workbench_Metabolites_2022","Proteomics_Drug_Atlas_2023","The_Kinase_Library_2023","GTEx_Tissues_V8_2023","GO_Biological_Process_2023","GO_Cellular_Component_2023","GO_Molecular_Function_2023","PFOCR_Pathways_2023","GWAS_Catalog_2023","GeDiPNet_2023","MAGNET_2023"],["http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/","http://jaspar.genereg.net/html/DOWNLOAD/","","http://amp.pharm.mssm.edu/lib/cheadownload.jsp","http://www.ncbi.nlm.nih.gov/geo/","http://genome.ucsc.edu/ENCODE/downloads.html","https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://www.reactome.org/download/index.html","http://www.wikipathways.org/index.php/Download_Pathways","http://www.ncbi.nlm.nih.gov/geo/","http://www.kegg.jp/kegg/download/","http://www.ncbi.nlm.nih.gov/geo/","http://www.targetscan.org/cgi-bin/targetscan/data_download.cgi?db=vert_61","http://amp.pharm.mssm.edu/X2K","http://www.geneontology.org/GO.downloads.annotations.shtml","https://pubmed.ncbi.nlm.nih.gov/22110038/","http://software.broadinstitute.org/gsea/msigdb/index.jsp","http://biogps.org/downloads/","http://biogps.org/downloads/","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.human-phenotype-ontology.org/","http://www.roadmapepigenomics.org/","http://amp.pharm.mssm.edu/lib/keacommandline.jsp","https://www.nursa.org/nursa/index.jsf","http://mips.helmholtz-muenchen.de/genre/proj/corum/","http://amp.pharm.mssm.edu/lib/keacommandline.jsp","http://www.informatics.jax.org/","http://www.informatics.jax.org/","http://www.broadinstitute.org/cmap/","http://www.broadinstitute.org/cmap/","http://www.omim.org/downloads","http://www.omim.org/downloads","http://mint.bio.uniroma2.it/download.html","http://www.broadinstitute.org/gsea/msigdb/collections.jsp","http://www.broadinstitute.org/gsea/msigdb/collections.jsp","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://portals.broadinstitute.org/ccle/home\n","http://biogps.org/downloads/","https://www.proteomicsdb.org/","http://www.humanproteomemap.org/index.php","http://www.hmdb.ca/downloads","ftp://ftp.ebi.ac.uk/pub/databases/interpro/","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.brain-map.org/","http://genome.ucsc.edu/ENCODE/downloads.html","http://genome.ucsc.edu/ENCODE/downloads.html","http://www.koehn.embl.de/depod/","http://www.brain-map.org/","http://genome.ucsc.edu/ENCODE/downloads.html","http://www.broadinstitute.org/achilles","http://www.broadinstitute.org/achilles","http://www.informatics.jax.org/","https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://humancyc.org/","http://www.kegg.jp/kegg/download/","http://pid.nci.nih.gov/","http://www.pantherdb.org/","http://www.wikipathways.org/index.php/Download_Pathways","http://www.reactome.org/download/index.html","http://www.maayanlab.net/ESCAPE/","http://www.ncbi.nlm.nih.gov/homologene","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://grants.nih.gov/grants/oer.htm\n","http://www.ncbi.nlm.nih.gov/geo/","http://amp.pharm.mssm.edu/Enrichr","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://amp.pharm.mssm.edu/Enrichr","http://www.ncbi.nlm.nih.gov/gap","https://clue.io/","https://clue.io/","http://www.gtexportal.org/","http://www.gtexportal.org/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://clue.io/","https://clue.io/","https://clue.io/","https://clue.io/","http://www.reactome.org/download/index.html","http://www.kegg.jp/kegg/download/","http://www.wikipathways.org/index.php/Download_Pathways","","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://humancyc.org/","http://pid.nci.nih.gov/","http://www.pantherdb.org/pathway/","https://ntp.niehs.nih.gov/drugmatrix/","http://amp.pharm.mssm.edu/Enrichr","http://proteincomplexes.org/","http://tissues.jensenlab.org/","http://www.ncbi.nlm.nih.gov/geo/","http://www.informatics.jax.org/","http://compartments.jensenlab.org/","http://diseases.jensenlab.org/","http://bioplex.hms.harvard.edu/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://sys-myo.rhcloud.com/","http://mirtarbase.mbc.nctu.edu.tw/","http://www.targetscan.org/","http://amp.pharm.mssm.edu/Enrichr","http://amp.pharm.mssm.edu/Enrichr","http://amp.pharm.mssm.edu/Enrichr","http://tanlab.ucdenver.edu/DSigDB/DSigDBv1.0/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.ncbi.nlm.nih.gov/geo/","http://hgdownload.cse.ucsc.edu/downloads.html","https://www.ncbi.nlm.nih.gov/pubmed/","https://www.ncbi.nlm.nih.gov/pubmed/","https://amp.pharm.mssm.edu/geneshot/","https://www.ncbi.nlm.nih.gov/gene/about-generif","https://www.ncbi.nlm.nih.gov/pubmed/","https://www.ncbi.nlm.nih.gov/pubmed/","https://www.ncbi.nlm.nih.gov/gene/about-generif","https://amp.pharm.mssm.edu/geneshot/","http://www.subcellbarcode.org/","https://www.ebi.ac.uk/gwas","https://www.wikipathways.org/","https://www.wikipathways.org/","https://www.grnpedia.org/trrust/","https://www.kegg.jp/","https://www.kegg.jp/","https://www.ebi.ac.uk/interpro/","https://pfam.xfam.org/","https://depmap.org/","https://depmap.org/","http://www.informatics.jax.org/","https://www.ukbiobank.ac.uk/tag/gwas/","https://tripod.nih.gov/bioplanet/","https://www.ncbi.nlm.nih.gov/clinvar/","http://pheweb.sph.umich.edu/","https://www.disgenet.org","http://lincs.hms.harvard.edu/kinomescan/","https://portals.broadinstitute.org/ccle","https://www.proteomicsdb.org/","https://amp.pharm.mssm.edu/lnchub/","http://phipster.org/","http://www.transgene.ru/disease-pathways/","","https://amp.pharm.mssm.edu/covid19","https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp ","https://maayanlab.cloud/Enrichr","https://toxico.nibiohn.go.jp/english/","https://portal.brain-map.org/","https://descartes.brotmanbaty.org/bbi/human-gene-expression-during-development/","https://www.kegg.jp/","https://www.wikipathways.org/","https://hubmapconsortium.github.io/ccf-asct-reporter/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.informatics.jax.org/","http://biocc.hrbmu.edu.cn/CellMarker/","http://www.orphadata.org/","https://maayanlab.cloud/covid19/","https://panglaodb.se/","https://azimuth.hubmapconsortium.org/","https://www.ncbi.nlm.nih.gov/gap/phegeni","https://maayanlab.cloud/archs4/","https://maayanlab.cloud/archs4/","https://maayanlab.cloud/archs4/","https://maayanlab.cloud/archs4/","https://gtexportal.org/","https://www.hdinhd.org/","https://www.hdinhd.org/","https://hubmapconsortium.github.io/ccf-asct-reporter/","https://fantom.gsc.riken.jp/6/","https://github.com/nybell/drugsets/tree/main/DATA/GENESETS","https://pfocr.wikipathways.org/","https://tabula-sapiens-portal.ds.czbiohub.org/","https://maayanlab.cloud/chea3/","https://appyters.maayanlab.cloud/#/Gene_Expression_T2D_Signatures","https://maayanlab.cloud/sigcom-lincs/#/Download","https://maayanlab.cloud/sigcom-lincs/#/Download","https://tabula-muris.ds.czbiohub.org/","https://reactome.org/download-data","https://www.syngoportal.org/","https://www.glygen.org/","https://drugcentral.org/","https://www.mousephenotype.org/","https://www.metabolomicsworkbench.org/","https://www.nature.com/articles/s41587-022-01539-0","https://kinase-library.phosphosite.org/site","https://gtexportal.org/home/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","https://pfocr.wikipathways.org/","https://www.ebi.ac.uk/gwas","http://gedipnet.bicnirrh.res.in/","https://magnet-winterlab.herokuapp.com/"],[615,326,290,353,701,498,249,78,199,142,200,269,222,385,1136,2139,386,84,96,641,5192,1779,383,474,1796,1658,84,71,476,6100,6100,90,187,85,858,189,142,323,323,967,93,207,30,3906,311,941,205,402,2192,816,412,59,2192,109,216,216,476,239,125,179,209,104,404,1389,315,12,839,839,906,32876,906,428,2460,2460,395,345,33132,33132,2918,2918,261,286,286,261,401,401,312,312,96,96,3644,3644,1530,293,437,104,285,285,237,152,209,112,7876,645,995,1842,1302,5231,2283,1811,3915,636,972,3166,816,3271,10125,108,125,352,498,1724,1135,3240,683,121,1722,12,4026,5103,446,1151,1958,36,5687,12558,3725,2244,12558,5684,2244,3725,104,1737,472,176,571,308,303,1071,608,558,325,5261,857,1510,182,1161,9828,148,378,913,3729,6715,1721,802,205,50,1482,1190,766,172,320,622,344,6036,511,1274,4601,1097,3774,478,178,341,950,4269,4269,4216,4216,270,2564,2579,777,350,1395,17326,469,757,601,10850,10424,106,1818,118,338,888,529,233,1748,303,511,5407,474,1147,21845,5271,2388,72],["ea115789fcbf12797fd692cec6df0ab4dbc79c6a","7d42eb43a64a4e3b20d721fc7148f685b53b6b30","849f222220618e2599d925b6b51868cf1dab3763","7ebe772afb55b63b41b79dd8d06ea0fdd9fa2630","ad270a6876534b7cb063e004289dcd4d3164f342","497787ebc418d308045efb63b8586f10c526af51","4a293326037a5229aedb1ad7b2867283573d8bcd","b343994a1b68483b0122b08650201c9b313d5c66","5c307674c8b97e098f8399c92f451c0ff21cbf68","248c4ed8ea28352795190214713c86a39fd7afab","eb26f55d3904cb0ea471998b6a932a9bf65d8e50","","f4029bf6a62c91ab29401348e51df23b8c44c90f","69c0cfe07d86f230a7ef01b365abcc7f6e52f138","f531ac2b6acdf7587a54b79b465a5f4aab8f00f9","6d655e0aa3408a7accb3311fbda9b108681a8486","8dab0f96078977223646ff63eb6187e0813f1433","0741451470203d7c40a06274442f25f74b345c9c","31191bfadded5f96983f93b2a113cf2110ff5ddb","e1d004d5797cbd2363ef54b1c3b361adb68795c6","bf120b6e11242b1a64c80910d8e89f87e618e235","17a138b0b70aa0e143fe63c14f82afb70bc3ed0a","e1bc8a398e9b21f9675fb11bef18087eda21b1bf","462045609440fa1e628a75716b81a1baa5bd9145","7d3566b12ebc23dd23d9ca9bb97650f826377b16","d047f6ead7831b00566d5da7a3b027ed9196e104","54dcd9438b33301deb219866e162b0f9da7e63a0","c3bfc90796cfca8f60cba830642a728e23a53565","0b09a9a1aa0af4fc7ea22d34a9ae644d45864bd6","9041f90cccbc18479138330228b336265e09021c","ebc0d905b3b3142f936d400c5f2a4ff926c81c37","cb2b92578a91e023d0498a334923ee84add34eca","27eca242904d8e12a38cf8881395bc50d57a03e1","5abad1fc36216222b0420cadcd9be805a0dda63e","e4cdcc7e259788fdf9b25586cce3403255637064","c76f5319c33c4833c71db86a30d7e33cd63ff8cf","aabdf7017ae55ae75a004270924bcd336653b986","45268b7fc680d05dd9a29743c2f2b2840a7620bf","5f531580ccd168ee4acc18b02c6bdf8200e19d08","eb38dbc3fb20adafa9d6f9f0b0e36f378e75284f","75c81676d8d6d99d262c9660edc024b78cfb07c9","","49351dc989f9e6ca97c55f8aca7778aa3bfb84b9","1905132115d22e4119bce543bdacaab074edb363","e2b4912cfb799b70d87977808c54501544e4cdc9","5216d1ade194ffa5a6c00f105e2b1899f64f45fe","fd1332a42395e0bc1dba82868b39be7983a48cc5","7e3e99e5aae02437f80b0697b197113ce3209ab0","3804715a63a308570e47aa1a7877f01147ca6202","56b6adb4dc8a2f540357ef992d6cd93dfa2907e5","55b56cd8cf2ff04b26a09b9f92904008b82f3a6f","d40701e21092b999f4720d1d2b644dd0257b6259","ea67371adec290599ddf484ced2658cfae259304","c209ae527bc8e98e4ccd27a668d36cd2c80b35b4","98366496a75f163164106e72439fb2bf2f77de4e","83a710c1ff67fd6b8af0d80fa6148c40dbd9bc64","a4c6e217a81a4a58ff5a1c9fc102b70beab298e9","70e4eb538daa7688691acfe5d9c3c19022be832b","711f0c02b23f5e02a01207174943cfeee9d3ea9c","e80d25c56de53c704791ddfdc6ab5eec28ae7243","47edfc012bcbb368a10b717d8dca103f7814b5a4","ab824aeeff0712bab61f372e43aebb870d1677a9","1f7eea2f339f37856522c1f1c70ec74c7b25325f","36e541bee015eddb8d53827579549e30fe7a3286","a7acc741440264717ff77751a7e5fed723307835","663b665b75a804ef98add689f838b68e612f0d2a","0f412e0802d76efa0374504c2c9f5e0624ff7f09","9ddc3902fb01fb9eaf1a2a7c2ff3acacbb48d37e","068623a05ecef3e4a5e0b4f8db64bb8faa3c897f","76fc5ec6735130e287e62bae6770a3c5ee068645","c9c2155b5ac81ac496854fa61ba566dcae06cc80","18a081774e6e0aaf60b1a4be7fd20afcf9e08399","53dedc29ce3100930d68e506f941ef59de05dc6b","499882af09c62dd6da545c15cb51c1dc5e234f78","712eb7b6edab04658df153605ec6079fa89fb5c7","010f1267055b1a1cb036e560ea525911c007a666","5e678b3debe8d8ea95187d0cd35c914017af5eb3","fedbf5e221f45ee60ebd944f92569b5eda7f2330","74b818bd299a9c42c1750ffe43616aa9f7929f02","103738763d89cae894bec9f145ac28167a90e611","1eb3c0426140340527155fd0ef67029db2a72191","cd95fe1b505ba6f28cd722cfba50fdea979d3b4c","74c4f0a0447777005b2a5c00c9882a56dfc62d7c","31baa39da2931ddd5f7aedf2d0bbba77d2ba7b46","555f68aef0a29a67b614a0d7e20b6303df9069c6","1bc2ba607f1ff0dda44e2a15f32a2c04767da18c","9e613dba78ef7e60676b13493a9dc49ccd3c8b3f","d0c3e2a68e8c611c669098df2c87b530cec3e132","957846cb05ef31fc8514120516b73cc65af7980e","3bd494146c98d8189898a947f5ef5710f1b7c4b2","1ccc5bce553e0c2279f8e3f4ddcfbabcf566623b","b54a0d4ba525eac4055c7314ca9d9312adcb220c","1f54638e8f45075fb79489f0e0ef906594cb0678","43f56da7540195ba3c94eb6e34c522a699b36da9","340be98b444cad50bb974df69018fd598e23e5e1","5426f7747965c23ef32cff46fabf906e2cd76bfa","bb9682d78b8fc43be842455e076166fcd02cefc3","78618915009cac3a0663d6f99d359e39a31b6660","13d9ab18921d5314a5b2b366f6142b78ab0ff6aa","d6a502ef9b4c789ed5e73ca5a8de372796e5c72a","3c1e1f7d1a651d9aaa198e73704030716fc09431","ca5f6abf7f75d9baae03396e84d07300bf1fd051","255c3db820d612f34310f22a6985dad50e9fe1fe","af271913344aa08e6a755af1d433ef15768d749a","249247d2f686d3eb4b9e4eb976c51159fac80a89","e8879ab9534794721614d78fe2883e9e564d7759","f0752e4d7f5198f86446678966b260c530d19d78","0705e59bff98deda6e9cbe00cfcdd871c85e7d04","56ec68c32d4e83edc2ee83bea0e9f6a3829b2279","3045dff8181367c1421627bb8e4c5a32c6d67f98","b8620b1a9d0d271d1a2747d8cfc63589dba39991","8fed21d22dfcc3015c05b31d942fdfc851cc8e04","b4018906e0a8b4e81a1b1afc51e0a2e7655403eb","d9da4dba4a3eb84d4a28a3835c06dfbbe5811f92","ecf39c41fa5bc7deb625a2b5761a708676e9db7c","8d8340361dd36a458f1f0a401f1a3141de1f3200","6404c38bffc2b3732de4e3fbe417b5043009fe34","4126374338235650ab158ba2c61cd2e2383b70df","5496ef9c9ae9429184d0b9485c23ba468ee522a8","ce60be284fdd5a9fc6240a355421a9e12b1ee84a","6721c5ed97b7772e4a19fdc3f797110df0164b75","8a468c3ae29fa68724f744cbef018f4f3b61c5ab","","6b7c7fe2a97b19aecbfba12d8644af6875ad99c4","79d13fb03d2fa6403f9be45c90eeda0f6822e269","e9b7d8ee237d0a690bd79d970a23a9fa849901ed","be2ca8ef5a8c8e17d7e7bd290e7cbfe0951396c0","17ce5192b9eba7d109b6d228772ea8ab222e01ef","287476538ab98337dbe727b3985a436feb6d192a","b5b77681c46ac58cd050e60bcd4ad5041a9ab0a9","e9ebe46188efacbe1056d82987ff1c70218fa7ae","79ff80ae9a69dd00796e52569e41422466fa0bee","34d08a4878c19584aaf180377f2ea96faa6a6eb1","fdab39c467ba6b0fb0288df1176d7dfddd7196d5","859b100fac3ca774ad84450b1fbb65a78fcc6b12","fc5bf033b932cf173633e783fc8c6228114211f8","375ff8cdd64275a916fa24707a67968a910329bb","0f7fb7f347534779ecc6c87498e96b5460a8d652","f77de51aaf0979dd6f56381cf67ba399b4640d28","25fa899b715cd6a9137f6656499f89cd25144029","0fb9ac92dbe52024661c088f71a1134f00567a8b","ee3adbac2da389959410260b280e7df1fd3730df","b50bb9480d8a77103fb75b331fd9dd927246939a","fef3864bcb5dd9e60cee27357eff30226116c49b","b0c9e9ebb9014f14561e896008087725a2db24b7","e7750958da20f585c8b6d5bc4451a5a4305514ba","5f8cf93e193d2bcefa5a37ccdf0eefac576861b0","3477bc578c4ea5d851dcb934fe2a41e9fd789bb4","187eb44b2d6fa154ebf628eba1f18537f64e797c","18dd5ec520fdf589a93d6a7911289c205e1ddf22","a6325ed264f9ac9e6518796076c46a1d885cca7a","0b08b32b20854ac8a738458728a9ea50c2e04800","b7c4ead26d0eb64f1697c030d31682b581c8bb56","f1bed632e89ebc054da44236c4815cdce03ef5ee","958fb52e6215626673a5acf6e9289a1b84d11b4a","e110851dfc763d30946f2abedcc2cd571ac357a0","0a95303f8059bec08836ecfe02ce3da951150547","6a7c7321b6b72c5285b722f7902d26a2611117cb","3c261626478ce9e6bf2c7f0a8014c5e901d43dc0","47ba06cdc92469ac79400fc57acd84ba343ba616","7094b097ae2301a1d6a5bd856a193b084cca993d","8c87c8346167bac2ba68195a32458aba9b1acfd1","45b597d7efa5693b7e4172b09c0ed2dda3305582","a592eed13e8e9496aedbab63003b965574e46a65","9196c760e3bcae9c9de1e3f87ad81f96bde24325","ad580f3864fa8ff69eaca11f6d2e7f9b86378d08","72b0346849570f66a77a6856722601e711596cb4","6952efda94663d4bd8db09bf6eeb4e67d21ef58c","8dc362703b38b30ac3b68b6401a9b20a58e7d3ef","9e32560437b11b4628b00ccf3d584360f7f7daee","46f8235cb585829331799a71aec3f7c082170219","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","",""],[1,1,1,7,7,7,7,7,7,7,7,1,7,2,7,4,7,5,5,7,7,3,1,7,2,2,2,7,7,7,7,4,4,4,4,4,7,4,4,5,5,7,5,6,6,7,7,7,5,1,1,2,5,7,4,4,7,7,7,7,7,7,7,7,5,6,8,8,8,6,8,2,8,8,7,4,7,7,5,5,8,8,8,8,8,8,8,8,7,7,7,7,7,7,7,1,2,2,2,2,2,2,4,7,2,3,8,7,3,3,2,7,7,7,7,7,7,5,5,4,2,1,8,1,1,6,1,6,4,7,7,7,1,6,6,6,4,4,6,6,4,4,2,7,7,7,1,7,7,6,6,4,4,7,4,2,4,4,4,2,5,5,1,2,2,6,7,2,6,4,5,5,2,2,5,7,7,7,3,5,4,4,5,5,4,8,8,8,8,4,4,4,5,1,4,7,5,1,4,4,4,5,2,3,2,4,3,2,4,2,5,3,3,3,2,4,4,5]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>geneCoverage<\/th>\n      <th>genesPerTerm<\/th>\n      <th>libraryName<\/th>\n      <th>link<\/th>\n      <th>numTerms<\/th>\n      <th>appyter<\/th>\n      <th>categoryId<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[1,2,5,7]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div class="section level2">
<h2 class="heading2" id="enrichment-analysis-using-enrichr">Enrichment analysis using Enrichr<a class="anchor" aria-label="anchor" href="#enrichment-analysis-using-enrichr"></a>
</h2>
<p>To perform enrichment analysis on your gene-set with Enrichr using
rbioapi, you can take two approaches. We will begin with the simple one.
But first, we create a vector of genes’ NCBI IDs to use as the input
example in this article.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a vector with our genes' NCBI IDs</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>,<span class="st">"CDK1"</span>,<span class="st">"KIF23"</span>,<span class="st">"PLK1"</span>,</span>
<span>           <span class="st">"RAC2"</span>,<span class="st">"RACGAP1"</span>,<span class="st">"RHOA"</span>,<span class="st">"RHOB"</span>, <span class="st">"PHF14"</span>, <span class="st">"RBM3"</span>, <span class="st">"MSL1"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 class="heading3" id="approach-1-using-the-wrapper-function">Approach 1: Using the one-step Wrapper
function<a class="anchor" aria-label="anchor" href="#approach-1-using-the-wrapper-function"></a>
</h3>
<p>The only required input for this function is to simply supply your
gene-set as a character vector. Optionally you can also select one or
more libraries. Please see <code><a href="../reference/rba_enrichr.html">rba_enrichr()</a></code> function’s manual
for more details on the arguments.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Request the enrichment analysis</span></span>
<span><span class="va">results_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr.html">rba_enrichr</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<p>Note that the default value for the argument
<code>gene_set_library</code> in the rba_enrichr function is “all”. This
means that if you call the function as above, all of the Enrichr
libraries will be used for the enrichment analysis of your uploaded gene
list. In this case, you will have a named list, where each of its
elements is a dataframe containing your genes’ analysis results using
that Enrichr library.</p>
<p>Alternatively, you can use the <code>gene_set_library</code> argument
to specify the library (or libraries) to use. Here we demonstrate using
“MSigDB_Hallmark_2020” library:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Request the enrichment analysis by a specific library</span></span>
<span><span class="va">results_msig_hallmark</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr.html">rba_enrichr</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span>,</span>
<span>                                     gene_set_library <span class="op">=</span> <span class="st">"MSigDB_Hallmark_2020"</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-4154b27ad0cd365a3b1e" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-4154b27ad0cd365a3b1e">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18"],["Mitotic Spindle","G2-M Checkpoint","E2F Targets","Apoptosis","PI3K/AKT/mTOR  Signaling","UV Response Up","Apical Surface","TGF-beta Signaling","Myc Targets V2","Spermatogenesis","IL-2/STAT5 Signaling","TNF-alpha Signaling via NF-kB","Glycolysis","Allograft Rejection","Apical Junction","mTORC1 Signaling","Myc Targets V1","Epithelial Mesenchymal Transition"],["5/199","4/200","4/200","3/161","2/105","2/158","1/44","1/54","1/58","1/135","1/199","1/200","1/200","1/200","1/200","1/200","1/200","1/200"],[2.56816489767419e-07,1.215000145553094e-05,1.215000145553094e-05,0.0002169706135589574,0.002741529140779137,0.006086585361302025,0.03250762309645384,0.03975687424498529,0.04264235482763042,0.09663500980086953,0.1393349236139555,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619],[4.622696815813541e-06,7.290000873318565e-05,7.290000873318565e-05,0.0009763677610153084,0.009869504906804891,0.01825975608390608,0.08359103081945272,0.08528470965526085,0.08528470965526085,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619,0.139986908731619],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[51.00773195876289,36.71428571428572,36.71428571428572,31.37183544303797,29.69678864824496,19.55522682445759,33.12624584717608,26.86253369272237,24.97243107769424,10.58155650319829,7.138167388167388,7.101938262742283,7.101938262742283,7.101938262742283,7.101938262742283,7.101938262742283,7.101938262742283,7.101938262742283],[774.0374385221934,415.5389408534144,415.5389408534144,264.644918015469,175.1884666472352,99.76427591098042,113.499815495736,86.63093286520242,78.78570447797571,24.72713131619311,14.06843366455485,13.96387625016532,13.96387625016532,13.96387625016532,13.96387625016532,13.96387625016532,13.96387625016532,13.96387625016532],["CDC42;RACGAP1;PLK1;CDK1;KIF23","RACGAP1;PLK1;CDK1;KIF23","RACGAP1;PLK1;CDK1;BRCA1","CDK2;BRCA1;RHOB","CDK2;CDK1","CDK2;RHOB","BRCA1","RHOA","PLK1","CDK1","RHOB","RHOB","CDK1","BRCA1","RAC2","PLK1","CDK2","RHOB"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Term<\/th>\n      <th>Overlap<\/th>\n      <th>P.value<\/th>\n      <th>Adjusted.P.value<\/th>\n      <th>Old.P.value<\/th>\n      <th>Old.Adjusted.P.value<\/th>\n      <th>Odds.Ratio<\/th>\n      <th>Combined.Score<\/th>\n      <th>Genes<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[3,4,5,6,7,8]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>When supplying the <code>gene_set_library</code> argument, rbioapi
assumes you are entering a regex pattern. You can disable this by
setting <code>regex_library_name</code> to <code>FALSE</code>. However,
this feature is useful if you need -for example- partial matches in the
library names. Suppose you want to perform the enrichment analysis on
every library available in Enrichr that contains the name “MSig”. You
can do the following:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Request the enrichment analysis</span></span>
<span><span class="va">results_msig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr.html">rba_enrichr</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span>,</span>
<span>                            gene_set_library <span class="op">=</span> <span class="st">"msig"</span>,</span>
<span>                            regex_library_name <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># You can drop `regex_library_name = TRUE`, as it is TRUE by default.</span></span></code></pre></div>
<p>Note that when only one Enrichr library is selected, a data frame
with enrichment analysis result will be returned.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">results_msig_hallmark</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'data.frame':    18 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;  $ Term                : chr  "Mitotic Spindle" "G2-M Checkpoint" "E2F Targets" "Apoptosis" ...</span></span>
<span><span class="co">#&gt;  $ Overlap             : chr  "5/199" "4/200" "4/200" "3/161" ...</span></span>
<span><span class="co">#&gt;  $ P.value             : num  2.57e-07 1.22e-05 1.22e-05 2.17e-04 2.74e-03 ...</span></span>
<span><span class="co">#&gt;  $ Adjusted.P.value    : num  4.62e-06 7.29e-05 7.29e-05 9.76e-04 9.87e-03 ...</span></span>
<span><span class="co">#&gt;  $ Old.P.value         : int  0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;  $ Old.Adjusted.P.value: int  0 0 0 0 0 0 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;  $ Odds.Ratio          : num  51 36.7 36.7 31.4 29.7 ...</span></span>
<span><span class="co">#&gt;  $ Combined.Score      : num  774 416 416 265 175 ...</span></span>
<span><span class="co">#&gt;  $ Genes               : chr  "CDC42;RACGAP1;PLK1;CDK1;KIF23" "RACGAP1;PLK1;CDK1;KIF23" "RACGAP1;PLK1;CDK1;BRCA1" "CDK2;BRCA1;RHOB" ...</span></span></code></pre></div>
<p>But when multiple libraries have been selected, the function’s output
will be a list where each element is a data frame corresponding to one
of the selected libraries.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">results_msig</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ MSigDB_Computational       :'data.frame': 195 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;  $ MSigDB_Oncogenic_Signatures:'data.frame': 26 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;  $ MSigDB_Hallmark_2020       :'data.frame': 18 obs. of  9 variables:</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 class="heading3" id="approach-2-going-step-by-step">Approach 2: Going step-by-step<a class="anchor" aria-label="anchor" href="#approach-2-going-step-by-step"></a>
</h3>
<p><code><a href="../reference/rba_enrichr.html">rba_enrichr()</a></code> is a wrapper function. It internally
executes a sequence of functions necessary to run your analysis.
Alternatively, you could go step by step. We demonstrate these steps in
this section.</p>
<p>First, you need to retrieve the list of available Enrichr libraries.
This step is optional. You can skip it if you already know the name of
your desired libraries or if you want to run the analysis over every
available library.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get a list of available Enrichr libraries</span></span>
<span><span class="va">libs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_libs.html">rba_enrichr_libs</a></span><span class="op">(</span>store_in_options <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now, you need to upload your genes list to Enrichr. By this, an
identifier will be assigned to your submitted list, which is needed for
the next step.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Submit your gene-set to enrichr</span></span>
<span><span class="va">list_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_add_list.html">rba_enrichr_add_list</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<p>From the returned response, we need the numeric ID in the
“userListId” element.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">list_id</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ shortId   : chr "330e67926968f58836b5b8293a6f4e4e"</span></span>
<span><span class="co">#&gt;  $ userListId: int 65456914</span></span></code></pre></div>
<p>Finally, we are ready to submit the enrichment analysis request to
Enrichr. Same as explained above for the wrapper function
<code><a href="../reference/rba_enrichr.html">rba_enrichr()</a></code>, we can supply the “gene_set_library”
argument in different ways. Here we will only select the
“Table_Mining_of_CRISPR_Studies” library:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Request the analysis</span></span>
<span><span class="va">results_crispr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_enrich.html">rba_enrichr_enrich</a></span><span class="op">(</span>user_list_id <span class="op">=</span> <span class="va">list_id</span><span class="op">$</span><span class="va">userListId</span>,</span>
<span>                                      gene_set_library <span class="op">=</span> <span class="st">"Table_Mining_of_CRISPR_Studies"</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-2c300ba65b710d05f6eb" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-2c300ba65b710d05f6eb">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46"],["Reaction polymerase chain quantitative time, PMC6627898 (Table1)","Anticancer corresponding targets potential non, PMC5981615 (Table1)","Names alphabetical order corresponding listed, PMC6783930 (Table1)","Supplemental, PMC6879830 (Tablesupplemental)","Oligonucleotides study used, PMC5342713 (Table2)","Oligonucleotides study used, PMC6307807 (Table1)","Delivery involving trials current rna, PMC5485616 (Table2)","Supplemental, PMC5111403 (Tablesupplemental)","Spindle assembly homologues factors list, PMC5371999 (Table1)","Supplemental, PMC5416823 (Tablesupplemental)","Comparison knockout patients mouse models, PMC4860525 (Table1)","Training scanning funtrpmodel deep mutational, PMC6868392 (Table1)","Weaknesses strengths strategies various treatment, PMC6073662 (Table1)","Useful oncology generated animal crispr, PMC4460969 (Table1)","Leukemia acute lncrnas described examples, PMC6387068 (Table2)","Mentioned trials text overview clinical, PMC3840483 (Table1)","Mef immortalized driver lines mutations, PMC5666318 (Table1)","Supplemental, PMC6071863 (Tablesupplemental)","Pathogenic cutaneous melanoma likely primary, PMC6863855 (Table2)","Topology degs ppi network important, PMC6533882 (Table1)","Submitted vus highest number variants, PMC7055088 (Table5)","Small interfering rnas study sequences, PMC5720660 (Table1)","Syrian hamster antibodies tested list, PMC6781508 (Table3)","Tar clusters cloning chromosomal isolated, PMC6586605 (Table1)","Phenotypic modification gemms effects cancer, PMC4558888 (Table1)","Chlamydomonas budding arabidopsis humans homologs, PMC4409525 (Table1)","Micrornas breast regulation involved cancer, PMC6801796 (Table1)","Supplemental, PMC6002376 (Tablesupplemental)","Typically panels large testing breast, PMC6462925 (Table2)","Therapeutics na activities representative development, PMC6957827 (Table1)","Mpkccd inhibitable intact pka kinases, PMC6391403 (Table3)","Diffusion network shows ranking effect, PMC7034926 (Table2)","Gondii tolerance dependency screened immune, PMC6987080 (Table1)","Single inborn immunity errors discoveries, PMC4203950 (Table1)","Sdi scores highest human genes, PMC6985572 (Table2)","Biomarkers genetic suggesting relationship predictive, PMC4907009 (Table2)","Uterus reproductive impaired phenotype mice, PMC6194304 (Table1)","Peak literature curated ai loci, PMC6131244 (Table1)","Genotoxic agents overexpression responsive stress, PMC5874662 (Table1)","Coding silenced non cancer genes, PMC5627056 (Table1)","Supplemental, PMC5338947 (Tablesupplemental)","Supplemental, PMC4654807 (Tablesupplemental)","Supplemental, PMC6542691 (Tablesupplemental)","Supplemental, PMC5906580 (Tablesupplemental)","Supplemental, PMC5859395 (Tablesupplemental)","Supplemental, PMC6303340 (Tablesupplemental)"],["4/6","3/25","3/31","10/2992","2/16","2/16","2/21","8/2739","2/54","3/316","1/6","1/7","1/7","1/7","1/8","1/8","1/9","1/9","1/9","1/10","1/10","1/12","1/14","1/14","1/15","1/16","1/16","1/16","1/17","1/18","1/19","1/20","1/23","1/28","1/30","1/30","1/30","1/33","1/35","1/67","1/405","1/421","1/961","1/1552","1/1845","1/2075"],[3.068439472952938e-12,7.771823542026143e-07,1.514799809161465e-06,8.048214146840604e-06,6.261919708622724e-05,6.261919708622724e-05,0.000109346950025726,0.0003205110949164696,0.0007345555571872974,0.001543908757415917,0.004492068393214533,0.005238916894144343,0.005238916894144343,0.005238916894144343,0.005985242836445906,0.005985242836445906,0.006731046493869444,0.006731046493869444,0.006731046493869444,0.007476328156448588,0.007476328156448588,0.008965326705510297,0.0104522409709381,0.0104522409709381,0.01119491729246133,0.01193707350635959,0.01193707350635959,0.01193707350635959,0.01267870993363306,0.01341982690512007,0.01416042474706076,0.01490050378889858,0.01711763141867997,0.02080249902141888,0.02227283302445056,0.02227283302445056,0.02227283302445056,0.024474471226601,0.02593965870865676,0.049105124607314,0.2643314089029322,0.2732937454721643,0.5223634530816231,0.7024233989376346,0.7659741470189,0.8067249461186436],[1.411482157558351e-10,1.787519414666013e-05,2.322693040714246e-05,9.255446268866694e-05,0.0004800805109944089,0.0004800805109944089,0.0007185656715976281,0.0018429387957697,0.003754395070068409,0.007101980284113216,0.01629621782726286,0.01629621782726286,0.01629621782726286,0.01629621782726286,0.01629621782726286,0.01629621782726286,0.01629621782726286,0.01629621782726286,0.01629621782726286,0.01637671881888738,0.01637671881888738,0.01874568311152153,0.01961090647473361,0.01961090647473361,0.01961090647473361,0.01961090647473361,0.01961090647473361,0.01961090647473361,0.02011105713610762,0.02057706792118411,0.02101224317305791,0.0214194741965417,0.02386094076543269,0.02769054916553313,0.02769054916553313,0.02769054916553313,0.02769054916553313,0.02962699148483279,0.03059549488713362,0.0564708932984111,0.2965669465740215,0.299321721231418,0.558807414924527,0.7343517352529816,0.782995794730431,0.8067249461186436],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[3633.272727272727,226.8522727272727,178.1875,11.4037558685446,219.4615384615385,219.4615384615385,161.668016194332,7.220379766699796,58.97337278106509,15.71246006389776,285.4285714285714,237.8452380952381,237.8452380952381,237.8452380952381,203.8571428571429,203.8571428571429,178.3660714285714,178.3660714285714,178.3660714285714,158.5396825396825,158.5396825396825,129.7012987012987,109.7362637362637,109.7362637362637,101.8928571428571,95.0952380952381,95.0952380952381,95.0952380952381,89.14732142857143,83.89915966386555,79.23412698412699,75.06015037593986,64.81493506493507,52.7989417989418,49.15270935960591,49.15270935960591,49.15270935960591,44.53794642857143,41.91386554621849,21.55735930735931,3.461987270155587,3.327380952380953,1.415550595238095,0.8489453808602745,0.7027037496126434,0.6168549386967902],[96317.52227042026,3191.264950226428,2387.752996121,133.7667443979848,2124.04503914007,2124.04503914007,1474.571502686574,58.09224170753369,425.5663020854597,101.7136348579706,1542.86759275583,1249.0776854949,1249.0776854949,1249.0776854949,1043.434298452164,1043.434298452164,892.013120036697,892.013120036697,892.013120036697,776.212425317664,776.212425317664,611.4626002996463,500.5003914455875,500.5003914455875,457.7328151534797,421.0918230074216,421.0918230074216,421.0918230074216,389.3804408242713,361.6911269349327,337.3237811891663,315.7300332957057,263.6442288281798,204.4735196670539,186.9959576603662,186.9959576603662,186.9959576603662,165.2413349284794,153.0686933026361,64.96939440904092,4.606352792738551,4.31630542844973,0.9192467552606759,0.2998635742829445,0.1873456405030657,0.1324834796990502],["CDC42;CDK2;CDK1;RHOA","PLK1;CDK2;CDK1","CDK2;CDK1;BRCA1","CDC42;RBM3;RACGAP1;PLK1;CDK2;CDK1;PHF14;KIF23;BRCA1;MSL1","P53;RHOB","P53;RHOB","PLK1;P53","CDC42;RBM3;RACGAP1;PLK1;CDK1;PHF14;KIF23;RHOA","PLK1;CDK1","PLK1;CDK2;CDK1","BRCA1","BRCA1","P53","P53","P53","P53","BRCA1","P53","BRCA1","PLK1","BRCA1","CDK2","P53","BRCA1","P53","PLK1","P53","BRCA1","BRCA1","P53","CDK2","PLK1","CDC42","RAC2","RHOA","CDK2","P53","RHOB","P53","BRCA1","RACGAP1","MSL1","CDC42","KIF23","RHOB","CDK2"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Term<\/th>\n      <th>Overlap<\/th>\n      <th>P.value<\/th>\n      <th>Adjusted.P.value<\/th>\n      <th>Old.P.value<\/th>\n      <th>Old.Adjusted.P.value<\/th>\n      <th>Odds.Ratio<\/th>\n      <th>Combined.Score<\/th>\n      <th>Genes<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[3,4,5,6,7,8]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><hr>
</div>
</div>
<div class="section level2">
<h2 class="heading2" id="other-species">Working with Other Species<a class="anchor" aria-label="anchor" href="#other-species"></a>
</h2>
<p>Enrichr also provides libraries for model organisms. The following
functions have an <code>organism</code> argument that allows you to
perform the analysis on species other than humans:</p>
<ol style="list-style-type: decimal">
<li><p><code><a href="../reference/rba_enrichr.html">rba_enrichr()</a></code></p></li>
<li><p><code><a href="../reference/rba_enrichr_enrich.html">rba_enrichr_enrich()</a></code></p></li>
<li><p><code><a href="../reference/rba_enrichr_gene_map.html">rba_enrichr_gene_map()</a></code></p></li>
<li><p><code><a href="../reference/rba_enrichr_libs.html">rba_enrichr_libs()</a></code></p></li>
</ol>
<p>The available options for the organism argument are human”, (H.
sapiens &amp; M. musculus), fly” (D. melanogaster), “yeast” (S.
cerevisiae), “worm” (C. elegans), and “fish” (D. rerio).</p>
<hr>
</div>
<div class="section level2">
<h2 class="heading2" id="see-also-in-functions-manuals">See also in Functions’ manuals<a class="anchor" aria-label="anchor" href="#see-also-in-functions-manuals"></a>
</h2>
<p>Some rbioapi Enrichr functions were not covered in this vignette, be
sure to check their manuals:</p>
<ul>
<li><p><code><a href="../reference/rba_enrichr_gene_map.html">rba_enrichr_gene_map()</a></code></p></li>
<li><p><code><a href="../reference/rba_enrichr_view_list.html">rba_enrichr_view_list()</a></code></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 class="heading2" id="citations">How to Cite?<a class="anchor" aria-label="anchor" href="#citations"></a>
</h2>
<p>To cite Enrichr (Please see <a href="https://maayanlab.cloud/Enrichr/help#terms" class="external-link uri">https://maayanlab.cloud/Enrichr/help#terms</a>):</p>
<ul>
<li><p>Chen, E.Y., Tan, C.M., Kou, Y. <em>et al.</em> Enrichr:
interactive and collaborative HTML5 gene list enrichment analysis tool.
<em>Bioinformatics</em> <strong>14,</strong> 128 (2013). <a href="https://doi.org/10.1186/1471-2105-14-128" class="external-link uri">https://doi.org/10.1186/1471-2105-14-128</a></p></li>
<li><p>Maxim V. Kuleshov, Matthew R. Jones, Andrew D. Rouillard, Nicolas
F. Fernandez, Qiaonan Duan, Zichen Wang, Simon Koplev, Sherry L.
Jenkins, Kathleen M. Jagodnik, Alexander Lachmann, Michael G. McDermott,
Caroline D. Monteiro, Gregory W. Gundersen, Avi Ma’ayan, Enrichr: a
comprehensive gene set enrichment analysis web server 2016 update,
<em>Nucleic Acids Research</em>, Volume 44, Issue W1, 8 July 2016, Pages
W90–W97, <a href="https://doi.org/10.1093/nar/gkw377" class="external-link uri">https://doi.org/10.1093/nar/gkw377</a></p></li>
<li><p>Xie, Z., Bailey, A., Kuleshov, M. V., Clarke, D. J. B.,
Evangelista, J. E., Jenkins, S. L., Lachmann, A., Wojciechowicz, M. L.,
Kropiwnicki, E., Jagodnik, K. M., Jeon, M., &amp; Ma’ayan, A. (2021).
Gene set knowledge discovery with Enrichr. <em>Current Protocols</em>,
1, e90. doi: 10.1002/cpz1.90</p></li>
</ul>
<p>To cite rbioapi:</p>
<ul>
<li>Moosa Rezwani, Ali Akbar Pourfathollah, Farshid Noorbakhsh, rbioapi:
user-friendly R interface to biologic web services’ API, Bioinformatics,
Volume 38, Issue 10, 15 May 2022, Pages 2952–2953, <a href="https://doi.org/10.1093/bioinformatics/btac172" class="external-link uri">https://doi.org/10.1093/bioinformatics/btac172</a>
</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 class="heading2" id="other-services">Over-representation analysis Using Other
Services<a class="anchor" aria-label="anchor" href="#other-services"></a>
</h2>
<p>Other services supported by rbioapi also provide Over-representation
analysis tools. Please see the vignette article <a href="rbioapi_do_enrich.html">Do with rbioapi: Over-Representation
(Enrichment) Analysis in R</a> (<a href="https://rbioapi.moosa-r.com/articles/rbioapi_do_enrich.html">link
to the documentation site</a>) for an in-depth review.</p>
<hr>
</div>
<div class="section level2">
<h2 class="heading2" id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_enrichr.html" title="https://rbioapi.moosa-r.com/articles/rbioapi_enrichr.html">This
article in rbioapi documentation site</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html#section-enrichr-rba-enrichr-" title="rbioapi reference">Functions references in rbioapi
documentation site</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi
vignette index</a></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 class="heading2" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] rbioapi_0.7.9</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] vctrs_0.6.3       httr_1.4.7        cli_3.6.1         knitr_1.43       </span></span>
<span><span class="co">#&gt;  [5] rlang_1.1.1       xfun_0.40         stringi_1.7.12    purrr_1.0.2      </span></span>
<span><span class="co">#&gt;  [9] textshaping_0.3.6 jsonlite_1.8.7    glue_1.6.2        DT_0.29          </span></span>
<span><span class="co">#&gt; [13] rprojroot_2.0.3   htmltools_0.5.6   ragg_1.2.5        sass_0.4.7       </span></span>
<span><span class="co">#&gt; [17] rmarkdown_2.24    crosstalk_1.2.0   evaluate_0.21     jquerylib_0.1.4  </span></span>
<span><span class="co">#&gt; [21] ellipsis_0.3.2    fastmap_1.1.1     yaml_2.3.7        lifecycle_1.0.3  </span></span>
<span><span class="co">#&gt; [25] memoise_2.0.1     stringr_1.5.0     compiler_4.3.1    fs_1.6.3         </span></span>
<span><span class="co">#&gt; [29] htmlwidgets_1.6.2 systemfonts_1.0.4 digest_0.6.33     R6_2.5.1         </span></span>
<span><span class="co">#&gt; [33] curl_5.0.2        magrittr_2.0.3    bslib_0.5.1       tools_4.3.1      </span></span>
<span><span class="co">#&gt; [37] pkgdown_2.0.7     cachem_1.0.8      desc_1.4.2</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://moosa-r.com/" class="external-link">Moosa Rezwani</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
