<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Connect to STRING in R with rbioapi package.
">
<title>2.F: STRING &amp; rbioapi • rbioapi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2.F: STRING &amp; rbioapi">
<meta property="og:description" content="Connect to STRING in R with rbioapi package.
">
<meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/rbioapi.html">Get Started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles-amp-supported-services">Articles &amp; Supported Services</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles-amp-supported-services">
    <a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://moosa-r.com/" aria-label="Moosa Rezwani Portfolio Website">
    <span class="fa fa-solid fa-id-card"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="rbioapi_string_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="rbioapi_string_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_string_files/datatables-binding-0.27/datatables.js"></script><link href="rbioapi_string_files/dt-core-1.12.1/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_string_files/dt-core-1.12.1/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_string_files/dt-core-1.12.1/js/jquery.dataTables.min.js"></script><link href="rbioapi_string_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="rbioapi_string_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>2.F: STRING &amp; rbioapi</h1>
                        <h4 data-toc-skip class="author">Moosa
Rezwani</h4>
            
            <h4 data-toc-skip class="date">2023-05-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/HEAD/vignettes/rbioapi_string.Rmd" class="external-link"><code>vignettes/rbioapi_string.Rmd</code></a></small>
      <div class="d-none name"><code>rbioapi_string.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level3">
<h3 class="heading2" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>STRING is a comprehensive database of protein-protein interactions
(PPI) that in version 11.0, covers 24,584,628 proteins from 5,090
organisms. Directly quoting from their paper:</p>
<blockquote>
<p>The STRING database aims to collect, score and integrate all publicly
available sources of protein–protein interaction information, and to
complement these with computational predictions. Its goal is to achieve
a comprehensive and objective global network, including direct
(physical) as well as indirect (functional) interactions.</p>
<p>(source: Szklarczyk, Damian, et al. “STRING v11: protein–protein
association networks with increased coverage, supporting functional
discovery in genome-wide experimental datasets.” <em>Nucleic acids
research</em> 47.D1 (2019): D607-D613. )</p>
</blockquote>
<hr>
</div>
<div class="section level3">
<h3 class="heading2" id="note-about-species-argument">Note about species argument<a class="anchor" aria-label="anchor" href="#note-about-species-argument"></a>
</h3>
<p>You can find an argument named “species” in every rbioapi STRING
function. Providing the species argument is not mandatory, but it has
been recommended in STRING API’s documentation to always specify the
species. An exception is when your input proteins’ vector length is more
than 100; In such cases, the species argument is required. Otherwise,
calling the function without providing the species will produce an
ERROR.</p>
<hr>
</div>
<div class="section level3">
<h3 class="heading2" id="map-your-ids-to-string-ids">Map your IDs to STRING IDs<a class="anchor" aria-label="anchor" href="#map-your-ids-to-string-ids"></a>
</h3>
<p>Although STRING API resources will handle and recognize a variety of
identifiers, it is recommended that you first map your IDs to STRING IDs
before using them in other rbioapi STRING functions.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1 We create a variable with our genes' NCBI IDs</span></span>
<span><span class="va">proteins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>,<span class="st">"CDK1"</span>,<span class="st">"KIF23"</span>,</span>
<span>              <span class="st">"PLK1"</span>,<span class="st">"RAC2"</span>,<span class="st">"RACGAP1"</span>,<span class="st">"RHOA"</span>,<span class="st">"RHOB"</span>, <span class="st">"PHF14"</span>, <span class="st">"RBM3"</span><span class="op">)</span></span>
<span><span class="co">## 2 Now we map our protein IDs</span></span>
<span><span class="va">proteins_mapped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_map_ids.html">rba_string_map_ids</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">proteins</span>,</span>
<span>                                      species <span class="op">=</span> <span class="fl">9606</span><span class="op">)</span></span>
<span><span class="co">## 3 What we need and will use for the rest of this vignette is the `stringId` column</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 class="heading2" id="get-interaction-network-of-a-protein-set">Get interaction network of a protein set<a class="anchor" aria-label="anchor" href="#get-interaction-network-of-a-protein-set"></a>
</h3>
<p>You can retrieve a list of interactions that the proteins in your set
have with each other along with the STRING annotations of each
interaction. You may filter the results by using required_score and
network_type arguments.</p>
<p>See the ‘values’ section <code>rba_string_interactions_network</code>
function’s manual for information on the returned columns.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">int_net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_interactions_network.html">rba_string_interactions_network</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>                                          species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>                                          required_score <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-1819227ed20757fff0b8" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-1819227ed20757fff0b8">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70"],["9606.ENSP00000249071","9606.ENSP00000249071","9606.ENSP00000249071","9606.ENSP00000249071","9606.ENSP00000249071","9606.ENSP00000249071","9606.ENSP00000249071","9606.ENSP00000249071","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000272233","9606.ENSP00000272233","9606.ENSP00000272233","9606.ENSP00000272233","9606.ENSP00000272233","9606.ENSP00000272233","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000383118","9606.ENSP00000383118","9606.ENSP00000383118","9606.ENSP00000383118","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000404190","9606.ENSP00000404190"],["9606.ENSP00000404190","9606.ENSP00000404190","9606.ENSP00000272233","9606.ENSP00000272233","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000383118","9606.ENSP00000383118","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000418960","9606.ENSP00000418960","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000404190","9606.ENSP00000404190","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000418960","9606.ENSP00000418960","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000383118","9606.ENSP00000383118","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000418960","9606.ENSP00000418960","9606.ENSP00000404190","9606.ENSP00000404190","9606.ENSP00000383118","9606.ENSP00000383118","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000418960","9606.ENSP00000418960","9606.ENSP00000404190","9606.ENSP00000404190","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000383118","9606.ENSP00000383118","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000418960","9606.ENSP00000418960","9606.ENSP00000404190","9606.ENSP00000404190","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000404190","9606.ENSP00000404190","9606.ENSP00000404190","9606.ENSP00000404190","9606.ENSP00000418960","9606.ENSP00000418960"],["RAC2","RAC2","RAC2","RAC2","RAC2","RAC2","RAC2","RAC2","KIF23","KIF23","KIF23","KIF23","KIF23","KIF23","KIF23","KIF23","KIF23","KIF23","KIF23","KIF23","CDK2","CDK2","CDK2","CDK2","CDK2","CDK2","CDK2","CDK2","CDK2","CDK2","TP53","TP53","TP53","TP53","TP53","TP53","TP53","TP53","TP53","TP53","RHOB","RHOB","RHOB","RHOB","RHOB","RHOB","PLK1","PLK1","PLK1","PLK1","PLK1","PLK1","PLK1","PLK1","CDK1","CDK1","CDK1","CDK1","CDK1","CDK1","CDK1","CDK1","CDC42","CDC42","CDC42","CDC42","RHOA","RHOA","RACGAP1","RACGAP1"],["RACGAP1","RACGAP1","RHOB","RHOB","RHOA","RHOA","CDC42","CDC42","CDK2","CDK2","BRCA1","BRCA1","RHOA","RHOA","CDK1","CDK1","RACGAP1","RACGAP1","PLK1","PLK1","PLK1","PLK1","RHOA","RHOA","CDK1","CDK1","BRCA1","BRCA1","TP53","TP53","CDC42","CDC42","RHOA","RHOA","CDK1","CDK1","PLK1","PLK1","BRCA1","BRCA1","RACGAP1","RACGAP1","CDC42","CDC42","RHOA","RHOA","RHOA","RHOA","BRCA1","BRCA1","RACGAP1","RACGAP1","CDK1","CDK1","CDC42","CDC42","RHOA","RHOA","BRCA1","BRCA1","RACGAP1","RACGAP1","RHOA","RHOA","RACGAP1","RACGAP1","RACGAP1","RACGAP1","BRCA1","BRCA1"],["9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606"],[0.627,0.627,0.702,0.702,0.886,0.886,0.918,0.918,0.509,0.509,0.707,0.707,0.974,0.974,0.995,0.995,0.999,0.999,0.999,0.999,0.601,0.601,0.955,0.955,0.987,0.987,0.995,0.995,0.999,0.999,0.563,0.563,0.606,0.606,0.977,0.977,0.989,0.989,0.999,0.999,0.58,0.58,0.587,0.587,0.902,0.902,0.757,0.757,0.94,0.94,0.99,0.99,0.996,0.996,0.564,0.564,0.668,0.668,0.906,0.906,0.99,0.99,0.978,0.978,0.994,0.994,0.98,0.98,0.697,0.697],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0.43,0.43,0.436,0.436,0.442,0.442,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.441,0.441,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.405,0.405,0.448,0.448,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.418,0.418,0,0,0,0,0,0],[0.063,0.063,0.08,0.08,0.109,0.109,0.246,0.246,0.307,0.307,0.381,0.381,0.049,0.049,0.979,0.979,0.947,0.947,0.928,0.928,0.283,0.283,0.052,0.052,0.68,0.68,0.159,0.159,0.095,0.095,0.049,0.049,0,0,0.096,0.096,0.096,0.096,0.076,0.076,0,0,0.08,0.08,0,0,0,0,0.259,0.259,0.877,0.877,0.963,0.963,0.062,0.062,0.052,0.052,0.445,0.445,0.962,0.962,0.268,0.268,0.062,0.062,0.062,0.062,0.356,0.356],[0.117,0.117,0.37,0.37,0.37,0.37,0.481,0.481,0.152,0.152,0.459,0.459,0.086,0.086,0.152,0.152,0.993,0.993,0.621,0.621,0.23,0.23,0.102,0.102,0.617,0.617,0.838,0.838,0.79,0.79,0,0,0,0,0.68,0.68,0.738,0.738,0.9,0.9,0.128,0.128,0.37,0.37,0,0,0.148,0.148,0.763,0.763,0.721,0.721,0.23,0.23,0.102,0.102,0.102,0.102,0.487,0.487,0.27,0.27,0.705,0.705,0.82,0.82,0.139,0.139,0.27,0.27],[0,0,0.5,0.5,0.8,0.8,0.8,0.8,0,0,0,0,0.9,0.9,0,0,0.9,0.9,0.9,0.9,0,0,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0.9,0,0,0,0,0,0,0,0,0.9,0.9,0,0,0.3,0.3,0.9,0.9,0,0,0,0,0,0,0.8,0.8,0,0,0,0,0,0,0,0,0.9,0.9,0.9,0.9,0.9,0.9,0,0],[0.585,0.585,0.628,0.628,0.928,0.928,0.877,0.877,0.232,0.232,0.196,0.196,0.745,0.745,0.777,0.777,0.995,0.995,0.935,0.935,0.838,0.838,0.541,0.541,0.978,0.978,0.694,0.694,0.97,0.97,0.559,0.559,0.606,0.606,0.927,0.927,0.958,0.958,0.989,0.989,0.538,0.538,0.783,0.783,0.827,0.827,0.726,0.726,0.687,0.687,0.741,0.741,0.989,0.989,0.524,0.524,0.641,0.641,0.697,0.697,0.685,0.685,0.958,0.958,0.724,0.724,0.785,0.785,0.407,0.407]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stringId_A<\/th>\n      <th>stringId_B<\/th>\n      <th>preferredName_A<\/th>\n      <th>preferredName_B<\/th>\n      <th>ncbiTaxonId<\/th>\n      <th>score<\/th>\n      <th>nscore<\/th>\n      <th>fscore<\/th>\n      <th>pscore<\/th>\n      <th>ascore<\/th>\n      <th>escore<\/th>\n      <th>dscore<\/th>\n      <th>tscore<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[6,7,8,9,10,11,12,13]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div class="section level3">
<h3 class="heading2" id="get-interaction-partners-of-a-protein-set">Get interaction partners of a protein set<a class="anchor" aria-label="anchor" href="#get-interaction-partners-of-a-protein-set"></a>
</h3>
<p>In the last example, we only obtained the interaction which our
proteins have among themselves, what if we wanted to get a list of every
protein which interact with our protein(s)?</p>
<p>To do that, we can use
<code>rba_string_interaction_partners</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Although we supply only one protein ID here (CD40 protein), you can provide a vector of proteins as the input</span></span>
<span><span class="va">int_partners</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_interaction_partners.html">rba_string_interaction_partners</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="st">"9606.ENSP00000361359"</span>,</span>
<span>                                               species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>                                               required_score <span class="op">=</span> <span class="fl">900</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-70b1dc37046b5843a384" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-70b1dc37046b5843a384">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10"],["9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359"],["9606.ENSP00000359663","9606.ENSP00000247668","9606.ENSP00000362994","9606.ENSP00000433623","9606.ENSP00000454207","9606.ENSP00000261464","9606.ENSP00000384273","9606.ENSP00000391676","9606.ENSP00000477613","9606.ENSP00000295025"],["CD40","CD40","CD40","CD40","CD40","CD40","CD40","CD40","CD40","CD40"],["CD40LG","TRAF2","TRAF1","TRAF6","TRAF3","TRAF5","RELA","JAK3","BIRC2","REL"],[9606,9606,9606,9606,9606,9606,9606,9606,9606,9606],[0.999,0.999,0.999,0.999,0.999,0.998,0.998,0.997,0.991,0.989],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0.062,0.084,0.175,0.062,0.062,0.062,0,0.085,0.064,0.063],[0.932,0.978,0.747,0.918,0.983,0.876,0,0.243,0.719,0],[0.9,0.9,0.9,0.9,0.9,0.54,0.9,0.9,0.9,0],[0.995,0.985,0.958,0.989,0.99,0.98,0.989,0.973,0.725,0.989]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stringId_A<\/th>\n      <th>stringId_B<\/th>\n      <th>preferredName_A<\/th>\n      <th>preferredName_B<\/th>\n      <th>ncbiTaxonId<\/th>\n      <th>score<\/th>\n      <th>nscore<\/th>\n      <th>fscore<\/th>\n      <th>pscore<\/th>\n      <th>ascore<\/th>\n      <th>escore<\/th>\n      <th>dscore<\/th>\n      <th>tscore<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[5,6,7,8,9,10,11,12,13]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div class="section level3">
<h3 class="heading2" id="get-network-image-of-a-protein-set">Get network image of a protein set<a class="anchor" aria-label="anchor" href="#get-network-image-of-a-protein-set"></a>
</h3>
<p>Let’s go back to the interaction network. As you must have seen in
the STRING webpages, STRING plots the interaction network of your
proteins with many customizations available. You can also do that with
STRING API services. <code>rba_string_network_image</code> function is
very flexible and you have a variety of options; see the function’s
manual.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Example 1:</span></span>
<span><span class="va">graph_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_network_image.html">rba_string_network_image</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>                                   image_format <span class="op">=</span> <span class="st">"image"</span>,</span>
<span>                                   species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>                                   save_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                   required_score <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                   network_flavor <span class="op">=</span> <span class="st">"confidence"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="rbioapi_string_files/figure-html/rba_string_network_image_ex1_image-1.png" alt="Network images - Example 1" width="700"><p class="caption">
Network images - Example 1
</p>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Example 2:</span></span>
<span><span class="va">graph_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_network_image.html">rba_string_network_image</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>                                    image_format <span class="op">=</span> <span class="st">"image"</span>,</span>
<span>                                    species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>                                    save_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                    required_score <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                                    add_color_nodes <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                    add_white_nodes <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                   network_flavor <span class="op">=</span> <span class="st">"actions"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="rbioapi_string_files/figure-html/rba_string_network_image_ex2_image-1.png" alt="Network images - Example 2" width="700"><p class="caption">
Network images - Example 2
</p>
</div>
<hr>
</div>
<div class="section level3">
<h3 class="heading2" id="enrichment-using-string">Enrichment using STRING<a class="anchor" aria-label="anchor" href="#enrichment-using-string"></a>
</h3>
<p>STRING let you perform two types of enrichments. See <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3531103/" title="STRING v9.1: protein-protein interaction networks, with increased coverage and integration" class="external-link">STRING’s
paper</a> for more information.</p>
<div class="section level4">
<h4 class="heading3" id="functional-enrichment">Functional enrichment<a class="anchor" aria-label="anchor" href="#functional-enrichment"></a>
</h4>
<p>The first type is the conventional type, which statistically tests
your supplied gene sets against some sets of annotation. Currently,
STRING supports Gene Ontology, KEGG pathways, UniProt Keywords, PubMed
publications, Pfam domains, InterPro domains, and SMART domains. (<a href="https://version11.string-db.org/help/api/#getting-functional-enrichment" title="STRING API - Getting functional enrichment" class="external-link">source</a>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enriched</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_enrichment.html">rba_string_enrichment</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>                                  species <span class="op">=</span> <span class="fl">9606</span>, <span class="op">)</span></span></code></pre></div>
<p>As usual, we inspect the output using the <code><a href="https://rdrr.io/r/utils/str.html" class="external-link">str()</a></code>
function. As you can see below, the enrichment results of each category
can be found as the returned list’s elements.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">enriched</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 14</span></span>
<span><span class="co">#&gt;  $ COMPARTMENTS     :'data.frame':   23 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Component        :'data.frame':   21 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ DISEASES         :'data.frame':   5 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Function         :'data.frame':   14 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ InterPro         :'data.frame':   3 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ KEGG             :'data.frame':   45 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Keyword          :'data.frame':   14 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ NetworkNeighborAL:'data.frame':   6 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Pfam             :'data.frame':   3 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ PMID             :'data.frame':   100 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Process          :'data.frame':   204 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ RCTM             :'data.frame':   55 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ TISSUES          :'data.frame':   14 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ WikiPathways     :'data.frame':   42 obs. of  10 variables:</span></span></code></pre></div>
<p>Let us see the “DISEASES” results as an example. Below, we can see
which terms of the <a href="https://disease-ontology.org/" title="Human Disease Ontology" class="external-link">Human Disease Ontology</a> were
over-represented:</p>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-4824ed06e364ef7e129c" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-4824ed06e364ef7e129c">{"x":{"filter":"none","vertical":false,"data":[["277","278","279","280","281"],["DISEASES","DISEASES","DISEASES","DISEASES","DISEASES"],["DOID:3012","DOID:3114","DOID:0060071","DOID:4001","DOID:768"],[2,2,2,2,2],[4,4,17,16,15],[9606,9606,9606,9606,9606],[["9606.ENSP00000269305","9606.ENSP00000418960"],["9606.ENSP00000269305","9606.ENSP00000418960"],["9606.ENSP00000269305","9606.ENSP00000418960"],["9606.ENSP00000269305","9606.ENSP00000418960"],["9606.ENSP00000266970","9606.ENSP00000269305"]],[["TP53","BRCA1"],["TP53","BRCA1"],["TP53","BRCA1"],["TP53","BRCA1"],["CDK2","TP53"]],[7.11e-06,7.11e-06,8.06e-05,7.22e-05,6.42e-05],[0.031,0.031,0.0467,0.0467,0.0467],["Li-Fraumeni syndrome","Serous cystadenocarcinoma","Pre-malignant neoplasm","Ovarian carcinoma","Retinoblastoma"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>category<\/th>\n      <th>term<\/th>\n      <th>number_of_genes<\/th>\n      <th>number_of_genes_in_background<\/th>\n      <th>ncbiTaxonId<\/th>\n      <th>inputGenes<\/th>\n      <th>preferredNames<\/th>\n      <th>p_value<\/th>\n      <th>fdr<\/th>\n      <th>description<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[3,4,5,8,9]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p><strong>Please Note:</strong> Other services supported by rbioapi
also provide Over-representation analysis tools. Please see the vignette
article <a href="rbioapi_do_enrich.html">Do with rbioapi:
Over-Representation (Enrichment) Analysis in R</a> (<a href="https://rbioapi.moosa-r.com/articles/rbioapi_do_enrich.html">link
to the documentation site</a>) for an in-depth review.</p>
</div>
<div class="section level4">
<h4 class="heading3" id="protein-protein-interaction-enrichment">Protein-protein interaction enrichment<a class="anchor" aria-label="anchor" href="#protein-protein-interaction-enrichment"></a>
</h4>
<p>Even without incorporating annotation data, STRING can calculate if
your proteins are functionally related. Briefly, STRING accomplishes
this by comparing the interactions’ distribution in your protein-set to
the interactions’ distribution in the proteome. Read <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3531103/#__sec4title" title="STRING v9.1: protein-protein interaction networks, with increased coverage and integration" class="external-link">STRING’s
paper</a> for more information.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rba_string_enrichment_ppi.html">rba_string_enrichment_ppi</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>                          species <span class="op">=</span> <span class="fl">9606</span><span class="op">)</span></span>
<span><span class="co">#&gt; $number_of_nodes</span></span>
<span><span class="co">#&gt; [1] 14</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $number_of_edges</span></span>
<span><span class="co">#&gt; [1] 40</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $average_node_degree</span></span>
<span><span class="co">#&gt; [1] 5.71</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $local_clustering_coefficient</span></span>
<span><span class="co">#&gt; [1] 0.617</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $expected_number_of_edges</span></span>
<span><span class="co">#&gt; [1] 16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $p_value</span></span>
<span><span class="co">#&gt; [1] 2.58e-07</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level3">
<h3 class="heading2" id="get-functional-annotations">Get functional annotations<a class="anchor" aria-label="anchor" href="#get-functional-annotations"></a>
</h3>
<p>As you have seen above, STRING maps the proteins to multiple
annotation sources. You can obtain any annotation associated with your
proteins without performing enrichment analysis and retrieving just the
significant portion.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_annotations.html">rba_string_annotations</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="st">"9606.ENSP00000269305"</span>,</span>
<span>                                     species <span class="op">=</span> <span class="fl">9606</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## This function returns large results, so the results was not shown in this vignette.</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 class="heading2" id="see-also-in-functions-manuals">See also in Functions’ manuals<a class="anchor" aria-label="anchor" href="#see-also-in-functions-manuals"></a>
</h3>
<p>Some rbioapi STRING functions were not covered in this vignette,
please check their manuals:</p>
<ul>
<li><p><code><a href="../reference/rba_string_homology_intra.html">rba_string_homology_intra()</a></code></p></li>
<li><p><code><a href="../reference/rba_string_homology_inter.html">rba_string_homology_inter()</a></code></p></li>
<li><p><code><a href="../reference/rba_string_version.html">rba_string_version()</a></code></p></li>
</ul>
<hr>
</div>
<div class="section level3">
<h3 class="heading2" id="citations">How to Cite?<a class="anchor" aria-label="anchor" href="#citations"></a>
</h3>
<p>To cite STRING (Please see <a href="https://string-db.org/cgi/about?footer_active_subpage=references" class="external-link uri">https://string-db.org/cgi/about?footer_active_subpage=references</a>):</p>
<ul>
<li>Damian Szklarczyk, Annika L Gable, Katerina C Nastou, David Lyon,
Rebecca Kirsch, Sampo Pyysalo, Nadezhda T Doncheva, Marc Legeay, Tao
Fang, Peer Bork, Lars J Jensen, Christian von Mering, The STRING
database in 2021: customizable protein–protein networks, and functional
characterization of user-uploaded gene/measurement sets, <em>Nucleic
Acids Research</em>, Volume 49, Issue D1, 8 January 2021, Pages
D605–D612, <a href="https://doi.org/10.1093/nar/gkaa1074" class="external-link uri">https://doi.org/10.1093/nar/gkaa1074</a>
</li>
</ul>
<p>To cite rbioapi:</p>
<ul>
<li>Moosa Rezwani, Ali Akbar Pourfathollah, Farshid Noorbakhsh, rbioapi:
user-friendly R interface to biologic web services’ API, Bioinformatics,
Volume 38, Issue 10, 15 May 2022, Pages 2952–2953, <a href="https://doi.org/10.1093/bioinformatics/btac172" class="external-link uri">https://doi.org/10.1093/bioinformatics/btac172</a>
</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 class="heading2" id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_string.html" title="2.E: STRING &amp; rbioapi">This article in rbioapi
documentation site</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html#section-string-rba-string-" title="rbioapi reference">Functions references in rbioapi
documentation site</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi
vignette index</a></p></li>
</ul>
</div>
<div class="section level2">
<h2 class="heading2" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] rbioapi_0.7.7</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] jsonlite_1.8.4    compiler_4.3.0    highr_0.10        stringr_1.5.0    </span></span>
<span><span class="co">#&gt;  [5] jquerylib_0.1.4   systemfonts_1.0.4 textshaping_0.3.6 png_0.1-8        </span></span>
<span><span class="co">#&gt;  [9] yaml_2.3.7        fastmap_1.1.1     R6_2.5.1          curl_5.0.0       </span></span>
<span><span class="co">#&gt; [13] knitr_1.42        htmlwidgets_1.6.2 desc_1.4.2        rprojroot_2.0.3  </span></span>
<span><span class="co">#&gt; [17] bslib_0.4.2       rlang_1.1.0       DT_0.27           cachem_1.0.7     </span></span>
<span><span class="co">#&gt; [21] stringi_1.7.12    xfun_0.39         fs_1.6.2          sass_0.4.5       </span></span>
<span><span class="co">#&gt; [25] memoise_2.0.1     cli_3.6.1         pkgdown_2.0.7     magrittr_2.0.3   </span></span>
<span><span class="co">#&gt; [29] crosstalk_1.2.0   digest_0.6.31     grid_4.3.0        lifecycle_1.0.3  </span></span>
<span><span class="co">#&gt; [33] vctrs_0.6.2       evaluate_0.20     glue_1.6.2        ragg_1.2.5       </span></span>
<span><span class="co">#&gt; [37] rmarkdown_2.21    purrr_1.0.1       httr_1.4.5        tools_4.3.0      </span></span>
<span><span class="co">#&gt; [41] ellipsis_0.3.2    htmltools_0.5.5</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://moosa-r.com/" class="external-link">Moosa Rezwani</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
