<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>2.F: STRING &amp; rbioapi • rbioapi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2.F: STRING &amp; rbioapi">
<meta name="description" content="Connect to STRING in R with rbioapi package.
">
<meta property="og:description" content="Connect to STRING in R with rbioapi package.
">
<meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rbioapi.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles-supported-services" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles &amp; Supported Services</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles-supported-services">
<li><a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://moosa-r.com/" aria-label="Moosa Rezwani Portfolio Website"><span class="fa fa-solid fa-id-card"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="rbioapi_string_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="rbioapi_string_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="rbioapi_string_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_string_files/datatables-binding-0.33/datatables.js"></script><link href="rbioapi_string_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_string_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_string_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="rbioapi_string_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="rbioapi_string_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>2.F: STRING &amp; rbioapi</h1>
                        <h4 data-toc-skip class="author">Moosa
Rezwani</h4>
            
            <h4 data-toc-skip class="date">2025-02-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/master/vignettes/rbioapi_string.Rmd" class="external-link"><code>vignettes/rbioapi_string.Rmd</code></a></small>
      <div class="d-none name"><code>rbioapi_string.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>STRING is a comprehensive database of protein-protein interactions
(PPI) that in version 11.0, covers 24,584,628 proteins from 5,090
organisms. Directly quoting from their paper:</p>
<blockquote>
<p>The STRING database aims to collect, score and integrate all publicly
available sources of protein–protein interaction information, and to
complement these with computational predictions. Its goal is to achieve
a comprehensive and objective global network, including direct
(physical) as well as indirect (functional) interactions.</p>
<p>(source: Szklarczyk, Damian, et al. “STRING v11: protein–protein
association networks with increased coverage, supporting functional
discovery in genome-wide experimental datasets.” <em>Nucleic acids
research</em> 47.D1 (2019): D607-D613. )</p>
</blockquote>
<hr>
</div>
<div class="section level2">
<h2 id="note-about-species-argument">Note about species argument<a class="anchor" aria-label="anchor" href="#note-about-species-argument"></a>
</h2>
<p>You can find an argument named “species” in every rbioapi STRING
function. Providing the species argument is not mandatory, but it has
been recommended in STRING API’s documentation to always specify the
species. An exception is when your input proteins’ vector length is more
than 100; In such cases, the species argument is required. Otherwise,
calling the function without providing the species will produce an
ERROR.</p>
<hr>
</div>
<div class="section level2">
<h2 id="map-your-ids-to-string-ids">Map your IDs to STRING IDs<a class="anchor" aria-label="anchor" href="#map-your-ids-to-string-ids"></a>
</h2>
<p>Although STRING API resources will handle and recognize a variety of
identifiers, it is recommended that you first map your IDs to STRING IDs
before using them in other rbioapi STRING functions.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1 We create a variable with our genes' NCBI IDs</span></span>
<span><span class="va">proteins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>,<span class="st">"CDK1"</span>,<span class="st">"KIF23"</span>,</span>
<span>  <span class="st">"PLK1"</span>,<span class="st">"RAC2"</span>,<span class="st">"RACGAP1"</span>,<span class="st">"RHOA"</span>,<span class="st">"RHOB"</span>, <span class="st">"PHF14"</span>, <span class="st">"RBM3"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 2 Now we map our protein IDs</span></span>
<span><span class="va">proteins_mapped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_map_ids.html">rba_string_map_ids</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="va">proteins</span>, species <span class="op">=</span> <span class="fl">9606</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 3 What we need and will use for the rest of this vignette is the `stringId` column</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14"],[0,1,2,3,4,5,6,7,8,9,10,11,12,13],["p53","BRCA1","cdk2","Q99835","CDC42","CDK1","KIF23","PLK1","RAC2","RACGAP1","RHOA","RHOB","PHF14","RBM3"],["9606.ENSP00000269305","9606.ENSP00000418960","9606.ENSP00000266970","9606.ENSP00000249373","9606.ENSP00000497733","9606.ENSP00000378699","9606.ENSP00000260363","9606.ENSP00000300093","9606.ENSP00000249071","9606.ENSP00000404190","9606.ENSP00000400175","9606.ENSP00000272233","9606.ENSP00000385795","9606.ENSP00000365950"],[9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["TP53","BRCA1","CDK2","SMO","CDC42","CDK1","KIF23","PLK1","RAC2","RACGAP1","RHOA","RHOB","PHF14","RBM3"],["Cellular tumor antigen p53; Acts as a tumor suppressor in many tumor types; induces growth arrest or apoptosis depending on the physiological circumstances and cell type. Involved in cell cycle regulation as a trans-activator that acts to negatively regulate cell division by controlling a set of genes required for this process. One of the activated genes is an inhibitor of cyclin-dependent kinases. Apoptosis induction seems to be mediated either by stimulation of BAX and FAS antigen expression, or by repression of Bcl-2 expression. Its pro-apoptotic activity is activated via its intera [...] ","Breast cancer type 1 susceptibility protein; E3 ubiquitin-protein ligase that specifically mediates the formation of 'Lys-6'-linked polyubiquitin chains and plays a central role in DNA repair by facilitating cellular responses to DNA damage. It is unclear whether it also mediates the formation of other types of polyubiquitin chains. The E3 ubiquitin-protein ligase activity is required for its tumor suppressor function. The BRCA1-BARD1 heterodimer coordinates a diverse range of cellular pathways such as DNA damage repair, ubiquitination and transcriptional regulation to maintain genomic [...] ","Cyclin-dependent kinase 2; Serine/threonine-protein kinase involved in the control of the cell cycle; essential for meiosis, but dispensable for mitosis. Phosphorylates CTNNB1, USP37, p53/TP53, NPM1, CDK7, RB1, BRCA2, MYC, NPAT, EZH2. Triggers duplication of centrosomes and DNA. Acts at the G1-S transition to promote the E2F transcriptional program and the initiation of DNA synthesis, and modulates G2 progression; controls the timing of entry into mitosis/meiosis by controlling the subsequent activation of cyclin B/CDK1 by phosphorylation, and coordinates the activation of cyclin B/CDK [...] ","Smoothened homolog; G protein-coupled receptor that probably associates with the patched protein (PTCH) to transduce the hedgehog's proteins signal. Binding of sonic hedgehog (SHH) to its receptor patched is thought to prevent normal inhibition by patched of smoothened (SMO). Required for the accumulation of KIF7, GLI2 and GLI3 in the cilia. Interacts with DLG5 at the ciliary base to induce the accumulation of KIF7 and GLI2 at the ciliary tip for GLI2 activation (By similarity).","Cell division control protein 42 homolog; Plasma membrane-associated small GTPase which cycles between an active GTP-bound and an inactive GDP-bound state. In active state binds to a variety of effector proteins to regulate cellular responses. Involved in epithelial cell polarization processes. Regulates the bipolar attachment of spindle microtubules to kinetochores before chromosome congression in metaphase. Regulates cell migration. In neurons, plays a role in the extension and maintenance of the formation of filopodia, thin and actin-rich surface projections. Required for DOCK10-med [...] ","Cyclin-dependent kinase 1; Plays a key role in the control of the eukaryotic cell cycle by modulating the centrosome cycle as well as mitotic onset; promotes G2-M transition, and regulates G1 progress and G1-S transition via association with multiple interphase cyclins. Required in higher cells for entry into S-phase and mitosis. Phosphorylates PARVA/actopaxin, APC, AMPH, APC, BARD1, Bcl-xL/BCL2L1, BRCA2, CALD1, CASP8, CDC7, CDC20, CDC25A, CDC25C, CC2D1A, CENPA, CSNK2 proteins/CKII, FZR1/CDH1, CDK7, CEBPB, CHAMP1, DMD/dystrophin, EEF1 proteins/EF-1, EZH2, KIF11/EG5, EGFR, FANCG, FOS, G [...] ","Kinesin-like protein KIF23; Component of the centralspindlin complex that serves as a microtubule-dependent and Rho-mediated signaling required for the myosin contractile ring formation during the cell cycle cytokinesis. Essential for cytokinesis in Rho-mediated signaling. Required for the localization of ECT2 to the central spindle. Plus-end-directed motor enzyme that moves antiparallel microtubules in vitro. Belongs to the TRAFAC class myosin-kinesin ATPase superfamily. Kinesin family.","Serine/threonine-protein kinase PLK1; Serine/threonine-protein kinase that performs several important functions throughout M phase of the cell cycle, including the regulation of centrosome maturation and spindle assembly, the removal of cohesins from chromosome arms, the inactivation of anaphase- promoting complex/cyclosome (APC/C) inhibitors, and the regulation of mitotic exit and cytokinesis. Polo-like kinase proteins acts by binding and phosphorylating proteins are that already phosphorylated on a specific motif recognized by the POLO box domains. Phosphorylates BORA, BUB1B/BUBR1, C [...] ","Ras-related C3 botulinum toxin substrate 2; Plasma membrane-associated small GTPase which cycles between an active GTP-bound and inactive GDP-bound state. In active state binds to a variety of effector proteins to regulate cellular responses, such as secretory processes, phagocytose of apoptotic cells and epithelial cell polarization. Augments the production of reactive oxygen species (ROS) by NADPH oxidase.","Rac GTPase-activating protein 1; Component of the centralspindlin complex that serves as a microtubule-dependent and Rho-mediated signaling required for the myosin contractile ring formation during the cell cycle cytokinesis. Required for proper attachment of the midbody to the cell membrane during cytokinesis. Plays key roles in controlling cell growth and differentiation of hematopoietic cells through mechanisms other than regulating Rac GTPase activity. Also involved in the regulation of growth-related processes in adipocytes and myoblasts. May be involved in regulating spermatogene [...] ","Transforming protein RhoA; Small GTPase which cycles between an active GTP-bound and an inactive GDP-bound state. Mainly associated with cytoskeleton organization, in active state binds to a variety of effector proteins to regulate cellular responses such cytoskeletal dynamics, cell migration and cell cycle. Regulates a signal transduction pathway linking plasma membrane receptors to the assembly of focal adhesions and actin stress fibers. Involved in a microtubule-dependent signal that is required for the myosin contractile ring formation during cell cycle cytokinesis. Plays an essent [...] ","Rho-related GTP-binding protein RhoB; Mediates apoptosis in neoplastically transformed cells after DNA damage. Not essential for development but affects cell adhesion and growth factor signaling in transformed cells. Plays a negative role in tumorigenesis as deletion causes tumor formation. Involved in intracellular protein trafficking of a number of proteins. Targets PKN1 to endosomes and is involved in trafficking of the EGF receptor from late endosomes to lysosomes. Also required for stability and nuclear trafficking of AKT1/AKT which promotes endothelial cell survival during vascul [...] ","PHD finger protein 14.","RNA-binding protein 3; Cold-inducible mRNA binding protein that enhances global protein synthesis at both physiological and mild hypothermic temperatures. Reduces the relative abundance of microRNAs, when overexpressed. Enhances phosphorylation of translation initiation factors and active polysome formation (By similarity)."]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>queryIndex<\/th>\n      <th>queryItem<\/th>\n      <th>stringId<\/th>\n      <th>ncbiTaxonId<\/th>\n      <th>taxonName<\/th>\n      <th>preferredName<\/th>\n      <th>annotation<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[1,4]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"queryIndex","targets":1},{"name":"queryItem","targets":2},{"name":"stringId","targets":3},{"name":"ncbiTaxonId","targets":4},{"name":"taxonName","targets":5},{"name":"preferredName","targets":6},{"name":"annotation","targets":7}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## 3 What we need and will use for the rest of this vignette is the `stringId` column</span></span>
<span><span class="va">proteins_mapped</span> <span class="op">&lt;-</span> <span class="va">proteins_mapped</span><span class="op">$</span><span class="va">stringId</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="get-interaction-network-of-a-protein-set">Get interaction network of a protein set<a class="anchor" aria-label="anchor" href="#get-interaction-network-of-a-protein-set"></a>
</h2>
<p>You can retrieve a list of interactions that the proteins in your set
have with each other along with the STRING annotations of each
interaction. You may filter the results by using required_score and
network_type arguments.</p>
<p>See the ‘values’ section <code>rba_string_interactions_network</code>
function’s manual for information on the returned columns.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">int_net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_interactions_network.html">rba_string_interactions_network</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  required_score <span class="op">=</span> <span class="fl">500</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33"],["9606.ENSP00000249071","9606.ENSP00000249071","9606.ENSP00000249071","9606.ENSP00000249373","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000260363","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000266970","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000269305","9606.ENSP00000272233","9606.ENSP00000272233","9606.ENSP00000272233","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000300093","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000378699","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000404190","9606.ENSP00000404190"],["9606.ENSP00000404190","9606.ENSP00000400175","9606.ENSP00000497733","9606.ENSP00000269305","9606.ENSP00000418960","9606.ENSP00000400175","9606.ENSP00000378699","9606.ENSP00000300093","9606.ENSP00000404190","9606.ENSP00000400175","9606.ENSP00000300093","9606.ENSP00000418960","9606.ENSP00000269305","9606.ENSP00000378699","9606.ENSP00000400175","9606.ENSP00000497733","9606.ENSP00000378699","9606.ENSP00000300093","9606.ENSP00000418960","9606.ENSP00000497733","9606.ENSP00000404190","9606.ENSP00000400175","9606.ENSP00000400175","9606.ENSP00000418960","9606.ENSP00000404190","9606.ENSP00000378699","9606.ENSP00000400175","9606.ENSP00000418960","9606.ENSP00000404190","9606.ENSP00000497733","9606.ENSP00000404190","9606.ENSP00000418960","9606.ENSP00000497733"],["RAC2","RAC2","RAC2","SMO","KIF23","KIF23","KIF23","KIF23","KIF23","CDK2","CDK2","CDK2","CDK2","CDK2","TP53","TP53","TP53","TP53","TP53","RHOB","RHOB","RHOB","PLK1","PLK1","PLK1","PLK1","CDK1","CDK1","CDK1","RHOA","RHOA","RACGAP1","RACGAP1"],["RACGAP1","RHOA","CDC42","TP53","BRCA1","RHOA","CDK1","PLK1","RACGAP1","RHOA","PLK1","BRCA1","TP53","CDK1","RHOA","CDC42","CDK1","PLK1","BRCA1","CDC42","RACGAP1","RHOA","RHOA","BRCA1","RACGAP1","CDK1","RHOA","BRCA1","RACGAP1","CDC42","RACGAP1","BRCA1","CDC42"],["9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606"],[0.819,0.9399999999999999,0.9429999999999999,0.525,0.713,0.714,0.914,0.988,0.999,0.514,0.742,0.984,0.987,0.994,0.662,0.76,0.962,0.981,0.999,0.604,0.62,0.821,0.512,0.96,0.961,0.993,0.516,0.878,0.96,0.955,0.979,0.552,0.979],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0.062,0.055,0,0,0,0,0,0,0,0,0,0,0.056,0,0,0,0,0,0.063,0,0.05,0,0,0,0,0,0,0,0.062,0,0,0],[0.058,0.159,0.218,0,0.41,0.042,0.8080000000000001,0.592,0.828,0.093,0.368,0.194,0.093,0.349,0.091,0.06,0.095,0.073,0.067,0.1,0.052,0.138,0.068,0.235,0.587,0.8139999999999999,0.1,0.402,0.6919999999999999,0.361,0.052,0.289,0.052],[0.07099999999999999,0.295,0.206,0,0.439,0.05,0,0.378,0.995,0.08500000000000001,0.15,0.735,0.778,0.47,0,0.16,0.626,0.737,0.895,0.295,0.114,0.399,0.126,0.781,0.644,0.15,0.08500000000000001,0.51,0.292,0.295,0.806,0.292,0.903],[0.4,0.9,0.9,0,0,0,0,0.75,0.9,0,0,0.5,0.5,0.4,0,0,0,0,0.5,0,0.4,0.5,0,0,0,0.9,0,0,0,0.9,0.4,0,0.4],[0.696,0.091,0.181,0.525,0.202,0.711,0.57,0.836,0.999,0.461,0.5580000000000001,0.872,0.891,0.977,0.644,0.721,0.899,0.93,0.999,0.411,0.336,0.385,0.449,0.783,0.761,0.66,0.459,0.619,0.832,0.103,0.836,0.182,0.671]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stringId_A<\/th>\n      <th>stringId_B<\/th>\n      <th>preferredName_A<\/th>\n      <th>preferredName_B<\/th>\n      <th>ncbiTaxonId<\/th>\n      <th>score<\/th>\n      <th>nscore<\/th>\n      <th>fscore<\/th>\n      <th>pscore<\/th>\n      <th>ascore<\/th>\n      <th>escore<\/th>\n      <th>dscore<\/th>\n      <th>tscore<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[6,7,8,9,10,11,12,13]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"stringId_A","targets":1},{"name":"stringId_B","targets":2},{"name":"preferredName_A","targets":3},{"name":"preferredName_B","targets":4},{"name":"ncbiTaxonId","targets":5},{"name":"score","targets":6},{"name":"nscore","targets":7},{"name":"fscore","targets":8},{"name":"pscore","targets":9},{"name":"ascore","targets":10},{"name":"escore","targets":11},{"name":"dscore","targets":12},{"name":"tscore","targets":13}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div class="section level2">
<h2 id="get-interaction-partners-of-a-protein-set">Get interaction partners of a protein set<a class="anchor" aria-label="anchor" href="#get-interaction-partners-of-a-protein-set"></a>
</h2>
<p>In the last example, we only obtained the interaction which our
proteins have among themselves, what if we wanted to get a list of every
protein which interact with our protein(s)?</p>
<p>To do that, we can use
<code>rba_string_interaction_partners</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Although we supply only one protein ID here (CD40 protein), you can provide a vector of proteins as the input</span></span>
<span><span class="va">int_partners</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_interaction_partners.html">rba_string_interaction_partners</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="st">"9606.ENSP00000361359"</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  required_score <span class="op">=</span> <span class="fl">900</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-36aa3d2a04d42bbc2145" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-36aa3d2a04d42bbc2145">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10"],["9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359","9606.ENSP00000361359"],["9606.ENSP00000264246","9606.ENSP00000376500","9606.ENSP00000433623","9606.ENSP00000359663","9606.ENSP00000324890","9606.ENSP00000247668","9606.ENSP00000261464","9606.ENSP00000295025","9606.ENSP00000398698","9606.ENSP00000302961"],["CD40","CD40","CD40","CD40","CD40","CD40","CD40","CD40","CD40","CD40"],["CD80","TRAF3","TRAF6","CD40LG","CD28","TRAF2","TRAF5","REL","TNF","HSPA4"],[9606,9606,9606,9606,9606,9606,9606,9606,9606,9606],[0.999,0.999,0.999,0.999,0.997,0.996,0.993,0.992,0.992,0.991],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[0.258,0.122,0,0.058,0.058,0.127,0.08699999999999999,0.089,0.135,0.055],[0,0.991,0.895,0.948,0,0.927,0.845,0,0,0.292],[0,0.9,0.9,0.9,0,0.9,0.54,0,0.5,0],[0.999,0.992,0.996,0.999,0.997,0.5580000000000001,0.906,0.992,0.984,0.988]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stringId_A<\/th>\n      <th>stringId_B<\/th>\n      <th>preferredName_A<\/th>\n      <th>preferredName_B<\/th>\n      <th>ncbiTaxonId<\/th>\n      <th>score<\/th>\n      <th>nscore<\/th>\n      <th>fscore<\/th>\n      <th>pscore<\/th>\n      <th>ascore<\/th>\n      <th>escore<\/th>\n      <th>dscore<\/th>\n      <th>tscore<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[5,6,7,8,9,10,11,12,13]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"stringId_A","targets":1},{"name":"stringId_B","targets":2},{"name":"preferredName_A","targets":3},{"name":"preferredName_B","targets":4},{"name":"ncbiTaxonId","targets":5},{"name":"score","targets":6},{"name":"nscore","targets":7},{"name":"fscore","targets":8},{"name":"pscore","targets":9},{"name":"ascore","targets":10},{"name":"escore","targets":11},{"name":"dscore","targets":12},{"name":"tscore","targets":13}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div class="section level2">
<h2 id="get-network-image-of-a-protein-set">Get network image of a protein set<a class="anchor" aria-label="anchor" href="#get-network-image-of-a-protein-set"></a>
</h2>
<p>Let’s go back to the interaction network. As you must have seen in
the STRING webpages, STRING plots the interaction network of your
proteins with many customizations available. You can also do that with
STRING API services. <code>rba_string_network_image</code> function is
very flexible and you have a variety of options; see the function’s
manual.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Example 1:</span></span>
<span><span class="va">graph_ppi1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_network_image.html">rba_string_network_image</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>  image_format <span class="op">=</span> <span class="st">"image"</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  save_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  required_score <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  network_flavor <span class="op">=</span> <span class="st">"confidence"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="rbioapi_string_files/figure-html/rba_string_network_image_ex1_image-1.png" alt="Network images - Example 1" width="672"><p class="caption">
Network images - Example 1
</p>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Example 2:</span></span>
<span><span class="va">graph_ppi2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_network_image.html">rba_string_network_image</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>  image_format <span class="op">=</span> <span class="st">"image"</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  save_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  required_score <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  add_color_nodes <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  add_white_nodes <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  network_flavor <span class="op">=</span> <span class="st">"actions"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="rbioapi_string_files/figure-html/rba_string_network_image_ex2_image-1.png" alt="Network images - Example 2" width="672"><p class="caption">
Network images - Example 2
</p>
</div>
<hr>
</div>
<div class="section level2">
<h2 id="enrichment-using-string">Enrichment analysis using STRING<a class="anchor" aria-label="anchor" href="#enrichment-using-string"></a>
</h2>
<p>STRING let you perform two types of enrichment analysis. See <a href="https://doi.org/10.1093/nar/gks1094" title="STRING v9.1: protein-protein interaction networks, with increased coverage and integration" class="external-link">STRING’s
paper</a> for more information.</p>
<div class="section level3">
<h3 id="functional-enrichment">Functional enrichment<a class="anchor" aria-label="anchor" href="#functional-enrichment"></a>
</h3>
<p>The first type is the conventional type, which statistically tests
your supplied gene sets against some sets of annotation. Currently,
STRING supports Gene Ontology, KEGG pathways, UniProt Keywords, PubMed
publications, Pfam domains, InterPro domains, and SMART domains. (<a href="https://version11.string-db.org/help/api/#getting-functional-enrichment" title="STRING API - Getting functional enrichment" class="external-link">source</a>).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enriched</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_enrichment.html">rba_string_enrichment</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>As usual, we inspect the output using the <code><a href="https://rdrr.io/r/utils/str.html" class="external-link">str()</a></code>
function. As you can see below, the enrichment results of each category
can be found as the returned list’s elements.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">enriched</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 14</span></span>
<span><span class="co">#&gt;  $ COMPARTMENTS     :'data.frame':   25 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Component        :'data.frame':   17 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ DISEASES         :'data.frame':   11 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Function         :'data.frame':   12 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ InterPro         :'data.frame':   3 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ KEGG             :'data.frame':   45 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Keyword          :'data.frame':   14 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ NetworkNeighborAL:'data.frame':   5 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ PMID             :'data.frame':   100 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Process          :'data.frame':   148 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ RCTM             :'data.frame':   59 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ SMART            :'data.frame':   1 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ TISSUES          :'data.frame':   12 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ WikiPathways     :'data.frame':   46 obs. of  10 variables:</span></span></code></pre></div>
<p>Let us see the “KEGG” results as an example. Below, we can see which
terms of the <a href="https://www.genome.jp/kegg/pathway.html" class="external-link">KEGG
pathways database</a> were over-represented:</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-febe03efa1a2d8d52a86" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-febe03efa1a2d8d52a86">{"x":{"filter":"none","vertical":false,"data":[["240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284"],["KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG","KEGG"],["hsa05203","hsa05200","hsa04110","hsa05206","hsa04360","hsa05205","hsa04115","hsa04520","hsa05212","hsa05210","hsa04914","hsa04071","hsa04670","hsa04722","hsa04114","hsa05135","hsa05418","hsa04218","hsa04310","hsa04062","hsa04510","hsa04015","hsa04810","hsa05132","hsa05131","hsa05163","hsa04014","hsa04370","hsa05217","hsa04010","hsa01524","hsa05100","hsa05165","hsa04151","hsa04666","hsa05222","hsa05215","hsa04660","hsa04068","hsa05162","hsa05224","hsa05226","hsa04530","hsa05160","hsa05161"],[5,6,4,4,4,4,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,2,3,2,2,3,3,2,2,2,2,2,2,2,2,2,2,2],[183,515,120,159,176,194,72,69,71,82,95,116,111,112,121,124,129,150,154,186,195,201,209,209,218,217,225,56,63,286,70,70,324,349,90,92,97,100,126,137,146,146,157,157,158],[9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606,9606],[["9606.ENSP00000266970","9606.ENSP00000269305","9606.ENSP00000378699","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000249071","9606.ENSP00000249373","9606.ENSP00000266970","9606.ENSP00000269305","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000266970","9606.ENSP00000269305","9606.ENSP00000300093","9606.ENSP00000378699"],["9606.ENSP00000260363","9606.ENSP00000269305","9606.ENSP00000400175","9606.ENSP00000418960"],["9606.ENSP00000249071","9606.ENSP00000249373","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000249373","9606.ENSP00000269305","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000266970","9606.ENSP00000269305","9606.ENSP00000378699"],["9606.ENSP00000249071","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000249071","9606.ENSP00000269305","9606.ENSP00000497733"],["9606.ENSP00000249071","9606.ENSP00000269305","9606.ENSP00000400175"],["9606.ENSP00000266970","9606.ENSP00000300093","9606.ENSP00000378699"],["9606.ENSP00000249071","9606.ENSP00000269305","9606.ENSP00000400175"],["9606.ENSP00000249071","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000269305","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000266970","9606.ENSP00000300093","9606.ENSP00000378699"],["9606.ENSP00000249071","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000249071","9606.ENSP00000269305","9606.ENSP00000400175"],["9606.ENSP00000266970","9606.ENSP00000269305","9606.ENSP00000378699"],["9606.ENSP00000249071","9606.ENSP00000269305","9606.ENSP00000400175"],["9606.ENSP00000249071","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000249071","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000249071","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000249071","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000272233","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000269305","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000249071","9606.ENSP00000269305","9606.ENSP00000400175"],["9606.ENSP00000249071","9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000249071","9606.ENSP00000497733"],["9606.ENSP00000249373","9606.ENSP00000269305"],["9606.ENSP00000249071","9606.ENSP00000269305","9606.ENSP00000497733"],["9606.ENSP00000269305","9606.ENSP00000418960"],["9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000266970","9606.ENSP00000269305","9606.ENSP00000497733"],["9606.ENSP00000266970","9606.ENSP00000269305","9606.ENSP00000418960"],["9606.ENSP00000249071","9606.ENSP00000497733"],["9606.ENSP00000266970","9606.ENSP00000269305"],["9606.ENSP00000266970","9606.ENSP00000269305"],["9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000266970","9606.ENSP00000300093"],["9606.ENSP00000266970","9606.ENSP00000269305"],["9606.ENSP00000269305","9606.ENSP00000418960"],["9606.ENSP00000266970","9606.ENSP00000269305"],["9606.ENSP00000400175","9606.ENSP00000497733"],["9606.ENSP00000266970","9606.ENSP00000269305"],["9606.ENSP00000266970","9606.ENSP00000269305"]],[["CDK2","TP53","CDK1","RHOA","CDC42"],["RAC2","SMO","CDK2","TP53","RHOA","CDC42"],["CDK2","TP53","PLK1","CDK1"],["KIF23","TP53","RHOA","BRCA1"],["RAC2","SMO","RHOA","CDC42"],["SMO","TP53","RHOA","CDC42"],["CDK2","TP53","CDK1"],["RAC2","RHOA","CDC42"],["RAC2","TP53","CDC42"],["RAC2","TP53","RHOA"],["CDK2","PLK1","CDK1"],["RAC2","TP53","RHOA"],["RAC2","RHOA","CDC42"],["TP53","RHOA","CDC42"],["CDK2","PLK1","CDK1"],["RAC2","RHOA","CDC42"],["RAC2","TP53","RHOA"],["CDK2","TP53","CDK1"],["RAC2","TP53","RHOA"],["RAC2","RHOA","CDC42"],["RAC2","RHOA","CDC42"],["RAC2","RHOA","CDC42"],["RAC2","RHOA","CDC42"],["RHOB","RHOA","CDC42"],["TP53","RHOA","CDC42"],["RAC2","TP53","RHOA"],["RAC2","RHOA","CDC42"],["RAC2","CDC42"],["SMO","TP53"],["RAC2","TP53","CDC42"],["TP53","BRCA1"],["RHOA","CDC42"],["CDK2","TP53","CDC42"],["CDK2","TP53","BRCA1"],["RAC2","CDC42"],["CDK2","TP53"],["CDK2","TP53"],["RHOA","CDC42"],["CDK2","PLK1"],["CDK2","TP53"],["TP53","BRCA1"],["CDK2","TP53"],["RHOA","CDC42"],["CDK2","TP53"],["CDK2","TP53"]],[1.4e-07,8.3e-07,1.42e-06,4.23e-06,6.27e-06,9.130000000000001e-06,1.87e-05,1.65e-05,1.79e-05,2.72e-05,4.17e-05,7.440000000000001e-05,6.550000000000001e-05,6.719999999999999e-05,8.41e-05,9.02e-05,0.0001,0.00016,0.00017,0.00029,0.00033,0.00037,0.00041,0.00041,0.00046,0.00046,0.00051,0.00076,0.00095,0.001,0.0012,0.0012,0.0014,0.0018,0.0019,0.002,0.0022,0.0023,0.0036,0.0042,0.0048,0.0048,0.0055,0.0055,0.0056],[4.69e-05,0.00014,0.00016,0.00036,0.00042,0.00051,0.00079,0.00079,0.00079,0.00091,0.0013,0.0018,0.0018,0.0018,0.0019,0.0019,0.002,0.0029,0.003,0.0049,0.0054,0.0056,0.006,0.006,0.0061,0.0061,0.0063,0.0091,0.011,0.0113,0.0126,0.0126,0.0146,0.0175,0.0181,0.0184,0.0198,0.0205,0.0312,0.0357,0.0394,0.0394,0.0431,0.0431,0.0431],["Viral carcinogenesis","Pathways in cancer","Cell cycle","MicroRNAs in cancer","Axon guidance","Proteoglycans in cancer","p53 signaling pathway","Adherens junction","Pancreatic cancer","Colorectal cancer","Progesterone-mediated oocyte maturation","Sphingolipid signaling pathway","Leukocyte transendothelial migration","Neurotrophin signaling pathway","Oocyte meiosis","Yersinia infection","Fluid shear stress and atherosclerosis","Cellular senescence","Wnt signaling pathway","Chemokine signaling pathway","Focal adhesion","Rap1 signaling pathway","Regulation of actin cytoskeleton","Salmonella infection","Shigellosis","Human cytomegalovirus infection","Ras signaling pathway","VEGF signaling pathway","Basal cell carcinoma","MAPK signaling pathway","Platinum drug resistance","Bacterial invasion of epithelial cells","Human papillomavirus infection","PI3K-Akt signaling pathway","Fc gamma R-mediated phagocytosis","Small cell lung cancer","Prostate cancer","T cell receptor signaling pathway","FoxO signaling pathway","Measles","Breast cancer","Gastric cancer","Tight junction","Hepatitis C","Hepatitis B"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>category<\/th>\n      <th>term<\/th>\n      <th>number_of_genes<\/th>\n      <th>number_of_genes_in_background<\/th>\n      <th>ncbiTaxonId<\/th>\n      <th>inputGenes<\/th>\n      <th>preferredNames<\/th>\n      <th>p_value<\/th>\n      <th>fdr<\/th>\n      <th>description<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[3,4,5,8,9]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"category","targets":1},{"name":"term","targets":2},{"name":"number_of_genes","targets":3},{"name":"number_of_genes_in_background","targets":4},{"name":"ncbiTaxonId","targets":5},{"name":"inputGenes","targets":6},{"name":"preferredNames","targets":7},{"name":"p_value","targets":8},{"name":"fdr","targets":9},{"name":"description","targets":10}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p><strong>Please Note:</strong> Other services supported by rbioapi
also provide Over-representation analysis tools. Please see the vignette
article <a href="rbioapi_do_enrich.html">Do with rbioapi:
Over-Representation (Enrichment) Analysis in R</a> (<a href="https://rbioapi.moosa-r.com/articles/rbioapi_do_enrich.html">link
to the documentation site</a>) for an in-depth review.</p>
</div>
<div class="section level3">
<h3 id="functional-enrichment-plot">Functional enrichment Plot<a class="anchor" aria-label="anchor" href="#functional-enrichment-plot"></a>
</h3>
<p>In addition to a data frame, you can also get a plot summarizing the
enrichment results. This API endpoint supports extensive customization
of the plot; please refer to the
<code><a href="../reference/rba_string_enrichment_image.html">rba_string_enrichment_image()</a></code> function’s manual for
detailed instructions. Here we perform the exact enrichment analysis
done above, and retrieve a plot of the results.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graph_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_enrichment_image.html">rba_string_enrichment_image</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  category <span class="op">=</span> <span class="st">"KEGG"</span>,</span>
<span>  image_format <span class="op">=</span> <span class="st">"image"</span>,</span>
<span>  save_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  group_by_similarity <span class="op">=</span> <span class="fl">0.6</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="rbioapi_string_files/figure-html/rba_string_enrichment_plot_image-1.png" alt="Visualization of enrichment analysis results" width="672"><p class="caption">
Visualization of enrichment analysis results
</p>
</div>
</div>
<div class="section level3">
<h3 id="protein-protein-interaction-enrichment">Protein-protein interaction enrichment<a class="anchor" aria-label="anchor" href="#protein-protein-interaction-enrichment"></a>
</h3>
<p>Even without incorporating annotation data, STRING can calculate if
your proteins are functionally related. Briefly, STRING accomplishes
this by comparing the interactions’ distribution in your protein-set to
the interactions’ distribution in the proteome. Read <a href="https://doi.org/10.1093/nar/gks1094" title="STRING v9.1: protein-protein interaction networks, with increased coverage and integration" class="external-link">STRING’s
paper</a> for more information.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rba_string_enrichment_ppi.html">rba_string_enrichment_ppi</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="va">proteins_mapped</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; $number_of_nodes</span></span>
<span><span class="co">#&gt; [1] 14</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $number_of_edges</span></span>
<span><span class="co">#&gt; [1] 40</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $average_node_degree</span></span>
<span><span class="co">#&gt; [1] 5.71</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $local_clustering_coefficient</span></span>
<span><span class="co">#&gt; [1] 0.694</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $expected_number_of_edges</span></span>
<span><span class="co">#&gt; [1] 19</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $p_value</span></span>
<span><span class="co">#&gt; [1] 1.35e-05</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="get-functional-annotations">Get functional annotations<a class="anchor" aria-label="anchor" href="#get-functional-annotations"></a>
</h2>
<p>As you have seen above, STRING maps the proteins to multiple
annotation sources. You can obtain any annotation associated with your
proteins without performing enrichment analysis and retrieving just the
significant portion.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_annotations.html">rba_string_annotations</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="st">"9606.ENSP00000269305"</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## This function returns large results, so the results are not shown in this vignette.</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="see-also-in-functions-manuals">See also in Functions’ manuals<a class="anchor" aria-label="anchor" href="#see-also-in-functions-manuals"></a>
</h2>
<p>Some rbioapi STRING functions were not covered in this vignette,
please check their manuals:</p>
<ul>
<li><p><code><a href="../reference/rba_string_homology_intra.html">rba_string_homology_intra()</a></code></p></li>
<li><p><code><a href="../reference/rba_string_homology_inter.html">rba_string_homology_inter()</a></code></p></li>
<li><p><code><a href="../reference/rba_string_version.html">rba_string_version()</a></code></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="citations">How to Cite?<a class="anchor" aria-label="anchor" href="#citations"></a>
</h2>
<p>To cite STRING (Please see <a href="https://string-db.org/cgi/about?footer_active_subpage=references" class="external-link uri">https://string-db.org/cgi/about?footer_active_subpage=references</a>):</p>
<ul>
<li>Damian Szklarczyk, Rebecca Kirsch, Mikaela Koutrouli, Katerina
Nastou, Farrokh Mehryary, Radja Hachilif, Annika L Gable, Tao Fang,
Nadezhda T Doncheva, Sampo Pyysalo, Peer Bork, Lars J Jensen, Christian
von Mering, The STRING database in 2023: protein–protein association
networks and functional enrichment analyses for any sequenced genome of
interest, Nucleic Acids Research, Volume 51, Issue D1, 6 January 2023,
Pages D638–D646, <a href="https://doi.org/10.1093/nar/gkac1000" class="external-link uri">https://doi.org/10.1093/nar/gkac1000</a>
</li>
</ul>
<p>To cite rbioapi:</p>
<ul>
<li>Moosa Rezwani, Ali Akbar Pourfathollah, Farshid Noorbakhsh, rbioapi:
user-friendly R interface to biologic web services’ API, Bioinformatics,
Volume 38, Issue 10, 15 May 2022, Pages 2952–2953, <a href="https://doi.org/10.1093/bioinformatics/btac172" class="external-link uri">https://doi.org/10.1093/bioinformatics/btac172</a>
</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_string.html" title="2.E: STRING &amp; rbioapi">This article in rbioapi
documentation site</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html#section-string-rba-string-" title="rbioapi reference">Functions references in rbioapi
documentation site</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi
vignette index</a></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] rbioapi_0.8.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] httr_1.4.7        cli_3.6.3         knitr_1.49        rlang_1.1.5      </span></span>
<span><span class="co">#&gt;  [5] xfun_0.50         png_0.1-8         textshaping_1.0.0 jsonlite_1.8.9   </span></span>
<span><span class="co">#&gt;  [9] DT_0.33           htmltools_0.5.8.1 ragg_1.3.3        sass_0.4.9       </span></span>
<span><span class="co">#&gt; [13] rmarkdown_2.29    grid_4.4.2        crosstalk_1.2.1   evaluate_1.0.3   </span></span>
<span><span class="co">#&gt; [17] jquerylib_0.1.4   fastmap_1.2.0     yaml_2.3.10       lifecycle_1.0.4  </span></span>
<span><span class="co">#&gt; [21] compiler_4.4.2    fs_1.6.5          htmlwidgets_1.6.4 systemfonts_1.2.1</span></span>
<span><span class="co">#&gt; [25] digest_0.6.37     R6_2.5.1          curl_6.2.0        magrittr_2.0.3   </span></span>
<span><span class="co">#&gt; [29] bslib_0.9.0       tools_4.4.2       pkgdown_2.1.1     cachem_1.1.0     </span></span>
<span><span class="co">#&gt; [33] desc_1.4.3</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://moosa-r.com/" class="external-link">Moosa Rezwani</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
