<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>2.D: PANTHER &amp; rbioapi • rbioapi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2.D: PANTHER &amp; rbioapi">
<meta name="description" content="Connect to PANTHER in R with rbioapi package.
">
<meta property="og:description" content="Connect to PANTHER in R with rbioapi package.
">
<meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rbioapi.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles-supported-services" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles &amp; Supported Services</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles-supported-services">
<li><a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://moosa-r.com/" aria-label="Moosa Rezwani Portfolio Website"><span class="fa fa-solid fa-id-card"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="rbioapi_panther_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="rbioapi_panther_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="rbioapi_panther_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_panther_files/datatables-binding-0.33/datatables.js"></script><link href="rbioapi_panther_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_panther_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_panther_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="rbioapi_panther_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="rbioapi_panther_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>2.D: PANTHER &amp; rbioapi</h1>
                        <h4 data-toc-skip class="author">Moosa
Rezwani</h4>
            
            <h4 data-toc-skip class="date">2025-01-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/master/vignettes/rbioapi_panther.Rmd" class="external-link"><code>vignettes/rbioapi_panther.Rmd</code></a></small>
      <div class="d-none name"><code>rbioapi_panther.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Directly quoting the the paper published by <a href="https://www.pantherdb.org" title="Protein Analysis THrough Evolutionary Relationships (PANTHER)" class="external-link">PANTHER</a>
(Protein Analysis THrough Evolutionary Relationships) authors:</p>
<blockquote>
<p>The PANTHER classification system (<a href="https://www.pantherdb.org" class="external-link uri">https://www.pantherdb.org</a>) a comprehensive system that
combines genomes, gene function , pathways and statistical analysis
tools to enable to analyze large-scale genome-wide experimental data.
The system (PANTHER v.14.0) covers 131 complete genomes organized gene
families and subfamilies; evolutionary relationships between are
represented in phylogenetic trees, multiple sequence and statistical
models (hidden Markov models (HMMs)). The families and subfamilies are
annotated with Gene Ontology (GO) terms, sequences are assigned to
PANTHER pathways. A suite of tools has built to allow users to browse
and query gene functions and analyze-scale experimental data with a
number of statistical tests. is widely used by bench scientists,
bioinformaticians, computer and systems biologists.</p>
<p>(source: Mi, Huaiyu, et al. “Protocol Update for large-scale genome
and gene function analysis with the PANTHER classification system (v.
14.0).” <em>Nature protocols</em> 14.3 (2019): 703-721)</p>
</blockquote>
<p>The available tools in PANTHER’s <strong>RESTful API
services</strong> can be divided into 3 broad categories: Mapping genes,
retrieving information, and research tools. Herein, we provide a very
short introduction; you can always check functions’ manuals for detailed
guides and examples.</p>
<hr>
</div>
<div class="section level2">
<h2 id="map-genes">Map genes<a class="anchor" aria-label="anchor" href="#map-genes"></a>
</h2>
<ul>
<li><p><code><a href="../reference/rba_panther_mapping.html">rba_panther_mapping()</a></code>: map your gene-set to PANTHER
database and retrieve attributes and annotations associated with your
genes</p></li>
<li><p><code><a href="../reference/rba_panther_ortholog.html">rba_panther_ortholog()</a></code>: Retrieve Orthologs of your
genes</p></li>
<li><p><code><a href="../reference/rba_panther_homolog.html">rba_panther_homolog()</a></code>: Retrieve Homologs of your
genes</p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="get-information">Get information<a class="anchor" aria-label="anchor" href="#get-information"></a>
</h2>
<ul>
<li><p><code>rba_panther_info</code>(): Retrieve a list of PANTHER’s
supported organisms, datasets, families, or pathways</p></li>
<li><p><code>rba_panther_family</code>(): Retrieve Orthologs, MSA, or
Tree topology of a given PANTHER family.</p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="gene-list-analysis">Gene List Analysis<a class="anchor" aria-label="anchor" href="#gene-list-analysis"></a>
</h2>
<p><code><a href="../reference/rba_panther_enrich.html">rba_panther_enrich()</a></code> is equivalent to <a href="https://www.pantherdb.org/index.jsp" title="PANTHER Gene List Analysis" class="external-link">Gene List analysis tool’s
webpage</a>. Depending on the provided input’s class, PANTHER will
perform either over-representation analysis or statistical enrichment
analysis. Below we demonstrate how to perform such analyses.</p>
<div class="section level3">
<h3 id="analysis-get-available-annotation-datasets">Get the available annotation datasets<a class="anchor" aria-label="anchor" href="#analysis-get-available-annotation-datasets"></a>
</h3>
<p>First, we need to select an annotation dataset to conduct the
analysis based on it. Each annotation dataset contains a collection of
terms, where each term is associated with a group of genes.</p>
<p>To retrieve the list of available annotation datasets in PANTHER, use
the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_panther_info.html">rba_panther_info</a></span><span class="op">(</span>what <span class="op">=</span> <span class="st">"datasets"</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9"],["2024-11-03","2024-11-03","2024-11-03","2024-06-20","2024-06-20","2024-06-20","2024-06-20","2024-06-20","2023-09-07"],["Gene Ontology Molecular Function annotations including both manually curated and electronic annotations.","Gene Ontology Biological Process annotations including both manually curated and electronic annotations.","Gene Ontology Cellular Component annotations including both manually curated and electronic annotations.","A molecular process that can be carried out by the action of a single macromolecular machine, usually via direct physical interactions with other molecular entities. Function in this sense denotes an action, or activity, that a gene product (or a complex) performs.","A biological process is the execution of a genetically-encoded biological module or program. It consists of all the steps required to achieve the specific biological objective of the module. A biological process is accomplished by a particular set of molecular functions carried out by specific gene products (or macromolecular complexes), often in a highly regulated manner and in a particular temporal sequence.","A location, relative to cellular compartments and structures, occupied by a macromolecular machine. There are three types of cellular components described in the gene ontology: (1) the cellular anatomical entity where a gene product carries out a molecular function (e.g., plasma membrane, cytoskeleton) or membrane-enclosed compartments (e.g., mitochondrion); (2) virion components, where viral proteins act, and (3) the stable macromolecular complexes of which gene product are parts (e.g., the clathrin complex).","","Panther Pathways","Reactome Pathways"],["GO:0003674","GO:0008150","GO:0005575","ANNOT_TYPE_ID_PANTHER_GO_SLIM_MF","ANNOT_TYPE_ID_PANTHER_GO_SLIM_BP","ANNOT_TYPE_ID_PANTHER_GO_SLIM_CC","ANNOT_TYPE_ID_PANTHER_PC","ANNOT_TYPE_ID_PANTHER_PATHWAY","ANNOT_TYPE_ID_REACTOME_PATHWAY"],["molecular_function","biological_process","cellular_component","PANTHER GO Slim Molecular Function","PANTHER GO Slim Biological Process","PANTHER GO Slim Cellular Location","protein class","ANNOT_TYPE_PANTHER_PATHWAY","ANNOT_TYPE_REACTOME_PATHWAY"],["10.5281/zenodo.14083199","10.5281/zenodo.14083199","10.5281/zenodo.14083199","19","19","19","19","19","86"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>release_date<\/th>\n      <th>description<\/th>\n      <th>id<\/th>\n      <th>label<\/th>\n      <th>version<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"release_date","targets":1},{"name":"description","targets":2},{"name":"id","targets":3},{"name":"label","targets":4},{"name":"version","targets":5}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>Please note that you should use the ID of the desired annotation
dataset, not its label. For example, using
<code>"biological_process"</code> is incorrect; you should rather use
<code>"GO:0008150"</code>.</p>
</div>
<div class="section level3">
<h3 id="submit-the-analysis-request">Submit the analysis request<a class="anchor" aria-label="anchor" href="#submit-the-analysis-request"></a>
</h3>
<p>Depending on the provided input, PANTHER will conduct two types of
analysis:</p>
<ol style="list-style-type: decimal">
<li><p>If a character vector is supplied, over-representation analysis
will be performed using either Fisher’s exact or binomial test.</p></li>
<li><p>If a data frame with gene identifiers and their corresponding
expression values is supplied, statistical enrichment test is performed
using Mann-Whitney U (Wilcoxon Rank-Sum) test.</p></li>
</ol>
<p>rbioapi determines the proper analysis based on the class of the
<code>genes</code> parameter. Please refer to the details section of
<code><a href="../reference/rba_panther_enrich.html">rba_panther_enrich()</a></code> function manual for more
information.</p>
<div class="section level4">
<h4 id="over-representation-analysis">Over-representation analysis<a class="anchor" aria-label="anchor" href="#over-representation-analysis"></a>
</h4>
<p>Now, suppose we want to perform an over-representation analysis
against the ‘GO biological process’ annotation dataset. In this example,
we only provide the gene names, thus over-representation analysis will
be conducted:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a variable to store the genes vector</span></span>
<span><span class="va">my_genes_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>, <span class="st">"CDK1"</span>,<span class="st">"KIF23"</span>,<span class="st">"PLK1"</span>,</span>
<span>  <span class="st">"RAC2"</span>,<span class="st">"RACGAP1"</span>,<span class="st">"RHOA"</span>, <span class="st">"RHOB"</span>, <span class="st">"PHF14"</span>, <span class="st">"RBM3"</span>, <span class="st">"MSL1"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Submit the analysis request.</span></span>
<span><span class="va">enriched</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_panther_enrich.html">rba_panther_enrich</a></span><span class="op">(</span></span>
<span>  genes <span class="op">=</span> <span class="va">my_genes_vec</span>,</span>
<span>  organism <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  annot_dataset <span class="op">=</span> <span class="st">"GO:0008150"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Performing PANTHER over-representation analysis (Fisher's exact test) on 15 genes from `organism 9606` against `GO:0008150` datasets.</span></span>
<span></span>
<span><span class="co"># Note that we didn't supply the `test_type` parameter.</span></span>
<span><span class="co"># In this case, the function will default to using Fisher's exact test # (i.e. `test_type = "FISHER"`).</span></span>
<span><span class="co"># You may also use binomial test for the over-representation analysis # (i.e. `test_type = "BINOMIAL"`).</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; [1] "Vignette building failed. It is probably because the web service was down during the building."</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="statistical-enrichment-analysis">Statistical enrichment analysis<a class="anchor" aria-label="anchor" href="#statistical-enrichment-analysis"></a>
</h4>
<p>As you can see in the above example, only a vector of gene names was
used. We can also use the corresponding expression values of the genes.
In this case, PANTHER will perform a statistical enrichment
analysis.</p>
<p>To do so, the only change will be to supply a data frame to the
<code>genes</code> parameter. Note that in this case, Mann-Whitney U
Test will be performed. The data frame should have two columns: the
first column should contain the gene identifiers as a character vector;
the second column should contain the corresponding expression values as
a numeric vector.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a variable to store the data frame</span></span>
<span><span class="va">my_genes_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>, <span class="st">"CDK1"</span>,<span class="st">"KIF23"</span>,<span class="st">"PLK1"</span>,</span>
<span>    <span class="st">"RAC2"</span>,<span class="st">"RACGAP1"</span>,<span class="st">"RHOA"</span>, <span class="st">"RHOB"</span>, <span class="st">"PHF14"</span>, <span class="st">"RBM3"</span>, <span class="st">"MSL1"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="co">## generate random expression values</span></span>
<span>  expression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">0</span>, <span class="fl">10</span><span class="op">)</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Submit the analysis request.</span></span>
<span><span class="va">enriched</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_panther_enrich.html">rba_panther_enrich</a></span><span class="op">(</span></span>
<span>  genes <span class="op">=</span> <span class="va">my_genes_df</span>,</span>
<span>  organism <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  annot_dataset <span class="op">=</span> <span class="st">"GO:0008150"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Note that we didn't supply the `test_type` parameter.</span></span>
<span><span class="co"># In this case, the function will default to Mann-Whitney U Test</span></span>
<span><span class="co"># (i.e. `test_type = "Mann-Whitney"`).</span></span>
<span><span class="co"># This is the only valid value for the statistical enrichment analysis test,</span></span>
<span><span class="co"># thus ommiting or supplying it will not make a difference.</span></span></code></pre></div>
<p><strong>Please Note:</strong> Other services supported by rbioapi
also provide Over-representation analysis tools. Please see the vignette
article <a href="rbioapi_do_enrich.html">Do with rbioapi:
Over-Representation (Enrichment) Analysis in R</a> (<a href="https://rbioapi.moosa-r.com/articles/rbioapi_do_enrich.html">link
to the documentation site</a>) for an in-depth review.</p>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="tree-grafter">Tree grafter<a class="anchor" aria-label="anchor" href="#tree-grafter"></a>
</h2>
<p><code><a href="../reference/rba_panther_tree_grafter.html">rba_panther_tree_grafter()</a></code> is an equivalent to the “<a href="https://www.pantherdb.org/tools/sequenceSearchForm.jsp" class="external-link">Graft
sequence into PANTHER library of trees</a>” tool.</p>
<hr>
</div>
<div class="section level2">
<h2 id="citations">How to Cite?<a class="anchor" aria-label="anchor" href="#citations"></a>
</h2>
<p>To cite PANTHER (Please see <a href="https://www.pantherdb.org/publications.jsp#HowToCitePANTHER" class="external-link uri">https://www.pantherdb.org/publications.jsp#HowToCitePANTHER</a>):</p>
<ul>
<li>Huaiyu Mi, Dustin Ebert, Anushya Muruganujan, Caitlin Mills,
Laurent-Philippe Albou, Tremayne Mushayamaha, Paul D Thomas, PANTHER
version 16: a revised family classification, tree-based classification
tool, enhancer regions and extensive API, <em>Nucleic Acids
Research</em>, Volume 49, Issue D1, 8 January 2021, Pages D394–D403, <a href="https://doi.org/10.1093/nar/gkaa1106" class="external-link uri">https://doi.org/10.1093/nar/gkaa1106</a>
</li>
</ul>
<p>To cite rbioapi:</p>
<ul>
<li>Moosa Rezwani, Ali Akbar Pourfathollah, Farshid Noorbakhsh, rbioapi:
user-friendly R interface to biologic web services’ API, Bioinformatics,
Volume 38, Issue 10, 15 May 2022, Pages 2952–2953, <a href="https://doi.org/10.1093/bioinformatics/btac172" class="external-link uri">https://doi.org/10.1093/bioinformatics/btac172</a>
</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_panther.html" title="2.C: PANTHER &amp; rbioapi">This article in rbioapi
documentation site</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html#section-enrichr-rba-enrichr-" title="rbioapi reference">Functions references in rbioapi
documentation site</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi
vignette index</a></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] rbioapi_0.8.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] httr_1.4.7        cli_3.6.3         knitr_1.49        rlang_1.1.5      </span></span>
<span><span class="co">#&gt;  [5] xfun_0.50         textshaping_1.0.0 jsonlite_1.8.9    DT_0.33          </span></span>
<span><span class="co">#&gt;  [9] htmltools_0.5.8.1 ragg_1.3.3        sass_0.4.9        rmarkdown_2.29   </span></span>
<span><span class="co">#&gt; [13] crosstalk_1.2.1   evaluate_1.0.3    jquerylib_0.1.4   fastmap_1.2.0    </span></span>
<span><span class="co">#&gt; [17] yaml_2.3.10       lifecycle_1.0.4   compiler_4.4.2    fs_1.6.5         </span></span>
<span><span class="co">#&gt; [21] htmlwidgets_1.6.4 systemfonts_1.2.1 digest_0.6.37     R6_2.5.1         </span></span>
<span><span class="co">#&gt; [25] curl_6.2.0        magrittr_2.0.3    bslib_0.8.0       tools_4.4.2      </span></span>
<span><span class="co">#&gt; [29] pkgdown_2.1.1     cachem_1.1.0      desc_1.4.3</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://moosa-r.com/" class="external-link">Moosa Rezwani</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
