<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>3.A: Do with rbioapi: Enrichment (Over-Representation) Analysis in R • rbioapi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="3.A: Do with rbioapi: Enrichment (Over-Representation) Analysis in R">
<meta name="description" content="Enrichment (Over-Representation) Analysis in R. Perform enrichment analysis using Enrichr, miEAA, PANTHER, Reactome, and STRING with rbioapi package.
">
<meta property="og:description" content="Enrichment (Over-Representation) Analysis in R. Perform enrichment analysis using Enrichr, miEAA, PANTHER, Reactome, and STRING with rbioapi package.
">
<meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-secondary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rbioapi.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles-supported-services" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles &amp; Supported Services</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles-supported-services">
<li><a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://moosa-r.com/" aria-label="Moosa Rezwani Portfolio Website"><span class="fa fa-solid fa-id-card"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="rbioapi_do_enrich_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="rbioapi_do_enrich_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="rbioapi_do_enrich_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_do_enrich_files/datatables-binding-0.33/datatables.js"></script><link href="rbioapi_do_enrich_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_do_enrich_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_do_enrich_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="rbioapi_do_enrich_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="rbioapi_do_enrich_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>3.A: Do with rbioapi: Enrichment (Over-Representation) Analysis in R</h1>
                        <h4 data-toc-skip class="author">Moosa
Rezwani</h4>
            
            <h4 data-toc-skip class="date">2025-07-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/master/vignettes/rbioapi_do_enrich.Rmd" class="external-link"><code>vignettes/rbioapi_do_enrich.Rmd</code></a></small>
      <div class="d-none name"><code>rbioapi_do_enrich.Rmd</code></div>
    </div>

    
    
<p>Among the databases and web services currently supported by rbioapi
<a href="https://maayanlab.cloud/Enrichr/" title="Enrichr" class="external-link">Enrichr</a>,
<a href="https://ccb-compute2.cs.uni-saarland.de/mieaa2" title="miRNA Enrichment Analysis and Annotation Tool (miEAA)" class="external-link">miEAA</a>,
<a href="http://www.pantherdb.org" title="Protein Analysis THrough Evolutionary Relationships (PANTHER)" class="external-link">PANTHER</a>,
<a href="https://reactome.org/" class="external-link">Reactome</a>, and <a href="https://string-db.org/" title="STRING: Protein-Protein Interaction Networks Functional Enrichment Analysis" class="external-link">STRING</a>
offer Enrichment (Over-Representation) analysis API endpoints. In this
article, after a brief introduction, we will demonstrate how to perform
Enrichment (Over-Representation) analysis in R, consistently and easily,
using the package rbioapi.</p>
<p>Before continuing, note that rbioapi is an interface to the databases
and web services. Hence it is the user’s responsibility to cite the
services they have used. We provide suggested citations in each
function’s manual. However, it is the user’s responsibility to fully and
adequately cite the databases or web services used in their published
research.</p>
<hr>
<div class="section level2">
<h2 id="how-to-install">How to install?<a class="anchor" aria-label="anchor" href="#how-to-install"></a>
</h2>
<p>You can install the <a href="https://cran.r-project.org/package=rbioapi" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API" class="external-link">stable
release version of rbioapi</a> from <a href="https://cran.r-project.org/" title="The Comprehensive R Archive Network" class="external-link">CRAN</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"rbioapi"</span><span class="op">)</span></span></code></pre></div>
<p>However, the CRAN version is released at longer intervals, You can
install the most recent -development- version from <a href="https://github.com/moosa-r/rbioapi/" title="rbioapi repository on GitHub" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"moosa-r/rbioapi"</span><span class="op">)</span></span></code></pre></div>
<p>Now, we can load the package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rbioapi.moosa-r.com">rbioapi</a></span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="what-is-over-representation-analysis">What is Over-representation analysis?<a class="anchor" aria-label="anchor" href="#what-is-over-representation-analysis"></a>
</h2>
<p>Enrichment (Over-Representation) analysis is commonly performed when
one’s analysis yields a list of genes, proteins, or other entities. We
will not delve into details here and refer you to <a href="https://doi.org/10.1093/nar/gkn923" title="Bioinformatics enrichment tools: paths toward the comprehensive functional analysis of large gene lists" class="external-link">D.
W. Huang et al.: Nucleic Acids Res. 37, 1 (2009)</a> for a thorough
review of the subject.</p>
<p>In short, to perform such analysis, in addition to some algorithms,
two data are required:</p>
<ol style="list-style-type: decimal">
<li><p>A gene list which is derived from your analysis.</p></li>
<li><p>Gene sets, where genes are grouped based on a shared property or
concept (such as pathway, role in diseases, physical location,
etc.).</p></li>
</ol>
<p>With over-representation analysis, you can compare your gene list to
find if the member of any gene set are over-represented, meaning that
they are present in your list in a manner that can not be explained by
chance alone.</p>
<hr>
</div>
<div class="section level2">
<h2 id="setting-up-data">Setting up the data<a class="anchor" aria-label="anchor" href="#setting-up-data"></a>
</h2>
<p>Many web services supported by rbioapi provide tools to perform
over-representation analysis against the gene sets curated in their
databases. Among the databases and web services currently supported by
rbioapi, <a href="https://maayanlab.cloud/Enrichr/" title="Enrichr" class="external-link">Enrichr</a>, <a href="https://ccb-compute2.cs.uni-saarland.de/mieaa2" title="miRNA Enrichment Analysis and Annotation Tool (miEAA)" class="external-link">miEAA</a>,
<a href="https://www.pantherdb.org" title="Protein Analysis THrough Evolutionary Relationships (PANTHER)" class="external-link">PANTHER</a>,
<a href="https://reactome.org/" class="external-link">Reactome</a>, and <a href="https://string-db.org/" title="STRING: Protein-Protein Interaction Networks Functional Enrichment Analysis" class="external-link">STRING</a>
provide such services. Each service has a dedicated vignette article in
rbioapi. Here we will focus on performing Enrichment
(Over-Representation) analysis.</p>
<p>As a demonstration, we will use the results of <a href="https://doi.org/10.1038/s41590-020-00840-x" title="Untuned antiviral immunity in COVID-19 revealed by temporal type I/III interferon patterns and flu comparison" class="external-link">Galani
et al.: Nat. Immunol. 22, 32 (2021)</a> where they have compared the
gene expression of critically ill and non-critically ill COVID patients
to healthy individuals. We assign a variable with 1214 differentially
expressed genes (DEGs) exclusively induced in critically ill COVID
patients. Next, we will see what insights we can extract from this
list.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">covid_critical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TXNDC5"</span>, <span class="st">"GABRR2"</span>, <span class="st">"MGAM2"</span>, <span class="st">"LOC200772"</span>, <span class="st">"LYPD2"</span>, <span class="st">"IFI27"</span>, <span class="st">"RPH3A"</span>, <span class="st">"OTOF"</span>, <span class="st">"NBPF8"</span>, <span class="st">"CLEC4F"</span>, <span class="st">"CNGB1"</span>, <span class="st">"HIST1H2BF"</span>, <span class="st">"HIF1A"</span>, <span class="st">"SDC1"</span>, <span class="st">"TUBB8"</span>, <span class="st">"FBXO39"</span>, <span class="st">"TPSB2"</span>, <span class="st">"CD177"</span>, <span class="st">"LRRN3"</span>, <span class="st">"EBLN2"</span>, <span class="st">"PCSK9"</span>, <span class="st">"ELK2AP"</span>, <span class="st">"UCHL1"</span>, <span class="st">"C22orf15"</span>, <span class="st">"LPO"</span>, <span class="st">"C3orf20"</span>, <span class="st">"CLRN1-AS1"</span>, <span class="st">"GPR75"</span>, <span class="st">"CA12"</span>, <span class="st">"RAB19"</span>, <span class="st">"CHRFAM7A"</span>, <span class="st">"CRYGN"</span>, <span class="st">"DLGAP5"</span>, <span class="st">"BTBD8"</span>, <span class="st">"LOC100272216"</span>, <span class="st">"PRG3"</span>, <span class="st">"CYP46A1"</span>, <span class="st">"LOC102723604"</span>, <span class="st">"PPAP2B"</span>, <span class="st">"C4BPA"</span>, <span class="st">"SPESP1"</span>, <span class="st">"LILRP2"</span>, <span class="st">"UBE2Q1-AS1"</span>, <span class="st">"MIR3945"</span>, <span class="st">"NOMO3"</span>, <span class="st">"MEG3"</span>, <span class="st">"LOC400927-CSNK1E"</span>, <span class="st">"MIR6732"</span>, <span class="st">"MIR590"</span>, <span class="st">"PPP1R3G"</span>, <span class="st">"PYCR1"</span>, <span class="st">"ARHGAP42"</span>, <span class="st">"MMP8"</span>, <span class="st">"HMMR"</span>, <span class="st">"P3H2"</span>, <span class="st">"ACER1"</span>, <span class="st">"NOG"</span>, <span class="st">"RAB39A"</span>, <span class="st">"ANTXRLP1"</span>, <span class="st">"LINC00266-3"</span>, <span class="st">"GPRC5D"</span>, <span class="st">"MCM10"</span>, <span class="st">"TSPY26P"</span>, <span class="st">"ANKRD36BP1"</span>, <span class="st">"GBP1P1"</span>, <span class="st">"PRL"</span>, <span class="st">"CYP1A1"</span>, <span class="st">"KIF4A"</span>, <span class="st">"LOC102724323"</span>, <span class="st">"SERPINB10"</span>, <span class="st">"GSTA4"</span>, <span class="st">"TRIM51"</span>, <span class="st">"MIXL1"</span>, <span class="st">"RNASE1"</span>, <span class="st">"CASC8"</span>, <span class="st">"MAOA"</span>, <span class="st">"XCL1"</span>, <span class="st">"ADAMTS2"</span>, <span class="st">"LOC101929125"</span>, <span class="st">"DCANP1"</span>, <span class="st">"BHLHA15"</span>, <span class="st">"ANOS1"</span>, <span class="st">"SLC18A1"</span>, <span class="st">"CCDC150"</span>, <span class="st">"CAV1"</span>, <span class="st">"SH3BP5"</span>, <span class="st">"LINC00398"</span>, <span class="st">"NCOA2"</span>, <span class="st">"SPATC1"</span>, <span class="st">"SHROOM2"</span>, <span class="st">"GPR27"</span>, <span class="st">"LRRC26"</span>, <span class="st">"RNF169"</span>, <span class="st">"USP3-AS1"</span>, <span class="st">"VWA7"</span>, <span class="st">"ROCK2"</span>, <span class="st">"FSTL4"</span>, <span class="st">"METTL7B"</span>, <span class="st">"CYP4F29P"</span>, <span class="st">"LEF1-AS1"</span>, <span class="st">"HORMAD2-AS1"</span>, <span class="st">"FBXO15"</span>, <span class="st">"PPTC7"</span>, <span class="st">"TYMS"</span>, <span class="st">"PPP4R2"</span>, <span class="st">"ZNF608"</span>, <span class="st">"FAM46B"</span>, <span class="st">"PCSK1N"</span>, <span class="st">"LINC00623"</span>, <span class="st">"CASC5"</span>, <span class="st">"ZNF224"</span>, <span class="st">"DENND2C"</span>, <span class="st">"WDR86"</span>, <span class="st">"PTGR1"</span>, <span class="st">"SPATA3-AS1"</span>, <span class="st">"LOC101927412"</span>, <span class="st">"KIF14"</span>, <span class="st">"MMP28"</span>, <span class="st">"PBK"</span>, <span class="st">"VMO1"</span>, <span class="st">"ADCY3"</span>, <span class="st">"HIST1H2BO"</span>, <span class="st">"FTO-IT1"</span>, <span class="st">"MIR342"</span>, <span class="st">"FANK1"</span>, <span class="st">"CRIP2"</span>, <span class="st">"TIAF1"</span>, <span class="st">"LOC344887"</span>, <span class="st">"OLFM4"</span>, <span class="st">"MKNK1-AS1"</span>, <span class="st">"ZDHHC19"</span>, <span class="st">"SEPT14"</span>, <span class="st">"EPSTI1"</span>, <span class="st">"FOXC1"</span>, <span class="st">"MIR616"</span>, <span class="st">"KRT72"</span>, <span class="st">"LINC01347"</span>, <span class="st">"LOC101928100"</span>, <span class="st">"KIAA0895"</span>, <span class="st">"BOK"</span>, <span class="st">"HIST1H2AI"</span>, <span class="st">"DBH-AS1"</span>, <span class="st">"ADORA2A-AS1"</span>, <span class="st">"MED12L"</span>, <span class="st">"SAMD15"</span>, <span class="st">"TARM1"</span>, <span class="st">"SMTNL1"</span>, <span class="st">"POU5F1P3"</span>, <span class="st">"LINC00968"</span>, <span class="st">"OAS3"</span>, <span class="st">"LOC103091866"</span>, <span class="st">"SH3RF3-AS1"</span>, <span class="st">"NBPF10"</span>, <span class="st">"APAF1"</span>, <span class="st">"SLC2A14"</span>, <span class="st">"SYT17"</span>, <span class="st">"ETV3"</span>, <span class="st">"SHROOM4"</span>, <span class="st">"AOAH-IT1"</span>, <span class="st">"NAIP"</span>, <span class="st">"ALAS2"</span>, <span class="st">"GLIS3"</span>, <span class="st">"ADAM17"</span>, <span class="st">"OLFM1"</span>, <span class="st">"PCAT29"</span>, <span class="st">"TNFRSF18"</span>, <span class="st">"DNASE1L3"</span>, <span class="st">"IGF2BP3"</span>, <span class="st">"LINC01271"</span>, <span class="st">"AP3B2"</span>, <span class="st">"TXNDC2"</span>, <span class="st">"CEP55"</span>, <span class="st">"SIGLEC1"</span>, <span class="st">"RAB3IL1"</span>, <span class="st">"PLD4"</span>, <span class="st">"KIFC1"</span>, <span class="st">"LINC00487"</span>, <span class="st">"ABHD12B"</span>, <span class="st">"ITGA7"</span>, <span class="st">"GJB6"</span>, <span class="st">"CARD14"</span>, <span class="st">"LRRN2"</span>, <span class="st">"MPO"</span>, <span class="st">"KLRC3"</span>, <span class="st">"LOC100507487"</span>, <span class="st">"SCGB3A1"</span>, <span class="st">"CD38"</span>, <span class="st">"LRRN1"</span>, <span class="st">"SYCP2L"</span>, <span class="st">"ANLN"</span>, <span class="st">"ASPM"</span>, <span class="st">"OAS1"</span>, <span class="st">"IFI44L"</span>, <span class="st">"CDCA3"</span>, <span class="st">"HLA-DQB2"</span>, <span class="st">"ANO9"</span>, <span class="st">"NUDT11"</span>, <span class="st">"HMP19"</span>, <span class="st">"DEPDC1"</span>, <span class="st">"GPR84"</span>, <span class="st">"PLEKHF1"</span>, <span class="st">"PI16"</span>, <span class="st">"RDH5"</span>, <span class="st">"TMED8"</span>, <span class="st">"LINC00824"</span>, <span class="st">"SEPT4"</span>, <span class="st">"CLIC3"</span>, <span class="st">"B4GALNT3"</span>, <span class="st">"OLAH"</span>, <span class="st">"ITGA1"</span>, <span class="st">"FOXI1"</span>, <span class="st">"LOC100506142"</span>, <span class="st">"CDC20"</span>, <span class="st">"HAGHL"</span>, <span class="st">"GTSCR1"</span>, <span class="st">"B3GALNT1"</span>, <span class="st">"SOCS3"</span>, <span class="st">"PCDH1"</span>, <span class="st">"TAS2R20"</span>, <span class="st">"CDC25A"</span>, <span class="st">"NPDC1"</span>, <span class="st">"LOC100505915"</span>, <span class="st">"XCL2"</span>, <span class="st">"TIAM2"</span>, <span class="st">"LOC100288069"</span>, <span class="st">"IL34"</span>, <span class="st">"IL18R1"</span>, <span class="st">"CYYR1"</span>, <span class="st">"ZNF888"</span>, <span class="st">"FAM20A"</span>, <span class="st">"MDS2"</span>, <span class="st">"ABCA13"</span>, <span class="st">"KCNJ1"</span>, <span class="st">"SLC4A9"</span>, <span class="st">"EXO1"</span>, <span class="st">"LAIR2"</span>, <span class="st">"IQSEC3"</span>, <span class="st">"SCT"</span>, <span class="st">"SLC26A8"</span>, <span class="st">"ESCO2"</span>, <span class="st">"ZDHHC1"</span>, <span class="st">"SNORA63"</span>, <span class="st">"FBLN5"</span>, <span class="st">"PRUNE2"</span>, <span class="st">"CPNE7"</span>, <span class="st">"CDHR2"</span>, <span class="st">"GNLY"</span>, <span class="st">"APOBEC3B"</span>, <span class="st">"MFSD9"</span>, <span class="st">"SYNGR3"</span>, <span class="st">"PTGDS"</span>, <span class="st">"COL17A1"</span>, <span class="st">"TBC1D14"</span>, <span class="st">"AIM2"</span>, <span class="st">"TMEM204"</span>, <span class="st">"FAM157B"</span>, <span class="st">"ZBED6"</span>, <span class="st">"EME1"</span>, <span class="st">"ATF3"</span>, <span class="st">"KBTBD4"</span>, <span class="st">"LUC7L2"</span>, <span class="st">"KIF20A"</span>, <span class="st">"LCNL1"</span>, <span class="st">"DOCK9-AS2"</span>, <span class="st">"KCNE1"</span>, <span class="st">"BTN2A3P"</span>, <span class="st">"KL"</span>, <span class="st">"IDI2-AS1"</span>, <span class="st">"EBF4"</span>, <span class="st">"SCARNA21"</span>, <span class="st">"CEACAM6"</span>, <span class="st">"KLRB1"</span>, <span class="st">"C5orf58"</span>, <span class="st">"ASXL2"</span>, <span class="st">"RPLP0P2"</span>, <span class="st">"CYTL1"</span>, <span class="st">"DIAPH3"</span>, <span class="st">"DOK7"</span>, <span class="st">"RFFL"</span>, <span class="st">"KIAA1107"</span>, <span class="st">"TAS2R40"</span>, <span class="st">"CCDC186"</span>, <span class="st">"COL6A2"</span>, <span class="st">"METRN"</span>, <span class="st">"SNHG25"</span>, <span class="st">"RRM2"</span>, <span class="st">"CYP1B1"</span>, <span class="st">"NEURL1"</span>, <span class="st">"MATK"</span>, <span class="st">"SLC28A3"</span>, <span class="st">"JCHAIN"</span>, <span class="st">"TNFRSF4"</span>, <span class="st">"FXYD1"</span>, <span class="st">"PLLP"</span>, <span class="st">"ARHGAP23"</span>, <span class="st">"SNORA4"</span>, <span class="st">"MYO18A"</span>, <span class="st">"ZNF429"</span>, <span class="st">"NRN1"</span>, <span class="st">"HJURP"</span>, <span class="st">"TCN1"</span>, <span class="st">"CDC6"</span>, <span class="st">"ATP5EP2"</span>, <span class="st">"SOAT2"</span>, <span class="st">"LOC101928034"</span>, <span class="st">"EIF2AK2"</span>, <span class="st">"ARG1"</span>, <span class="st">"SLC16A11"</span>, <span class="st">"TPPP3"</span>, <span class="st">"TMEM38A"</span>, <span class="st">"TOP2A"</span>, <span class="st">"LINC00999"</span>, <span class="st">"DGCR9"</span>, <span class="st">"RCAN3AS"</span>, <span class="st">"CACNG6"</span>, <span class="st">"LINC01550"</span>, <span class="st">"TDRD9"</span>, <span class="st">"CARD17"</span>, <span class="st">"GBP6"</span>, <span class="st">"LY6E"</span>, <span class="st">"RSAD2"</span>, <span class="st">"LOC100506258"</span>, <span class="st">"PRKAR2A"</span>, <span class="st">"LTF"</span>, <span class="st">"IGLL5"</span>, <span class="st">"FAM157A"</span>, <span class="st">"LOC101927018"</span>, <span class="st">"CCR7"</span>, <span class="st">"FLJ42351"</span>, <span class="st">"IFIT3"</span>, <span class="st">"GPRIN1"</span>, <span class="st">"ANXA9"</span>, <span class="st">"TMEM119"</span>, <span class="st">"PARGP1"</span>, <span class="st">"A3GALT2"</span>, <span class="st">"CYP4F22"</span>, <span class="st">"PLBD1-AS1"</span>, <span class="st">"SELM"</span>, <span class="st">"NCR3"</span>, <span class="st">"PNPLA1"</span>, <span class="st">"BMX"</span>, <span class="st">"LOC440461"</span>, <span class="st">"GAMT"</span>, <span class="st">"CEACAM8"</span>, <span class="st">"TP53I13"</span>, <span class="st">"C1orf226"</span>, <span class="st">"SKA3"</span>, <span class="st">"DHRS3"</span>, <span class="st">"VRK2"</span>, <span class="st">"BTBD19"</span>, <span class="st">"ETNK2"</span>, <span class="st">"LOC728323"</span>, <span class="st">"NRIR"</span>, <span class="st">"ADORA2A"</span>, <span class="st">"GPR162"</span>, <span class="st">"CMPK2"</span>, <span class="st">"USP18"</span>, <span class="st">"P4HA2"</span>, <span class="st">"TSSK4"</span>, <span class="st">"EP300"</span>, <span class="st">"CLEC4D"</span>, <span class="st">"LCN10"</span>, <span class="st">"GPR141"</span>, <span class="st">"SH2D2A"</span>, <span class="st">"GOLGA7B"</span>, <span class="st">"TMIGD2"</span>, <span class="st">"DLL4"</span>, <span class="st">"HP"</span>, <span class="st">"CXCR6"</span>, <span class="st">"MAL"</span>, <span class="st">"C12orf57"</span>, <span class="st">"CLIC5"</span>, <span class="st">"IL4"</span>, <span class="st">"IGFBP6"</span>, <span class="st">"ERG"</span>, <span class="st">"HERC3"</span>, <span class="st">"KEL"</span>, <span class="st">"MSX2P1"</span>, <span class="st">"INE1"</span>, <span class="st">"PRKCQ-AS1"</span>, <span class="st">"FHIT"</span>, <span class="st">"SLC1A7"</span>, <span class="st">"KIAA1958"</span>, <span class="st">"SARDH"</span>, <span class="st">"PFKFB3"</span>, <span class="st">"SNORD89"</span>, <span class="st">"IFI44"</span>, <span class="st">"DDX60"</span>, <span class="st">"TMEM238"</span>, <span class="st">"HAR1A"</span>, <span class="st">"EGR1"</span>, <span class="st">"VNN1"</span>, <span class="st">"TRIM9"</span>, <span class="st">"TAF13"</span>, <span class="st">"AP3S2"</span>, <span class="st">"TMEM56"</span>, <span class="st">"KLRC2"</span>, <span class="st">"CACNA1C-AS2"</span>, <span class="st">"ALX3"</span>, <span class="st">"FCGBP"</span>, <span class="st">"CD247"</span>, <span class="st">"ALDH1L2"</span>, <span class="st">"HIST1H2AC"</span>, <span class="st">"RTP5"</span>, <span class="st">"PPARG"</span>, <span class="st">"AMPH"</span>, <span class="st">"LINC00861"</span>, <span class="st">"CDKL5"</span>, <span class="st">"MYBL2"</span>, <span class="st">"LOC101927051"</span>, <span class="st">"TLR5"</span>, <span class="st">"TMEM121"</span>, <span class="st">"BMP8B"</span>, <span class="st">"AK5"</span>, <span class="st">"RBP5"</span>, <span class="st">"LINC01355"</span>, <span class="st">"PITRM1-AS1"</span>, <span class="st">"CMTM1"</span>, <span class="st">"BIRC5"</span>, <span class="st">"C10orf10"</span>, <span class="st">"TNFRSF25"</span>, <span class="st">"ZAP70"</span>, <span class="st">"LOC105373383"</span>, <span class="st">"DSP"</span>, <span class="st">"WDR86-AS1"</span>, <span class="st">"RPLP2"</span>, <span class="st">"KREMEN1"</span>, <span class="st">"LOC101927550"</span>, <span class="st">"PDZD4"</span>, <span class="st">"LOC100130451"</span>, <span class="st">"MKI67"</span>, <span class="st">"LOC728743"</span>, <span class="st">"CXCL10"</span>, <span class="st">"LINC01547"</span>, <span class="st">"LOC645513"</span>, <span class="st">"ATOH8"</span>, <span class="st">"S1PR5"</span>, <span class="st">"KISS1R"</span>, <span class="st">"MELK"</span>, <span class="st">"TCEA3"</span>, <span class="st">"SLC22A20"</span>, <span class="st">"RAMP1"</span>, <span class="st">"FOLR2"</span>, <span class="st">"GGT6"</span>, <span class="st">"CACNA1E"</span>, <span class="st">"FABP6"</span>, <span class="st">"RAP2C-AS1"</span>, <span class="st">"PRRG4"</span>, <span class="st">"FAM63B"</span>, <span class="st">"BLK"</span>, <span class="st">"P3H3"</span>, <span class="st">"WIPF2"</span>, <span class="st">"TROAP"</span>, <span class="st">"FAHD2CP"</span>, <span class="st">"CA6"</span>, <span class="st">"LINC00892"</span>, <span class="st">"LRCH3"</span>, <span class="st">"BUB1B"</span>, <span class="st">"LOC100996286"</span>, <span class="st">"NT5E"</span>, <span class="st">"PASK"</span>, <span class="st">"BATF2"</span>, <span class="st">"TTC39C-AS1"</span>, <span class="st">"ACTA2"</span>, <span class="st">"CHIT1"</span>, <span class="st">"LAMC1"</span>, <span class="st">"TNFAIP3"</span>, <span class="st">"ANKRD22"</span>, <span class="st">"C12orf42"</span>, <span class="st">"SPON2"</span>, <span class="st">"SLX4IP"</span>, <span class="st">"TNNC1"</span>, <span class="st">"ZNF771"</span>, <span class="st">"SPEG"</span>, <span class="st">"HIST1H4H"</span>, <span class="st">"HTR6"</span>, <span class="st">"TNFRSF17"</span>, <span class="st">"FSD1"</span>, <span class="st">"LINC00266-1"</span>, <span class="st">"CD6"</span>, <span class="st">"PRSS30P"</span>, <span class="st">"SFTPD"</span>, <span class="st">"COPZ2"</span>, <span class="st">"BPI"</span>, <span class="st">"CCIN"</span>, <span class="st">"CDK1"</span>, <span class="st">"ATP2C2"</span>, <span class="st">"GOLGA8H"</span>, <span class="st">"USP44"</span>, <span class="st">"SLC14A1"</span>, <span class="st">"ZDHHC14"</span>, <span class="st">"SNORD50A"</span>, <span class="st">"LTK"</span>, <span class="st">"KCNG2"</span>, <span class="st">"MAPK11"</span>, <span class="st">"ACOT4"</span>, <span class="st">"CTSW"</span>, <span class="st">"C1orf106"</span>, <span class="st">"CDKN3"</span>, <span class="st">"UPB1"</span>, <span class="st">"CAMK2N1"</span>, <span class="st">"RBM5-AS1"</span>, <span class="st">"DNLZ"</span>, <span class="st">"GZMM"</span>, <span class="st">"PLA2G7"</span>, <span class="st">"ZCCHC2"</span>, <span class="st">"PRSS41"</span>, <span class="st">"RPL13"</span>, <span class="st">"OLR1"</span>, <span class="st">"BCAS4"</span>, <span class="st">"EPB41"</span>, <span class="st">"TRABD2A"</span>, <span class="st">"PBX4"</span>, <span class="st">"GZMK"</span>, <span class="st">"PNPLA7"</span>, <span class="st">"B4GALT5"</span>, <span class="st">"ARHGEF17"</span>, <span class="st">"APOA1"</span>, <span class="st">"GPA33"</span>, <span class="st">"KLRG1"</span>, <span class="st">"GBP5"</span>, <span class="st">"LOC102723701"</span>, <span class="st">"CENPA"</span>, <span class="st">"LOC285696"</span>, <span class="st">"KCNK5"</span>, <span class="st">"SUSD4"</span>, <span class="st">"RPS28"</span>, <span class="st">"ESPL1"</span>, <span class="st">"ITGB4"</span>, <span class="st">"SPAG16"</span>, <span class="st">"CYP4F35P"</span>, <span class="st">"CD3D"</span>, <span class="st">"BMS1P20"</span>, <span class="st">"CBS"</span>, <span class="st">"ETV2"</span>, <span class="st">"SPATS2L"</span>, <span class="st">"DUSP13"</span>, <span class="st">"FAM19A1"</span>, <span class="st">"VSIG10L"</span>, <span class="st">"CD2"</span>, <span class="st">"NHSL2"</span>, <span class="st">"FAAH2"</span>, <span class="st">"FXYD7"</span>, <span class="st">"NTN5"</span>, <span class="st">"FICD"</span>, <span class="st">"GLDC"</span>, <span class="st">"LOC101927865"</span>, <span class="st">"LINC00944"</span>, <span class="st">"PTPRCAP"</span>, <span class="st">"GNB3"</span>, <span class="st">"TRIM22"</span>, <span class="st">"PRRT1"</span>, <span class="st">"NR1I3"</span>, <span class="st">"BTN3A1"</span>, <span class="st">"STX16-NPEPL1"</span>, <span class="st">"TMEM191A"</span>, <span class="st">"SERPINB2"</span>, <span class="st">"UST"</span>, <span class="st">"GALNT12"</span>, <span class="st">"BUB1"</span>, <span class="st">"GTF2I"</span>, <span class="st">"FOSL2"</span>, <span class="st">"COL13A1"</span>, <span class="st">"RGS9"</span>, <span class="st">"NCALD"</span>, <span class="st">"DNAH10"</span>, <span class="st">"CA11"</span>, <span class="st">"CKB"</span>, <span class="st">"HSPB1"</span>, <span class="st">"CDC45"</span>, <span class="st">"ATP7B"</span>, <span class="st">"WNT5B"</span>, <span class="st">"ZNF699"</span>, <span class="st">"PRRT4"</span>, <span class="st">"GALNT14"</span>, <span class="st">"ZNF319"</span>, <span class="st">"DNAH17"</span>, <span class="st">"LOC283710"</span>, <span class="st">"CPEB4"</span>, <span class="st">"LY6G6C"</span>, <span class="st">"PPAN"</span>, <span class="st">"WASH5P"</span>, <span class="st">"HIST1H3H"</span>, <span class="st">"IL32"</span>, <span class="st">"TEPP"</span>, <span class="st">"CNR1"</span>, <span class="st">"YJEFN3"</span>, <span class="st">"FAM159A"</span>, <span class="st">"FGFBP2"</span>, <span class="st">"FKBP5"</span>, <span class="st">"BIN1"</span>, <span class="st">"VPREB3"</span>, <span class="st">"E2F8"</span>, <span class="st">"HK3"</span>, <span class="st">"CDH2"</span>, <span class="st">"HFE"</span>, <span class="st">"BEAN1-AS1"</span>, <span class="st">"KIFC3"</span>, <span class="st">"HELB"</span>, <span class="st">"HLA-DPB1"</span>, <span class="st">"GZMH"</span>, <span class="st">"LMNB1"</span>, <span class="st">"CC2D2A"</span>, <span class="st">"IQCH-AS1"</span>, <span class="st">"CTSF"</span>, <span class="st">"TMEM132D"</span>, <span class="st">"LEF1"</span>, <span class="st">"REL"</span>, <span class="st">"CBR3"</span>, <span class="st">"WNT10B"</span>, <span class="st">"LOC100289473"</span>, <span class="st">"APBA2"</span>, <span class="st">"KIF23"</span>, <span class="st">"LOC646471"</span>, <span class="st">"PNMA6A"</span>, <span class="st">"EPC2"</span>, <span class="st">"HBEGF"</span>, <span class="st">"S100B"</span>, <span class="st">"C14orf132"</span>, <span class="st">"KCND1"</span>, <span class="st">"SIPA1L2"</span>, <span class="st">"NCAPG"</span>, <span class="st">"OAS2"</span>, <span class="st">"C3AR1"</span>, <span class="st">"CSF1"</span>, <span class="st">"CHCHD6"</span>, <span class="st">"AUTS2"</span>, <span class="st">"IL24"</span>, <span class="st">"CDC42EP2"</span>, <span class="st">"LTBP3"</span>, <span class="st">"PXT1"</span>, <span class="st">"ADAMTS10"</span>, <span class="st">"BLZF1"</span>, <span class="st">"TPST1"</span>, <span class="st">"ID3"</span>, <span class="st">"GYG1"</span>, <span class="st">"EFCAB2"</span>, <span class="st">"MYO10"</span>, <span class="st">"TLE2"</span>, <span class="st">"SLC30A1"</span>, <span class="st">"CCNI"</span>, <span class="st">"WNT7A"</span>, <span class="st">"RBL2"</span>, <span class="st">"ERICH6-AS1"</span>, <span class="st">"IGFBP3"</span>, <span class="st">"DTL"</span>, <span class="st">"SORBS3"</span>, <span class="st">"RPS19"</span>, <span class="st">"SMA4"</span>, <span class="st">"RPS27"</span>, <span class="st">"BEAN1"</span>, <span class="st">"MIRLET7BHG"</span>, <span class="st">"ABCA1"</span>, <span class="st">"ZKSCAN7"</span>, <span class="st">"GPR34"</span>, <span class="st">"GPR153"</span>, <span class="st">"TRIM52"</span>, <span class="st">"GRASP"</span>, <span class="st">"RNASE2"</span>, <span class="st">"CENPF"</span>, <span class="st">"ACTG1P20"</span>, <span class="st">"MMP9"</span>, <span class="st">"ZFP82"</span>, <span class="st">"RTN1"</span>, <span class="st">"RPS21"</span>, <span class="st">"PCBP1-AS1"</span>, <span class="st">"PVRL3"</span>, <span class="st">"RPS10"</span>, <span class="st">"FER1L4"</span>, <span class="st">"SEPT1"</span>, <span class="st">"DEPDC1B"</span>, <span class="st">"LINC00926"</span>, <span class="st">"SEC14L2"</span>, <span class="st">"LOC100420587"</span>, <span class="st">"PCSK4"</span>, <span class="st">"HPS3"</span>, <span class="st">"KIR3DL1"</span>, <span class="st">"SPRN"</span>, <span class="st">"HRH2"</span>, <span class="st">"FAHD2B"</span>, <span class="st">"P3H4"</span>, <span class="st">"TMEM160"</span>, <span class="st">"ENGASE"</span>, <span class="st">"RAD51AP1"</span>, <span class="st">"ZNF233"</span>, <span class="st">"RAP1GAP"</span>, <span class="st">"TTBK2"</span>, <span class="st">"PINK1"</span>, <span class="st">"SH2D1B"</span>, <span class="st">"BAIAP2L1"</span>, <span class="st">"BECN1"</span>, <span class="st">"RPL13A"</span>, <span class="st">"INHBB"</span>, <span class="st">"HPCAL4"</span>, <span class="st">"TPM2"</span>, <span class="st">"ACVR2B"</span>, <span class="st">"CEP126"</span>, <span class="st">"RPL18"</span>, <span class="st">"B3GNT7"</span>, <span class="st">"ROBO3"</span>, <span class="st">"MYOM2"</span>, <span class="st">"SLC51A"</span>, <span class="st">"CCL5"</span>, <span class="st">"PROSER3"</span>, <span class="st">"RPL18A"</span>, <span class="st">"SERINC4"</span>, <span class="st">"ATP11B"</span>, <span class="st">"SMARCD3"</span>, <span class="st">"STMN3"</span>, <span class="st">"FAM173A"</span>, <span class="st">"CDK5RAP2"</span>, <span class="st">"GGH"</span>, <span class="st">"EIF4G3"</span>, <span class="st">"C18orf32"</span>, <span class="st">"SHCBP1"</span>, <span class="st">"CCM2L"</span>, <span class="st">"LAT"</span>, <span class="st">"PRF1"</span>, <span class="st">"ARL15"</span>, <span class="st">"WHAMMP1"</span>, <span class="st">"NBN"</span>, <span class="st">"ARHGEF11"</span>, <span class="st">"ZWINT"</span>, <span class="st">"BTBD11"</span>, <span class="st">"MCOLN2"</span>, <span class="st">"EYS"</span>, <span class="st">"GRAP"</span>, <span class="st">"TAP1"</span>, <span class="st">"FAM157C"</span>, <span class="st">"ANXA3"</span>, <span class="st">"MCEMP1"</span>, <span class="st">"TCF7"</span>, <span class="st">"IGSF9B"</span>, <span class="st">"SDSL"</span>, <span class="st">"LNPEP"</span>, <span class="st">"NSUN5P1"</span>, <span class="st">"FAM110B"</span>, <span class="st">"SPNS3"</span>, <span class="st">"ACTL10"</span>, <span class="st">"GCSAM"</span>, <span class="st">"LOC101927482"</span>, <span class="st">"UGCG"</span>, <span class="st">"CD24"</span>, <span class="st">"SCO2"</span>, <span class="st">"GMCL1"</span>, <span class="st">"SAMD12"</span>, <span class="st">"NUSAP1"</span>, <span class="st">"WNT1"</span>, <span class="st">"NME3"</span>, <span class="st">"LOC728084"</span>, <span class="st">"RGS18"</span>, <span class="st">"PNMA3"</span>, <span class="st">"KLRF1"</span>, <span class="st">"ENO2"</span>, <span class="st">"SUPT3H"</span>, <span class="st">"ATP5D"</span>, <span class="st">"ZC3H13"</span>, <span class="st">"PSTPIP2"</span>, <span class="st">"SNORD81"</span>, <span class="st">"SESN2"</span>, <span class="st">"TK1"</span>, <span class="st">"LHFPL2"</span>, <span class="st">"SCARNA10"</span>, <span class="st">"SERPING1"</span>, <span class="st">"SYT2"</span>, <span class="st">"SNORA33"</span>, <span class="st">"JAK2"</span>, <span class="st">"CYB5D1"</span>, <span class="st">"SH3RF3"</span>, <span class="st">"ARL14EP"</span>, <span class="st">"UBASH3A"</span>, <span class="st">"ARL5B"</span>, <span class="st">"SPRED2"</span>, <span class="st">"NELL2"</span>, <span class="st">"PPP2R2B"</span>, <span class="st">"CDH12"</span>, <span class="st">"WDR54"</span>, <span class="st">"KLF4"</span>, <span class="st">"MTUS1"</span>, <span class="st">"ELF1"</span>, <span class="st">"CARNS1"</span>, <span class="st">"CLEC10A"</span>, <span class="st">"CD3E"</span>, <span class="st">"TTK"</span>, <span class="st">"KBTBD7"</span>, <span class="st">"TPX2"</span>, <span class="st">"ZNF69"</span>, <span class="st">"AICDA"</span>, <span class="st">"TAMM41"</span>, <span class="st">"LOC100288152"</span>, <span class="st">"AK1"</span>, <span class="st">"MMP24"</span>, <span class="st">"SAMD9L"</span>, <span class="st">"RAD51"</span>, <span class="st">"ACE"</span>, <span class="st">"COLGALT2"</span>, <span class="st">"NEDD4"</span>, <span class="st">"SPEF2"</span>, <span class="st">"CENPE"</span>, <span class="st">"CRISPLD2"</span>, <span class="st">"TSEN54"</span>, <span class="st">"RCVRN"</span>, <span class="st">"FRMD3"</span>, <span class="st">"SKAP1"</span>, <span class="st">"CYP4F3"</span>, <span class="st">"TMEM161B-AS1"</span>, <span class="st">"KANSL1L"</span>, <span class="st">"SLC29A2"</span>, <span class="st">"CEP97"</span>, <span class="st">"XAF1"</span>, <span class="st">"CDKN1C"</span>, <span class="st">"PRDM5"</span>, <span class="st">"HBP1"</span>, <span class="st">"CACNG8"</span>, <span class="st">"RP2"</span>, <span class="st">"OASL"</span>, <span class="st">"NPIPA1"</span>, <span class="st">"ZNF354A"</span>, <span class="st">"RPL32"</span>, <span class="st">"GLTSCR2"</span>, <span class="st">"CD52"</span>, <span class="st">"SBK1"</span>, <span class="st">"ZNF703"</span>, <span class="st">"GBP1"</span>, <span class="st">"C4orf29"</span>, <span class="st">"ERV3-1"</span>, <span class="st">"ABCC2"</span>, <span class="st">"EPB41L5"</span>, <span class="st">"FFAR3"</span>, <span class="st">"PLEKHB1"</span>, <span class="st">"LOC100507387"</span>, <span class="st">"RPL36"</span>, <span class="st">"PRC1"</span>, <span class="st">"C19orf60"</span>, <span class="st">"PRSS23"</span>, <span class="st">"CD160"</span>, <span class="st">"HOPX"</span>, <span class="st">"SAMD10"</span>, <span class="st">"RPSAP9"</span>, <span class="st">"CDCA2"</span>, <span class="st">"SWT1"</span>, <span class="st">"NPEPL1"</span>, <span class="st">"RORC"</span>, <span class="st">"MS4A3"</span>, <span class="st">"BCL2L15"</span>, <span class="st">"CXCR5"</span>, <span class="st">"EPHX2"</span>, <span class="st">"B3GNT9"</span>, <span class="st">"CDCA7L"</span>, <span class="st">"LMTK2"</span>, <span class="st">"UBAP1L"</span>, <span class="st">"CD79B"</span>, <span class="st">"OBSCN"</span>, <span class="st">"TMEM102"</span>, <span class="st">"ZNF540"</span>, <span class="st">"SPP1"</span>, <span class="st">"HABP4"</span>, <span class="st">"LOC100130872"</span>, <span class="st">"APBB1"</span>, <span class="st">"GPC2"</span>, <span class="st">"CD1C"</span>, <span class="st">"LONRF3"</span>, <span class="st">"NR2C2"</span>, <span class="st">"NSUN7"</span>, <span class="st">"SIGIRR"</span>, <span class="st">"SNORD38B"</span>, <span class="st">"FAM65C"</span>, <span class="st">"HELZ2"</span>, <span class="st">"ARL4D"</span>, <span class="st">"ENO3"</span>, <span class="st">"RPL9"</span>, <span class="st">"IFT81"</span>, <span class="st">"LSMEM2"</span>, <span class="st">"SCARNA2"</span>, <span class="st">"CD5"</span>, <span class="st">"AHNAK2"</span>, <span class="st">"RPL27A"</span>, <span class="st">"SOX8"</span>, <span class="st">"TMEM161A"</span>, <span class="st">"ZNF81"</span>, <span class="st">"PXK"</span>, <span class="st">"LOC441081"</span>, <span class="st">"RPS26"</span>, <span class="st">"CCNB1"</span>, <span class="st">"PLXDC1"</span>, <span class="st">"NFKBIA"</span>, <span class="st">"PKMYT1"</span>, <span class="st">"SLC27A5"</span>, <span class="st">"LRRC4"</span>, <span class="st">"TARP"</span>, <span class="st">"MAP2K6"</span>, <span class="st">"EIF1B-AS1"</span>, <span class="st">"SCARNA17"</span>, <span class="st">"NUDT8"</span>, <span class="st">"EAF1"</span>, <span class="st">"TBC1D8"</span>, <span class="st">"SESTD1"</span>, <span class="st">"CLEC12B"</span>, <span class="st">"ZNF776"</span>, <span class="st">"PDCD1"</span>, <span class="st">"FBXL16"</span>, <span class="st">"SLFN14"</span>, <span class="st">"ATP8B3"</span>, <span class="st">"ZNF396"</span>, <span class="st">"STOM"</span>, <span class="st">"CCDC30"</span>, <span class="st">"FAM86FP"</span>, <span class="st">"FAM129C"</span>, <span class="st">"TMEM42"</span>, <span class="st">"ZNF607"</span>, <span class="st">"IL11RA"</span>, <span class="st">"ECHDC2"</span>, <span class="st">"CCNE2"</span>, <span class="st">"PUS7L"</span>, <span class="st">"EFCAB12"</span>, <span class="st">"TGFA"</span>, <span class="st">"MAPK14"</span>, <span class="st">"NSG1"</span>, <span class="st">"C9orf142"</span>, <span class="st">"KLHL15"</span>, <span class="st">"LOC102723766"</span>, <span class="st">"VEGFB"</span>, <span class="st">"TJP3"</span>, <span class="st">"YAE1D1"</span>, <span class="st">"PROS1"</span>, <span class="st">"CSGALNACT2"</span>, <span class="st">"HLA-DOA"</span>, <span class="st">"TUBBP5"</span>, <span class="st">"RPL37A"</span>, <span class="st">"FAM195A"</span>, <span class="st">"LTB4R2"</span>, <span class="st">"NCAPH"</span>, <span class="st">"EVL"</span>, <span class="st">"CR1"</span>, <span class="st">"LTC4S"</span>, <span class="st">"ANKS3"</span>, <span class="st">"RPL35"</span>, <span class="st">"GLIS2"</span>, <span class="st">"TRAPPC6A"</span>, <span class="st">"LIPN"</span>, <span class="st">"RPS12"</span>, <span class="st">"RPL38"</span>, <span class="st">"LIME1"</span>, <span class="st">"PHGDH"</span>, <span class="st">"C4orf48"</span>, <span class="st">"SLED1"</span>, <span class="st">"BACH2"</span>, <span class="st">"EVA1C"</span>, <span class="st">"GATA3"</span>, <span class="st">"SLC25A38"</span>, <span class="st">"USP32"</span>, <span class="st">"RPL39"</span>, <span class="st">"EGR2"</span>, <span class="st">"USP35"</span>, <span class="st">"ZNF662"</span>, <span class="st">"PLEKHG1"</span>, <span class="st">"CRTAM"</span>, <span class="st">"AANAT"</span>, <span class="st">"OPLAH"</span>, <span class="st">"FBXO6"</span>, <span class="st">"LOXL1"</span>, <span class="st">"LOC101928786"</span>, <span class="st">"FRMD4B"</span>, <span class="st">"IL1R1"</span>, <span class="st">"ADM2"</span>, <span class="st">"DNASE1L1"</span>, <span class="st">"STAT4"</span>, <span class="st">"RPLP1"</span>, <span class="st">"TTC9"</span>, <span class="st">"BEND7"</span>, <span class="st">"F8"</span>, <span class="st">"FOSB"</span>, <span class="st">"SLAMF6"</span>, <span class="st">"STK17B"</span>, <span class="st">"PVT1"</span>, <span class="st">"ARVCF"</span>, <span class="st">"APOBEC3C"</span>, <span class="st">"CDK14"</span>, <span class="st">"RPS14"</span>, <span class="st">"POLQ"</span>, <span class="st">"CDC42BPG"</span>, <span class="st">"CCDC85B"</span>, <span class="st">"UBR5-AS1"</span>, <span class="st">"TNNT3"</span>, <span class="st">"TCL1A"</span>, <span class="st">"IL7R"</span>, <span class="st">"PTX3"</span>, <span class="st">"KLHL14"</span>, <span class="st">"MTRNR2L1"</span>, <span class="st">"ZBTB46"</span>, <span class="st">"RPL34"</span>, <span class="st">"VPS9D1-AS1"</span>, <span class="st">"KLF7"</span>, <span class="st">"RPL10A"</span>, <span class="st">"SLC2A11"</span>, <span class="st">"ACTN4"</span>, <span class="st">"RPSA"</span>, <span class="st">"SGK223"</span>, <span class="st">"PLSCR1"</span>, <span class="st">"SBF2"</span>, <span class="st">"LINC01420"</span>, <span class="st">"CDCA5"</span>, <span class="st">"F5"</span>, <span class="st">"ADGRB2"</span>, <span class="st">"RPL23A"</span>, <span class="st">"MAN1A2"</span>, <span class="st">"LLGL2"</span>, <span class="st">"GINS2"</span>, <span class="st">"NUDT14"</span>, <span class="st">"TG"</span>, <span class="st">"RSRC1"</span>, <span class="st">"TMEM256"</span>, <span class="st">"SYTL2"</span>, <span class="st">"TYSND1"</span>, <span class="st">"TPGS1"</span>, <span class="st">"ABHD14B"</span>, <span class="st">"LSM7"</span>, <span class="st">"ZNF281"</span>, <span class="st">"CEMP1"</span>, <span class="st">"M1AP"</span>, <span class="st">"CD82"</span>, <span class="st">"LOC101928150"</span>, <span class="st">"MRPL41"</span>, <span class="st">"IFIH1"</span>, <span class="st">"CCNB2"</span>, <span class="st">"IFITM10"</span>, <span class="st">"RPS3A"</span>, <span class="st">"TMEM8B"</span>, <span class="st">"CEP135"</span>, <span class="st">"SIRPB2"</span>, <span class="st">"IRS2"</span>, <span class="st">"SYCE1L"</span>, <span class="st">"SULT1B1"</span>, <span class="st">"ADAM9"</span>, <span class="st">"ZCCHC18"</span>, <span class="st">"DDIAS"</span>, <span class="st">"LINC01278"</span>, <span class="st">"MTHFD2"</span>, <span class="st">"TIGIT"</span>, <span class="st">"FUOM"</span>, <span class="st">"FOSL1"</span>, <span class="st">"RPS16"</span>, <span class="st">"BRAF"</span>, <span class="st">"CACNA1I"</span>, <span class="st">"EXT1"</span>, <span class="st">"GPR82"</span>, <span class="st">"LRRC45"</span>, <span class="st">"STAG1"</span>, <span class="st">"KIF11"</span>, <span class="st">"NCAM1"</span>, <span class="st">"RNF24"</span>, <span class="st">"KIAA2018"</span>, <span class="st">"SAMD9"</span>, <span class="st">"ABCA2"</span>, <span class="st">"PVRIG"</span>, <span class="st">"RASGRF2"</span>, <span class="st">"IRAK2"</span>, <span class="st">"ARHGEF4"</span>, <span class="st">"ZNF682"</span>, <span class="st">"RASSF7"</span>, <span class="st">"THAP7"</span>, <span class="st">"KIF1B"</span>, <span class="st">"GAS5"</span>, <span class="st">"KIF15"</span>, <span class="st">"RMRP"</span>, <span class="st">"GALM"</span>, <span class="st">"LYNX1"</span>, <span class="st">"C7orf50"</span>, <span class="st">"ATXN1L"</span>, <span class="st">"KCNN4"</span>, <span class="st">"LDLRAP1"</span>, <span class="st">"RETSAT"</span>, <span class="st">"RNF10"</span>, <span class="st">"NOXA1"</span>, <span class="st">"UBXN7"</span>, <span class="st">"UHRF1BP1L"</span>, <span class="st">"TESPA1"</span>, <span class="st">"MB21D1"</span>, <span class="st">"LINC00116"</span>, <span class="st">"FOXM1"</span>, <span class="st">"CCR6"</span>, <span class="st">"NMUR1"</span>, <span class="st">"RPL22"</span>, <span class="st">"LCK"</span>, <span class="st">"MEGF9"</span>, <span class="st">"CLSPN"</span>, <span class="st">"MIR4697HG"</span>, <span class="st">"PPP1R15A"</span>, <span class="st">"ADCK5"</span>, <span class="st">"SAMD3"</span>, <span class="st">"RPS27A"</span>, <span class="st">"TPPP"</span>, <span class="st">"PIWIL4"</span>, <span class="st">"ANKRD33B"</span>, <span class="st">"IL18RAP"</span>, <span class="st">"SNRPD2"</span>, <span class="st">"WASF1"</span>, <span class="st">"ALDH5A1"</span>, <span class="st">"FAM149B1"</span>, <span class="st">"RPS18"</span>, <span class="st">"SLCO4C1"</span>, <span class="st">"MGC16275"</span>, <span class="st">"ZNF837"</span>, <span class="st">"SAMHD1"</span>, <span class="st">"NSMF"</span>, <span class="st">"PDIA6"</span>, <span class="st">"SNX20"</span>, <span class="st">"ATF7IP2"</span>, <span class="st">"BPHL"</span>, <span class="st">"RPL35A"</span>, <span class="st">"NFKBIZ"</span>, <span class="st">"LMF1"</span>, <span class="st">"MGMT"</span>, <span class="st">"PPL"</span>, <span class="st">"FAM120C"</span>, <span class="st">"APRT"</span>, <span class="st">"RANGRF"</span>, <span class="st">"PHLDB3"</span>, <span class="st">"IDNK"</span>, <span class="st">"RPS15"</span>, <span class="st">"LGR6"</span>, <span class="st">"MOB1A"</span>, <span class="st">"CREB5"</span>, <span class="st">"THEM4"</span>, <span class="st">"ENDOG"</span>, <span class="st">"RPL7"</span>, <span class="st">"EDN1"</span>, <span class="st">"ZNF808"</span>, <span class="st">"SNHG15"</span>, <span class="st">"RPL3"</span>, <span class="st">"SMCR8"</span>, <span class="st">"FCMR"</span>, <span class="st">"ZNF449"</span>, <span class="st">"ZBTB37"</span>, <span class="st">"NHLRC4"</span>, <span class="st">"DHRS9"</span>, <span class="st">"C1QTNF3"</span>, <span class="st">"MS4A1"</span>, <span class="st">"AFF2"</span>, <span class="st">"FBXO32"</span>, <span class="st">"SLC6A6"</span>, <span class="st">"ZSCAN30"</span>, <span class="st">"CHPT1"</span>, <span class="st">"PARP14"</span>, <span class="st">"MZT2B"</span>, <span class="st">"FLT3LG"</span>, <span class="st">"TCTN2"</span>, <span class="st">"MAN1A1"</span>, <span class="st">"LOC728175"</span>, <span class="st">"ALOX5"</span>, <span class="st">"DUSP18"</span>, <span class="st">"ZC3H12D"</span>, <span class="st">"PECR"</span>, <span class="st">"LRRC37A3"</span>, <span class="st">"EPB41L4A-AS1"</span>, <span class="st">"SPAG5"</span>, <span class="st">"KDM7A"</span>, <span class="st">"DDX31"</span>, <span class="st">"NCOA6"</span>, <span class="st">"SLC6A12"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FBXL13"</span>, <span class="st">"AKR7A2"</span>, <span class="st">"PRR5"</span>, <span class="st">"PRKCE"</span>, <span class="st">"PPP1R3D"</span>, <span class="st">"SYNM"</span>, <span class="st">"GLI1"</span>, <span class="st">"NTNG2"</span>, <span class="st">"EEF1G"</span>, <span class="st">"LOC643802"</span>, <span class="st">"MAGEE1"</span>, <span class="st">"PIK3AP1"</span>, <span class="st">"RPS8"</span>, <span class="st">"SNHG19"</span>, <span class="st">"C19orf70"</span>, <span class="st">"GAS7"</span>, <span class="st">"PHF12"</span>, <span class="st">"ATP11A"</span>, <span class="st">"SLC22A23"</span>, <span class="st">"ATP6V0E2"</span>, <span class="st">"UCKL1-AS1"</span>, <span class="st">"TIPARP"</span>, <span class="st">"BEX2"</span>, <span class="st">"FAM102A"</span>, <span class="st">"RPS20"</span>, <span class="st">"TTC16"</span>, <span class="st">"PLP2"</span>, <span class="st">"HAPLN3"</span>, <span class="st">"C19orf24"</span>, <span class="st">"ZNF397"</span>, <span class="st">"FOXO4"</span>, <span class="st">"TRAPPC2B"</span>, <span class="st">"QTRT1"</span>, <span class="st">"CDK10"</span>, <span class="st">"PDZD8"</span>, <span class="st">"GIPC1"</span>, <span class="st">"CD274"</span>, <span class="st">"C9orf139"</span>, <span class="st">"SPIB"</span>, <span class="st">"KLHDC7B"</span>, <span class="st">"ZNF32"</span>, <span class="st">"ELMO2"</span>, <span class="st">"AK4"</span>, <span class="st">"SAMD4B"</span>, <span class="st">"MFSD3"</span>, <span class="st">"EXD3"</span>, <span class="st">"SLC37A3"</span>, <span class="st">"RPL8"</span>, <span class="st">"TRIM56"</span>, <span class="st">"NMB"</span>, <span class="st">"PIPOX"</span>, <span class="st">"PXN-AS1"</span>, <span class="st">"SLC2A5"</span><span class="op">)</span></span></code></pre></div>
<p><em><strong>Please Note</strong>: Over-representation services return
large responses. To reduce the size of this vignette article, we will
only display the first 10 rows of each Data Frame.</em></p>
<hr>
</div>
<div class="section level2">
<h2 id="enrichr">Enrichr<a class="anchor" aria-label="anchor" href="#enrichr"></a>
</h2>
<p><a href="https://maayanlab.cloud/Enrichr/" title="Enrichr" class="external-link">Enrichr</a> , developed in the Ma’ayan Lab, is a service
to perform enrichment analysis against a considerable number of curated
gene set libraries across various species.</p>
<p>We have thoroughly covered performing enrichment analysis with
Enrichr in the vignette article <a href="rbioapi_enrichr.html" title="2.A: Enrichr &amp; rbioapi">Enrichr &amp; rbioap</a><a href="https://rbioapi.moosa-r.com/articles/rbioapi_enrichr.html" title="2.A: Enrichr &amp; rbioapi">i</a>. Therefore, we will only
provide a brief demonstration here.</p>
<p>First, let us retrieve a list of available libraries in Enrichr using
the <code><a href="../reference/rba_enrichr_libs.html">rba_enrichr_libs()</a></code> function. You can find the available
Enrichr libraries in the <code>libraryName</code> column in the returned
data frame.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichr_libs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_libs.html">rba_enrichr_libs</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Retrieving List of available libraries and statistics from Enrichr human.</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"autoHideNavigation":true,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216"],[13362,27884,6002,47172,47107,21493,1295,2854,15057,4128,34061,7504,16399,23726,32740,13373,19270,3096,22288,4533,10231,2741,5655,11251,8695,1759,2178,851,10061,11250,15406,17711,17576,15797,12232,13572,6454,3723,7588,13121,26382,29065,280,13877,15852,4320,4271,1678,756,3800,1918,5863,25651,19129,23939,23561,23877,15886,24350,3102,31132,30832,48230,5613,9559,9448,16725,19249,15090,16129,15309,15103,15022,15676,15854,15015,3788,3357,12668,12638,7010,5966,15562,17850,17660,1348,934,2541,2041,5209,49238,2243,19586,22440,18329,15755,10271,21809,23601,20883,19612,25983,19500,14893,17598,5902,12486,1073,19513,19741,27360,13787,13929,10352,10471,12419,19378,6201,4558,3264,7802,8551,12444,9000,7744,6204,14148,9813,1397,9116,17464,394,11851,8189,18704,5605,5718,14156,16979,4383,54974,12118,12361,9763,8078,7173,5833,14937,11497,11936,9767,14167,17851,16853,6654,1683,20414,11980,31158,30006,13366,13697,2183,1509,18365,13525,9525,9440,3857,10489,1198,1882,1552,6713,936,8220,9021,8076,14698,10972,12126,13662,18290,12081,12853,3712,19178,19434,19379,10428,8044,9711,1555,12642,2513,11625,4494,8281,9122,16971,11671,14674,11501,11484,12033,10641,3741,10631,12865,15733,7724,14528,5961,6999,16211,18936],[275,1284,77,1370,509,3713,18,34,300,48,641,155,247,127,85,258,388,31,4368,37,158,5,342,100,100,25,89,15,106,166,300,300,300,176,343,301,301,47,35,305,1811,2123,9,304,912,129,128,21,12,48,39,51,807,1594,293,307,302,9,299,25,298,302,1429,36,73,63,1443,1443,282,292,308,318,290,310,279,321,159,153,300,300,87,51,887,300,300,19,13,39,42,300,1550,19,545,505,161,28,22,2316,2395,299,299,299,137,128,1208,109,299,100,117,270,802,200,200,58,76,491,37,45,54,22,92,98,23,20,363,387,122,49,13,22,63,73,586,421,100,39,31,40,295,146,483,448,124,139,102,43,100,33,80,34,33,80,102,360,136,10,112,250,805,815,103,343,18,100,1214,175,245,245,80,61,23,47,16,68,15,146,793,96,33,85,38,12,34,50,485,9,218,369,250,115,42,414,27,30,20,36,51,43,627,76,62,32,90,36,158,201,63,343,116,68,83,1232,643,68,737,752],["Genome_Browser_PWMs","TRANSFAC_and_JASPAR_PWMs","Transcription_Factor_PPIs","ChEA_2013","Drug_Perturbations_from_GEO_2014","ENCODE_TF_ChIP-seq_2014","BioCarta_2013","WikiPathways_2013","Disease_Signatures_from_GEO_up_2014","KEGG_2013","TF-LOF_Expression_from_GEO","TargetScan_microRNA","PPI_Hub_Proteins","GeneSigDB","Chromosome_Location","Human_Gene_Atlas","Mouse_Gene_Atlas","Human_Phenotype_Ontology","Epigenomics_Roadmap_HM_ChIP-seq","KEA_2013","NURSA_Human_Endogenous_Complexome","CORUM","SILAC_Phosphoproteomics","Old_CMAP_up","Old_CMAP_down","OMIM_Disease","OMIM_Expanded","VirusMINT","MSigDB_Computational","MSigDB_Oncogenic_Signatures","Disease_Signatures_from_GEO_down_2014","Virus_Perturbations_from_GEO_up","Virus_Perturbations_from_GEO_down","Cancer_Cell_Line_Encyclopedia","NCI-60_Cancer_Cell_Lines","Tissue_Protein_Expression_from_ProteomicsDB","Tissue_Protein_Expression_from_Human_Proteome_Map","HMDB_Metabolites","Pfam_InterPro_Domains","Allen_Brain_Atlas_up","ENCODE_TF_ChIP-seq_2015","ENCODE_Histone_Modifications_2015","Phosphatase_Substrates_from_DEPOD","Allen_Brain_Atlas_down","ENCODE_Histone_Modifications_2013","Achilles_fitness_increase","Achilles_fitness_decrease","BioCarta_2015","HumanCyc_2015","KEGG_2015","Panther_2015","WikiPathways_2015","ESCAPE","HomoloGene","Disease_Perturbations_from_GEO_down","Disease_Perturbations_from_GEO_up","Drug_Perturbations_from_GEO_down","Genes_Associated_with_NIH_Grants","Drug_Perturbations_from_GEO_up","KEA_2015","Gene_Perturbations_from_GEO_up","Gene_Perturbations_from_GEO_down","ChEA_2015","dbGaP","LINCS_L1000_Chem_Pert_up","LINCS_L1000_Chem_Pert_down","GTEx_Tissue_Expression_Down","GTEx_Tissue_Expression_Up","Ligand_Perturbations_from_GEO_down","Aging_Perturbations_from_GEO_down","Aging_Perturbations_from_GEO_up","Ligand_Perturbations_from_GEO_up","MCF7_Perturbations_from_GEO_down","MCF7_Perturbations_from_GEO_up","Microbe_Perturbations_from_GEO_down","Microbe_Perturbations_from_GEO_up","LINCS_L1000_Ligand_Perturbations_down","LINCS_L1000_Ligand_Perturbations_up","L1000_Kinase_and_GPCR_Perturbations_down","L1000_Kinase_and_GPCR_Perturbations_up","KEGG_2016","WikiPathways_2016","ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X","Kinase_Perturbations_from_GEO_down","Kinase_Perturbations_from_GEO_up","BioCarta_2016","HumanCyc_2016","NCI-Nature_2016","Panther_2016","DrugMatrix","ChEA_2016","huMAP","Jensen_TISSUES","RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO","Jensen_COMPARTMENTS","Jensen_DISEASES","BioPlex_2017","ARCHS4_Tissues","ARCHS4_Cell-lines","ARCHS4_IDG_Coexp","ARCHS4_Kinases_Coexp","ARCHS4_TFs_Coexp","SysMyo_Muscle_Gene_Sets","miRTarBase_2017","TargetScan_microRNA_2017","Enrichr_Libraries_Most_Popular_Genes","Enrichr_Submissions_TF-Gene_Coocurrence","Data_Acquisition_Method_Most_Popular_Genes","DSigDB","TF_Perturbations_Followed_by_Expression","Chromosome_Location_hg19","Rare_Diseases_AutoRIF_ARCHS4_Predictions","Rare_Diseases_GeneRIF_ARCHS4_Predictions","Rare_Diseases_GeneRIF_Gene_Lists","Rare_Diseases_AutoRIF_Gene_Lists","SubCell_BarCode","GWAS_Catalog_2019","WikiPathways_2019_Human","WikiPathways_2019_Mouse","TRRUST_Transcription_Factors_2019","KEGG_2019_Human","KEGG_2019_Mouse","InterPro_Domains_2019","Pfam_Domains_2019","DepMap_WG_CRISPR_Screens_Broad_CellLines_2019","DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019","UK_Biobank_GWAS_v1","BioPlanet_2019","ClinVar_2019","PheWeb_2019","DisGeNET","HMS_LINCS_KinomeScan","CCLE_Proteomics_2020","ProteomicsDB_2020","lncHUB_lncRNA_Co-Expression","Virus-Host_PPI_P-HIPSTer_2020","Elsevier_Pathway_Collection","Table_Mining_of_CRISPR_Studies","COVID-19_Related_Gene_Sets","MSigDB_Hallmark_2020","Enrichr_Users_Contributed_Lists_2020","TG_GATES_2020","Allen_Brain_Atlas_10x_scRNA_2021","Descartes_Cell_Types_and_Tissue_2021","KEGG_2021_Human","WikiPathway_2021_Human","HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression","GO_Biological_Process_2021","GO_Cellular_Component_2021","GO_Molecular_Function_2021","MGI_Mammalian_Phenotype_Level_4_2021","CellMarker_Augmented_2021","Orphanet_Augmented_2021","COVID-19_Related_Gene_Sets_2021","PanglaoDB_Augmented_2021","Azimuth_Cell_Types_2021","PhenGenI_Association_2021","GTEx_Aging_Signatures_2021","HDSigDB_Human_2021","HDSigDB_Mouse_2021","HuBMAP_ASCTplusB_augmented_2022","FANTOM6_lncRNA_KD_DEGs","MAGMA_Drugs_and_Diseases","Tabula_Sapiens","ChEA_2022","Diabetes_Perturbations_GEO_2022","LINCS_L1000_Chem_Pert_Consensus_Sigs","LINCS_L1000_CRISPR_KO_Consensus_Sigs","Tabula_Muris","Reactome_2022","SynGO_2022","GlyGen_Glycosylated_Proteins_2022","IDG_Drug_Targets_2022","KOMP2_Mouse_Phenotypes_2022","Metabolomics_Workbench_Metabolites_2022","Proteomics_Drug_Atlas_2023","The_Kinase_Library_2023","GTEx_Tissues_V8_2023","GO_Biological_Process_2023","GO_Cellular_Component_2023","GO_Molecular_Function_2023","PFOCR_Pathways_2023","GWAS_Catalog_2023","GeDiPNet_2023","MAGNET_2023","Azimuth_2023","Rummagene_kinases","Rummagene_signatures","Rummagene_transcription_factors","MoTrPAC_2023","WikiPathway_2023_Human","DepMap_CRISPR_GeneDependency_CellLines_2023","SynGO_2024","CellMarker_2024","DGIdb_Drug_Targets_2024","MGI_Mammalian_Phenotype_Level_4_2024","WikiPathways_2024_Mouse","WikiPathways_2024_Human","The_Kinase_Library_2024","PerturbAtlas","Reactome_Pathways_2024","GO_Biological_Process_2025","GO_Cellular_Component_2025","GO_Molecular_Function_2025","NIBR_DRUGseq_2025_up","NIBR_DRUGseq_2025_down","Jensen_DISEASES_Curated_2025","Jensen_DISEASES_Experimental_2025","COMPARTMENTS_Curated_2025","PerturbAtlas_MouseGenePerturbationSigs","PerturbSeq_ReplogleK562","TISSUES_Experimental_2025","COMPARTMENTS_Experimental_2025","PerturbSeq_ReplogleRPE1","TISSUES_Curated_2025","RummaGEO_DrugPerturbations_2025"],["http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/","http://jaspar.genereg.net/html/DOWNLOAD/","","http://amp.pharm.mssm.edu/lib/cheadownload.jsp","http://www.ncbi.nlm.nih.gov/geo/","http://genome.ucsc.edu/ENCODE/downloads.html","https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://www.wikipathways.org/index.php/Download_Pathways","http://www.ncbi.nlm.nih.gov/geo/","http://www.kegg.jp/kegg/download/","http://www.ncbi.nlm.nih.gov/geo/","http://www.targetscan.org/cgi-bin/targetscan/data_download.cgi?db=vert_61","http://amp.pharm.mssm.edu/X2K","https://pubmed.ncbi.nlm.nih.gov/22110038/","http://software.broadinstitute.org/gsea/msigdb/index.jsp","http://biogps.org/downloads/","http://biogps.org/downloads/","http://www.human-phenotype-ontology.org/","http://www.roadmapepigenomics.org/","http://amp.pharm.mssm.edu/lib/keacommandline.jsp","https://www.nursa.org/nursa/index.jsf","http://mips.helmholtz-muenchen.de/genre/proj/corum/","http://amp.pharm.mssm.edu/lib/keacommandline.jsp","http://www.broadinstitute.org/cmap/","http://www.broadinstitute.org/cmap/","http://www.omim.org/downloads","http://www.omim.org/downloads","http://mint.bio.uniroma2.it/download.html","http://www.broadinstitute.org/gsea/msigdb/collections.jsp","http://www.broadinstitute.org/gsea/msigdb/collections.jsp","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://portals.broadinstitute.org/ccle/home\n","http://biogps.org/downloads/","https://www.proteomicsdb.org/","http://www.humanproteomemap.org/index.php","http://www.hmdb.ca/downloads","ftp://ftp.ebi.ac.uk/pub/databases/interpro/","http://www.brain-map.org/","http://genome.ucsc.edu/ENCODE/downloads.html","http://genome.ucsc.edu/ENCODE/downloads.html","http://www.koehn.embl.de/depod/","http://www.brain-map.org/","http://genome.ucsc.edu/ENCODE/downloads.html","http://www.broadinstitute.org/achilles","http://www.broadinstitute.org/achilles","https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://humancyc.org/","http://www.kegg.jp/kegg/download/","http://www.pantherdb.org/","http://www.wikipathways.org/index.php/Download_Pathways","http://www.maayanlab.net/ESCAPE/","http://www.ncbi.nlm.nih.gov/homologene","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://grants.nih.gov/grants/oer.htm\n","http://www.ncbi.nlm.nih.gov/geo/","http://amp.pharm.mssm.edu/Enrichr","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://amp.pharm.mssm.edu/Enrichr","http://www.ncbi.nlm.nih.gov/gap","https://clue.io/","https://clue.io/","http://www.gtexportal.org/","http://www.gtexportal.org/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://clue.io/","https://clue.io/","https://clue.io/","https://clue.io/","http://www.kegg.jp/kegg/download/","http://www.wikipathways.org/index.php/Download_Pathways","","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://humancyc.org/","https://www.ndexbio.org/index.html#/networkset/7bc65b82-2a2f-11ed-ac45-0ac135e8bacf","http://www.pantherdb.org/pathway/","https://ntp.niehs.nih.gov/drugmatrix/","http://amp.pharm.mssm.edu/Enrichr","http://proteincomplexes.org/","http://tissues.jensenlab.org/","http://www.ncbi.nlm.nih.gov/geo/","http://compartments.jensenlab.org/","http://diseases.jensenlab.org/","http://bioplex.hms.harvard.edu/","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://sys-myo.rhcloud.com/","http://mirtarbase.mbc.nctu.edu.tw/","http://www.targetscan.org/","http://amp.pharm.mssm.edu/Enrichr","http://amp.pharm.mssm.edu/Enrichr","http://amp.pharm.mssm.edu/Enrichr","http://tanlab.ucdenver.edu/DSigDB/DSigDBv1.0/","http://www.ncbi.nlm.nih.gov/geo/","http://hgdownload.cse.ucsc.edu/downloads.html","https://amp.pharm.mssm.edu/geneshot/","https://www.ncbi.nlm.nih.gov/gene/about-generif","https://www.ncbi.nlm.nih.gov/gene/about-generif","https://amp.pharm.mssm.edu/geneshot/","http://www.subcellbarcode.org/","https://www.ebi.ac.uk/gwas","https://www.wikipathways.org/","https://www.wikipathways.org/","https://www.grnpedia.org/trrust/","https://www.kegg.jp/","https://www.kegg.jp/","https://www.ebi.ac.uk/interpro/","https://pfam.xfam.org/","https://depmap.org/","https://depmap.org/","https://www.ukbiobank.ac.uk/tag/gwas/","https://tripod.nih.gov/bioplanet/","https://www.ncbi.nlm.nih.gov/clinvar/","http://pheweb.sph.umich.edu/","https://www.disgenet.org","http://lincs.hms.harvard.edu/kinomescan/","https://portals.broadinstitute.org/ccle","https://www.proteomicsdb.org/","https://amp.pharm.mssm.edu/lnchub/","http://phipster.org/","http://www.transgene.ru/disease-pathways/","","https://amp.pharm.mssm.edu/covid19","https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp ","https://maayanlab.cloud/Enrichr","https://toxico.nibiohn.go.jp/english/","https://portal.brain-map.org/","https://descartes.brotmanbaty.org/bbi/human-gene-expression-during-development/","https://www.kegg.jp/","https://www.wikipathways.org/","https://hubmapconsortium.github.io/ccf-asct-reporter/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.informatics.jax.org/","http://biocc.hrbmu.edu.cn/CellMarker/","http://www.orphadata.org/","https://maayanlab.cloud/covid19/","https://panglaodb.se/","https://azimuth.hubmapconsortium.org/","https://www.ncbi.nlm.nih.gov/gap/phegeni","https://gtexportal.org/","https://www.hdinhd.org/","https://www.hdinhd.org/","https://hubmapconsortium.github.io/ccf-asct-reporter/","https://fantom.gsc.riken.jp/6/","https://github.com/nybell/drugsets/tree/main/DATA/GENESETS","https://tabula-sapiens-portal.ds.czbiohub.org/","https://maayanlab.cloud/chea3/","https://appyters.maayanlab.cloud/#/Gene_Expression_T2D_Signatures","https://maayanlab.cloud/sigcom-lincs/#/Download","https://maayanlab.cloud/sigcom-lincs/#/Download","https://tabula-muris.ds.czbiohub.org/","https://reactome.org/download-data","https://www.syngoportal.org/","https://www.glygen.org/","https://drugcentral.org/","https://www.mousephenotype.org/","https://www.metabolomicsworkbench.org/","https://www.nature.com/articles/s41587-022-01539-0","https://kinase-library.phosphosite.org/site","https://gtexportal.org/home/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","https://pfocr.wikipathways.org/","https://www.ebi.ac.uk/gwas","http://gedipnet.bicnirrh.res.in/","https://magnet-winterlab.herokuapp.com/","https://azimuth.hubmapconsortium.org/","https://rummagene.com/","https://rummagene.com/","https://rummagene.com/","https://motrpac-data.org/","https://www.wikipathways.org/","https://depmap.org/","https://www.syngoportal.org/","http://bio-bigdata.hrbmu.edu.cn/CellMarker/","https://dgidb.org/","http://www.informatics.jax.org/","https://www.wikipathways.org/","https://www.wikipathways.org/","https://kinase-library.phosphosite.org/site","https://perturbatlas.kratoss.site/#/","https://reactome.org/download-data","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","","","http://diseases.jensenlab.org/","http://diseases.jensenlab.org/","https://compartments.jensenlab.org/Search","https://perturbatlas.kratoss.site/#/","https://gwps.wi.mit.edu/","https://tissues.jensenlab.org/Search","https://compartments.jensenlab.org/Search","https://gwps.wi.mit.edu/","https://tissues.jensenlab.org/Search","https://rummageo.com/"],[615,326,290,353,701,498,249,199,142,200,269,222,385,2139,386,84,96,1779,383,474,1796,1658,84,6100,6100,90,187,85,858,189,142,323,323,967,93,207,30,3906,311,2192,816,412,59,2192,109,216,216,239,125,179,104,404,315,12,839,839,906,32876,906,428,2460,2460,395,345,33132,33132,2918,2918,261,286,286,261,401,401,312,312,96,96,3644,3644,293,437,104,285,285,237,152,209,112,7876,645,995,1842,1302,2283,1811,3915,108,125,352,498,1724,1135,3240,683,121,1722,12,4026,1958,36,3725,2244,2244,3725,104,1737,472,176,571,308,303,1071,608,558,325,857,1510,182,1161,9828,148,378,913,3729,6715,1721,802,205,50,1482,1190,766,172,320,622,344,6036,511,1274,4601,1097,3774,478,178,341,950,270,2564,2579,777,350,1395,469,757,601,10850,10424,106,1818,118,338,888,529,233,1748,303,511,5407,474,1147,21845,5271,2388,72,1425,3810,11130,7244,225,801,1093,134,1692,659,5003,188,829,392,12841,2105,5343,468,1174,5412,4503,574,500,1015,10729,4102,218,48,2708,417,5344],["ea115789fcbf12797fd692cec6df0ab4dbc79c6a","7d42eb43a64a4e3b20d721fc7148f685b53b6b30","849f222220618e2599d925b6b51868cf1dab3763","7ebe772afb55b63b41b79dd8d06ea0fdd9fa2630","ad270a6876534b7cb063e004289dcd4d3164f342","497787ebc418d308045efb63b8586f10c526af51","4a293326037a5229aedb1ad7b2867283573d8bcd","5c307674c8b97e098f8399c92f451c0ff21cbf68","248c4ed8ea28352795190214713c86a39fd7afab","eb26f55d3904cb0ea471998b6a932a9bf65d8e50","","f4029bf6a62c91ab29401348e51df23b8c44c90f","69c0cfe07d86f230a7ef01b365abcc7f6e52f138","6d655e0aa3408a7accb3311fbda9b108681a8486","8dab0f96078977223646ff63eb6187e0813f1433","0741451470203d7c40a06274442f25f74b345c9c","31191bfadded5f96983f93b2a113cf2110ff5ddb","17a138b0b70aa0e143fe63c14f82afb70bc3ed0a","e1bc8a398e9b21f9675fb11bef18087eda21b1bf","462045609440fa1e628a75716b81a1baa5bd9145","7d3566b12ebc23dd23d9ca9bb97650f826377b16","d047f6ead7831b00566d5da7a3b027ed9196e104","54dcd9438b33301deb219866e162b0f9da7e63a0","9041f90cccbc18479138330228b336265e09021c","ebc0d905b3b3142f936d400c5f2a4ff926c81c37","cb2b92578a91e023d0498a334923ee84add34eca","27eca242904d8e12a38cf8881395bc50d57a03e1","5abad1fc36216222b0420cadcd9be805a0dda63e","e4cdcc7e259788fdf9b25586cce3403255637064","c76f5319c33c4833c71db86a30d7e33cd63ff8cf","aabdf7017ae55ae75a004270924bcd336653b986","45268b7fc680d05dd9a29743c2f2b2840a7620bf","5f531580ccd168ee4acc18b02c6bdf8200e19d08","eb38dbc3fb20adafa9d6f9f0b0e36f378e75284f","75c81676d8d6d99d262c9660edc024b78cfb07c9","","49351dc989f9e6ca97c55f8aca7778aa3bfb84b9","1905132115d22e4119bce543bdacaab074edb363","e2b4912cfb799b70d87977808c54501544e4cdc9","3804715a63a308570e47aa1a7877f01147ca6202","56b6adb4dc8a2f540357ef992d6cd93dfa2907e5","55b56cd8cf2ff04b26a09b9f92904008b82f3a6f","d40701e21092b999f4720d1d2b644dd0257b6259","ea67371adec290599ddf484ced2658cfae259304","c209ae527bc8e98e4ccd27a668d36cd2c80b35b4","98366496a75f163164106e72439fb2bf2f77de4e","83a710c1ff67fd6b8af0d80fa6148c40dbd9bc64","70e4eb538daa7688691acfe5d9c3c19022be832b","711f0c02b23f5e02a01207174943cfeee9d3ea9c","e80d25c56de53c704791ddfdc6ab5eec28ae7243","ab824aeeff0712bab61f372e43aebb870d1677a9","1f7eea2f339f37856522c1f1c70ec74c7b25325f","a7acc741440264717ff77751a7e5fed723307835","663b665b75a804ef98add689f838b68e612f0d2a","0f412e0802d76efa0374504c2c9f5e0624ff7f09","9ddc3902fb01fb9eaf1a2a7c2ff3acacbb48d37e","068623a05ecef3e4a5e0b4f8db64bb8faa3c897f","76fc5ec6735130e287e62bae6770a3c5ee068645","c9c2155b5ac81ac496854fa61ba566dcae06cc80","18a081774e6e0aaf60b1a4be7fd20afcf9e08399","53dedc29ce3100930d68e506f941ef59de05dc6b","499882af09c62dd6da545c15cb51c1dc5e234f78","712eb7b6edab04658df153605ec6079fa89fb5c7","010f1267055b1a1cb036e560ea525911c007a666","5e678b3debe8d8ea95187d0cd35c914017af5eb3","fedbf5e221f45ee60ebd944f92569b5eda7f2330","74b818bd299a9c42c1750ffe43616aa9f7929f02","103738763d89cae894bec9f145ac28167a90e611","1eb3c0426140340527155fd0ef67029db2a72191","cd95fe1b505ba6f28cd722cfba50fdea979d3b4c","74c4f0a0447777005b2a5c00c9882a56dfc62d7c","31baa39da2931ddd5f7aedf2d0bbba77d2ba7b46","555f68aef0a29a67b614a0d7e20b6303df9069c6","1bc2ba607f1ff0dda44e2a15f32a2c04767da18c","9e613dba78ef7e60676b13493a9dc49ccd3c8b3f","d0c3e2a68e8c611c669098df2c87b530cec3e132","957846cb05ef31fc8514120516b73cc65af7980e","3bd494146c98d8189898a947f5ef5710f1b7c4b2","1ccc5bce553e0c2279f8e3f4ddcfbabcf566623b","b54a0d4ba525eac4055c7314ca9d9312adcb220c","43f56da7540195ba3c94eb6e34c522a699b36da9","340be98b444cad50bb974df69018fd598e23e5e1","5426f7747965c23ef32cff46fabf906e2cd76bfa","bb9682d78b8fc43be842455e076166fcd02cefc3","78618915009cac3a0663d6f99d359e39a31b6660","13d9ab18921d5314a5b2b366f6142b78ab0ff6aa","d6a502ef9b4c789ed5e73ca5a8de372796e5c72a","3c1e1f7d1a651d9aaa198e73704030716fc09431","ca5f6abf7f75d9baae03396e84d07300bf1fd051","255c3db820d612f34310f22a6985dad50e9fe1fe","af271913344aa08e6a755af1d433ef15768d749a","249247d2f686d3eb4b9e4eb976c51159fac80a89","e8879ab9534794721614d78fe2883e9e564d7759","f0752e4d7f5198f86446678966b260c530d19d78","56ec68c32d4e83edc2ee83bea0e9f6a3829b2279","3045dff8181367c1421627bb8e4c5a32c6d67f98","b8620b1a9d0d271d1a2747d8cfc63589dba39991","4126374338235650ab158ba2c61cd2e2383b70df","5496ef9c9ae9429184d0b9485c23ba468ee522a8","ce60be284fdd5a9fc6240a355421a9e12b1ee84a","6721c5ed97b7772e4a19fdc3f797110df0164b75","8a468c3ae29fa68724f744cbef018f4f3b61c5ab","","6b7c7fe2a97b19aecbfba12d8644af6875ad99c4","79d13fb03d2fa6403f9be45c90eeda0f6822e269","e9b7d8ee237d0a690bd79d970a23a9fa849901ed","be2ca8ef5a8c8e17d7e7bd290e7cbfe0951396c0","17ce5192b9eba7d109b6d228772ea8ab222e01ef","287476538ab98337dbe727b3985a436feb6d192a","34d08a4878c19584aaf180377f2ea96faa6a6eb1","fdab39c467ba6b0fb0288df1176d7dfddd7196d5","375ff8cdd64275a916fa24707a67968a910329bb","0f7fb7f347534779ecc6c87498e96b5460a8d652","0fb9ac92dbe52024661c088f71a1134f00567a8b","ee3adbac2da389959410260b280e7df1fd3730df","b50bb9480d8a77103fb75b331fd9dd927246939a","fef3864bcb5dd9e60cee27357eff30226116c49b","b0c9e9ebb9014f14561e896008087725a2db24b7","e7750958da20f585c8b6d5bc4451a5a4305514ba","5f8cf93e193d2bcefa5a37ccdf0eefac576861b0","3477bc578c4ea5d851dcb934fe2a41e9fd789bb4","187eb44b2d6fa154ebf628eba1f18537f64e797c","18dd5ec520fdf589a93d6a7911289c205e1ddf22","a6325ed264f9ac9e6518796076c46a1d885cca7a","0b08b32b20854ac8a738458728a9ea50c2e04800","b7c4ead26d0eb64f1697c030d31682b581c8bb56","958fb52e6215626673a5acf6e9289a1b84d11b4a","e110851dfc763d30946f2abedcc2cd571ac357a0","0a95303f8059bec08836ecfe02ce3da951150547","6a7c7321b6b72c5285b722f7902d26a2611117cb","3c261626478ce9e6bf2c7f0a8014c5e901d43dc0","47ba06cdc92469ac79400fc57acd84ba343ba616","7094b097ae2301a1d6a5bd856a193b084cca993d","8c87c8346167bac2ba68195a32458aba9b1acfd1","45b597d7efa5693b7e4172b09c0ed2dda3305582","a592eed13e8e9496aedbab63003b965574e46a65","9196c760e3bcae9c9de1e3f87ad81f96bde24325","ad580f3864fa8ff69eaca11f6d2e7f9b86378d08","72b0346849570f66a77a6856722601e711596cb4","6952efda94663d4bd8db09bf6eeb4e67d21ef58c","8dc362703b38b30ac3b68b6401a9b20a58e7d3ef","9e32560437b11b4628b00ccf3d584360f7f7daee","46f8235cb585829331799a71aec3f7c082170219","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","",""],[1,1,1,7,7,7,7,7,7,7,1,7,2,4,7,5,5,3,1,7,2,2,2,7,7,4,4,4,4,4,7,4,4,5,5,7,5,6,6,5,1,1,2,5,7,4,4,7,7,7,7,7,5,6,8,8,8,6,8,2,8,8,7,4,7,7,5,5,8,8,8,8,8,8,8,8,7,7,7,7,7,7,1,2,2,2,2,2,2,4,7,2,7,8,7,7,2,5,5,4,2,1,8,1,1,6,1,6,4,1,6,4,4,4,4,2,7,7,7,1,7,7,6,6,7,4,4,2,4,4,4,2,5,5,1,2,2,6,7,2,6,4,5,5,2,7,5,7,7,7,7,5,4,4,5,5,4,4,4,4,5,1,4,5,1,4,4,4,5,7,7,2,4,3,2,4,7,5,7,7,7,2,4,4,5,5,2,8,1,4,7,4,3,5,4,3,2,2,2,4,2,3,3,3,4,4,3,3,3,4,4,3,3,4,3,3]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>geneCoverage<\/th>\n      <th>genesPerTerm<\/th>\n      <th>libraryName<\/th>\n      <th>link<\/th>\n      <th>numTerms<\/th>\n      <th>appyter<\/th>\n      <th>categoryId<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[1,2,5,7]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"geneCoverage","targets":1},{"name":"genesPerTerm","targets":2},{"name":"libraryName","targets":3},{"name":"link","targets":4},{"name":"numTerms","targets":5},{"name":"appyter","targets":6},{"name":"categoryId","targets":7}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>While the only required argument of <code><a href="../reference/rba_enrichr.html">rba_enrichr()</a></code>
function is <code>gene_list</code>, here we use
<code>gene_set_library</code> argument to request an enrichment analysis
only against “KEGG_2021_Human” library.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichr_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr.html">rba_enrichr</a></span><span class="op">(</span></span>
<span>  gene_list <span class="op">=</span> <span class="va">covid_critical</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>  gene_set_library <span class="op">=</span> <span class="st">"KEGG_2021_Human"</span>,</span>
<span>  progress_bar <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># to avoid printing issues in the vignette</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; --Step 1/3:</span></span>
<span><span class="co">#&gt; Retrieving List of available libraries and statistics from Enrichr human.</span></span>
<span><span class="co">#&gt; --Step 2/3:</span></span>
<span><span class="co">#&gt; Uploading 1214 gene symbols to Enrichr human.</span></span>
<span><span class="co">#&gt; --Step 3/3:</span></span>
<span><span class="co">#&gt; Performing Enrichr analysis on gene-list 96027426 against Enrichr human library: KEGG_2021_Human.</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"autoHideNavigation":true,"data":[["1","2","3","4","5","6","7","8","9","10"],["Coronavirus disease","Ribosome","Hematopoietic cell lineage","Th1 and Th2 cell differentiation","Th17 cell differentiation","PD-L1 expression and PD-1 checkpoint pathway in cancer","Glycine, serine and threonine metabolism","Cell cycle","Primary immunodeficiency","Inflammatory bowel disease"],["51/232","38/158","18/99","16/92","17/107","15/89","9/40","17/124","7/38","9/65"],[5.457660615844306e-16,1.461442604133301e-13,2.542147408390094e-05,0.0001221586459325963,0.0002377387069538916,0.0002815676772171271,0.000536146158525187,0.00134153930817423,0.007179992090380879,0.01618833319043852],[1.555433275515627e-13,2.082555710889954e-11,0.002415040037970589,0.008703803522697489,0.01337446466781354,0.01337446466781354,0.02182880788281119,0.04779233785370694,0.2273664161953945,0.4326283249122099],[0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0],[4.507560462321202,5.026275510204082,3.475473801560758,3.287936033740444,2.95026455026455,3.163447016658026,4.518672199170124,2.479274510263197,3.508699254349627,2.498073503260225],[158.4152430865095,148.5474620954103,36.7702219398494,29.62492831099334,24.61800565105845,25.86161504020705,34.03058914531737,16.39776687159813,17.32054298640066,10.30071733320014],["RPL3;RPL32;RPL34;RPLP1;RPL8;RPL10A;RPL9;RPL7;RPS15;IFIH1;RPS14;RPS16;RPL18A;RPS19;RPS18;C3AR1;RPL36;RPL35;RPLP2;RPL38;RPL39;RPS10;RPS12;RPS8;RPL22;RPL13A;RPSA;RPS3A;ADAM17;OAS1;OAS2;OAS3;RPL37A;HBEGF;RPL13;RPL18;RPS27A;ACE;EIF2AK2;RPL35A;RPL23A;MAPK14;RPS26;NFKBIA;CXCL10;MAPK11;RPS28;RPS27;RPL27A;RPS20;RPS21","RPL3;RPL32;RPL34;RPLP1;RPL10A;RPL8;RPL9;RPL7;RPS15;RPS14;RPS16;RPL18A;RPS19;RPS18;RPL36;RPL35;RPLP2;RPL13;RPL38;RPL18;RPS27A;RPL39;RPS10;RPS12;RPS8;RPL22;RPL13A;RPL35A;RPSA;RPS3A;RPL23A;RPS26;RPS28;RPS27;RPL27A;RPL37A;RPS20;RPS21","CR1;CSF1;IL11RA;IL1R1;ITGA1;FLT3LG;CD1C;CD3E;CD3D;IL4;CD2;CD5;HLA-DPB1;CD38;HLA-DOA;CD24;IL7R;MS4A1","GATA3;CD3E;MAPK14;CD3D;DLL4;IL4;NFKBIA;ZAP70;MAPK11;LCK;STAT4;HLA-DPB1;CD247;HLA-DOA;JAK2;LAT","IL1R1;RORC;GATA3;CD3E;MAPK14;HIF1A;CD3D;IL4;NFKBIA;ZAP70;MAPK11;LCK;HLA-DPB1;CD247;HLA-DOA;JAK2;LAT","BATF2;CD274;CD3E;MAPK14;CD3D;HIF1A;NFKBIA;ZAP70;MAPK11;LCK;CD247;PDCD1;JAK2;LAT;MAP2K6","GAMT;ALAS2;GLDC;MAOA;CBS;SARDH;PIPOX;PHGDH;SDSL","CDKN1C;BUB1B;TTK;CDC6;PKMYT1;CDC25A;CDC20;RBL2;CCNB2;CCNB1;STAG1;CDC45;CCNE2;ESPL1;CDK1;EP300;BUB1","ZAP70;LCK;TAP1;CD3E;IL7R;CD3D;AICDA","IL4;IL18RAP;RORC;STAT4;HLA-DPB1;GATA3;HLA-DOA;TLR5;IL18R1"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Term<\/th>\n      <th>Overlap<\/th>\n      <th>P.value<\/th>\n      <th>Adjusted.P.value<\/th>\n      <th>Old.P.value<\/th>\n      <th>Old.Adjusted.P.value<\/th>\n      <th>Odds.Ratio<\/th>\n      <th>Combined.Score<\/th>\n      <th>Genes<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[3,4,5,6,7,8]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Term","targets":1},{"name":"Overlap","targets":2},{"name":"P.value","targets":3},{"name":"Adjusted.P.value","targets":4},{"name":"Old.P.value","targets":5},{"name":"Old.Adjusted.P.value","targets":6},{"name":"Odds.Ratio","targets":7},{"name":"Combined.Score","targets":8},{"name":"Genes","targets":9}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>In the <code>gene_set_library</code> parameter, you can also provide
multiple gene set libraries or leave it as the default parameter
<code>NULL</code> to perform the enrichment analysis using every gene
set library available in Enrichr. Also note that when
<code>regex_library_name</code> parameter is TRUE (the default value),
<code>gene_set_library</code> parameter will be considered as a regex
pattern, so partial matchs will be also selected.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichr_enrich_kegg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr.html">rba_enrichr</a></span><span class="op">(</span></span>
<span>  gene_list <span class="op">=</span> <span class="va">covid_critical</span>,</span>
<span>  gene_set_library <span class="op">=</span> <span class="st">"kegg"</span>,</span>
<span>  regex_library_name <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># default value</span></span>
<span>  progress_bar <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># to avoid printing issues in the vignette</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; --Step 1/3:</span></span>
<span><span class="co">#&gt; Retrieving List of available libraries and statistics from Enrichr human.</span></span>
<span><span class="co">#&gt; --Step 2/3:</span></span>
<span><span class="co">#&gt; Uploading 1214 gene symbols to Enrichr human.</span></span>
<span><span class="co">#&gt; --Step 3/3:</span></span>
<span><span class="co">#&gt; Performing Enrichr analysis on gene-list 96027429 using multiple Enrichr human libraries.</span></span>
<span><span class="co">#&gt; Note: You have selected '6' Enrichr human libraries. Note that for each library, a separate call should be sent to Enrichr server. Thus, this could take a while depending on the number of selected libraries.</span></span></code></pre></div>
<p>As you can see below, when than one library is selected, the results
will be a list where each of its elements is the enrichment results of
one library:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">enrichr_enrich_kegg</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 6</span></span>
<span><span class="co">#&gt;  $ KEGG_2013      :'data.frame': 139 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;  $ KEGG_2015      :'data.frame': 131 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;  $ KEGG_2016      :'data.frame': 260 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;  $ KEGG_2019_Human:'data.frame': 272 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;  $ KEGG_2019_Mouse:'data.frame': 267 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;  $ KEGG_2021_Human:'data.frame': 285 obs. of  9 variables:</span></span></code></pre></div>
<p>It is also possible to provide a background gene list for Enrichr
analysis. This feature is only supported when the organism is set to
“human”.</p>
<hr>
</div>
<div class="section level2">
<h2 id="reactome-analysis-services">Reactome Analysis services<a class="anchor" aria-label="anchor" href="#reactome-analysis-services"></a>
</h2>
<p>Reactome curates extensive and top-quality cellular pathways data
across various species. Given the fact that proteins operate in
organized units, which are the pathways, analyzing your list against
Reactome pathways gene sets can provide valuable insights into the
functional trajectories of your results.</p>
<p>You can refer to <a href="rbioapi_enrichr.html" title="2.E: Reactome &amp; rbioapi">Reactome &amp; rbioapi</a> article
for in-depth coverage of Reactome services. Here we will only
demonstrate the over-representation functionality</p>
<p>The function <code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code> accepts gene lists
or a table as input. If a table is supplied, the first column should
consist of identifiers, and the next column should be numerical gene
expression values. Refer to <code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code>
function’s manual for details. Here, we will use our gene list of
critical COVID:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reactome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">covid_critical</span><span class="op">)</span></span>
<span><span class="co">#&gt; Retrieving Reactome Analysis Results of your supplied Identifiers.</span></span></code></pre></div>
<p>Let us examine the response’s structure:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">reactome</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 8</span></span>
<span><span class="co">#&gt;  $ summary            :List of 7</span></span>
<span><span class="co">#&gt;  $ expression         :List of 1</span></span>
<span><span class="co">#&gt;  $ identifiersNotFound: int 557</span></span>
<span><span class="co">#&gt;  $ pathwaysFound      : int 1579</span></span>
<span><span class="co">#&gt;  $ pathways           :'data.frame': 1579 obs. of  19 variables:</span></span>
<span><span class="co">#&gt;  $ resourceSummary    :'data.frame': 4 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;  $ speciesSummary     :'data.frame': 1 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;  $ warnings           : list()</span></span></code></pre></div>
<p>The list’s element names are self-explanatory. The results are
returned as a data frame in the <code>pathways</code> element. Note that
Reactome will map your provided gene identifiers and some identifiers
may not be found. Hence, keep an eye on the
<code>identifiersNotFound</code> element.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-36aa3d2a04d42bbc2145" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-36aa3d2a04d42bbc2145">{"x":{"filter":"none","vertical":false,"autoHideNavigation":true,"data":[["1","2","3","4","5","6","7","8","9","10"],["R-HSA-975956","R-HSA-72689","R-HSA-192823","R-HSA-1799339","R-HSA-156902","R-HSA-156842","R-HSA-72764","R-HSA-9633012","R-HSA-2408557","R-HSA-72706"],[975956,72689,192823,1799339,156902,156842,72764,9633012,2408557,72706],["Nonsense Mediated Decay (NMD) independent of the Exon Junction Complex (EJC)","Formation of a pool of free 40S subunits","Viral mRNA Translation","SRP-dependent cotranslational protein targeting to membrane","Peptide chain elongation","Eukaryotic Translation Elongation","Eukaryotic Translation Termination","Response of EIF2AK4 (GCN2) to amino acid deficiency","Selenocysteine synthesis","GTP hydrolysis and joining of the 60S ribosomal subunit"],[true,true,true,true,true,true,true,true,true,true],[false,false,true,false,false,false,false,false,false,false],[48887,48887,48887,48887,48887,48887,48887,48887,48887,48887],["9606","9606","9606","9606","9606","9606","9606","9606","9606","9606"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL"],[101,106,114,119,97,102,106,115,112,120],[46,46,46,46,46,47,46,48,46,46],[0.006288916562889165,0.006600249066002491,0.007098381070983811,0.007409713574097135,0.006039850560398506,0.006351183063511831,0.006600249066002491,0.007160647571606476,0.006973848069738481,0.007471980074719801],[1.110223024625157e-16,1.110223024625157e-16,1.110223024625157e-16,1.110223024625157e-16,1.110223024625157e-16,1.110223024625157e-16,1.110223024625157e-16,1.110223024625157e-16,1.110223024625157e-16,2.220446049250313e-16],[2.120525977034049e-14,2.120525977034049e-14,2.120525977034049e-14,2.120525977034049e-14,2.120525977034049e-14,2.120525977034049e-14,2.120525977034049e-14,2.120525977034049e-14,2.120525977034049e-14,3.463895836830488e-14],[[],[],[],[],[],[],[],[],[],[]],["TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL"],[1,2,2,5,5,9,5,16,7,3],[1,2,2,5,4,6,3,6,2,3],[6.429214350006429e-05,0.0001285842870001286,0.0001285842870001286,0.0003214607175003214,0.0003214607175003214,0.0005786292915005787,0.0003214607175003214,0.001028674296001029,0.00045004500450045,0.0001928764305001929]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stId<\/th>\n      <th>dbId<\/th>\n      <th>name<\/th>\n      <th>llp<\/th>\n      <th>inDisease<\/th>\n      <th>species.dbId<\/th>\n      <th>species.taxId<\/th>\n      <th>species.name<\/th>\n      <th>entities.resource<\/th>\n      <th>entities.total<\/th>\n      <th>entities.found<\/th>\n      <th>entities.ratio<\/th>\n      <th>entities.pValue<\/th>\n      <th>entities.fdr<\/th>\n      <th>entities.exp<\/th>\n      <th>reactions.resource<\/th>\n      <th>reactions.total<\/th>\n      <th>reactions.found<\/th>\n      <th>reactions.ratio<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[2,6,10,11,12,13,14,17,18,19]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"stId","targets":1},{"name":"dbId","targets":2},{"name":"name","targets":3},{"name":"llp","targets":4},{"name":"inDisease","targets":5},{"name":"species.dbId","targets":6},{"name":"species.taxId","targets":7},{"name":"species.name","targets":8},{"name":"entities.resource","targets":9},{"name":"entities.total","targets":10},{"name":"entities.found","targets":11},{"name":"entities.ratio","targets":12},{"name":"entities.pValue","targets":13},{"name":"entities.fdr","targets":14},{"name":"entities.exp","targets":15},{"name":"reactions.resource","targets":16},{"name":"reactions.total","targets":17},{"name":"reactions.found","targets":18},{"name":"reactions.ratio","targets":19}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>Naturally, you can alter the analysis parameters using the
<code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code> function’s arguments. The used
parameters are returned in the “summary” element.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">reactome</span><span class="op">$</span><span class="va">summary</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 7</span></span>
<span><span class="co">#&gt;  $ token         : chr "MjAyNTA2MjMxNTUwMTdfNDU%3D"</span></span>
<span><span class="co">#&gt;  $ projection    : logi TRUE</span></span>
<span><span class="co">#&gt;  $ interactors   : logi FALSE</span></span>
<span><span class="co">#&gt;  $ type          : chr "OVERREPRESENTATION"</span></span>
<span><span class="co">#&gt;  $ sampleName    : chr "Gene names"</span></span>
<span><span class="co">#&gt;  $ text          : logi TRUE</span></span>
<span><span class="co">#&gt;  $ includeDisease: logi TRUE</span></span></code></pre></div>
<p>In the above code chunk, note the “token” associated with your
analysis. You can use this token to perform the following.</p>
<p>Retrieve the analysis results:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reactome_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_analysis_token.html">rba_reactome_analysis_token</a></span><span class="op">(</span><span class="va">reactome</span><span class="op">$</span><span class="va">summary</span><span class="op">$</span><span class="va">token</span><span class="op">)</span></span></code></pre></div>
<p>In addition to the main response, you can download other data
associated with your request. see the argument <code>request</code> in
the <code><a href="../reference/rba_reactome_analysis_download.html">rba_reactome_analysis_download()</a></code> function’s manual for
more information. Here, we will download a CSV file with the input
identifiers which has not been found:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rba_reactome_analysis_download.html">rba_reactome_analysis_download</a></span><span class="op">(</span></span>
<span>  token <span class="op">=</span> <span class="va">reactome</span><span class="op">$</span><span class="va">summary</span><span class="op">$</span><span class="va">token</span>,</span>
<span>  request <span class="op">=</span> <span class="st">"not_found_ids"</span>,</span>
<span>  save_to <span class="op">=</span> <span class="st">"my_analysis.csv"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Reactome also generates a nice report of your analysis in pdf format,
make sure to try it:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rba_reactome_analysis_pdf.html">rba_reactome_analysis_pdf</a></span><span class="op">(</span></span>
<span>  token <span class="op">=</span> <span class="va">reactome</span><span class="op">$</span><span class="va">summary</span><span class="op">$</span><span class="va">token</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span>, <span class="co">#Homo sapiens</span></span>
<span>  save_to <span class="op">=</span> <span class="st">"my_analysis.pdf"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="panther">PANTHER<a class="anchor" aria-label="anchor" href="#panther"></a>
</h2>
<p>PANTHER (protein analysis through evolutionary relationships) is a
project that provides classification systems of genes and proteins.
PANTHER also provides an enrichment service. In fact, the enrichment
tool available on the <a href="https://geneontology.org/" title="Gene Ontology Resource - Unifying Biology" class="external-link">Gene Ontology (GO)
website</a> is powered by <a href="https://pantherdb.org/" class="external-link">PANTHER</a>.
Just like the previous sections, we will only review the enrichment
functionality here; for an in-depth review of PANTHER refer to the
vignette article “<a href="rbioapi_panther.html" title="2.D: PANTHER &amp; rbioapi">PANTHER &amp; rbioapi</a>”.</p>
<p>To perform the analysis, first, we need to choose the annotation
datasets (i.e. a collection of gene sets) that we want to compare our
results against it. To retrieve a list of available annotation datasets,
we call the following function:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">panther_sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_panther_info.html">rba_panther_info</a></span><span class="op">(</span>what <span class="op">=</span> <span class="st">"datasets"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Retrieving available annotation datasets.</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-febe03efa1a2d8d52a86" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-febe03efa1a2d8d52a86">{"x":{"filter":"none","vertical":false,"autoHideNavigation":true,"data":[["1","2","3","4","5","6","7","8","9"],["2025-03-16","2025-03-16","2025-03-16","2024-06-20","2024-06-20","2024-06-20","2024-06-20","2024-06-20","2023-09-07"],["GO:0003674","GO:0008150","GO:0005575","ANNOT_TYPE_ID_PANTHER_GO_SLIM_MF","ANNOT_TYPE_ID_PANTHER_GO_SLIM_BP","ANNOT_TYPE_ID_PANTHER_GO_SLIM_CC","ANNOT_TYPE_ID_PANTHER_PC","ANNOT_TYPE_ID_PANTHER_PATHWAY","ANNOT_TYPE_ID_REACTOME_PATHWAY"],["molecular_function","biological_process","cellular_component","PANTHER GO Slim Molecular Function","PANTHER GO Slim Biological Process","PANTHER GO Slim Cellular Location","protein class","ANNOT_TYPE_PANTHER_PATHWAY","ANNOT_TYPE_REACTOME_PATHWAY"],["10.5281/zenodo.15066566","10.5281/zenodo.15066566","10.5281/zenodo.15066566","19","19","19","19","19","86"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>release_date<\/th>\n      <th>id<\/th>\n      <th>label<\/th>\n      <th>version<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"release_date","targets":1},{"name":"id","targets":2},{"name":"label","targets":3},{"name":"version","targets":4}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>Note that you should enter the “id” of the datasets, not its label.
For example, entering <code>"biological_process"</code> is incorrect,
you should rather use <code>"GO:0008150"</code>.</p>
<p>Here, we demonstrate the enrichment analysis using the Biological
Process annotations. The Gene Ontology (GO) project is one of the
pinnacles of scientists’ collective effort in bioinformatics. The GO
Consortium provides a comprehensive model of biological systems. In
short, GO curates a thoroughly designed directed acyclic graph (DAG) of
ontologies. You may think of it as a tree of terms, where as it branches
out, the terms become more specific). Each protein may be annotated with
one or more terms. The terms are organized in three domains: “Molecular
Function,” “Biological Process,” and “Cellular Component”. GO slim
datasets refer to subsets which are a cut-down version of GO terms. If
you are not familiar with GO, I strongly encourage you to see this page
and follow the links it provides: <a href="https://geneontology.org/docs/introduction-to-go-resource/" title="About the GO resource" class="external-link">About the GO resource</a>.</p>
<p>Depending on the provided input, PANTHER will conduct two types of
analysis:</p>
<ol style="list-style-type: decimal">
<li><p>If a character vector is supplied, over-representation analysis
will be performed using either Fisher’s exact or binomial test.</p></li>
<li><p>If a data frame with gene identifiers and their corresponding
expression values is supplied, statistical enrichment test is performed
using Mann-Whitney U (Wilcoxon Rank-Sum) test.</p></li>
</ol>
<p>rbioapi determines the proper analysis based on the class of the
<code>genes</code> parameter. Please refer to the details section of
<code><a href="../reference/rba_panther_enrich.html">rba_panther_enrich()</a></code> function manual and <a href="rbioapi_panther.html#submit-the-analysis-request">rbioapi &amp;
PANTHER vignette article</a> for more information.</p>
<p>Here, we only demonstrate using a gene list, without the expression
values:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">panther_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_panther_enrich.html">rba_panther_enrich</a></span><span class="op">(</span></span>
<span>  genes <span class="op">=</span> <span class="va">covid_critical</span>,</span>
<span>  organism <span class="op">=</span> <span class="fl">9606</span>, <span class="co">#Homo sapiens</span></span>
<span>  annot_dataset <span class="op">=</span> <span class="st">"GO:0008150"</span> <span class="co">#Biological Process</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Performing PANTHER over-representation analysis (Fisher's exact test) on 1214 genes from `organism 9606` against `GO:0008150` datasets.</span></span></code></pre></div>
<p>In addition to the results table, PANTHER returns other useful
information about your analysis. The names are self-explanatory:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">panther_enrich</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 9</span></span>
<span><span class="co">#&gt;  $ reference                 :List of 3</span></span>
<span><span class="co">#&gt;   ..$ organism      : chr "Homo sapiens"</span></span>
<span><span class="co">#&gt;   ..$ mapped_count  :List of 2</span></span>
<span><span class="co">#&gt;   ..$ unmapped_count: int 0</span></span>
<span><span class="co">#&gt;  $ input_list                :List of 5</span></span>
<span><span class="co">#&gt;   ..$ organism      : chr "Homo sapiens"</span></span>
<span><span class="co">#&gt;   ..$ mapped_ids    : chr "SCT,C4BPA,FAM159A,TSEN54,VPREB3,ZC3H12D,SPIB,HTR6,SAMD15,PPP4R2,FAM110B,LIPN,C3AR1,EPSTI1,UBASH3A,SAMD12,SAMD10"| __truncated__</span></span>
<span><span class="co">#&gt;   ..$ mapped_count  : int 1040</span></span>
<span><span class="co">#&gt;   ..$ unmapped_ids  : chr "LOC200772,ELK2AP,CLRN1-AS1,LOC100272216,LOC102723604,LILRP2,UBE2Q1-AS1,MIR3945,MEG3,LOC400927-CSNK1E,MIR6732,MI"| __truncated__</span></span>
<span><span class="co">#&gt;   ..$ unmapped_count: int 174</span></span>
<span><span class="co">#&gt;  $ result                    :'data.frame':  14816 obs. of  9 variables:</span></span>
<span><span class="co">#&gt;   ..$ number_in_list     : int [1:14816] 38 109 199 104 136 145 296 150 135 76 ...</span></span>
<span><span class="co">#&gt;   ..$ fold_enrichment    : num [1:14816] 5.57 2.02 1.6 1.96 1.76 ...</span></span>
<span><span class="co">#&gt;   ..$ fdr                : num [1:14816] 1.84e-14 1.04e-08 4.84e-08 1.45e-07 1.53e-07 ...</span></span>
<span><span class="co">#&gt;   ..$ expected           : num [1:14816] 6.82 53.87 124.55 53.15 77.09 ...</span></span>
<span><span class="co">#&gt;   ..$ number_in_reference: int [1:14816] 132 1042 2409 1028 1491 1653 4107 1741 1515 689 ...</span></span>
<span><span class="co">#&gt;   ..$ pValue             : num [1:14816] 1.24e-18 1.40e-12 9.80e-12 3.91e-11 5.16e-11 ...</span></span>
<span><span class="co">#&gt;   ..$ plus_minus         : chr [1:14816] "+" "+" "+" "+" ...</span></span>
<span><span class="co">#&gt;   ..$ term.id            : chr [1:14816] "GO:0002181" "GO:0002684" "GO:0002376" "GO:0050776" ...</span></span>
<span><span class="co">#&gt;   ..$ term.label         : chr [1:14816] "cytoplasmic translation" "positive regulation of immune system process" "immune system process" "regulation of immune response" ...</span></span>
<span><span class="co">#&gt;  $ search                    :List of 1</span></span>
<span><span class="co">#&gt;   ..$ search_type: chr "overrepresentation"</span></span>
<span><span class="co">#&gt;  $ tool_release_date         : int 20240807</span></span>
<span><span class="co">#&gt;  $ enrichment_test_type      : chr "FISHER"</span></span>
<span><span class="co">#&gt;  $ annotDataSet              : chr "GO:0008150"</span></span>
<span><span class="co">#&gt;  $ annot_version_release_date: chr "GO Ontology database DOI:  10.5281/zenodo.15066566 Released 2025-03-16"</span></span>
<span><span class="co">#&gt;  $ correction                : chr "FDR"</span></span></code></pre></div>
<p>The analysis results are returned as a Data Frame with an element
named result:</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-1fb4450895fe099f74a1" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-1fb4450895fe099f74a1">{"x":{"filter":"none","vertical":false,"autoHideNavigation":true,"data":[["1","2","3","4","5","6","7","8","9","10"],[38,109,199,104,136,145,296,150,135,76],[5.568181818181818,2.023310435397515,1.597791177820017,1.956788859307802,1.764269829503336,1.696675900277008,1.39402560455192,1.666465128933764,1.723553934340803,2.133526850507983],[1.836369850118467e-14,1.036555129530243e-08,4.839973174510567e-08,1.449685776843505e-07,1.527781441648867e-07,4.630105085701153e-07,5.253216008197239e-07,6.371036953711031e-07,5.868229018934767e-07,5.92505379659376e-07],[6.824489795918367,53.87210884353741,124.5469387755102,53.1482993197279,77.08571428571427,85.46122448979591,212.334693877551,90.0108843537415,78.32653061224489,35.62176870748299],[132,1042,2409,1028,1491,1653,4107,1741,1515,689],[1.239450492790542e-18,1.399237485866959e-12,9.800161665450663e-12,3.913838490398231e-11,5.155849897573121e-11,1.875042556304463e-10,2.48194600819254e-10,3.440084748224099e-10,3.5646639558864e-10,3.999091385389957e-10],["+","+","+","+","+","+","+","+","+","+"],["GO:0002181","GO:0002684","GO:0002376","GO:0050776","GO:0009607","GO:0002682","GO:0048583","GO:0006955","GO:0006952","GO:0050778"],["cytoplasmic translation","positive regulation of immune system process","immune system process","regulation of immune response","response to biotic stimulus","regulation of immune system process","regulation of response to stimulus","immune response","defense response","positive regulation of immune response"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>number_in_list<\/th>\n      <th>fold_enrichment<\/th>\n      <th>fdr<\/th>\n      <th>expected<\/th>\n      <th>number_in_reference<\/th>\n      <th>pValue<\/th>\n      <th>plus_minus<\/th>\n      <th>term.id<\/th>\n      <th>term.label<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"number_in_list","targets":1},{"name":"fold_enrichment","targets":2},{"name":"fdr","targets":3},{"name":"expected","targets":4},{"name":"number_in_reference","targets":5},{"name":"pValue","targets":6},{"name":"plus_minus","targets":7},{"name":"term.id","targets":8},{"name":"term.label","targets":9}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div class="section level2">
<h2 id="string">STRING<a class="anchor" aria-label="anchor" href="#string"></a>
</h2>
<p>In addition to proteins interaction data, STRING also curates
proteins/genes annotations and provides enrichment analysis services.
You can perform the analysis against multiple gene sets. Some of the
gene sets can be accessed using other services but two gene sets are
exclusive to STRING. See “<a href="https://doi.org/10.1093/nar/gkaa1074" title="The STRING database in 2021: customizable protein–protein networks, and functional characterization of user-uploaded gene/measurement sets" class="external-link">D.
Szklarczyk et al.: Nucleic Acids Res. 49, D1 (2021)</a>” for further
information. Again. We will only review the enrichment functionality
here; for an in-depth review of STRING refer to the vignette article “<a href="rbioapi_string.html" title="2.F: STRING &amp; rbioapi">STRING
&amp; rbioapi</a>”.</p>
<ul>
<li>
<p>Gene sets exclusively in STRING:</p>
<ol style="list-style-type: decimal">
<li><p>Reference Publications (<a href="https://pubmed.ncbi.nlm.nih.gov/" class="external-link">PubMed</a>)</p></li>
<li><p><a href="https://string-db.org/help/faq/#what-are-local-string-network-clusters" title="What are local STRING network clusters?" class="external-link">Hierarchical clustering
of the STRING network</a></p></li>
</ol>
</li>
<li>
<p>Gene sets also available in other services:</p>
<ol style="list-style-type: decimal">
<li><p><a href="https://geneontology.org/" title="Gene Ontology Resource - Unifying Biology" class="external-link">Gene Ontology</a>
domains: Molecular Function, Biological Process, and Cellular
Component</p></li>
<li><p><a href="https://doi.org/10.1093/database/bau012" title="Subcellular localization database" class="external-link">COMPARTMENTS</a></p></li>
<li><p><a href="https://disease-ontology.org/" title="Disease Ontology" class="external-link">Human Disease Ontology</a></p></li>
<li><p><a href="https://www.ebi.ac.uk/interpro/" title="Classification of protein families" class="external-link">InterPro</a></p></li>
<li><p><a href="https://www.genome.jp/kegg/pathway.html" title="Wiring diagrams of molecular interactions, reactions and relations" class="external-link">KEG1G
pathways</a></p></li>
<li><p><a href="https://www.uniprot.org/keywords/" title="UniProt keyword: a controlled vocabulary with a hierarchical structure" class="external-link">UniProt
keyword</a></p></li>
<li><p><a href="https://www.ebi.ac.uk/interpro/" title="InterPro: Classification of protein families" class="external-link">Pfam</a></p></li>
<li><p><a href="https://reactome.org/PathwayBrowser/" title="free, open-source, curated and peer-reviewed pathway database" class="external-link">Reactome
pathways</a></p></li>
<li><p><a href="https://bioportal.bioontology.org/ontologies/BTO/" class="external-link">BRENDA Tissue
and Enzyme Source Ontology</a></p></li>
<li><p><a href="http://smart.embl-heidelberg.de/" title="Genetically mobile domains and the analysis of domain architectures" class="external-link">SMART</a></p></li>
</ol>
</li>
</ul>
<p>We can directly supply our <code>covid_critical</code> variable to
STRING’s enrichment function:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">string_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_enrichment.html">rba_string_enrichment</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="va">covid_critical</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span> <span class="co">#Homo sapiens</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Performing functional enrichment of 1214 Input Identifiers.</span></span></code></pre></div>
<p>As you can see, by default a list is returned where each element is a
data Frame with enrichment analysis results against one of the gene sets
listed above:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">string_enrich</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 13</span></span>
<span><span class="co">#&gt;  $ COMPARTMENTS     :'data.frame':   18 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Component        :'data.frame':   19 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ DISEASES         :'data.frame':   10 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Function         :'data.frame':   1 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ HPO              :'data.frame':   3 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ KEGG             :'data.frame':   6 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Keyword          :'data.frame':   18 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ NetworkNeighborAL:'data.frame':   46 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ PMID             :'data.frame':   100 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ Process          :'data.frame':   77 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ RCTM             :'data.frame':   49 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ TISSUES          :'data.frame':   14 obs. of  10 variables:</span></span>
<span><span class="co">#&gt;  $ WikiPathways     :'data.frame':   3 obs. of  10 variables:</span></span></code></pre></div>
<p>In the “PMID” element, we see which PubMed papers are
over-represented with our gene list. To keep the table tidy, we are not
displaying the <code>inputGenes</code> and <code>preferredNames</code>
columns.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-10b3b7155e8045a1b2ad" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-10b3b7155e8045a1b2ad">{"x":{"filter":"none","vertical":false,"autoHideNavigation":true,"data":[["164","165","166","167","168","169","170","171","172","173"],["PMID","PMID","PMID","PMID","PMID","PMID","PMID","PMID","PMID","PMID"],["PMID:36201886","PMID:35190725","PMID:30337941","PMID:35456421","PMID:35480888","PMID:31474370","PMID:36776306","PMID:33560295","PMID:37456231","PMID:37158921"],[55,61,53,48,38,70,36,33,30,51],[136,202,151,121,84,327,80,67,55,199],[9606,9606,9606,9606,9606,9606,9606,9606,9606,9606],[6.53e-27,3.52e-24,1.23e-23,2.36e-23,2.6e-20,2.79e-20,2.96e-19,1.1e-18,4.76e-18,7.509999999999999e-18],[3.33e-20,8.969999999999999e-18,2.08e-17,3e-17,2.65e-14,2.65e-14,2.15e-13,7.01e-13,2.69e-12,3.83e-12],["(2022) Analysis and identification of potential type II helper T cell (Th2)-Related key genes and therapeutic agents for COVID-19.","(2022) Ulcerative colitis is characterized by a plasmablast-skewed humoral response associated with disease activity.","(2018) Meta-Analysis Identification of Highly Robust and Differential Immune-Metabolic Signatures of Systemic Host Response to Acute and Latent Tuberculosis in Children and Adults.","(2022) Transcriptional Profiling and Deriving a Seven-Gene Signature That Discriminates Active and Latent Tuberculosis: An Integrative Bioinformatics Approach.","(2022) Identification of Novel Immune Cell-Relevant Therapeutic Targets and Validation of Roles of TK1 in BMSCs of Systemic Lupus Erythematosus.","(2019) Single-Cell Analysis of Crohns Disease Lesions Identifies a Pathogenic Cellular Module Associated with Resistance to Anti-TNF Therapy.","(2023) KNL1 is a prognostic and diagnostic biomarker related to immune infiltration in patients with uterine corpus endometrial carcinoma.","(2021) An OMICs-based meta-analysis to support infection state stratification.","(2023) Pan-cancer evaluation of clinical value of mitotic network activity index (MNAI) and its predictive value for immunotherapy.","(2023) Identification of potential resistance mechanisms and therapeutic targets for the relapse of BCMA CAR-T therapy in relapsedrefractory multiple myeloma through single-cell sequencing."]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>category<\/th>\n      <th>term<\/th>\n      <th>number_of_genes<\/th>\n      <th>number_of_genes_in_background<\/th>\n      <th>ncbiTaxonId<\/th>\n      <th>p_value<\/th>\n      <th>fdr<\/th>\n      <th>description<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[3,4,5,6,7]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"category","targets":1},{"name":"term","targets":2},{"name":"number_of_genes","targets":3},{"name":"number_of_genes_in_background","targets":4},{"name":"ncbiTaxonId","targets":5},{"name":"p_value","targets":6},{"name":"fdr","targets":7},{"name":"description","targets":8}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>You can retrieve the annotations associated with your input
protein(s). The difference here is that no statistical analysis will be
performed and the full annotations of your input gene will be returned.
Because of potentially unwieldy response, <code>allow_pubmed</code> is
set fo FALSE by default. Here we set it to TRUE to also retrieve the
PubMed paper abstracts that include our gene.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">string_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_annotations.html">rba_string_annotations</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="st">"CD177"</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span>, <span class="co">#Homo sapiens</span></span>
<span>  allow_pubmed <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Retrieving functional annotations of 1 Input Identifiers.</span></span></code></pre></div>
<p>As you can see, Our input gene has a large number of annotations. Let
us take a look at the PMID Data Frame, which contains the PubMed
Abstract paper terms:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">string_annot</span>, max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 13</span></span>
<span><span class="co">#&gt;  $ COMPARTMENTS     :'data.frame':   31 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ Component        :'data.frame':   43 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ DISEASES         :'data.frame':   6 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ Function         :'data.frame':   9 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ HPO              :'data.frame':   8 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ InterPro         :'data.frame':   2 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ Keyword          :'data.frame':   13 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ NetworkNeighborAL:'data.frame':   5 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ Pfam             :'data.frame':   1 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ PMID             :'data.frame':   2476 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ Process          :'data.frame':   92 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ RCTM             :'data.frame':   7 obs. of  8 variables:</span></span>
<span><span class="co">#&gt;  $ TISSUES          :'data.frame':   13 obs. of  8 variables:</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-4018eef1a407a0df6b52" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-4018eef1a407a0df6b52">{"x":{"filter":"none","vertical":false,"autoHideNavigation":true,"data":[["211","212","213","214","215","216","217","218","219","220"],["PMID","PMID","PMID","PMID","PMID","PMID","PMID","PMID","PMID","PMID"],["PMID:10349111","PMID:10474090","PMID:10494766","PMID:10536185","PMID:10571272","PMID:10753836","PMID:10864984","PMID:10925144","PMID:10938941","PMID:10980605"],[1,1,1,1,1,1,1,1,1,1],[1,1,1,1,1,1,1,1,1,1],[9606,9606,9606,9606,9606,9606,9606,9606,9606,9606],[["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"]],[["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"],["CD177"]],["(1999) [Bone marrow transplantation in chronic myeloid leukemia].","(1999) Anti-granulocyte antibody screening with extracted granulocyte antigens by a micro-mixed passive hemagglutination method.","(1999) Inhibition of thrombin-induced neuronal cell death by recombinant thrombomodulin and E5510, a synthetic thrombin receptor signaling inhibitor.","(1999) Bovine seminal ribonuclease selectively kills human multidrug-resistant neuroblastoma cells via induction of apoptosis.","(1999) NB1-C16-insulin: site-specific synthesis, purification, and biological activity.","(2000) Cloning of PRV-1, a novel member of the uPAR receptor superfamily, which is overexpressed in polycythemia rubra vera.","(2000) Variations in the expression of granulocyte antigen NB1.","(2000) Pituitary adelylate cyclase-activating peptide is an activator of vasoactive intestinal polypeptide gene transcription in human neuroblastoma cells.","(2000) Molecular genetics of granulocyte polymorphisms.","(2000) Identification and characterization of a 500-kb homozygously deleted region at 1p36.2-p36.3 in a neuroblastoma cell line."]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>category<\/th>\n      <th>term<\/th>\n      <th>number_of_genes<\/th>\n      <th>ratio_in_set<\/th>\n      <th>ncbiTaxonId<\/th>\n      <th>inputGenes<\/th>\n      <th>preferredNames<\/th>\n      <th>description<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[3,4,5]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"category","targets":1},{"name":"term","targets":2},{"name":"number_of_genes","targets":3},{"name":"ratio_in_set","targets":4},{"name":"ncbiTaxonId","targets":5},{"name":"inputGenes","targets":6},{"name":"preferredNames","targets":7},{"name":"description","targets":8}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>In addition to a data frame, you can also get a plot summarizing the
enrichment results. This API endpoint supports extensive customization
of the plot; please refer to the
<code><a href="../reference/rba_string_enrichment_image.html">rba_string_enrichment_image()</a></code> function’s manual for
detailed instructions. Here we perform an enrichment analysis against
the KEGG database, and retrieve a plot of the results.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrich_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_string_enrichment_image.html">rba_string_enrichment_image</a></span><span class="op">(</span></span>
<span>  ids <span class="op">=</span> <span class="va">covid_critical</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  category <span class="op">=</span> <span class="st">"KEGG"</span>,</span>
<span>  image_format <span class="op">=</span> <span class="st">"image"</span>,</span>
<span>  save_image <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  group_by_similarity <span class="op">=</span> <span class="fl">0.6</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="rbioapi_do_enrich_files/figure-html/string_enrichment_plot_image-1.png" alt="Visualization of enrichment analysis results" width="672"><p class="caption">
Visualization of enrichment analysis results
</p>
</div>
<hr>
</div>
<div class="section level2">
<h2 id="mieaa">miEAA<a class="anchor" aria-label="anchor" href="#mieaa"></a>
</h2>
<p>The miRNA Enrichment Analysis and Annotation Tool (<a href="https://ccb-compute2.cs.uni-saarland.de/mieaa2/" title="https://ccb-compute2.cs.uni-saarland.de/mieaa2" class="external-link">miEAA</a>) is a
service provided by the <a href="https://www.ccb.uni-saarland.de/" class="external-link">Chair
for Clinical Bioinformatics at Saarland University</a>. What makes miEAA
unique among other services presented here is that miEAA curates miRNA
sets. Hence you can directly perform miRNA enrichment analysis across
various species with the services miEAA provides.</p>
<p>We have in-depth covered performing enrichment analysis with miEAA in
the vignette article “<a href="rbioapi_mieaa.html" title="2.C: miEEA &amp; rbioapi">miEEA &amp; rbioapi</a><a href="https://rbioapi.moosa-r.com/articles/rbioapi_enrichr.html" title="2.A: Enrichr &amp; rbioapi">i</a>”. Therefore, we will only
provide a brief demonstration here.</p>
<p>We can not use our <code>covid_critical</code> gene list here. We
rather use a miRNA list from the paper “<a href="https://doi.org/10.3390/vaccines9101056" title="SnoRNAs and miRNAs Networks Underlying COVID-19 Disease Severity" class="external-link">A.
Parray et al.: Vaccines 9, 1056 (2021)</a>”. They used microarray to
find differentially expressed miRNA in the severe versus mild COVID 19
patients’ blood.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">covid_mirna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"hsa-miR-3609"</span>, <span class="st">"hsa-miR-199a-5p"</span>, <span class="st">"hsa-miR-139-5p"</span>, <span class="st">"hsa-miR-145-5p"</span>,</span>
<span>  <span class="st">"hsa-miR-3651"</span>, <span class="st">"hsa-miR-1273h-3p"</span>, <span class="st">"hsa-miR-4632-5p"</span>, <span class="st">"hsa-miR-6861-5p"</span>,</span>
<span>  <span class="st">"hsa-miR-6802-5p"</span>, <span class="st">"hsa-miR-5196-5p"</span>, <span class="st">"hsa-miR-92b-5p"</span>, <span class="st">"hsa-miR-6805-5p"</span>,</span>
<span>  <span class="st">"hsa-miR-98-5p"</span>, <span class="st">"hsa-miR-3185"</span>, <span class="st">"hsa-miR-572"</span>, <span class="st">"hsa-miR-371b-5p"</span>,</span>
<span>  <span class="st">"hsa-miR-3180"</span>, <span class="st">"hsa-miR-8073"</span>, <span class="st">"hsa-miR-4750-5p"</span>, <span class="st">"hsa-miR-6075"</span>,</span>
<span>  <span class="st">"hsa-let-7i-5p"</span>, <span class="st">"hsa-miR-1231"</span>, <span class="st">"hsa-miR-885-3p"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>While not necessary, we can first see what enrichment categories are
available for our mature human miRNA:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rba_mieaa_cats.html">rba_mieaa_cats</a></span><span class="op">(</span></span>
<span>  mirna_type <span class="op">=</span> <span class="st">"mature"</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span> <span class="co">#Homo sapiens</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Retrieving available enrichment categories of miRNA for Homo sapiens.</span></span>
<span><span class="co">#&gt; Annotations derived over miRTarBase (Gene Ontology) </span></span>
<span><span class="co">#&gt;                    "GO_Annotations_indirect_mature" </span></span>
<span><span class="co">#&gt;                          Annotation (Gene Ontology) </span></span>
<span><span class="co">#&gt;                             "GO_Annotations_mature" </span></span>
<span><span class="co">#&gt;                                     Pathways (KEGG) </span></span>
<span><span class="co">#&gt;                                       "KEGG_mature" </span></span>
<span><span class="co">#&gt;                             exRNA forms (miRandola) </span></span>
<span><span class="co">#&gt;                                  "miRandola_mature" </span></span>
<span><span class="co">#&gt;                      Chromosomal location (miRBase) </span></span>
<span><span class="co">#&gt;                        "miRBase_Chromosomes_mature" </span></span>
<span><span class="co">#&gt;                              Conservation (miRBase) </span></span>
<span><span class="co">#&gt;       "miRBase_Conserved_miRNAs_5_organisms_mature" </span></span>
<span><span class="co">#&gt;                                Confidence (miRBase) </span></span>
<span><span class="co">#&gt;                    "miRBase_High_confidence_mature" </span></span>
<span><span class="co">#&gt;                               Seed family (miRBase) </span></span>
<span><span class="co">#&gt;                        "miRBase_Seed_family_mature" </span></span>
<span><span class="co">#&gt;                   GO Biological process (miRPathDB) </span></span>
<span><span class="co">#&gt;            "miRPathDB_GO_Biological_process_mature" </span></span>
<span><span class="co">#&gt;                   GO Cellular component (miRPathDB) </span></span>
<span><span class="co">#&gt;            "miRPathDB_GO_Cellular_component_mature" </span></span>
<span><span class="co">#&gt;                   GO Molecular function (miRPathDB) </span></span>
<span><span class="co">#&gt;            "miRPathDB_GO_Molecular_function_mature" </span></span>
<span><span class="co">#&gt;                                    KEGG (miRPathDB) </span></span>
<span><span class="co">#&gt;                             "miRPathDB_KEGG_mature" </span></span>
<span><span class="co">#&gt;                                Reactome (miRPathDB) </span></span>
<span><span class="co">#&gt;                         "miRPathDB_Reactome_mature" </span></span>
<span><span class="co">#&gt;                            WikiPathways (miRPathDB) </span></span>
<span><span class="co">#&gt;                     "miRPathDB_WikiPathways_mature" </span></span>
<span><span class="co">#&gt;                           Target genes (miRTarBase) </span></span>
<span><span class="co">#&gt;                                 "miRTarBase_mature" </span></span>
<span><span class="co">#&gt;                                  Diseases (miRWalk) </span></span>
<span><span class="co">#&gt;                           "miRWalk_Diseases_mature" </span></span>
<span><span class="co">#&gt;                             Gene Ontology (miRWalk) </span></span>
<span><span class="co">#&gt;                                 "miRWalk_GO_mature" </span></span>
<span><span class="co">#&gt;                                    Organs (miRWalk) </span></span>
<span><span class="co">#&gt;                             "miRWalk_Organs_mature" </span></span>
<span><span class="co">#&gt;                                  Pathways (miRWalk) </span></span>
<span><span class="co">#&gt;                           "miRWalk_Pathways_mature" </span></span>
<span><span class="co">#&gt;                                     Diseases (MNDR) </span></span>
<span><span class="co">#&gt;                                       "MNDR_mature" </span></span>
<span><span class="co">#&gt;                              Interactions (NPInter) </span></span>
<span><span class="co">#&gt;                                    "NPInter_mature" </span></span>
<span><span class="co">#&gt;                                          Gender/Age </span></span>
<span><span class="co">#&gt;                       "Published_Age_gender_mature" </span></span>
<span><span class="co">#&gt;                          Cell-type specific (Atlas) </span></span>
<span><span class="co">#&gt;                    "Published_cell_specific_mature" </span></span>
<span><span class="co">#&gt;           Cell-type specific (Cellular microRNAome) </span></span>
<span><span class="co">#&gt;                                   "miRNAome_mature" </span></span>
<span><span class="co">#&gt;                                  Published Diseases </span></span>
<span><span class="co">#&gt;                         "Published_Diseases_mature" </span></span>
<span><span class="co">#&gt;                                        Immune cells </span></span>
<span><span class="co">#&gt;                     "Published_Immune_cells_mature" </span></span>
<span><span class="co">#&gt;                            Localization (RNALocate) </span></span>
<span><span class="co">#&gt;                                  "RNALocate_mature" </span></span>
<span><span class="co">#&gt;                                      Drugs (SM2miR) </span></span>
<span><span class="co">#&gt;                                     "SM2miR_mature" </span></span>
<span><span class="co">#&gt;                  Expressed in tissue (Tissue Atlas) </span></span>
<span><span class="co">#&gt;                                "TissueAtlas_mature" </span></span>
<span><span class="co">#&gt;                          Tissue specific (isomiRdb) </span></span>
<span><span class="co">#&gt;                   "isomiRdb_tissue_specific_mature" </span></span>
<span><span class="co">#&gt;                         High confidence (MirGeneDB) </span></span>
<span><span class="co">#&gt;                  "MirGeneDB_High_confidence_mature" </span></span>
<span><span class="co">#&gt;                                  isomiRs (isomiRdb) </span></span>
<span><span class="co">#&gt;                                   "isomiRdb_mature"</span></span></code></pre></div>
<p>Next, we will use the wrapper function available to perform the
necessary steps in one function call. Note that we have supplied the
categories parameter. If left NULL, the enrichment will be performed
against every available category.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mieaa_enrich</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_mieaa_enrich.html">rba_mieaa_enrich</a></span><span class="op">(</span></span>
<span>  test_set <span class="op">=</span> <span class="va">covid_mirna</span>,</span>
<span>  mirna_type <span class="op">=</span> <span class="st">"mature"</span>,</span>
<span>  test_type <span class="op">=</span> <span class="st">"ORA"</span>,</span>
<span>  species <span class="op">=</span> <span class="fl">9606</span>,</span>
<span>  categories <span class="op">=</span> <span class="st">"miRPathDB_GO_Biological_process_mature"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt;  -- Step 1/3: Submitting Enrichment analysis request:</span></span>
<span><span class="co">#&gt; Submitting ORA enrichment request for 23 miRNA IDs of species Homo sapiens to miEAA servers.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  -- Step 2/3: Checking for Submitted enrichment analysis's status every 5 seconds.</span></span>
<span><span class="co">#&gt;     Your submitted job ID is: ed779d5e-fc1a-4478-a588-292ba91c7a72</span></span>
<span><span class="co">#&gt; .</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  -- Step 3/3: Retrieving the results.</span></span>
<span><span class="co">#&gt; Retrieving results of submitted enrichment request with ID: ed779d5e-fc1a-4478-a588-292ba91c7a72</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-5b1b2f4ad92281566982" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-5b1b2f4ad92281566982">{"x":{"filter":"none","vertical":false,"autoHideNavigation":true,"data":[["1","2","3","4","5","6","7","8","9","10"],["GO Biological process (miRPathDB)","GO Biological process (miRPathDB)","GO Biological process (miRPathDB)","GO Biological process (miRPathDB)","GO Biological process (miRPathDB)","GO Biological process (miRPathDB)","GO Biological process (miRPathDB)","GO Biological process (miRPathDB)","GO Biological process (miRPathDB)","GO Biological process (miRPathDB)"],["B cell proliferation","DNA damage response signal transduction by p53 class mediator","L-serine metabolic process","amyloid-beta clearance","angiotensin-activated signaling pathway","cell-cell adhesion mediated by integrin","cell-cell adhesion via plasma-membrane adhesion molecules","cellular response to calcium ion","cerebral cortex development","chondrocyte differentiation involved in endochondral bone morphogenesis"],["over-represented","over-represented","over-represented","over-represented","over-represented","over-represented","over-represented","over-represented","over-represented","over-represented"],["5.44362e-05","0.000105875","0.000335649","0.000557165","0.000832385","0.000557165","0.000335649","0.0011607","0.0011607","0.000557165"],["0.0294189","0.0331389","0.0415979","0.0415979","0.0452719","0.0415979","0.0415979","0.0490924","0.0490924","0.0415979"],["0.0294189","0.0331389","0.0415979","0.0415979","0.0452719","0.0415979","0.0415979","0.0490924","0.0490924","0.0415979"],["0.0882648","0.609829","0.0320963","0.0401204","0.0481444","0.0401204","0.0320963","0.0561685","0.0561685","0.0401204"],["3","5","2","2","2","2","2","2","2","2"],["hsa-miR-145-5p; hsa-miR-98-5p; hsa-miR-572","hsa-miR-199a-5p; hsa-miR-145-5p; hsa-miR-6805-5p; hsa-miR-98-5p; hsa-miR-572","hsa-miR-145-5p; hsa-miR-6805-5p","hsa-miR-199a-5p; hsa-miR-145-5p","hsa-miR-199a-5p; hsa-miR-145-5p","hsa-miR-199a-5p; hsa-miR-145-5p","hsa-miR-199a-5p; hsa-miR-145-5p","hsa-miR-199a-5p; hsa-miR-139-5p","hsa-miR-199a-5p; hsa-miR-145-5p","hsa-miR-199a-5p; hsa-miR-145-5p"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Category<\/th>\n      <th>Subcategory<\/th>\n      <th>Enrichment<\/th>\n      <th>P-value<\/th>\n      <th>P-adjusted<\/th>\n      <th>Q-value<\/th>\n      <th>Expected<\/th>\n      <th>Observed<\/th>\n      <th>miRNAs/precursors<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Category","targets":1},{"name":"Subcategory","targets":2},{"name":"Enrichment","targets":3},{"name":"P-value","targets":4},{"name":"P-adjusted","targets":5},{"name":"Q-value","targets":6},{"name":"Expected","targets":7},{"name":"Observed","targets":8},{"name":"miRNAs/precursors","targets":9}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div class="section level2">
<h2 id="how-to-cite">How to cite?<a class="anchor" aria-label="anchor" href="#how-to-cite"></a>
</h2>
<p>rbioapi is an interface between you and other databases and services.
Thus, if you have used rbioapi in published research, <strong>in
addition to kindly citing rbioapi, <u><em>ensure to fully and properly
cite the databases/services you have used</em></u></strong>. Suggested
citations have been added in the functions’ manuals, under the
“references” section; Nevertheless, it is the user’s responsibility to
check for proper citations and to properly cite the database/services
that they have used.</p>
<p>Please see the <a href="rbioapi.html#how-to-cite">rbioapi’s main
vignette article</a> for more details.</p>
<hr>
</div>
<div class="section level2">
<h2 id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_do_enrich.html" title="Over-Representation Analysis with rbioapi">This article in
rbioapi documentation site</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html" title="rbioapi reference">Functions references in rbioapi
documentation site</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi
vignette index</a></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] rbioapi_0.8.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] httr_1.4.7        cli_3.6.5         knitr_1.50        rlang_1.1.6      </span></span>
<span><span class="co">#&gt;  [5] xfun_0.52         png_0.1-8         textshaping_1.0.1 jsonlite_2.0.0   </span></span>
<span><span class="co">#&gt;  [9] DT_0.33           htmltools_0.5.8.1 ragg_1.4.0        sass_0.4.10      </span></span>
<span><span class="co">#&gt; [13] rmarkdown_2.29    grid_4.5.1        crosstalk_1.2.1   evaluate_1.0.4   </span></span>
<span><span class="co">#&gt; [17] jquerylib_0.1.4   fastmap_1.2.0     yaml_2.3.10       lifecycle_1.0.4  </span></span>
<span><span class="co">#&gt; [21] compiler_4.5.1    fs_1.6.6          htmlwidgets_1.6.4 systemfonts_1.2.3</span></span>
<span><span class="co">#&gt; [25] digest_0.6.37     R6_2.6.1          curl_6.4.0        magrittr_2.0.3   </span></span>
<span><span class="co">#&gt; [29] bslib_0.9.0       tools_4.5.1       pkgdown_2.1.3     cachem_1.1.0     </span></span>
<span><span class="co">#&gt; [33] desc_1.4.3</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://moosa-r.com/" class="external-link">Moosa Rezwani</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
