<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2.A: Enrichr &amp; rbioapi • rbioapi</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2.A: Enrichr &amp; rbioapi">
<meta property="og:description" content="rbioapi">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rbioapi</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/rbioapi.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Supported Services
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/rbioapi.html">1: rbioapi: Get Started</a>
    </li>
    <li>
      <a href="../articles/rbioapi_mieaa.html">2.B: miEEA &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_panther.html">2.C: PANTHER &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_reactome.html">2.D: Reactome &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_string.html">2.E: STRING &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_uniprot.html">2.F: UniProt &amp; rbioapi</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="rbioapi_enrichr_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="rbioapi_enrichr_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="rbioapi_enrichr_files/jquery-3.5.1/jquery.min.js"></script><link href="rbioapi_enrichr_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_enrichr_files/datatables-binding-0.18/datatables.js"></script><link href="rbioapi_enrichr_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_enrichr_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_enrichr_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="rbioapi_enrichr_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="rbioapi_enrichr_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>2.A: Enrichr &amp; rbioapi</h1>
                        <h4 class="author">Moosa Rezwani</h4>
            
            <h4 class="date">2021-05-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/master/vignettes/rbioapi_enrichr.Rmd"><code>vignettes/rbioapi_enrichr.Rmd</code></a></small>
      <div class="hidden name"><code>rbioapi_enrichr.Rmd</code></div>

    </div>

    
    
<hr>
<div id="introduction" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><a href="https://maayanlab.cloud/Enrichr">Enrichr</a> is a gene-set enrichment analysis tool developed in the <a href="https://icahn.mssm.edu/research/labs/maayan-laboratory">Ma’ayan lab</a>.</p>
</div>
<div id="gene-set-library-concept-in-enrichr" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#gene-set-library-concept-in-enrichr" class="anchor"></a>Gene set library concept in Enrichr</h1>
<p>Directly quoting from Enrichr’s help page;</p>
<blockquote>
<p>A <em>gene set library</em> is a set of related gene sets or enrichment terms. Each enrichment term in Enrichr’s results pages is organized by its gene set library. These libraries have been constructed from many sources such as published studies and major biological and biomedical online databases. Others have been created for and only available through Enrichr. For example, the ChEA 2015 library is a set of functional terms representing transcription factors profiled by ChIP-seq in mammalian cells. Each term is associated with a collection of putative targets inferred from the peaks identified in each ChIP-seq study.</p>
<p>(source: <a href="https://maayanlab.cloud/Enrichr/help#background" class="uri">https://maayanlab.cloud/Enrichr/help#background</a>)</p>
</blockquote>
<p>To get a list of available Enrichr libraries, use:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">enrichr_libs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_libs.html">rba_enrichr_libs</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>In the returned data frame, you can find the names of available Enrichr libraries in “libraryName” column:</p>
<div id="htmlwidget-8f18ec91e57e88d21e39" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-8f18ec91e57e88d21e39">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172"],[13362,27884,6002,47172,47107,21493,1295,3185,2854,15057,4128,34061,7504,16399,12753,23726,32740,13373,19270,13236,14264,3096,22288,4533,10231,2741,5655,10406,10493,11251,8695,1759,2178,851,10061,11250,15406,17711,17576,15797,12232,13572,6454,3723,7588,7682,7324,8469,13121,26382,29065,280,13877,15852,4320,4271,10496,1678,756,3800,2541,1918,5863,6768,25651,19129,23939,23561,23877,15886,24350,3102,31132,30832,48230,5613,9559,9448,16725,19249,15090,16129,15309,15103,15022,15676,15854,15015,3788,3357,12668,12638,8973,7010,5966,15562,17850,17660,1348,934,2541,2041,5209,49238,2243,19586,22440,8184,18329,15755,10271,10427,10601,13822,8002,10089,13247,21809,23601,20883,19612,25983,19500,14893,17598,5902,12486,1073,19513,14433,8655,11459,19741,27360,13072,13464,13787,13929,16964,17258,10352,10471,12419,19378,6201,4558,3264,7802,8551,12444,9000,7744,6204,13420,14148,9813,1397,9116,17464,394,11851,8189,18704,5605,5718,14156,16979,4383,54974,12118,12361,9763],[275,1284,77,1370,509,3713,18,73,34,300,48,641,155,247,57,127,85,258,388,82,58,31,4368,37,158,5,342,715,200,100,100,25,89,15,106,166,300,300,300,176,343,301,301,47,35,78,172,122,305,1811,2123,9,304,912,129,128,201,21,12,48,39,39,51,47,807,1594,293,307,302,9,299,25,298,302,1429,36,73,63,1443,1443,282,292,308,318,290,310,279,321,159,153,300,300,64,87,51,887,300,300,19,13,39,42,300,1550,19,545,505,24,161,28,22,38,25,21,143,45,49,2316,2395,299,299,299,137,128,1208,109,299,100,117,36,61,39,270,802,26,45,200,200,200,200,58,76,491,37,45,54,22,92,98,23,20,363,387,32,122,49,13,22,63,73,586,421,100,39,31,40,295,146,483,448,124,139],["Genome_Browser_PWMs","TRANSFAC_and_JASPAR_PWMs","Transcription_Factor_PPIs","ChEA_2013","Drug_Perturbations_from_GEO_2014","ENCODE_TF_ChIP-seq_2014","BioCarta_2013","Reactome_2013","WikiPathways_2013","Disease_Signatures_from_GEO_up_2014","KEGG_2013","TF-LOF_Expression_from_GEO","TargetScan_microRNA","PPI_Hub_Proteins","GO_Molecular_Function_2015","GeneSigDB","Chromosome_Location","Human_Gene_Atlas","Mouse_Gene_Atlas","GO_Cellular_Component_2015","GO_Biological_Process_2015","Human_Phenotype_Ontology","Epigenomics_Roadmap_HM_ChIP-seq","KEA_2013","NURSA_Human_Endogenous_Complexome","CORUM","SILAC_Phosphoproteomics","MGI_Mammalian_Phenotype_Level_3","MGI_Mammalian_Phenotype_Level_4","Old_CMAP_up","Old_CMAP_down","OMIM_Disease","OMIM_Expanded","VirusMINT","MSigDB_Computational","MSigDB_Oncogenic_Signatures","Disease_Signatures_from_GEO_down_2014","Virus_Perturbations_from_GEO_up","Virus_Perturbations_from_GEO_down","Cancer_Cell_Line_Encyclopedia","NCI-60_Cancer_Cell_Lines","Tissue_Protein_Expression_from_ProteomicsDB","Tissue_Protein_Expression_from_Human_Proteome_Map","HMDB_Metabolites","Pfam_InterPro_Domains","GO_Biological_Process_2013","GO_Cellular_Component_2013","GO_Molecular_Function_2013","Allen_Brain_Atlas_up","ENCODE_TF_ChIP-seq_2015","ENCODE_Histone_Modifications_2015","Phosphatase_Substrates_from_DEPOD","Allen_Brain_Atlas_down","ENCODE_Histone_Modifications_2013","Achilles_fitness_increase","Achilles_fitness_decrease","MGI_Mammalian_Phenotype_2013","BioCarta_2015","HumanCyc_2015","KEGG_2015","NCI-Nature_2015","Panther_2015","WikiPathways_2015","Reactome_2015","ESCAPE","HomoloGene","Disease_Perturbations_from_GEO_down","Disease_Perturbations_from_GEO_up","Drug_Perturbations_from_GEO_down","Genes_Associated_with_NIH_Grants","Drug_Perturbations_from_GEO_up","KEA_2015","Gene_Perturbations_from_GEO_up","Gene_Perturbations_from_GEO_down","ChEA_2015","dbGaP","LINCS_L1000_Chem_Pert_up","LINCS_L1000_Chem_Pert_down","GTEx_Tissue_Sample_Gene_Expression_Profiles_down","GTEx_Tissue_Sample_Gene_Expression_Profiles_up","Ligand_Perturbations_from_GEO_down","Aging_Perturbations_from_GEO_down","Aging_Perturbations_from_GEO_up","Ligand_Perturbations_from_GEO_up","MCF7_Perturbations_from_GEO_down","MCF7_Perturbations_from_GEO_up","Microbe_Perturbations_from_GEO_down","Microbe_Perturbations_from_GEO_up","LINCS_L1000_Ligand_Perturbations_down","LINCS_L1000_Ligand_Perturbations_up","L1000_Kinase_and_GPCR_Perturbations_down","L1000_Kinase_and_GPCR_Perturbations_up","Reactome_2016","KEGG_2016","WikiPathways_2016","ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X","Kinase_Perturbations_from_GEO_down","Kinase_Perturbations_from_GEO_up","BioCarta_2016","HumanCyc_2016","NCI-Nature_2016","Panther_2016","DrugMatrix","ChEA_2016","huMAP","Jensen_TISSUES","RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO","MGI_Mammalian_Phenotype_2017","Jensen_COMPARTMENTS","Jensen_DISEASES","BioPlex_2017","GO_Cellular_Component_2017","GO_Molecular_Function_2017","GO_Biological_Process_2017","GO_Cellular_Component_2017b","GO_Molecular_Function_2017b","GO_Biological_Process_2017b","ARCHS4_Tissues","ARCHS4_Cell-lines","ARCHS4_IDG_Coexp","ARCHS4_Kinases_Coexp","ARCHS4_TFs_Coexp","SysMyo_Muscle_Gene_Sets","miRTarBase_2017","TargetScan_microRNA_2017","Enrichr_Libraries_Most_Popular_Genes","Enrichr_Submissions_TF-Gene_Coocurrence","Data_Acquisition_Method_Most_Popular_Genes","DSigDB","GO_Biological_Process_2018","GO_Cellular_Component_2018","GO_Molecular_Function_2018","TF_Perturbations_Followed_by_Expression","Chromosome_Location_hg19","NIH_Funded_PIs_2017_Human_GeneRIF","NIH_Funded_PIs_2017_Human_AutoRIF","Rare_Diseases_AutoRIF_ARCHS4_Predictions","Rare_Diseases_GeneRIF_ARCHS4_Predictions","NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions","NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions","Rare_Diseases_GeneRIF_Gene_Lists","Rare_Diseases_AutoRIF_Gene_Lists","SubCell_BarCode","GWAS_Catalog_2019","WikiPathways_2019_Human","WikiPathways_2019_Mouse","TRRUST_Transcription_Factors_2019","KEGG_2019_Human","KEGG_2019_Mouse","InterPro_Domains_2019","Pfam_Domains_2019","DepMap_WG_CRISPR_Screens_Broad_CellLines_2019","DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019","MGI_Mammalian_Phenotype_Level_4_2019","UK_Biobank_GWAS_v1","BioPlanet_2019","ClinVar_2019","PheWeb_2019","DisGeNET","HMS_LINCS_KinomeScan","CCLE_Proteomics_2020","ProteomicsDB_2020","lncHUB_lncRNA_Co-Expression","Virus-Host_PPI_P-HIPSTer_2020","Elsevier_Pathway_Collection","Table_Mining_of_CRISPR_Studies","COVID-19_Related_Gene_Sets","MSigDB_Hallmark_2020","Enrichr_Users_Contributed_Lists_2020","TG_GATES_2020","Allen_Brain_Atlas_10x_scRNA_2021","Descartes_Cell_Types_and_Tissue_2021"],["http://hgdownload.cse.ucsc.edu/goldenPath/hg18/database/","http://jaspar.genereg.net/html/DOWNLOAD/","","http://amp.pharm.mssm.edu/lib/cheadownload.jsp","http://www.ncbi.nlm.nih.gov/geo/","http://genome.ucsc.edu/ENCODE/downloads.html","https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://www.reactome.org/download/index.html","http://www.wikipathways.org/index.php/Download_Pathways","http://www.ncbi.nlm.nih.gov/geo/","http://www.kegg.jp/kegg/download/","http://www.ncbi.nlm.nih.gov/geo/","http://www.targetscan.org/cgi-bin/targetscan/data_download.cgi?db=vert_61","http://amp.pharm.mssm.edu/X2K","http://www.geneontology.org/GO.downloads.annotations.shtml","https://maayanlab.cloud/Enrichr/","http://software.broadinstitute.org/gsea/msigdb/index.jsp","http://biogps.org/downloads/","http://biogps.org/downloads/","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.human-phenotype-ontology.org/","http://www.roadmapepigenomics.org/","http://amp.pharm.mssm.edu/lib/keacommandline.jsp","https://www.nursa.org/nursa/index.jsf","http://mips.helmholtz-muenchen.de/genre/proj/corum/","http://amp.pharm.mssm.edu/lib/keacommandline.jsp","http://www.informatics.jax.org/","http://www.informatics.jax.org/","http://www.broadinstitute.org/cmap/","http://www.broadinstitute.org/cmap/","http://www.omim.org/downloads","http://www.omim.org/downloads","http://mint.bio.uniroma2.it/download.html","http://www.broadinstitute.org/gsea/msigdb/collections.jsp","http://www.broadinstitute.org/gsea/msigdb/collections.jsp","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://portals.broadinstitute.org/ccle/home\n","http://biogps.org/downloads/","https://www.proteomicsdb.org/","http://www.humanproteomemap.org/index.php","http://www.hmdb.ca/downloads","ftp://ftp.ebi.ac.uk/pub/databases/interpro/","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.geneontology.org/GO.downloads.annotations.shtml","http://www.brain-map.org/","http://genome.ucsc.edu/ENCODE/downloads.html","http://genome.ucsc.edu/ENCODE/downloads.html","http://www.koehn.embl.de/depod/","http://www.brain-map.org/","http://genome.ucsc.edu/ENCODE/downloads.html","http://www.broadinstitute.org/achilles","http://www.broadinstitute.org/achilles","http://www.informatics.jax.org/","https://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://humancyc.org/","http://www.kegg.jp/kegg/download/","http://pid.nci.nih.gov/","http://www.pantherdb.org/","http://www.wikipathways.org/index.php/Download_Pathways","http://www.reactome.org/download/index.html","http://www.maayanlab.net/ESCAPE/","http://www.ncbi.nlm.nih.gov/homologene","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://grants.nih.gov/grants/oer.htm\n","http://www.ncbi.nlm.nih.gov/geo/","http://amp.pharm.mssm.edu/Enrichr","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://amp.pharm.mssm.edu/Enrichr","http://www.ncbi.nlm.nih.gov/gap","https://clue.io/","https://clue.io/","http://www.gtexportal.org/","http://www.gtexportal.org/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","https://clue.io/","https://clue.io/","https://clue.io/","https://clue.io/","http://www.reactome.org/download/index.html","http://www.kegg.jp/kegg/download/","http://www.wikipathways.org/index.php/Download_Pathways","","http://www.ncbi.nlm.nih.gov/geo/","http://www.ncbi.nlm.nih.gov/geo/","http://cgap.nci.nih.gov/Pathways/BioCarta_Pathways","http://humancyc.org/","http://pid.nci.nih.gov/","http://www.pantherdb.org/pathway/","https://ntp.niehs.nih.gov/drugmatrix/","http://amp.pharm.mssm.edu/Enrichr","http://proteincomplexes.org/","http://tissues.jensenlab.org/","http://www.ncbi.nlm.nih.gov/geo/","http://www.informatics.jax.org/","http://compartments.jensenlab.org/","http://diseases.jensenlab.org/","http://bioplex.hms.harvard.edu/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://amp.pharm.mssm.edu/archs4","http://sys-myo.rhcloud.com/","http://mirtarbase.mbc.nctu.edu.tw/","http://www.targetscan.org/","http://amp.pharm.mssm.edu/Enrichr","http://amp.pharm.mssm.edu/Enrichr","http://amp.pharm.mssm.edu/Enrichr","http://tanlab.ucdenver.edu/DSigDB/DSigDBv1.0/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.geneontology.org/","http://www.ncbi.nlm.nih.gov/geo/","http://hgdownload.cse.ucsc.edu/downloads.html","https://www.ncbi.nlm.nih.gov/pubmed/","https://www.ncbi.nlm.nih.gov/pubmed/","https://amp.pharm.mssm.edu/geneshot/","https://www.ncbi.nlm.nih.gov/gene/about-generif","https://www.ncbi.nlm.nih.gov/pubmed/","https://www.ncbi.nlm.nih.gov/pubmed/","https://www.ncbi.nlm.nih.gov/gene/about-generif","https://amp.pharm.mssm.edu/geneshot/","http://www.subcellbarcode.org/","https://www.ebi.ac.uk/gwas","https://www.wikipathways.org/","https://www.wikipathways.org/","https://www.grnpedia.org/trrust/","https://www.kegg.jp/","https://www.kegg.jp/","https://www.ebi.ac.uk/interpro/","https://pfam.xfam.org/","https://depmap.org/","https://depmap.org/","http://www.informatics.jax.org/","https://www.ukbiobank.ac.uk/tag/gwas/","https://tripod.nih.gov/bioplanet/","https://www.ncbi.nlm.nih.gov/clinvar/","http://pheweb.sph.umich.edu/","https://www.disgenet.org","http://lincs.hms.harvard.edu/kinomescan/","https://portals.broadinstitute.org/ccle","https://www.proteomicsdb.org/","https://amp.pharm.mssm.edu/lnchub/","http://phipster.org/","http://www.transgene.ru/disease-pathways/","","https://amp.pharm.mssm.edu/covid19","https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp ","https://maayanlab.cloud/Enrichr","https://toxico.nibiohn.go.jp/english/","https://portal.brain-map.org/","https://descartes.brotmanbaty.org/bbi/human-gene-expression-during-development/"],[615,326,290,353,701,498,249,78,199,142,200,269,222,385,1136,2139,386,84,96,641,5192,1779,383,474,1796,1658,84,71,476,6100,6100,90,187,85,858,189,142,323,323,967,93,207,30,3906,311,941,205,402,2192,816,412,59,2192,109,216,216,476,239,125,179,209,104,404,1389,315,12,839,839,906,32876,906,428,2460,2460,395,345,33132,33132,2918,2918,261,286,286,261,401,401,312,312,96,96,3644,3644,1530,293,437,104,285,285,237,152,209,112,7876,645,995,1842,1302,5231,2283,1811,3915,636,972,3166,816,3271,10125,108,125,352,498,1724,1135,3240,683,121,1722,12,4026,5103,446,1151,1958,36,5687,12558,3725,2244,12558,5684,2244,3725,104,1737,472,176,571,308,303,1071,608,558,325,5261,857,1510,182,1161,9828,148,378,913,3729,6715,1721,802,205,50,1482,1190,766,172]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>geneCoverage<\/th>\n      <th>genesPerTerm<\/th>\n      <th>libraryName<\/th>\n      <th>link<\/th>\n      <th>numTerms<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[1,2,5]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div id="enrichment-analysis-using-enrichr" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#enrichment-analysis-using-enrichr" class="anchor"></a>Enrichment analysis using Enrichr</h1>
<p>To enrich your gene-set with Enrichr using rbioapi, you can take two approaches. we will begin with the simple one.</p>
<div id="approach-1-using-the-wrapper-function" class="section level2 heading3">
<h2 class="hasAnchor">
<a href="#approach-1-using-the-wrapper-function" class="anchor"></a>Approach 1: Using the Wrapper function</h2>
<p>Just fill the arguments of <code>rba_enrichr</code> according to the function’s manual; Simply provide your gene-set as a character vector and select the libraries.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 1 We create a variable with our genes' NCBI IDs</span>
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>,<span class="st">"CDK1"</span>,<span class="st">"KIF23"</span>,<span class="st">"PLK1"</span>,
           <span class="st">"RAC2"</span>,<span class="st">"RACGAP1"</span>,<span class="st">"RHOA"</span>,<span class="st">"RHOB"</span>, <span class="st">"PHF14"</span>, <span class="st">"RBM3"</span>, <span class="st">"MSL1"</span><span class="op">)</span>

<span class="co"># 2.a Enrich your genes using "MSigDB_Hallmark_2020" library</span>
<span class="va">enrichr_msig_hallmark</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr.html">rba_enrichr</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span>,
                                         gene_set_library <span class="op">=</span> <span class="st">"MSigDB_Hallmark_2020"</span><span class="op">)</span>
<span class="co"># 2.b Maybe you want to enrich using every library that contains the word "msig":</span>
<span class="va">enrichr_msig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr.html">rba_enrichr</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span>,
                                gene_set_library <span class="op">=</span> <span class="st">"msig"</span>,
                                regex_library_name <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co"># 2.c Or maybe you want to enrich using every library available at Enrichr:</span>
<span class="co"># enrichr_all &lt;- rba_enrichr(gene_list = genes,</span>
<span class="co">#                            gene_set_library = "all")</span></code></pre></div>
<p>Note that when only one Enrichr library is selected, a data frame with enrichment analysis result will be returned:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">enrichr_msig_hallmark</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    18 obs. of  9 variables:</span>
<span class="co">#&gt;  $ Term                : chr  "Mitotic Spindle" "G2-M Checkpoint" "E2F Targets" "Apoptosis" ...</span>
<span class="co">#&gt;  $ Overlap             : chr  "5/199" "4/200" "4/200" "3/161" ...</span>
<span class="co">#&gt;  $ P.value             : num  2.57e-07 1.22e-05 1.22e-05 2.17e-04 2.74e-03 ...</span>
<span class="co">#&gt;  $ Adjusted.P.value    : num  4.62e-06 7.29e-05 7.29e-05 9.76e-04 9.87e-03 ...</span>
<span class="co">#&gt;  $ Old.P.value         : int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ Old.Adjusted.P.value: int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ Odds.Ratio          : num  51 36.7 36.7 31.4 29.7 ...</span>
<span class="co">#&gt;  $ Combined.Score      : num  774 416 416 265 175 ...</span>
<span class="co">#&gt;  $ Genes               : chr  "CDC42;RACGAP1;PLK1;CDK1;KIF23" "RACGAP1;PLK1;CDK1;KIF23" "RACGAP1;PLK1;CDK1;BRCA1" "CDK2;BRCA1;RHOB" ...</span></code></pre></div>
<p>But when multiple libraries have been selected, the function’s output will be a list where each element is a data frame corresponding to one of the selected libraries:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">enrichr_msig</span>, <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; List of 3</span>
<span class="co">#&gt;  $ MSigDB_Computational       :'data.frame': 195 obs. of  9 variables:</span>
<span class="co">#&gt;  $ MSigDB_Oncogenic_Signatures:'data.frame': 26 obs. of  9 variables:</span>
<span class="co">#&gt;  $ MSigDB_Hallmark_2020       :'data.frame': 18 obs. of  9 variables:</span></code></pre></div>
</div>
<div id="approach-2-going-step-by-step" class="section level2 heading3">
<h2 class="hasAnchor">
<a href="#approach-2-going-step-by-step" class="anchor"></a>Approach 2: Going step-by-step</h2>
<p>As you can see in <code><a href="../reference/rba_enrichr.html">rba_enrichr()</a></code>’s name, it is a wrapper function. It basically executes the following sequence of functions:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 1 Get a list of available Enrichr libraries</span>
<span class="va">libs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_libs.html">rba_enrichr_libs</a></span><span class="op">(</span>store_in_options <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># 2 Submit your gene-set to enrichr</span>
<span class="va">list_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_add_list.html">rba_enrichr_add_list</a></span><span class="op">(</span>gene_list <span class="op">=</span> <span class="va">genes</span><span class="op">)</span>

<span class="co"># 3 Perform Enrichment analysis with your uploaded gene-set</span>
<span class="va">enriched</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_enrichr_enrich.html">rba_enrichr_enrich</a></span><span class="op">(</span>user_list_id <span class="op">=</span> <span class="va">list_id</span><span class="op">$</span><span class="va">userListId</span>,
                               gene_set_library <span class="op">=</span> <span class="st">"Table_Mining_of_CRISPR_Studies"</span><span class="op">)</span>

<span class="co">## As always, use str() to see what you have:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">enriched</span>, <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; 'data.frame':    46 obs. of  9 variables:</span>
<span class="co">#&gt;  $ Term                : chr  "Reaction polymerase chain quantitative time, PMC6627898 (Table1)" "Anticancer corresponding targets potential non, PMC5981615 (Table1)" "Names alphabetical order corresponding listed, PMC6783930 (Table1)" "Supplemental, PMC6879830 (Tablesupplemental)" ...</span>
<span class="co">#&gt;  $ Overlap             : chr  "4/6" "3/25" "3/31" "10/2992" ...</span>
<span class="co">#&gt;  $ P.value             : num  3.07e-12 7.77e-07 1.51e-06 8.05e-06 6.26e-05 ...</span>
<span class="co">#&gt;  $ Adjusted.P.value    : num  1.41e-10 1.79e-05 2.32e-05 9.26e-05 4.80e-04 ...</span>
<span class="co">#&gt;  $ Old.P.value         : int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ Old.Adjusted.P.value: int  0 0 0 0 0 0 0 0 0 0 ...</span>
<span class="co">#&gt;  $ Odds.Ratio          : num  3633.3 226.9 178.2 11.4 219.5 ...</span>
<span class="co">#&gt;  $ Combined.Score      : num  96318 3191 2388 134 2124 ...</span>
<span class="co">#&gt;  $ Genes               : chr  "CDC42;CDK2;CDK1;RHOA" "PLK1;CDK2;CDK1" "CDK2;CDK1;BRCA1" "CDC42;RBM3;RACGAP1;PLK1;CDK2;CDK1;PHF14;KIF23;BRCA1;MSL1" ...</span></code></pre></div>
</div>
</div>
<div id="see-also-in-functions-manuals" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#see-also-in-functions-manuals" class="anchor"></a>See also in Functions’ manuals</h1>
<p>Some rbioapi Enrichr functions were not covered in this vignette, be sure to check their manuals:</p>
<ul>
<li><p><code><a href="../reference/rba_enrichr_gene_map.html">rba_enrichr_gene_map()</a></code></p></li>
<li><p><code><a href="../reference/rba_enrichr_view_list.html">rba_enrichr_view_list()</a></code></p></li>
</ul>
<hr>
</div>
<div id="session-info" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<pre><code>#&gt; R version 4.0.5 (2021-03-31)
#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)
#&gt; Running under: macOS Catalina 10.15.7
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] rbioapi_0.7.1
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] knitr_1.33        magrittr_2.0.1    R6_2.5.0          ragg_1.1.2       
#&gt;  [5] rlang_0.4.11      fastmap_1.1.0     stringr_1.4.0     httr_1.4.2       
#&gt;  [9] tools_4.0.5       DT_0.18           xfun_0.22         crosstalk_1.1.1  
#&gt; [13] htmltools_0.5.1.1 systemfonts_1.0.2 yaml_2.2.1        digest_0.6.27    
#&gt; [17] rprojroot_2.0.2   pkgdown_1.6.1     crayon_1.4.1      textshaping_0.3.4
#&gt; [21] htmlwidgets_1.5.3 fs_1.5.0          curl_4.3.1        memoise_2.0.0    
#&gt; [25] cachem_1.0.4      evaluate_0.14     rmarkdown_2.8     stringi_1.6.1    
#&gt; [29] compiler_4.0.5    desc_1.3.0        jsonlite_1.7.2</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
