<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Connect to Reactome in R with rbioapi package. 
">
<title>2.E: Reactome &amp; rbioapi • rbioapi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2.E: Reactome &amp; rbioapi">
<meta property="og:description" content="Connect to Reactome in R with rbioapi package. 
">
<meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/rbioapi.html">Get Started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles-amp-supported-services">Articles &amp; Supported Services</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles-amp-supported-services">
    <a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://moosa-r.com/" aria-label="Moosa Rezwani Portfolio Website">
    <span class="fa fa-solid fa-id-card"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="rbioapi_reactome_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="rbioapi_reactome_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_reactome_files/datatables-binding-0.27/datatables.js"></script><link href="rbioapi_reactome_files/dt-core-1.12.1/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_reactome_files/dt-core-1.12.1/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_reactome_files/dt-core-1.12.1/js/jquery.dataTables.min.js"></script><link href="rbioapi_reactome_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="rbioapi_reactome_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>2.E: Reactome &amp; rbioapi</h1>
                        <h4 data-toc-skip class="author">Moosa
Rezwani</h4>
            
            <h4 data-toc-skip class="date">2023-04-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/HEAD/vignettes/rbioapi_reactome.Rmd" class="external-link"><code>vignettes/rbioapi_reactome.Rmd</code></a></small>
      <div class="d-none name"><code>rbioapi_reactome.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level3">
<h3 class="heading2" id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>Directly quoting from <a href="https://reactome.org/" class="external-link">Reactome</a>:</p>
<blockquote>
<p>REACTOME is an open-source, open access, manually curated and
peer-reviewed pathway database. Our goal is to provide intuitive
bioinformatics tools for the visualization, interpretation and analysis
of pathway knowledge to support basic and clinical research, genome
analysis, modeling, systems biology and education. Founded in 2003, the
Reactome project is led by Lincoln Stein of <a href="https://oicr.on.ca/" class="external-link">OICR</a>, Peter D’Eustachio of <a href="https://nyulangone.org/" class="external-link">NYULMC</a>, Henning Hermjakob of <a href="https://www.ebi.ac.uk/" class="external-link">EMBL-EBI</a>, and Guanming Wu of <a href="https://www.ohsu.edu/" class="external-link">OHSU</a>.</p>
<p>(source: <a href="https://reactome.org/what-is-reactome" class="external-link uri">https://reactome.org/what-is-reactome</a>)</p>
</blockquote>
<p>Reactome provides two RESTful API services: Reactome content services
and Reactome analysis services. In rbioapi, the naming schema is that
any function which belongs to analysis services starts with
rba_reactome_analysis* . Other rba_reactome_* functions without the
‘analysis’ infix correspond to content services API.</p>
<p>Before continuing reading this article, it is a good idea to read <a href="https://reactome.org/documentation/data-model" class="external-link">Reactome Data
Model</a> page.</p>
<hr>
</div>
<div class="section level3">
<h3 class="heading2" id="reactome-analysis-services">Reactome analysis services<a class="anchor" aria-label="anchor" href="#reactome-analysis-services"></a>
</h3>
<p>This section mostly revolves around
<code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code> function. So, naturally, we will
start with that. As explained in the function’s manual, you have
considerable freedom in providing the main input for this function; You
can supply an R object (as a data frame, matrix, or simple vector), a
URL, or a local file path. Note that the type of analysis will be
decided based on whether your input is 1-dimensional or 2-dimensional.
This has been explained in detail in the manual of
<code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code>, see that for more
information.<br><code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code> is the API equivalent of Reactome’s
<a href="https://reactome.org/PathwayBrowser/#TOOL=AT" class="external-link">analyse gene
list</a> tool. You can see that the function’s arguments correspond to
what would you choose in the webpage’s wizard.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1 We create a simple vector with our genes</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>, <span class="st">"CDK1"</span>, <span class="st">"KIF23"</span>, <span class="st">"PLK1"</span>, <span class="st">"RAC2"</span>, <span class="st">"RACGAP1"</span>, <span class="st">"RHOA"</span>, <span class="st">"RHOB"</span>, <span class="st">"MSL1"</span>, <span class="st">"PHF21A"</span>, <span class="st">"INSR"</span>, <span class="st">"JADE2"</span>, <span class="st">"P2RX7"</span>, <span class="st">"CCDC101"</span>, <span class="st">"PPM1B"</span>, <span class="st">"ANAPC16"</span>, <span class="st">"CDH8"</span>, <span class="st">"HSPA1L"</span>, <span class="st">"CUL2"</span>, <span class="st">"ZNF302"</span>, <span class="st">"CUX1"</span>, <span class="st">"CYTH2"</span>, <span class="st">"SEC22C"</span>, <span class="st">"EIF4E3"</span>, <span class="st">"ROBO2"</span>, <span class="st">"CXXC1"</span>, <span class="st">"LINC01314"</span>, <span class="st">"ATP5F1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 2 We call reactome analysis with the default parameters</span></span>
<span><span class="va">analyzed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">genes</span>,</span>
<span>                                  projection <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                  p_value <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## 3 As always, we use str() to inspect the resutls</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">analyzed</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 8</span></span>
<span><span class="co">#&gt;  $ summary            :List of 7</span></span>
<span><span class="co">#&gt;  $ expression         :List of 1</span></span>
<span><span class="co">#&gt;  $ identifiersNotFound: int 1</span></span>
<span><span class="co">#&gt;  $ pathwaysFound      : int 73</span></span>
<span><span class="co">#&gt;  $ pathways           :'data.frame': 73 obs. of  19 variables:</span></span>
<span><span class="co">#&gt;  $ resourceSummary    :'data.frame': 3 obs. of  3 variables:</span></span>
<span><span class="co">#&gt;  $ speciesSummary     :'data.frame': 1 obs. of  5 variables:</span></span>
<span><span class="co">#&gt;  $ warnings           : chr "Missing header. Using a default one."</span></span>
<span></span>
<span><span class="co">## 4 Note that in the summary element: (analyzed$summary)</span></span>
<span><span class="co">### 4.a because we supplied a simple vector, the analysis type was: over-representation</span></span>
<span><span class="co">### 4.b You need the token for other rba_reactome_analysis_* functions</span></span>
<span></span>
<span><span class="co">## 5 Analsis results are in the pathways data frame:</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-32d262f8bedafd695a37" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-32d262f8bedafd695a37">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73"],["R-HSA-3371568","R-HSA-3371571","R-HSA-3371511","R-HSA-114604","R-HSA-6806003","R-HSA-3371453","R-HSA-983231","R-HSA-176412","R-HSA-109582","R-HSA-2262752","R-HSA-3371556","R-HSA-8953897","R-HSA-6804754","R-HSA-176408","R-HSA-69273","R-HSA-6796648","R-HSA-174143","R-HSA-453276","R-HSA-1538133","R-HSA-69620","R-HSA-5633007","R-HSA-176417","R-HSA-6804757","R-HSA-2980767","R-HSA-5689896","R-HSA-5666185","R-HSA-1640170","R-HSA-8853884","R-HSA-6811438","R-HSA-69278","R-HSA-69275","R-HSA-453274","R-HSA-6811442","R-HSA-5625970","R-HSA-3700989","R-HSA-376176","R-HSA-68884","R-HSA-6791312","R-HSA-5663220","R-HSA-162658","R-HSA-69481","R-HSA-6804116","R-HSA-1362300","R-HSA-176814","R-HSA-6804114","R-HSA-69473","R-HSA-8852276","R-HSA-76002","R-HSA-5625900","R-HSA-156711","R-HSA-422475","R-HSA-174048","R-HSA-5627117","R-HSA-416572","R-HSA-9723905","R-HSA-9723907","R-HSA-9013404","R-HSA-9675108","R-HSA-6804756","R-HSA-195258","R-HSA-400685","R-HSA-3214847","R-HSA-9687139","R-HSA-9699150","R-HSA-9663199","R-HSA-9675126","R-HSA-397795","R-HSA-9772755","R-HSA-4420097","R-HSA-69205","R-HSA-3247509","R-HSA-4839726","R-HSA-8953750"],[3371568,3371571,3371511,114604,6806003,3371453,983231,176412,109582,2262752,3371556,8953897,6804754,176408,69273,6796648,174143,453276,1538133,69620,5633007,176417,6804757,2980767,5689896,5666185,1640170,8853884,6811438,69278,69275,453274,6811442,5625970,3700989,376176,68884,6791312,5663220,162658,69481,6804116,1362300,176814,6804114,69473,8852276,76002,5625900,156711,422475,174048,5627117,416572,9723905,9723907,9013404,9675108,6804756,195258,400685,3214847,9687139,9699150,9663199,9675126,397795,9772755,4420097,69205,3247509,4839726,8953750],["Attenuation phase","HSF1-dependent transactivation","HSF1 activation","GPVI-mediated activation cascade","Regulation of TP53 Expression and Degradation","Regulation of HSF1-mediated heat shock response","Factors involved in megakaryocyte development and platelet production","Phosphorylation of the APC/C","Hemostasis","Cellular responses to stress","Cellular response to heat stress","Cellular responses to stimuli","Regulation of TP53 Expression","Regulation of APC/C activators between G1/S and early anaphase","Cyclin A/B1/B2 associated events during G2/M transition","TP53 Regulates Transcription of DNA Repair Genes","APC/C-mediated degradation of cell cycle proteins","Regulation of mitotic cell cycle","G0 and Early G1","Cell Cycle Checkpoints","Regulation of TP53 Activity","Phosphorylation of Emi1","Regulation of TP53 Degradation","Activation of NIMA Kinases NEK9, NEK6, NEK7","Ovarian tumor domain proteases","RHO GTPases Activate Rhotekin and Rhophilins","Cell Cycle","Transcriptional Regulation by VENTX","Intra-Golgi traffic","Cell Cycle, Mitotic","G2/M Transition","Mitotic G2-G2/M phases","Intra-Golgi and retrograde Golgi-to-ER traffic","RHO GTPases activate KTN1","Transcriptional Regulation by TP53","Signaling by ROBO receptors","Mitotic Telophase/Cytokinesis","TP53 Regulates Transcription of Cell Cycle Genes","RHO GTPases Activate Formins","Golgi Cisternae Pericentriolar Stack Reorganization","G2/M Checkpoints","TP53 Regulates Transcription of Genes Involved in G1 Cell Cycle Arrest","Transcription of E2F targets under negative control by p107 (RBL1) and p130 (RBL2) in complex with HDAC1","Activation of APC/C and APC/C:Cdc20 mediated degradation of mitotic proteins","TP53 Regulates Transcription of Genes Involved in G2 Cell Cycle Arrest","G2/M DNA damage checkpoint","The role of GTSE1 in G2/M progression after G2 checkpoint","Platelet activation, signaling and aggregation","RHO GTPases activate CIT","Polo-like kinase mediated events","Axon guidance","APC/C:Cdc20 mediated degradation of Cyclin B","RHO GTPases Activate ROCKs","Sema4D induced cell migration and growth-cone collapse","Loss of function of TP53 in cancer due to loss of tetramerization ability","Loss of Function of TP53 in Cancer","RAC2 GTPase cycle","Nervous system development","Regulation of TP53 Activity through Phosphorylation","RHO GTPase Effectors","Sema4D in semaphorin signaling","HATs acetylate histones","Aberrant regulation of mitotic cell cycle due to RB1 defects","Defective DNA double strand break response due to BARD1 loss of function","Defective DNA double strand break response due to BRCA1 loss of function","Diseases of mitotic cell cycle","G-protein beta:gamma signalling","Formation of WDR5-containing histone-modifying complexes","VEGFA-VEGFR2 Pathway","G1/S-Specific Transcription","Chromatin modifying enzymes","Chromatin organization","Transcriptional Regulation by E2F6"],[true,true,true,true,false,true,true,true,false,false,false,false,true,true,true,true,false,false,true,false,false,true,true,true,true,true,false,true,true,false,false,false,false,true,false,true,true,false,true,true,true,true,true,false,true,true,true,false,true,true,false,true,true,true,true,false,true,false,true,false,false,true,false,true,true,false,false,true,true,true,true,false,true],[false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,true,true,false,false,false,false,false,false,true,true,true,true,false,false,false,false,false,false,false],[48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887],["9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL"],[47,59,43,43,46,113,194,20,803,1007,135,1025,4,83,32,86,92,92,38,279,178,8,43,9,44,10,733,48,49,595,212,214,219,12,484,235,15,65,149,17,153,20,20,77,21,80,83,293,23,23,585,24,24,25,1,1,92,621,95,325,31,110,36,2,2,38,39,42,129,43,254,254,46],[5,5,4,4,4,5,6,3,11,12,5,12,2,4,3,4,4,4,3,6,5,2,3,2,3,2,9,3,3,8,5,5,5,2,7,5,2,3,4,2,4,2,2,3,2,3,3,5,2,2,7,2,2,2,1,1,3,7,3,5,2,3,2,1,1,2,2,2,3,2,4,4,2],[0.00309760759243393,0.00388848612667238,0.00283398141435445,0.00283398141435445,0.00303170104791406,0.0074474395307454,0.0127858696368549,0.00131813089039742,0.0529229552494563,0.0663678903315099,0.00889738351018256,0.0675542081328676,0.000263626178079483,0.00547024319514928,0.00210900942463587,0.00566796282870889,0.00606340209582812,0.00606340209582812,0.00250444869175509,0.018387925921044,0.011731364924537,0.000527252356158967,0.00283398141435445,0.000593158900678837,0.00289988795887432,0.000659065445198708,0.0483094971330653,0.0031635141369538,0.00322942068147367,0.0392143939893231,0.0139721874382126,0.0141040005272524,0.0144335332498517,0.00079087853423845,0.0318987675476175,0.0154880379621696,0.000988598167798062,0.0042839253937916,0.00982007513346075,0.0011204112568378,0.0100837013115402,0.00131813089039742,0.00131813089039742,0.00507480392803005,0.00138403743491729,0.00527252356158967,0.00547024319514928,0.0193106175443222,0.00151585052395703,0.00151585052395703,0.0385553285441244,0.0015817570684769,0.0015817570684769,0.00164766361299677,6.59065445198708e-05,6.59065445198708e-05,0.00606340209582812,0.0409279641468398,0.00626112172938773,0.021419626968958,0.002043102880116,0.00724971989718579,0.00237263560271535,0.000131813089039742,0.000131813089039742,0.00250444869175509,0.00257035523627496,0.00276807486983457,0.00850194424306334,0.00283398141435445,0.0167402623080472,0.0167402623080472,0.00303170104791406],[4.85450536436005e-07,1.47004602502054e-06,1.23425368514507e-05,1.23425368514507e-05,1.60500701897881e-05,3.32553498518795e-05,3.81283311712277e-05,4.03208281644085e-05,4.0723697813716e-05,6.61800226030573e-05,7.67527036369398e-05,7.84567203016007e-05,8.10584865421626e-05,0.00015586418859237,0.000160720192474706,0.000178381083232426,0.000230329592718848,0.000230329592718848,0.000265501680787539,0.000274364004015704,0.000275747901407497,0.000321570046249997,0.000380368575844559,0.00040614899227398,0.000406606348955951,0.000500386168572176,0.000506395320263975,0.00052312841919322,0.000555252110297388,0.000597360501357946,0.000608876940645042,0.000635079464878952,0.000704291851855832,0.00071759305318353,0.000889788606804776,0.000964159467996772,0.00111433385363291,0.00125006947777606,0.00138532762727595,0.00142542042567939,0.00152574618691925,0.00196076272317847,0.00196076272317847,0.00202256701581216,0.0021573005708001,0.00225300550619278,0.00249913130735768,0.00252688745395324,0.00257716573626299,0.00257716573626299,0.0026209237059277,0.00280037859679816,0.00280037859679816,0.00303236910954863,0.00322431779284615,0.00322431779284615,0.00333512088194121,0.00364629049006104,0.00364741587678552,0.00392909750318093,0.00460549301411128,0.00547443176742202,0.00614759510860741,0.00643845085796158,0.00643845085796158,0.00682162434271916,0.0071706790643129,0.0082653681832906,0.00846066702541692,0.008645931124037,0.00914270728856292,0.00914270728856292,0.00983389112705291],[0.000237385312317206,0.000358691230105013,0.00150578949587699,0.00150578949587699,0.00155685680840945,0.00219907968194066,0.00219907968194066,0.00219907968194066,0.00219907968194066,0.00299916400206002,0.00299916400206002,0.00299916400206002,0.00299916400206002,0.00514304615919059,0.00514304615919059,0.00535143249697279,0.00621889900340888,0.00621889900340888,0.00634220173237243,0.00634220173237243,0.00634220173237243,0.00707454101749994,0.00772552063016307,0.00772552063016307,0.00772552063016307,0.00888403376475821,0.00888403376475821,0.00888403376475821,0.00888403376475821,0.00913315410967563,0.00913315410967563,0.00952619197318427,0.00986008592598164,0.0100463027445694,0.0115672518884621,0.012534073083958,0.0144863400972278,0.0150008337333127,0.0166239315273113,0.0167832080561118,0.0167832080561118,0.0215683899549631,0.0215683899549631,0.021573005708001,0.021573005708001,0.0225300550619278,0.0231944916263669,0.0231944916263669,0.0231944916263669,0.0231944916263669,0.0235883133533493,0.0252034073711834,0.0252034073711834,0.0257945423427692,0.0257945423427692,0.0257945423427692,0.0266809670555297,0.0291703239204884,0.0291793270142842,0.0314327800254475,0.0368439441128903,0.0383210223719541,0.0430331657602518,0.045069156005731,0.045069156005731,0.0477513703990341,0.0501947534501903,0.051875586744222,0.051875586744222,0.051875586744222,0.0548562437313775,0.0548562437313775,0.0590033467623174],[[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]],["TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL"],[5,8,7,25,35,14,43,2,338,450,29,481,5,14,25,17,40,40,27,56,98,2,30,4,14,6,451,13,10,352,78,80,48,2,259,60,4,42,27,6,24,17,8,15,11,12,10,117,6,15,298,3,7,7,1,1,10,324,26,113,13,15,3,1,1,5,15,30,79,28,85,85,33],[3,4,1,3,17,7,7,2,34,48,12,48,5,7,19,15,32,32,6,27,64,2,12,2,2,6,183,3,4,156,49,51,6,2,172,18,2,27,10,4,11,10,2,15,7,5,6,24,5,7,34,3,4,4,1,1,10,34,23,48,5,3,2,1,1,2,5,7,12,2,5,5,2],[0.000352211890673429,0.000563539025077487,0.000493096646942801,0.00176105945336715,0.002465483234714,0.000986193293885602,0.00302902225979149,0.000140884756269372,0.0238095238095238,0.0316990701606086,0.00204282896590589,0.0338827838827839,0.000352211890673429,0.000986193293885602,0.00176105945336715,0.00119752042828966,0.00281769512538743,0.00281769512538743,0.00190194420963652,0.00394477317554241,0.00690335305719921,0.000140884756269372,0.00211327134404058,0.000281769512538743,0.000986193293885602,0.000422654268808115,0.0317695125387433,0.000915750915750916,0.000704423781346858,0.0247957171034094,0.00549450549450549,0.00563539025077487,0.00338123415046492,0.000140884756269372,0.0182445759368836,0.00422654268808115,0.000281769512538743,0.0029585798816568,0.00190194420963652,0.000422654268808115,0.00169061707523246,0.00119752042828966,0.000563539025077487,0.00105663567202029,0.000774866159481544,0.00084530853761623,0.000704423781346858,0.00824175824175824,0.000422654268808115,0.00105663567202029,0.0209918286841364,0.000211327134404057,0.000493096646942801,0.000493096646942801,7.04423781346858e-05,7.04423781346858e-05,0.000704423781346858,0.0228233305156382,0.00183150183150183,0.0079599887292195,0.000915750915750916,0.00105663567202029,0.000211327134404057,7.04423781346858e-05,7.04423781346858e-05,0.000352211890673429,0.00105663567202029,0.00211327134404058,0.00556494787264018,0.0019723865877712,0.0059876021414483,0.0059876021414483,0.00232459847844463]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stId<\/th>\n      <th>dbId<\/th>\n      <th>name<\/th>\n      <th>llp<\/th>\n      <th>inDisease<\/th>\n      <th>species.dbId<\/th>\n      <th>species.taxId<\/th>\n      <th>species.name<\/th>\n      <th>entities.resource<\/th>\n      <th>entities.total<\/th>\n      <th>entities.found<\/th>\n      <th>entities.ratio<\/th>\n      <th>entities.pValue<\/th>\n      <th>entities.fdr<\/th>\n      <th>entities.exp<\/th>\n      <th>reactions.resource<\/th>\n      <th>reactions.total<\/th>\n      <th>reactions.found<\/th>\n      <th>reactions.ratio<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[2,6,10,11,12,13,14,17,18,19]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p>As mentioned, some of <code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code>’s
arguments correspond to the wizard of <a href="https://reactome.org/PathwayBrowser/#TOOL=AT" class="external-link">analyse gene
list</a> tool; Other arguments corresponds to the contents of “Filter
your results” tab in the results page.</p>
<p>Having the analysis’s token, you can retrieve the analysis results in
many formats using <code><a href="../reference/rba_reactome_analysis_pdf.html">rba_reactome_analysis_pdf()</a></code> and
<code><a href="../reference/rba_reactome_analysis_download.html">rba_reactome_analysis_download()</a></code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download a full pdf report</span></span>
<span><span class="fu"><a href="../reference/rba_reactome_analysis_pdf.html">rba_reactome_analysis_pdf</a></span><span class="op">(</span>token <span class="op">=</span> <span class="va">analyzed</span><span class="op">$</span><span class="va">summary</span><span class="op">$</span><span class="va">token</span>,</span>
<span>                          species <span class="op">=</span> <span class="fl">9606</span><span class="op">)</span></span>
<span><span class="co"># download the result in compressed json.gz format</span></span>
<span><span class="fu"><a href="../reference/rba_reactome_analysis_download.html">rba_reactome_analysis_download</a></span><span class="op">(</span>token <span class="op">=</span> <span class="va">analyzed</span><span class="op">$</span><span class="va">summary</span><span class="op">$</span><span class="va">token</span>,</span>
<span>                               request <span class="op">=</span> <span class="st">"results"</span>,</span>
<span>                               save_to <span class="op">=</span> <span class="st">"reactome_results.json"</span><span class="op">)</span></span></code></pre></div>
<p>Your token is only guaranteed to be stored for 7 days. After that,
you can upload the JSON file you have downloaded using
<code>rba_reactome_analysis_download</code> and get a token for
that:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">re_uploaded</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_analysis_import.html">rba_reactome_analysis_import</a></span><span class="op">(</span>input <span class="op">=</span> <span class="st">"reactome_results.json"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Please Note:</strong> Other services supported by rbioapi
also provide Over-representation analysis tools. Please see the vignette
article <a href="rbioapi_do_enrich.html">Do with rbioapi:
Over-Representation (Enrichment) Analysis in R</a> (<a href="https://rbioapi.moosa-r.com/articles/rbioapi_do_enrich.html">link
to the documentation site</a>) for an in-depth review.</p>
<div class="section level4">
<h4 class="heading3" id="analysis_see-also-in-functions-manuals">See also in Functions’ manuals<a class="anchor" aria-label="anchor" href="#analysis_see-also-in-functions-manuals"></a>
</h4>
<p>Some rbioapi Reactome analysis functions were not covered in this
vignette, be sure to check their manuals:</p>
<ul>
<li><p><code><a href="../reference/rba_reactome_analysis_mapping.html">rba_reactome_analysis_mapping()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_analysis_species.html">rba_reactome_analysis_species()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_analysis_token.html">rba_reactome_analysis_token()</a></code></p></li>
</ul>
<hr>
</div>
</div>
<div class="section level3">
<h3 class="heading2" id="reactome-contents-services">Reactome contents services<a class="anchor" aria-label="anchor" href="#reactome-contents-services"></a>
</h3>
<p>rbioapi functions that correspond to Reactome content services are
those starting with rba_reactome_* but without “_analysis” infix. These
functions cover what you can do with objects in Reactome knowledge-base.
In simpler terms, most -but not all of them- correspond to what you can
find in <a href="https://reactome.org/PathwayBrowser/" title="Reactome Pathway Browser" class="external-link">Reactome Pathway Browser</a> and <a href="https://reactome.org/content/query?q=apoptosis&amp;species=Homo+sapiens" title="Search results for apoptosis" class="external-link">search results</a>. (e.g. a <a href="https://reactome.org/content/detail/R-HSA-109581" title="Apoptosis Homo sapiens" class="external-link">pathway</a>, a <a href="https://reactome.org/content/detail/R-HSA-202939" title="Caspase-mediated cleavage of E-Cadherin" class="external-link">reaction</a>, a <a href="https://reactome.org/content/detail/R-HSA-350870" title="Caspase-3 [cytosol]" class="external-link">physical Entity</a>, etc.)</p>
<div class="section level4">
<h4 class="heading3" id="retrieve-any-object-from-reactome-knowledge-base">Retrieve any object from Reactome
knowledge-base<a class="anchor" aria-label="anchor" href="#retrieve-any-object-from-reactome-knowledge-base"></a>
</h4>
<p>Using <code><a href="../reference/rba_reactome_query.html">rba_reactome_query()</a></code>, you can retrieve any object
from Reactome knowledge-base. In simpler terms, what I mean by the
object is roughly anything that Reactome associated an ID to it. This
can range from a person’s entry to proteins, reactions, pathways,
species, and many more! You can explore <a href="https://reactome.org/content/schema/DatabaseObject" title="Graph Database :: Data Schema" class="external-link">Reactome’s data schema</a> to
learn about Reactome knowledge-base objects and their organization. Here
are some examples, note that you are not limited to only one ID per
query. You can use a vector of inputs, the only limitation is that when
you supply more than one ID, you cannot have
<code>enhanced = TRUE</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1 query a pathway Entry</span></span>
<span><span class="va">pathway</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_query.html">rba_reactome_query</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="st">"R-HSA-109581"</span>, enhanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## 2 As always we use str() to inspect the output's structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">pathway</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 26</span></span>
<span><span class="co">#&gt;  $ dbId               : int 109581</span></span>
<span><span class="co">#&gt;  $ displayName        : chr "Apoptosis"</span></span>
<span><span class="co">#&gt;  $ stId               : chr "R-HSA-109581"</span></span>
<span><span class="co">#&gt;  $ stIdVersion        : chr "R-HSA-109581.4"</span></span>
<span><span class="co">#&gt;  $ created            :List of 5</span></span>
<span><span class="co">#&gt;   ..$ dbId       : int 109608</span></span>
<span><span class="co">#&gt;   ..$ displayName: chr "Alnemri, E, Hengartner, Michael, Tschopp, Jürg, Tsujimoto, Yoshihide, Hardwick, JM, 2004-01-16"</span></span>
<span><span class="co">#&gt;   ..$ dateTime   : chr "2004-01-16 21:01:51"</span></span>
<span><span class="co">#&gt;   ..$ className  : chr "InstanceEdit"</span></span>
<span><span class="co">#&gt;   ..$ schemaClass: chr "InstanceEdit"</span></span>
<span><span class="co">#&gt;  $ modified           :List of 6</span></span>
<span><span class="co">#&gt;   ..$ dbId       : int 10943396</span></span>
<span><span class="co">#&gt;   ..$ displayName: chr "Wright, Adam, 2023-03-09"</span></span>
<span><span class="co">#&gt;   ..$ dateTime   : chr "2023-03-09 11:19:01"</span></span>
<span><span class="co">#&gt;   ..$ note       : chr "Inserted by org.reactome.orthoinference"</span></span>
<span><span class="co">#&gt;   ..$ className  : chr "InstanceEdit"</span></span>
<span><span class="co">#&gt;   ..$ schemaClass: chr "InstanceEdit"</span></span>
<span><span class="co">#&gt;  $ isInDisease        : logi FALSE</span></span>
<span><span class="co">#&gt;  $ isInferred         : logi FALSE</span></span>
<span><span class="co">#&gt;  $ name               :List of 1</span></span>
<span><span class="co">#&gt;   ..$ : chr "Apoptosis"</span></span>
<span><span class="co">#&gt;  $ releaseDate        : chr "2004-09-20"</span></span>
<span><span class="co">#&gt;  $ speciesName        : chr "Homo sapiens"</span></span>
<span><span class="co">#&gt;  $ authored           :List of 1</span></span>
<span><span class="co">#&gt;   ..$ : int 109608</span></span>
<span><span class="co">#&gt;  $ edited             :List of 1</span></span>
<span><span class="co">#&gt;   ..$ :List of 5</span></span>
<span><span class="co">#&gt;  $ figure             :List of 1</span></span>
<span><span class="co">#&gt;   ..$ :List of 5</span></span>
<span><span class="co">#&gt;  $ goBiologicalProcess:List of 9</span></span>
<span><span class="co">#&gt;   ..$ dbId        : int 2273</span></span>
<span><span class="co">#&gt;   ..$ displayName : chr "apoptotic process"</span></span>
<span><span class="co">#&gt;   ..$ accession   : chr "0006915"</span></span>
<span><span class="co">#&gt;   ..$ databaseName: chr "GO"</span></span>
<span><span class="co">#&gt;   ..$ definition  : chr "A programmed cell death process which begins when a cell receives an internal (e.g. DNA damage) or external sig"| __truncated__</span></span>
<span><span class="co">#&gt;   ..$ name        : chr "apoptotic process"</span></span>
<span><span class="co">#&gt;   ..$ url         : chr "https://www.ebi.ac.uk/QuickGO/term/GO:0006915"</span></span>
<span><span class="co">#&gt;   ..$ className   : chr "GO_BiologicalProcess"</span></span>
<span><span class="co">#&gt;   ..$ schemaClass : chr "GO_BiologicalProcess"</span></span>
<span><span class="co">#&gt;  $ literatureReference:List of 7</span></span>
<span><span class="co">#&gt;   ..$ :List of 11</span></span>
<span><span class="co">#&gt;   ..$ :List of 11</span></span>
<span><span class="co">#&gt;   ..$ :List of 11</span></span>
<span><span class="co">#&gt;   ..$ :List of 11</span></span>
<span><span class="co">#&gt;   ..$ :List of 11</span></span>
<span><span class="co">#&gt;   ..$ :List of 11</span></span>
<span><span class="co">#&gt;   ..$ :List of 11</span></span>
<span><span class="co">#&gt;  $ orthologousEvent   :List of 14</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;  $ reviewed           :List of 1</span></span>
<span><span class="co">#&gt;   ..$ :List of 5</span></span>
<span><span class="co">#&gt;  $ species            :List of 1</span></span>
<span><span class="co">#&gt;   ..$ :List of 8</span></span>
<span><span class="co">#&gt;  $ summation          :List of 1</span></span>
<span><span class="co">#&gt;   ..$ :List of 5</span></span>
<span><span class="co">#&gt;  $ reviewStatus       :List of 6</span></span>
<span><span class="co">#&gt;   ..$ dbId       : int 9821382</span></span>
<span><span class="co">#&gt;   ..$ displayName: chr "five stars"</span></span>
<span><span class="co">#&gt;   ..$ definition : chr "externally reviewed"</span></span>
<span><span class="co">#&gt;   ..$ name       :List of 1</span></span>
<span><span class="co">#&gt;   ..$ className  : chr "ReviewStatus"</span></span>
<span><span class="co">#&gt;   ..$ schemaClass: chr "ReviewStatus"</span></span>
<span><span class="co">#&gt;  $ hasDiagram         : logi TRUE</span></span>
<span><span class="co">#&gt;  $ hasEHLD            : logi TRUE</span></span>
<span><span class="co">#&gt;  $ hasEvent           :List of 4</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;   ..$ :List of 16</span></span>
<span><span class="co">#&gt;   ..$ :List of 16</span></span>
<span><span class="co">#&gt;   ..$ :List of 15</span></span>
<span><span class="co">#&gt;  $ className          : chr "Pathway"</span></span>
<span><span class="co">#&gt;  $ schemaClass        : chr "Pathway"</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">## 3 You can compare it with the webpage of R-HSA-202939 entry:</span></span>
<span><span class="co"># https://reactome.org/content/detail/R-HSA-202939</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1 query a protein Entry</span></span>
<span><span class="va">protein</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_query.html">rba_reactome_query</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="fl">66247</span>, enhanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## 2 As always we use str() to inspect the output's structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">protein</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 27</span></span>
<span><span class="co">#&gt;  $ dbId               : int 66247</span></span>
<span><span class="co">#&gt;  $ displayName        : chr "UniProt:P25942-1 CD40"</span></span>
<span><span class="co">#&gt;  $ modified           :List of 6</span></span>
<span><span class="co">#&gt;  $ databaseName       : chr "UniProt"</span></span>
<span><span class="co">#&gt;  $ identifier         : chr "P25942"</span></span>
<span><span class="co">#&gt;  $ name               :List of 1</span></span>
<span><span class="co">#&gt;  $ otherIdentifier    :List of 99</span></span>
<span><span class="co">#&gt;  $ url                : chr "https://purl.uniprot.org/uniprot/P25942-1"</span></span>
<span><span class="co">#&gt;  $ crossReference     :List of 30</span></span>
<span><span class="co">#&gt;  $ referenceDatabase  :List of 8</span></span>
<span><span class="co">#&gt;  $ physicalEntity     :List of 1</span></span>
<span><span class="co">#&gt;  $ checksum           : chr "BC8776EC2C4A5680"</span></span>
<span><span class="co">#&gt;  $ comment            :List of 1</span></span>
<span><span class="co">#&gt;  $ description        :List of 1</span></span>
<span><span class="co">#&gt;  $ geneName           :List of 2</span></span>
<span><span class="co">#&gt;  $ isSequenceChanged  : logi FALSE</span></span>
<span><span class="co">#&gt;  $ keyword            :List of 16</span></span>
<span><span class="co">#&gt;  $ secondaryIdentifier:List of 8</span></span>
<span><span class="co">#&gt;  $ sequenceLength     : int 277</span></span>
<span><span class="co">#&gt;  $ species            : int 48887</span></span>
<span><span class="co">#&gt;  $ chain              :List of 2</span></span>
<span><span class="co">#&gt;  $ referenceGene      :List of 11</span></span>
<span><span class="co">#&gt;  $ referenceTranscript:List of 4</span></span>
<span><span class="co">#&gt;  $ variantIdentifier  : chr "P25942-1"</span></span>
<span><span class="co">#&gt;  $ isoformParent      :List of 1</span></span>
<span><span class="co">#&gt;  $ className          : chr "ReferenceIsoform"</span></span>
<span><span class="co">#&gt;  $ schemaClass        : chr "ReferenceIsoform"</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">## 3 You can compare it with the webpage of R-HSA-202939 entry:</span></span>
<span><span class="co"># https://reactome.org/content/detail/R-HSA-202939</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 class="heading3" id="find-cross-reference-ids-in-reactome">Find Cross-Reference IDs in Reactome<a class="anchor" aria-label="anchor" href="#find-cross-reference-ids-in-reactome"></a>
</h4>
<p>As you can see in the second example usage of we used Reactome’s dbID
<code>66247</code> to query CD40 protein. How did we obtain that in the
first place? You can use <code>rba_reactome_xref</code> to map any
cross-reference (external) IDs to Reactome IDs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1 We Supply HGNC ID to find what is the corresponding database ID in Reactome</span></span>
<span><span class="va">xref_protein</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_xref.html">rba_reactome_xref</a></span><span class="op">(</span><span class="st">"CD40"</span><span class="op">)</span></span>
<span><span class="co">## 2 As always use str() to inspect the output's structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">xref_protein</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 19</span></span>
<span><span class="co">#&gt;  $ dbId               : int 66247</span></span>
<span><span class="co">#&gt;  $ displayName        : chr "UniProt:P25942-1 CD40"</span></span>
<span><span class="co">#&gt;  $ databaseName       : chr "UniProt"</span></span>
<span><span class="co">#&gt;  $ identifier         : chr "P25942"</span></span>
<span><span class="co">#&gt;  $ name               :List of 1</span></span>
<span><span class="co">#&gt;  $ otherIdentifier    :List of 1</span></span>
<span><span class="co">#&gt;  $ url                : chr "https://purl.uniprot.org/uniprot/P25942-1"</span></span>
<span><span class="co">#&gt;  $ checksum           : chr "BC8776EC2C4A5680"</span></span>
<span><span class="co">#&gt;  $ comment            :List of 1</span></span>
<span><span class="co">#&gt;  $ description        :List of 1</span></span>
<span><span class="co">#&gt;  $ geneName           :List of 1</span></span>
<span><span class="co">#&gt;  $ isSequenceChanged  : logi FALSE</span></span>
<span><span class="co">#&gt;  $ keyword            :List of 1</span></span>
<span><span class="co">#&gt;  $ secondaryIdentifier:List of 1</span></span>
<span><span class="co">#&gt;  $ sequenceLength     : int 277</span></span>
<span><span class="co">#&gt;  $ chain              :List of 1</span></span>
<span><span class="co">#&gt;  $ variantIdentifier  : chr "P25942-1"</span></span>
<span><span class="co">#&gt;  $ className          : chr "ReferenceIsoform"</span></span>
<span><span class="co">#&gt;  $ schemaClass        : chr "ReferenceIsoform"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 class="heading3" id="map-cross-reference-ids-to-reactome">Map Cross-Reference IDs to Reactome<a class="anchor" aria-label="anchor" href="#map-cross-reference-ids-to-reactome"></a>
</h4>
<p>While we are at the cross-reference topic, here is another useful
resource. Using <code>rba_reactome_mapping</code> you can find the
Reactome pathways or reactions which include your external ID:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1 Again, consider CD40 protein:</span></span>
<span><span class="va">xref_mapping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_mapping.html">rba_reactome_mapping</a></span><span class="op">(</span>id <span class="op">=</span> <span class="st">"CD40"</span>,</span>
<span>                                    resource <span class="op">=</span> <span class="st">"hgnc"</span>,</span>
<span>                                    map_to <span class="op">=</span> <span class="st">"pathways"</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-5f90f9aa490164fedfec" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-5f90f9aa490164fedfec">{"x":{"filter":"none","vertical":false,"data":[["1","2","3"],[198933,5668541,5676594],["Immunoregulatory interactions between a Lymphoid and a non-Lymphoid cell","TNFR2 non-canonical NF-kB pathway","TNF receptor superfamily (TNFSF) members mediating non-canonical NF-kB pathway"],["R-HSA-198933","R-HSA-5668541","R-HSA-5676594"],["R-HSA-198933.7","R-HSA-5668541.3","R-HSA-5676594.2"],[false,false,false],[false,false,false],[["Immunoregulatory interactions between a Lymphoid and a non-Lymphoid cell"],["TNFR2 non-canonical NF-kB pathway","TNFRSF mediated non-canonical NF-kB pathway"],["TNF receptor superfamily (TNFSF) members mediating non-canonical NF-kB pathway"]],["2007-08-21","2015-06-17","2015-06-17"],["Homo sapiens","Homo sapiens","Homo sapiens"],["10.3180/r-hsa-198933.2","10.3180/r-hsa-5668541.1",null],[true,true,false],[false,false,false],["Pathway","Pathway","Pathway"],["Pathway","Pathway","Pathway"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>dbId<\/th>\n      <th>displayName<\/th>\n      <th>stId<\/th>\n      <th>stIdVersion<\/th>\n      <th>isInDisease<\/th>\n      <th>isInferred<\/th>\n      <th>name<\/th>\n      <th>releaseDate<\/th>\n      <th>speciesName<\/th>\n      <th>doi<\/th>\n      <th>hasDiagram<\/th>\n      <th>hasEHLD<\/th>\n      <th>className<\/th>\n      <th>schemaClass<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":1},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><hr>
</div>
</div>
<div class="section level3">
<h3 class="heading2" id="see-also-in-functions-manuals">See also in Functions’ manuals<a class="anchor" aria-label="anchor" href="#see-also-in-functions-manuals"></a>
</h3>
<p>There are still more rbioapi f Reactome content functions that were
not covered in this vignette. Here is a brief overview, see the
functions’ manual for detailed guides and examples.</p>
<div class="section level4">
<h4 class="heading3" id="retrieve-reactome-database-information">Retrieve Reactome Database information<a class="anchor" aria-label="anchor" href="#retrieve-reactome-database-information"></a>
</h4>
<ul>
<li><p><code><a href="../reference/rba_reactome_version.html">rba_reactome_version()</a></code>: Return current Reactome
version</p></li>
<li><p><code><a href="../reference/rba_reactome_diseases.html">rba_reactome_diseases()</a></code>: Retrieve a list of disease
annotated in Reactome.</p></li>
<li><p><code><a href="../reference/rba_reactome_species.html">rba_reactome_species()</a></code>: Retrieve a list of species
annotated in Reactome.</p></li>
</ul>
</div>
<div class="section level4">
<h4 class="heading3" id="general-mapping-querying">General Mapping/Querying<a class="anchor" aria-label="anchor" href="#general-mapping-querying"></a>
</h4>
<ul>
<li><p><code><a href="../reference/rba_reactome_query.html">rba_reactome_query()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_mapping.html">rba_reactome_mapping()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_xref.html">rba_reactome_xref()</a></code></p></li>
</ul>
</div>
<div class="section level4">
<h4 class="heading3" id="things-you-can-do-with-a-entities">Things you can do with a Entities<a class="anchor" aria-label="anchor" href="#things-you-can-do-with-a-entities"></a>
</h4>
<ul>
<li><p><code>reactome_complex_list()</code>: Get a list of complexes
that have your molecule in them.</p></li>
<li><p><code><a href="../reference/rba_reactome_complex_subunits.html">rba_reactome_complex_subunits()</a></code>: Get the list of
subunits in your complex</p></li>
<li><p><code><a href="../reference/rba_reactome_participant_of.html">rba_reactome_participant_of()</a></code>: Get a list of
Reactome sets and complexes that your entity (event, molecule, reaction,
pathway etc.) is a participant in them.</p></li>
<li><p><code><a href="../reference/rba_reactome_entity_other_forms.html">rba_reactome_entity_other_forms()</a></code></p></li>
</ul>
</div>
<div class="section level4">
<h4 class="heading3" id="things-you-can-do-with-events">Things you can do with Events<a class="anchor" aria-label="anchor" href="#things-you-can-do-with-events"></a>
</h4>
<ul>
<li><p><code><a href="../reference/rba_reactome_event_ancestors.html">rba_reactome_event_ancestors()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_participants.html">rba_reactome_participants()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_pathways_events.html">rba_reactome_pathways_events()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_event_ancestors.html">rba_reactome_event_ancestors()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_orthology.html">rba_reactome_orthology()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_event_hierarchy.html">rba_reactome_event_hierarchy()</a></code>: Retrieve full event
hierarchy of an species.</p></li>
</ul>
</div>
<div class="section level4">
<h4 class="heading3" id="pathways">Pathways<a class="anchor" aria-label="anchor" href="#pathways"></a>
</h4>
<ul>
<li><p><code><a href="../reference/rba_reactome_pathways_low.html">rba_reactome_pathways_low()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_pathways_events.html">rba_reactome_pathways_events()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_pathways_top.html">rba_reactome_pathways_top()</a></code></p></li>
</ul>
</div>
<div class="section level4">
<h4 class="heading3" id="interactors">Interactors<a class="anchor" aria-label="anchor" href="#interactors"></a>
</h4>
<ul>
<li><p><code><a href="../reference/rba_reactome_interactors_psicquic.html">rba_reactome_interactors_psicquic()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_interactors_static.html">rba_reactome_interactors_static()</a></code></p></li>
</ul>
</div>
<div class="section level4">
<h4 class="heading3" id="peaple">People<a class="anchor" aria-label="anchor" href="#peaple"></a>
</h4>
<ul>
<li><p><code><a href="../reference/rba_reactome_people_name.html">rba_reactome_people_name()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_people_id.html">rba_reactome_people_id()</a></code></p></li>
</ul>
</div>
<div class="section level4">
<h4 class="heading3" id="export-diagrams-and-events">Export diagrams and events<a class="anchor" aria-label="anchor" href="#export-diagrams-and-events"></a>
</h4>
<ul>
<li><p><code><a href="../reference/rba_reactome_exporter_diagram.html">rba_reactome_exporter_diagram()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_exporter_overview.html">rba_reactome_exporter_overview()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_exporter_reaction.html">rba_reactome_exporter_reaction()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_exporter_event.html">rba_reactome_exporter_event()</a></code></p></li>
</ul>
<hr>
</div>
</div>
<div class="section level3">
<h3 class="heading2" id="citations">How to Cite?<a class="anchor" aria-label="anchor" href="#citations"></a>
</h3>
<p>To cite Reactome (Please see <a href="https://reactome.org/cite" class="external-link uri">https://reactome.org/cite</a>):</p>
<ul>
<li>Marc Gillespie, Bijay Jassal, Ralf Stephan, Marija Milacic, Karen
Rothfels, Andrea Senff-Ribeiro, Johannes Griss, Cristoffer Sevilla, Lisa
Matthews, Chuqiao Gong, Chuan Deng, Thawfeek Varusai, Eliot Ragueneau,
Yusra Haider, Bruce May, Veronica Shamovsky, Joel Weiser, Timothy
Brunson, Nasim Sanati, Liam Beckman, Xiang Shao, Antonio Fabregat,
Konstantinos Sidiropoulos, Julieth Murillo, Guilherme Viteri, Justin
Cook, Solomon Shorser, Gary Bader, Emek Demir, Chris Sander, Robin Haw,
Guanming Wu, Lincoln Stein, Henning Hermjakob, Peter D’Eustachio, The
reactome pathway knowledgebase 2022, <em>Nucleic Acids Research</em>,
2021;, kab1028, <a href="https://doi.org/10.1093/nar/gkab1028" class="external-link uri">https://doi.org/10.1093/nar/gkab1028</a>
</li>
<li>Griss J, Viteri G, Sidiropoulos K, Nguyen V, Fabregat A, Hermjakob
H. ReactomeGSA - Efficient Multi-Omics Comparative Pathway Analysis. Mol
Cell Proteomics. 2020 Sep 9. doi: 10.1074/mcp. <a href="https://pubmed.ncbi.nlm.nih.gov/32907876" class="external-link">PubMed PMID:
32907876</a>.</li>
<li>Fabregat A, Korninger F, Viteri G, Sidiropoulos K, Marin-Garcia P,
Ping P, Wu G, Stein L, D’Eustachio P, Hermjakob H. Reactome graph
database: Efficient access to complex pathway data. <em>PLoS Comput
Biol.</em> 2018 Jan 29;14(1):e1005968. doi:
10.1371/journal.pcbi.1005968. eCollection 2018 Jan. <a href="https://pubmed.ncbi.nlm.nih.gov/29377902" class="external-link">PubMed PMID:
29377902</a>.</li>
<li>Fabregat A, Sidiropoulos K, Viteri G, Marin-Garcia P, Ping P, Stein
L, D’Eustachio P, Hermjakob H. Reactome diagram viewer: data structures
and strategies to boost performance. <em>Bioinformatics.</em> 2018 Apr
1;34(7):1208-1214. doi: 10.1093/bioinformatics/btx752. <a href="https://pubmed.ncbi.nlm.nih.gov/29186351" class="external-link">PubMed PMID:
29186351</a>.</li>
<li>Fabregat A, Sidiropoulos K, Viteri G, Forner O, Marin-Garcia P,
Arnau V, D’Eustachio P, Stein L, Hermjakob H. Reactome pathway analysis:
a high-performance in-memory approach. <em>BMC Bioinformatics.</em> 2017
Mar 2;18(1):142. doi: 10.1186/s12859-017-1559-2. <a href="https://pubmed.ncbi.nlm.nih.gov/28249561" class="external-link">PubMed PMID:
28249561</a>.</li>
<li>Wu G, Haw R. Functional Interaction Network Construction and
Analysis for Disease Discovery. <em>Methods Mol Biol.</em>
2017;1558:235-253. doi: 10.1007/978-1-4939-6783-4_11. <a href="https://pubmed.ncbi.nlm.nih.gov/28150241" class="external-link">PubMed PMID:
28150241</a>. </li>
</ul>
<p>To cite rbioapi:</p>
<ul>
<li>Moosa Rezwani, Ali Akbar Pourfathollah, Farshid Noorbakhsh, rbioapi:
user-friendly R interface to biologic web services’ API, Bioinformatics,
Volume 38, Issue 10, 15 May 2022, Pages 2952–2953, <a href="https://doi.org/10.1093/bioinformatics/btac172" class="external-link uri">https://doi.org/10.1093/bioinformatics/btac172</a>
</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 class="heading2" id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_reactome.html" title="2.D: Reactome &amp; rbioapi">This article in rbioapi
documentation site</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html#section-reactome-analysis-services-rba-reactome-" title="rbioapi reference">Functions references in rbioapi
documentation site</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi
vignette index</a></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 class="heading2" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] rbioapi_0.7.7</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] vctrs_0.6.2       httr_1.4.5        cli_3.6.1         knitr_1.42       </span></span>
<span><span class="co">#&gt;  [5] rlang_1.1.0       xfun_0.39         stringi_1.7.12    purrr_1.0.1      </span></span>
<span><span class="co">#&gt;  [9] textshaping_0.3.6 jsonlite_1.8.4    glue_1.6.2        DT_0.27          </span></span>
<span><span class="co">#&gt; [13] rprojroot_2.0.3   htmltools_0.5.5   ragg_1.2.5        sass_0.4.5       </span></span>
<span><span class="co">#&gt; [17] rmarkdown_2.21    crosstalk_1.2.0   evaluate_0.20     jquerylib_0.1.4  </span></span>
<span><span class="co">#&gt; [21] ellipsis_0.3.2    fastmap_1.1.1     yaml_2.3.7        lifecycle_1.0.3  </span></span>
<span><span class="co">#&gt; [25] memoise_2.0.1     stringr_1.5.0     compiler_4.3.0    fs_1.6.2         </span></span>
<span><span class="co">#&gt; [29] htmlwidgets_1.6.2 systemfonts_1.0.4 digest_0.6.31     R6_2.5.1         </span></span>
<span><span class="co">#&gt; [33] curl_5.0.0        magrittr_2.0.3    bslib_0.4.2       tools_4.3.0      </span></span>
<span><span class="co">#&gt; [37] pkgdown_2.0.7     cachem_1.0.7      desc_1.4.2</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://moosa-r.com/" class="external-link">Moosa Rezwani</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
