<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2.B: JASPAR &amp; rbioapi • rbioapi</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2.B: JASPAR &amp; rbioapi">
<meta property="og:description" content="rbioapi">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-1HHD9Z0414"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-1HHD9Z0414');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rbioapi</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/rbioapi.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Supported Services
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/rbioapi.html">1: rbioapi: Get Started</a>
    </li>
    <li>
      <a href="../articles/rbioapi_enrichr.html">2.A: Enrichr &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_mieaa.html">2.B: miEEA &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_panther.html">2.C: PANTHER &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_reactome.html">2.D: Reactome &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_string.html">2.E: STRING &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_uniprot.html">2.F: UniProt &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_jaspar.html">2.G: JASPAR &amp; rbioapi</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="rbioapi_jaspar_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>2.B: JASPAR &amp; rbioapi</h1>
                        <h4 class="author">Moosa Rezwani</h4>
            
            <h4 class="date">2021-06-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/master/vignettes/rbioapi_jaspar.Rmd"><code>vignettes/rbioapi_jaspar.Rmd</code></a></small>
      <div class="hidden name"><code>rbioapi_jaspar.Rmd</code></div>

    </div>

    
    
<hr>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Directly quoting from Fornes O, Castro-Mondragon JA, Khan A, et al:</p>
<blockquote>
<p>JASPAR (<a href="http://jaspar.genereg.net">http://jaspar.genereg.net)</a> is an open-access database of curated, non-redundant transcription factor (TF)-binding profiles stored as position frequency matrices (PFMs) for TFs across multiple species in six taxonomic groups. In this 8th release of JASPAR, the CORE collection has been expanded with 245 new PFMs (169 for vertebrates, 42 for plants, 17 for nematodes, 10 for insects, and 7 for fungi), and 156 PFMs were updated (125 for vertebrates, 28 for plants and 3 for insects). These new profiles represent an 18% expansion compared to the previous release.</p>
<p>source:<br>
Fornes O, Castro-Mondragon JA, Khan A, et al. JASPAR 2020: update of the open-access database of transcription factor binding profiles. <em>Nucleic Acids Res</em>. 2019; doi: <a href="https://doi.org/10.1093/nar/gkz1001">10.1093/nar/gkz1001</a></p>
</blockquote>
<hr>
</div>
<div id="data-organization-in-jaspar" class="section level1">
<h1 class="hasAnchor">
<a href="#data-organization-in-jaspar" class="anchor"></a>Data Organization in JASPAR</h1>
<p>JASPAR is a database of transcription factor binding matrices with annotations and metadata. These entities are organized in a hierarchical fashion that we will explore next.</p>
<div id="releases" class="section level2">
<h2 class="hasAnchor">
<a href="#releases" class="anchor"></a>Releases</h2>
<p>In addition to the latest JASPAR database release (2020), other active releases are also available. Most of the rbioapi JASPAR functions have a <code>release</code> argument that allows you to use other database releases.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Call the function without any arguments to get a list of releases</span>
<span class="va">releases</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_releases.html">rba_jaspar_releases</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## Supply a release number for details:</span>
<span class="va">release_7_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_releases.html">rba_jaspar_releases</a></span><span class="op">(</span><span class="fl">7</span><span class="op">)</span></code></pre></div>
</div>
<div id="collections" class="section level2">
<h2 class="hasAnchor">
<a href="#collections" class="anchor"></a>Collections</h2>
<p>Within a release, Matrix profiles are organized into collections, You can use <code><a href="../reference/rba_jaspar_collections.html">rba_jaspar_collections()</a></code> to get a list of available collections, or read “JASPAR Collections” section in <a href="http://jaspar.genereg.net/docs/" title="JASPAR Documentation">documentation page in JASPAR web-site</a> for a thorough review.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## To get a list of available collection in release 2020:</span>
<span class="fu"><a href="../reference/rba_jaspar_collections.html">rba_jaspar_collections</a></span><span class="op">(</span>release <span class="op">=</span> <span class="fl">2020</span><span class="op">)</span>
<span class="co">#&gt;                                                          url        name</span>
<span class="co">#&gt; 1          http://jaspar.genereg.net/api/v1/collections/CNE/         CNE</span>
<span class="co">#&gt; 2         http://jaspar.genereg.net/api/v1/collections/CORE/        CORE</span>
<span class="co">#&gt; 3          http://jaspar.genereg.net/api/v1/collections/FAM/         FAM</span>
<span class="co">#&gt; 4          http://jaspar.genereg.net/api/v1/collections/PBM/         PBM</span>
<span class="co">#&gt; 5      http://jaspar.genereg.net/api/v1/collections/PBM_HLH/     PBM_HLH</span>
<span class="co">#&gt; 6    http://jaspar.genereg.net/api/v1/collections/PBM_HOMEO/   PBM_HOMEO</span>
<span class="co">#&gt; 7   http://jaspar.genereg.net/api/v1/collections/PHYLOFACTS/  PHYLOFACTS</span>
<span class="co">#&gt; 8        http://jaspar.genereg.net/api/v1/collections/POLII/       POLII</span>
<span class="co">#&gt; 9       http://jaspar.genereg.net/api/v1/collections/SPLICE/      SPLICE</span>
<span class="co">#&gt; 10 http://jaspar.genereg.net/api/v1/collections/UNVALIDATED/ UNVALIDATED</span>


<span class="co">## You can list information of all matrices available in a collection:</span>
<span class="va">mat_in_core_2020</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_collections_matrices.html">rba_jaspar_collections_matrices</a></span><span class="op">(</span>collection <span class="op">=</span> <span class="st">"CORE"</span><span class="op">)</span></code></pre></div>
</div>
<div id="taxonomic-groups" class="section level2">
<h2 class="hasAnchor">
<a href="#taxonomic-groups" class="anchor"></a>Taxonomic Groups</h2>
<p>Within each collection, the matrix profiles are organized based on main taxonomic groups:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## To get a list of taxonomic groups in release 2020:</span>
<span class="fu"><a href="../reference/rba_jaspar_taxons.html">rba_jaspar_taxons</a></span><span class="op">(</span>release <span class="op">=</span> <span class="fl">2020</span><span class="op">)</span>
<span class="co">#&gt;                                                    url         name</span>
<span class="co">#&gt; 1       http://jaspar.genereg.net/api/v1/taxon/plants/       plants</span>
<span class="co">#&gt; 2  http://jaspar.genereg.net/api/v1/taxon/vertebrates/  vertebrates</span>
<span class="co">#&gt; 3      http://jaspar.genereg.net/api/v1/taxon/insects/      insects</span>
<span class="co">#&gt; 4 http://jaspar.genereg.net/api/v1/taxon/urochordates/ urochordates</span>
<span class="co">#&gt; 5    http://jaspar.genereg.net/api/v1/taxon/nematodes/    nematodes</span>
<span class="co">#&gt; 6        http://jaspar.genereg.net/api/v1/taxon/fungi/        fungi</span>
<span class="co">#&gt; 7   http://jaspar.genereg.net/api/v1/taxon/trematodes/   trematodes</span>
<span class="co">#&gt; 8     http://jaspar.genereg.net/api/v1/taxon/protozoa/     protozoa</span>
<span class="co">#&gt; 9     http://jaspar.genereg.net/api/v1/taxon/cnidaria/     cnidaria</span>


<span class="co">## You can list information of all matrices available in a taxonomic group:</span>
<span class="va">mat_in_insects</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_taxons_matrices.html">rba_jaspar_taxons_matrices</a></span><span class="op">(</span>tax_group <span class="op">=</span> <span class="st">"insects"</span><span class="op">)</span></code></pre></div>
</div>
<div id="species" class="section level2">
<h2 class="hasAnchor">
<a href="#species" class="anchor"></a>Species</h2>
<p>As we go down in the data organization hierarchy, Each taxonomic group consist of species:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## To get a list of species in release 2020:</span>
<span class="va">species</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_species.html">rba_jaspar_species</a></span><span class="op">(</span>release <span class="op">=</span> <span class="fl">2020</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">species</span><span class="op">)</span>
<span class="co">#&gt;   tax_id                          species</span>
<span class="co">#&gt; 1   5037           Ajellomyces capsulatus</span>
<span class="co">#&gt; 2   4151                Antirrhinum majus</span>
<span class="co">#&gt; 3  81972 Arabidopsis lyrata subsp. lyrata</span>
<span class="co">#&gt; 4   3702             Arabidopsis thaliana</span>
<span class="co">#&gt; 5   9913                       Bos taurus</span>
<span class="co">#&gt; 6   6238          Caenorhabditis briggsae</span>
<span class="co">#&gt;                                               url</span>
<span class="co">#&gt; 1  http://jaspar.genereg.net/api/v1/species/5037/</span>
<span class="co">#&gt; 2  http://jaspar.genereg.net/api/v1/species/4151/</span>
<span class="co">#&gt; 3 http://jaspar.genereg.net/api/v1/species/81972/</span>
<span class="co">#&gt; 4  http://jaspar.genereg.net/api/v1/species/3702/</span>
<span class="co">#&gt; 5  http://jaspar.genereg.net/api/v1/species/9913/</span>
<span class="co">#&gt; 6  http://jaspar.genereg.net/api/v1/species/6238/</span>
<span class="co">#&gt;                                        matrix_url</span>
<span class="co">#&gt; 1  http://jaspar.genereg.net/api/v1/species/5037/</span>
<span class="co">#&gt; 2  http://jaspar.genereg.net/api/v1/species/4151/</span>
<span class="co">#&gt; 3 http://jaspar.genereg.net/api/v1/species/81972/</span>
<span class="co">#&gt; 4  http://jaspar.genereg.net/api/v1/species/3702/</span>
<span class="co">#&gt; 5  http://jaspar.genereg.net/api/v1/species/9913/</span>
<span class="co">#&gt; 6  http://jaspar.genereg.net/api/v1/species/6238/</span>

<span class="co">## You can list information of all matrices available in a specie:</span>
<span class="va">mat_in_human</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_species_matrices.html">rba_jaspar_species_matrices</a></span><span class="op">(</span>tax_id <span class="op">=</span> <span class="fl">9606</span><span class="op">)</span></code></pre></div>
<hr>
</div>
</div>
<div id="matrix-profiles" class="section level1">
<h1 class="hasAnchor">
<a href="#matrix-profiles" class="anchor"></a>Matrix Profiles</h1>
<div id="search-matrix-profiles" class="section level2">
<h2 class="hasAnchor">
<a href="#search-matrix-profiles" class="anchor"></a>Search Matrix Profiles</h2>
<p>Retrieving a list of every matrix available in a given category is not the only option. You can also build a search query using <code>rba_jaspar_matrix_search</code>. Note that this is a search function, you are not required to fill every argument. You may use any combination of arguments you see fit to build your query. You can even call the function without any argument to get a list of all the matrix profiles. For instance:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get a list of all the available matrix profile:</span>
<span class="va">all_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix_search.html">rba_jaspar_matrix_search</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">## Search FOX:</span>
<span class="va">FOX_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix_search.html">rba_jaspar_matrix_search</a></span><span class="op">(</span>term <span class="op">=</span> <span class="st">"FOX"</span><span class="op">)</span>

<span class="co">## Transcription factors named FOXP3</span>
<span class="va">FOXP3_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix_search.html">rba_jaspar_matrix_search</a></span><span class="op">(</span>term <span class="op">=</span> <span class="st">"FOXP3"</span><span class="op">)</span>

<span class="co">## Transcription factors of Zipper-Type Class</span>
<span class="va">zipper_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix_search.html">rba_jaspar_matrix_search</a></span><span class="op">(</span>tf_class <span class="op">=</span> <span class="st">"Zipper-Type"</span><span class="op">)</span>

<span class="co">## Transcription factors of Zipper-Type Class in PBM collection</span>
<span class="va">zipper_pbm_matrices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix_search.html">rba_jaspar_matrix_search</a></span><span class="op">(</span>tf_class <span class="op">=</span> <span class="st">"Zipper-Type"</span>,
                                                collection <span class="op">=</span> <span class="st">"PBM"</span><span class="op">)</span></code></pre></div>
</div>
<div id="list-matrix-profiles-associated-to-a-base-identifier" class="section level2">
<h2 class="hasAnchor">
<a href="#list-matrix-profiles-associated-to-a-base-identifier" class="anchor"></a>List Matrix Profiles Associated to a Base identifier</h2>
<p>Since JASPAR release 2010, the matrix profiles are versioned. A matrix profile Identifier has a “base_id.version” naming schema; for example “MA0600.2” corresponds to the second version of a matrix with base ID MA0600. You can Use <code>rba_jaspar_matrix_versions</code> to get a list of matrix profiles with a given base ID. Also note that some functions, generally those that are used to list available matrices, have an argument called <code>only_last_version</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get matrix profiles versions associated to a base id</span>
<span class="va">MA0600_versions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix_versions.html">rba_jaspar_matrix_versions</a></span><span class="op">(</span><span class="st">"MA0600"</span><span class="op">)</span></code></pre></div>
</div>
<div id="get-a-matrix-profile" class="section level2">
<h2 class="hasAnchor">
<a href="#get-a-matrix-profile" class="anchor"></a>Get a Matrix Profile</h2>
<p>Now that you listed or searched for matrix profiles, you can use <code>rba_jaspar_matrix</code> to retrieve matrix profiles. There are two ways in which you can use this function:</p>
<div id="get-matrix-and-annotations-as-an-r-object" class="section level3">
<h3 class="hasAnchor">
<a href="#get-matrix-and-annotations-as-an-r-object" class="anchor"></a>Get Matrix and Annotations as an R Object</h3>
<p>To do that, only fill in the <code>matrix_id</code> argument in <code>rba_jaspar_matrix</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pfm_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix.html">rba_jaspar_matrix</a></span><span class="op">(</span>matrix_id <span class="op">=</span> <span class="st">"MA0600.2"</span><span class="op">)</span>

<span class="co">## you can find the matrix in the pfm element along with</span>
<span class="co">## other elements which correspond to annotations and details</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">pfm_matrix</span><span class="op">)</span>
<span class="co">#&gt; List of 23</span>
<span class="co">#&gt;  $ comment      : chr "Data is from Taipale HTSELEX DBD (2013)"</span>
<span class="co">#&gt;  $ pubmed_ids   : chr "8754849"</span>
<span class="co">#&gt;  $ family       : chr "RFX-related factors"</span>
<span class="co">#&gt;  $ pfm          : num [1:4, 1:16] 1381 5653 4042 2336 270 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : chr [1:4] "A" "C" "G" "T"</span>
<span class="co">#&gt;   .. ..$ : NULL</span>
<span class="co">#&gt;  $ tax_group    : chr "vertebrates"</span>
<span class="co">#&gt;  $ matrix_id    : chr "MA0600.2"</span>
<span class="co">#&gt;  $ sequence_logo: chr "http://jaspar.genereg.net/static/logos/svg/MA0600.2.svg"</span>
<span class="co">#&gt;  $ remap_tf_name: chr "RFX2"</span>
<span class="co">#&gt;  $ pazar_tf_ids : list()</span>
<span class="co">#&gt;  $ versions_url : chr "http://jaspar.genereg.net/api/v1/matrix/MA0600/versions"</span>
<span class="co">#&gt;  $ collection   : chr "CORE"</span>
<span class="co">#&gt;  $ base_id      : chr "MA0600"</span>
<span class="co">#&gt;  $ class        : chr "Fork head / winged helix factors"</span>
<span class="co">#&gt;  $ tffm         :List of 7</span>
<span class="co">#&gt;   ..$ tffm_url       : chr "http://jaspar.genereg.net/api/v1/tffm/TFFM0576.1"</span>
<span class="co">#&gt;   ..$ log_p_1st_order: num -6275</span>
<span class="co">#&gt;   ..$ experiment_name: chr "CistromeDB_58298"</span>
<span class="co">#&gt;   ..$ version        : int 1</span>
<span class="co">#&gt;   ..$ log_p_detailed : num -6660</span>
<span class="co">#&gt;   ..$ base_id        : chr "TFFM0576"</span>
<span class="co">#&gt;   ..$ tffm_id        : chr "TFFM0576.1"</span>
<span class="co">#&gt;  $ tfe_ids      : list()</span>
<span class="co">#&gt;  $ name         : chr "RFX2"</span>
<span class="co">#&gt;  $ uniprot_ids  : chr "P48378"</span>
<span class="co">#&gt;  $ sites_url    : NULL</span>
<span class="co">#&gt;  $ species      :'data.frame':   1 obs. of  2 variables:</span>
<span class="co">#&gt;   ..$ name  : chr "Homo sapiens"</span>
<span class="co">#&gt;   ..$ tax_id: int 9606</span>
<span class="co">#&gt;  $ source       : chr "23332764"</span>
<span class="co">#&gt;  $ version      : int 2</span>
<span class="co">#&gt;  $ unibind      : chr "1"</span>
<span class="co">#&gt;  $ type         : chr "HT-SELEX"</span></code></pre></div>
</div>
<div id="save-a-matrix-a-file-in-specific-format" class="section level3">
<h3 class="hasAnchor">
<a href="#save-a-matrix-a-file-in-specific-format" class="anchor"></a>Save a Matrix a File in Specific Format</h3>
<p>JASPAR provides position frequency matrices (PFM) formatted as <strong>Raw PFM</strong>, <strong>JASPAR</strong>, <strong>TRANSFAC</strong>, <strong>YAML</strong>, and <strong>MEME</strong>. You can download a matrix profile as a file with any of these formats. To do that, You should use the <code>file_format</code> and <code>save_to</code> arguments available in <code>rba_jaspar_matrix</code>. There are two notes here:</p>
<ol style="list-style-type: decimal">
<li><p>In this case, the function will save your matrix as a file and returns the un-parsed content of the file as a character string.</p></li>
<li>
<p>The <code>save_to</code> argument in this function, and in fact through any rbioapi function can be used in many ways:<br>
2.1. save_to = NA: rbioapi will automatically generate a file path under your working directory, save the file in that path , and informs you with a message.<br>
2.2 save_to = file_name without path: rbioapi will save the file with your supplied name in your working directory.<br>
2.3. save_to = a directory path (without file): rbioapi will save the file with a proper name in that directory.<br>
2.4. save_to = a file path (i.e. ending with .extension): rbioapi will save the file exactly to this path. Make sure that the file extension of the path matches your requested file format. If this was not the case, rbioapi will save the file with the extension supplied in the path, but issues a warning to inform you about that.</p>
<p>In any of the aforementioned cases, the file path can be absolute or relative.</p>
</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Different wqays in which you can save the matrix file:</span>
<span class="va">meme_matrix1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix.html">rba_jaspar_matrix</a></span><span class="op">(</span>matrix_id <span class="op">=</span> <span class="st">"MA0600.2"</span>,
                                  file_format <span class="op">=</span> <span class="st">"meme"</span><span class="op">)</span>

<span class="va">meme_matrix2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix.html">rba_jaspar_matrix</a></span><span class="op">(</span>matrix_id <span class="op">=</span> <span class="st">"MA0600.2"</span>,
                                  file_format <span class="op">=</span> <span class="st">"meme"</span>,
                                  save_to <span class="op">=</span> <span class="st">"my_matrix.meme"</span><span class="op">)</span>

<span class="va">meme_matrix3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix.html">rba_jaspar_matrix</a></span><span class="op">(</span>matrix_id <span class="op">=</span> <span class="st">"MA0600.2"</span>,
                                  file_format <span class="op">=</span> <span class="st">"meme"</span>,
                                  save_to <span class="op">=</span> <span class="st">"c:/rbioapi"</span><span class="op">)</span>

<span class="va">meme_matrix4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_matrix.html">rba_jaspar_matrix</a></span><span class="op">(</span>matrix_id <span class="op">=</span> <span class="st">"MA0600.2"</span>,
                                  file_format <span class="op">=</span> <span class="st">"meme"</span>,
                                  save_to <span class="op">=</span> <span class="st">"c:/rbioapi/my_matrix.meme"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="get-binding-sites-of-a-matrix-profiles" class="section level2">
<h2 class="hasAnchor">
<a href="#get-binding-sites-of-a-matrix-profiles" class="anchor"></a>Get Binding Sites of a Matrix Profiles</h2>
<p>If available, you can retrieve information on binding sites associated with a matrix profile. The information includes a data frame of genomic coordination of the binding site, URL to FASTA and BED files, along with other annotations.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Get binding site of a matrix profile:</span>
<span class="va">binding_sites</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_sites.html">rba_jaspar_sites</a></span><span class="op">(</span>matrix_id <span class="op">=</span> <span class="st">"MA0600.2"</span><span class="op">)</span></code></pre></div>
<hr>
</div>
</div>
<div id="tf-flexible-models-tffms" class="section level1">
<h1 class="hasAnchor">
<a href="#tf-flexible-models-tffms" class="anchor"></a>TF flexible models (TFFMs)</h1>
<p>JASPAR also stores and assigns identifiers to TF flexible models (TFFMs). Just like PFM (position frequency matrices), you can search TFFMs or retrieve information and annotations using a TFFM Identifier. TFFM IDs are versioned, meaning that they are in base_id.version format.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Search TFFMs. This is a search function. Thus, what has been presented</span>
<span class="co">## in `Search Matrix Profiles` section also applies here:</span>

<span class="co">## Get a list of all the available matrix profile:</span>
<span class="va">all_tffms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_tffm_search.html">rba_jaspar_tffm_search</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">## Search FOX:</span>
<span class="va">FOX_tffms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_tffm_search.html">rba_jaspar_tffm_search</a></span><span class="op">(</span>term <span class="op">=</span> <span class="st">"FOX"</span><span class="op">)</span>

<span class="co">## Transcription factors named FOXP3</span>
<span class="va">FOXP3_tffms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_tffm_search.html">rba_jaspar_tffm_search</a></span><span class="op">(</span>term <span class="op">=</span> <span class="st">"FOXP3"</span><span class="op">)</span>

<span class="co">## Transcription factors of insects taxonomic group</span>
<span class="va">insects_tffms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_tffm_search.html">rba_jaspar_tffm_search</a></span><span class="op">(</span>tax_group <span class="op">=</span> <span class="st">"insects"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Now that you have a TFFM ID, you can retrieve it</span>
<span class="va">TFFM0056</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_jaspar_tffm.html">rba_jaspar_tffm</a></span><span class="op">(</span><span class="st">"TFFM0056.3"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">TFFM0056</span><span class="op">)</span>
<span class="co">#&gt; List of 10</span>
<span class="co">#&gt;  $ detailed       :List of 5</span>
<span class="co">#&gt;   ..$ xml         : chr "http://jaspar.genereg.net/static/TFFM/TFFM0056.3/TFFM_detailed_trained.xml"</span>
<span class="co">#&gt;   ..$ hits        : chr "http://jaspar.genereg.net/static/TFFM/TFFM0056.3/TFFM_detailed_trained.hits.svg"</span>
<span class="co">#&gt;   ..$ log_p       : num -6854</span>
<span class="co">#&gt;   ..$ dense_logo  : chr "http://jaspar.genereg.net/static/TFFM/TFFM0056.3/TFFM_detailed_trained_dense_logo.svg"</span>
<span class="co">#&gt;   ..$ summary_logo: chr "http://jaspar.genereg.net/static/TFFM/TFFM0056.3/TFFM_detailed_trained_summary_logo.svg"</span>
<span class="co">#&gt;  $ matrix_base_id : chr "MA0039"</span>
<span class="co">#&gt;  $ experiment_name: chr "CistromeDB_33718"</span>
<span class="co">#&gt;  $ version        : int 3</span>
<span class="co">#&gt;  $ matrix_version : int 4</span>
<span class="co">#&gt;  $ base_id        : chr "TFFM0056"</span>
<span class="co">#&gt;  $ tffm_id        : chr "TFFM0056.3"</span>
<span class="co">#&gt;  $ first_order    :List of 5</span>
<span class="co">#&gt;   ..$ xml         : chr "http://jaspar.genereg.net/static/TFFM/TFFM0056.3/TFFM_first_order_trained.xml"</span>
<span class="co">#&gt;   ..$ hits        : chr "http://jaspar.genereg.net/static/TFFM/TFFM0056.3/TFFM_first_order_trained.hits.svg"</span>
<span class="co">#&gt;   ..$ log_p       : num -7420</span>
<span class="co">#&gt;   ..$ dense_logo  : chr "http://jaspar.genereg.net/static/TFFM/TFFM0056.3/TFFM_first_order_trained_dense_logo.svg"</span>
<span class="co">#&gt;   ..$ summary_logo: chr "http://jaspar.genereg.net/static/TFFM/TFFM0056.3/TFFM_first_order_trained_summary_logo.svg"</span>
<span class="co">#&gt;  $ matrix_url     : chr "http://jaspar.genereg.net/api/v1/matrix/MA0039.4"</span>
<span class="co">#&gt;  $ matrix_id      : chr "MA0039.4"</span></code></pre></div>
<hr>
</div>
<div id="links" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#links" class="anchor"></a>Links</h1>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_jaspar.html" title="2.G: JASPAR &amp; rbioapi">This article in pkgdown website</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html#section-jaspar-rba-jaspar-" title="rbioapi reference">Functions references in pkgdown website</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi vignette index</a></p></li>
</ul>
<hr>
</div>
<div id="session-info" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<pre><code>#&gt; R version 4.1.0 (2021-05-18)
#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)
#&gt; Running under: macOS Catalina 10.15.7
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] rbioapi_0.7.4
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] rprojroot_2.0.2   digest_0.6.27     crayon_1.4.1      R6_2.5.0         
#&gt;  [5] jsonlite_1.7.2    magrittr_2.0.1    evaluate_0.14     httr_1.4.2       
#&gt;  [9] stringi_1.6.2     rlang_0.4.11      cachem_1.0.5      curl_4.3.1       
#&gt; [13] fs_1.5.0          ragg_1.1.3        rmarkdown_2.9     pkgdown_1.6.1    
#&gt; [17] textshaping_0.3.5 desc_1.3.0        tools_4.1.0       stringr_1.4.0    
#&gt; [21] yaml_2.2.1        xfun_0.24         fastmap_1.1.0     compiler_4.1.0   
#&gt; [25] systemfonts_1.0.2 memoise_2.0.0     htmltools_0.5.1.1 knitr_1.33</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
