<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2.D: Reactome &amp; rbioapi • rbioapi</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2.D: Reactome &amp; rbioapi">
<meta property="og:description" content="rbioapi">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-1HHD9Z0414"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-1HHD9Z0414');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rbioapi</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/rbioapi.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Supported Services
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/rbioapi.html">1: rbioapi: Get Started</a>
    </li>
    <li>
      <a href="../articles/rbioapi_enrichr.html">2.A: Enrichr &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_mieaa.html">2.B: miEEA &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_panther.html">2.C: PANTHER &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_reactome.html">2.D: Reactome &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_string.html">2.E: STRING &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_uniprot.html">2.F: UniProt &amp; rbioapi</a>
    </li>
    <li>
      <a href="../articles/rbioapi_jaspar.html">2.G: JASPAR &amp; rbioapi</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="rbioapi_reactome_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="rbioapi_reactome_files/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="rbioapi_reactome_files/jquery-3.5.1/jquery.min.js"></script><link href="rbioapi_reactome_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_reactome_files/datatables-binding-0.18/datatables.js"></script><link href="rbioapi_reactome_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_reactome_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_reactome_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="rbioapi_reactome_files/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="rbioapi_reactome_files/crosstalk-1.1.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>2.D: Reactome &amp; rbioapi</h1>
                        <h4 class="author">Moosa Rezwani</h4>
            
            <h4 class="date">2021-05-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/master/vignettes/rbioapi_reactome.Rmd"><code>vignettes/rbioapi_reactome.Rmd</code></a></small>
      <div class="hidden name"><code>rbioapi_reactome.Rmd</code></div>

    </div>

    
    
<hr>
<div id="introduction" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Directly quoting from <a href="https://reactome.org/">Reactome</a>:</p>
<blockquote>
<p>REACTOME is an open-source, open access, manually curated and peer-reviewed pathway database. Our goal is to provide intuitive bioinformatics tools for the visualization, interpretation and analysis of pathway knowledge to support basic and clinical research, genome analysis, modeling, systems biology and education. Founded in 2003, the Reactome project is led by Lincoln Stein of <a href="https://oicr.on.ca/">OICR</a>, Peter D’Eustachio of <a href="https://nyulangone.org/">NYULMC</a>, Henning Hermjakob of <a href="https://www.ebi.ac.uk/">EMBL-EBI</a>, and Guanming Wu of <a href="https://www.ohsu.edu/">OHSU</a>.</p>
<p>(source: <a href="https://reactome.org/what-is-reactome" class="uri">https://reactome.org/what-is-reactome</a>)</p>
</blockquote>
<p>Reactome provides two RESTful API services: Reactome content services and Reactome analysis services. In rbioapi, the naming schema is that any function which belongs to analysis services starts with rba_reactome_analysis* . Other rba_reactome_* functions without the ‘analysis’ infix correspond to content services API.</p>
<p>Before continuing reading this article, it is a good idea to read <a href="https://reactome.org/documentation/data-model">Reactome Data Model</a> page.</p>
<hr>
</div>
<div id="reactome-analysis-services" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#reactome-analysis-services" class="anchor"></a>Reactome analysis services</h1>
<p>This section mostly revolves around <code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code> function. So, naturally, we will start with that. As explained in the function’s manual, you have considerable freedom in providing the main input for this function; You can provide an R object (as a data frame, matrix, or simple vector), a URL, or a local file path. Note that the type of analysis will be decided based on whether your input is 1-dimensional or 2-dimensional. This has been explained in detail in the manual of <code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code>, so refer to that for more information.<br><code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code> is the API equivalent of Reactome’s <a href="https://reactome.org/PathwayBrowser/#TOOL=AT">analyse gene list</a> tool. You can see that the function’s arguments correspond to what would you choose in the webpage’s wizard.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## 1 We create a simple vector with our genes</span>
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>, <span class="st">"CDK1"</span>, <span class="st">"KIF23"</span>, <span class="st">"PLK1"</span>, <span class="st">"RAC2"</span>, <span class="st">"RACGAP1"</span>, <span class="st">"RHOA"</span>, <span class="st">"RHOB"</span>, <span class="st">"MSL1"</span>, <span class="st">"PHF21A"</span>, <span class="st">"INSR"</span>, <span class="st">"JADE2"</span>, <span class="st">"P2RX7"</span>, <span class="st">"CCDC101"</span>, <span class="st">"PPM1B"</span>, <span class="st">"ANAPC16"</span>, <span class="st">"CDH8"</span>, <span class="st">"HSPA1L"</span>, <span class="st">"CUL2"</span>, <span class="st">"ZNF302"</span>, <span class="st">"CUX1"</span>, <span class="st">"CYTH2"</span>, <span class="st">"SEC22C"</span>, <span class="st">"EIF4E3"</span>, <span class="st">"ROBO2"</span>, <span class="st">"CXXC1"</span>, <span class="st">"LINC01314"</span>, <span class="st">"ATP5F1"</span><span class="op">)</span>

<span class="co">## 2 We call reactome analysis with the default parameters</span>
<span class="va">analyzed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">genes</span>,
                                 projection <span class="op">=</span> <span class="cn">TRUE</span>,
                                 p_value <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>

<span class="co">## 3 As always, we use str() to inspect the resutls</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">analyzed</span>, <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; List of 8</span>
<span class="co">#&gt;  $ summary            :List of 7</span>
<span class="co">#&gt;  $ expression         :List of 1</span>
<span class="co">#&gt;  $ identifiersNotFound: int 1</span>
<span class="co">#&gt;  $ pathwaysFound      : int 67</span>
<span class="co">#&gt;  $ pathways           :'data.frame': 67 obs. of  8 variables:</span>
<span class="co">#&gt;  $ resourceSummary    :'data.frame': 3 obs. of  3 variables:</span>
<span class="co">#&gt;  $ speciesSummary     :'data.frame': 1 obs. of  5 variables:</span>
<span class="co">#&gt;  $ warnings           : chr "Missing header. Using a default one."</span>

<span class="co">## 4 Note that in the summary element:</span>
<span class="co">### 4.a because we provided a simple vector, the analysis type was: over-representation</span>
<span class="co">### 4.b You need the token for other rba_reactome_analysis_* functions</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">analyzed</span><span class="op">$</span><span class="va">summary</span><span class="op">)</span>
<span class="co">#&gt; List of 7</span>
<span class="co">#&gt;  $ token         : chr "MjAyMTA0MjUwMDU3MjJfMzAyMTM%3D"</span>
<span class="co">#&gt;  $ projection    : logi TRUE</span>
<span class="co">#&gt;  $ interactors   : logi FALSE</span>
<span class="co">#&gt;  $ type          : chr "OVERREPRESENTATION"</span>
<span class="co">#&gt;  $ sampleName    : chr ""</span>
<span class="co">#&gt;  $ text          : logi TRUE</span>
<span class="co">#&gt;  $ includeDisease: logi TRUE</span>

<span class="co">## 5 Analsis results are in the pathways data frame:</span></code></pre></div>
<div id="htmlwidget-1d7e7adf853ce50879f7" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-1d7e7adf853ce50879f7">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67"],["R-HSA-3371568","R-HSA-3371571","R-HSA-3371511","R-HSA-114604","R-HSA-6806003","R-HSA-3371453","R-HSA-983231","R-HSA-176412","R-HSA-3371556","R-HSA-6804754","R-HSA-176408","R-HSA-69273","R-HSA-6796648","R-HSA-109582","R-HSA-174143","R-HSA-453276","R-HSA-1538133","R-HSA-5633007","R-HSA-69620","R-HSA-176417","R-HSA-6804757","R-HSA-5689896","R-HSA-2980767","R-HSA-5666185","R-HSA-1640170","R-HSA-8853884","R-HSA-69278","R-HSA-69275","R-HSA-453274","R-HSA-6811442","R-HSA-5625970","R-HSA-3700989","R-HSA-376176","R-HSA-2262752","R-HSA-68884","R-HSA-8953897","R-HSA-6791312","R-HSA-5663220","R-HSA-162658","R-HSA-69481","R-HSA-6804116","R-HSA-1362300","R-HSA-176814","R-HSA-6804114","R-HSA-69473","R-HSA-8852276","R-HSA-5625900","R-HSA-156711","R-HSA-174048","R-HSA-416572","R-HSA-5627117","R-HSA-9013404","R-HSA-6804756","R-HSA-195258","R-HSA-400685","R-HSA-3214847","R-HSA-9687139","R-HSA-9675136","R-HSA-9663199","R-HSA-9699150","R-HSA-9675126","R-HSA-397795","R-HSA-9716542","R-HSA-69205","R-HSA-4839726","R-HSA-3247509","R-HSA-8953750"],[3371568,3371571,3371511,114604,6806003,3371453,983231,176412,3371556,6804754,176408,69273,6796648,109582,174143,453276,1538133,5633007,69620,176417,6804757,5689896,2980767,5666185,1640170,8853884,69278,69275,453274,6811442,5625970,3700989,376176,2262752,68884,8953897,6791312,5663220,162658,69481,6804116,1362300,176814,6804114,69473,8852276,5625900,156711,174048,416572,5627117,9013404,6804756,195258,400685,3214847,9687139,9675136,9663199,9699150,9675126,397795,9716542,69205,4839726,3247509,8953750],["Attenuation phase","HSF1-dependent transactivation","HSF1 activation","GPVI-mediated activation cascade","Regulation of TP53 Expression and Degradation","Regulation of HSF1-mediated heat shock response","Factors involved in megakaryocyte development and platelet production","Phosphorylation of the APC/C","Cellular response to heat stress","Regulation of TP53 Expression","Regulation of APC/C activators between G1/S and early anaphase","Cyclin A/B1/B2 associated events during G2/M transition","TP53 Regulates Transcription of DNA Repair Genes","Hemostasis","APC/C-mediated degradation of cell cycle proteins","Regulation of mitotic cell cycle","G0 and Early G1","Regulation of TP53 Activity","Cell Cycle Checkpoints","Phosphorylation of Emi1","Regulation of TP53 Degradation","Ovarian tumor domain proteases","Activation of NIMA Kinases NEK9, NEK6, NEK7","RHO GTPases Activate Rhotekin and Rhophilins","Cell Cycle","Transcriptional Regulation by VENTX","Cell Cycle, Mitotic","G2/M Transition","Mitotic G2-G2/M phases","Intra-Golgi and retrograde Golgi-to-ER traffic","RHO GTPases activate KTN1","Transcriptional Regulation by TP53","Signaling by ROBO receptors","Cellular responses to stress","Mitotic Telophase/Cytokinesis","Cellular responses to external stimuli","TP53 Regulates Transcription of Cell Cycle Genes","RHO GTPases Activate Formins","Golgi Cisternae Pericentriolar Stack Reorganization","G2/M Checkpoints","TP53 Regulates Transcription of Genes Involved in G1 Cell Cycle Arrest","Transcription of E2F targets under negative control by p107 (RBL1) and p130 (RBL2) in complex with HDAC1","Activation of APC/C and APC/C:Cdc20 mediated degradation of mitotic proteins","TP53 Regulates Transcription of Genes Involved in G2 Cell Cycle Arrest","G2/M DNA damage checkpoint","The role of GTSE1 in G2/M progression after G2 checkpoint","RHO GTPases activate CIT","Polo-like kinase mediated events","APC/C:Cdc20 mediated degradation of Cyclin B","Sema4D induced cell migration and growth-cone collapse","RHO GTPases Activate ROCKs","RAC2 GTPase cycle","Regulation of TP53 Activity through Phosphorylation","RHO GTPase Effectors","Sema4D in semaphorin signaling","HATs acetylate histones","Aberrant regulation of mitotic cell cycle due to RB1 defects","Diseases of DNA Double-Strand Break Repair","Defective DNA double strand break response due to BRCA1 loss of function","Defective DNA double strand break response due to BARD1 loss of function","Diseases of mitotic cell cycle","G-protein beta:gamma signalling","Signaling by Rho GTPases, Miro GTPases and RHOBTB3","G1/S-Specific Transcription","Chromatin organization","Chromatin modifying enzymes","Transcriptional Regulation by E2F6"],[true,true,true,true,false,true,true,true,false,true,true,true,true,false,false,false,true,false,false,true,true,true,true,true,false,true,false,false,false,false,true,false,true,false,true,false,false,true,true,true,true,true,false,true,true,true,true,true,true,true,true,true,true,false,false,true,false,false,true,true,false,false,false,true,false,true,true],[false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,false,true,true,true,true,true,false,false,false,false,false,false],[48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887,48887],["9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606","9606"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL"],[47,59,43,43,46,113,194,20,135,4,83,32,89,801,92,92,38,178,280,8,43,43,9,10,734,48,596,212,214,218,12,487,235,811,15,829,65,149,17,154,20,20,77,21,81,83,23,23,24,24,24,92,95,326,30,110,36,2,2,2,38,39,720,43,256,256,46],[5,5,4,4,4,5,6,3,5,2,4,3,4,10,4,4,3,5,6,2,3,3,2,2,9,3,8,5,5,5,2,7,5,9,2,9,3,4,2,4,2,2,3,2,3,3,2,2,2,2,2,3,3,5,2,3,2,1,1,1,2,2,7,2,4,4,2],[0.00323646880594959,0.00406280126704311,0.00296102465225176,0.00296102465225176,0.00316760776752513,0.00778129734196392,0.0133590414543451,0.00137722076848919,0.00929624018730203,0.000275444153697838,0.00571546618923013,0.0022035532295827,0.00612863241977689,0.055157691777992,0.00633521553505027,0.00633521553505027,0.00261671946012946,0.0122572648395538,0.0192810907588486,0.000550888307395676,0.00296102465225176,0.00296102465225176,0.000619749345820135,0.000688610384244594,0.0505440022035532,0.00330532984437405,0.0410411789009778,0.0145985401459854,0.0147362622228343,0.0150117063765322,0.000826332461093513,0.0335353257127117,0.016182344029748,0.0558463021622366,0.00103291557636689,0.0570858008538769,0.00447596749758986,0.0102602947252445,0.00117063765321581,0.0106045999173668,0.00137722076848919,0.00137722076848919,0.00530229995868338,0.00144608180691365,0.00557774411238121,0.00571546618923013,0.00158380388376257,0.00158380388376257,0.00165266492218703,0.00165266492218703,0.00165266492218703,0.00633521553505027,0.00654179865032365,0.0224486985263738,0.00206583115273378,0.00757471422669054,0.00247899738328054,0.000137722076848919,0.000137722076848919,0.000137722076848919,0.00261671946012946,0.00268558049855392,0.0495799476656108,0.00296102465225176,0.0176284258366616,0.0176284258366616,0.00316760776752513],[4.8638720018257e-07,1.47299000374801e-06,1.23845751334173e-05,1.23845751334173e-05,1.61049069064001e-05,3.33325068940793e-05,3.81517868893733e-05,4.04770524197184e-05,7.69402845764988e-05,8.13402644070926e-05,0.0001564164286324,0.000161345257870371,0.00020388301605323,0.000204599747499867,0.000231152455185679,0.000231152455185679,0.000266536117513927,0.000276486011851595,0.000279990918665374,0.00032268688457282,0.000381852714001818,0.000381852714001818,0.000407559255930945,0.000502123248883768,0.000508992324569135,0.000525172483307856,0.000602681923054904,0.000610613572108898,0.00063689723375504,0.000692006097243403,0.000720083002577643,0.000922537424995373,0.000967019685884973,0.00103703007932399,0.00111819770188293,0.00120968264511723,0.00125497622979942,0.00139051887564079,0.00143036057352985,0.00156822181282212,0.0019675532938086,0.0019675532938086,0.00203052941387571,0.00216476996794668,0.00234248078548316,0.00250898357507046,0.00258608446464015,0.00258608446464015,0.00281006738295808,0.00281006738295808,0.00281006738295808,0.00334829511944079,0.00366183294237432,0.00399382650594149,0.00433694620776348,0.00549613555682738,0.0061687992136189,0.00645247617758726,0.00645247617758726,0.00645247617758726,0.00684514080142684,0.00719539222624677,0.00808133426240376,0.00867569640807142,0.00942779052985343,0.00942779052985343,0.0098677182973127],[0.00019115016967175,0.00028870604073461,0.00121368836307489,0.00121368836307489,0.00125618273869921,0.0019833755685662,0.0019833755685662,0.0019833755685662,0.00317227031187661,0.00317227031187661,0.00516304825185188,0.00516304825185188,0.0055476589244563,0.0055476589244563,0.0055476589244563,0.0055476589244563,0.00559981837330747,0.00559981837330747,0.00559981837330747,0.00613105080688359,0.00649149613803091,0.00649149613803091,0.00692850735082606,0.00763488486853703,0.00763488486853703,0.00787758724961785,0.00827966403881553,0.00827966403881553,0.00827966403881553,0.00864099603093171,0.00864099603093171,0.0106372165447347,0.0106372165447347,0.0114073308725638,0.0120968264511723,0.0120968264511723,0.0125497622979942,0.0139051887564079,0.0141139963153991,0.0141139963153991,0.0173181597435734,0.0173181597435734,0.0173181597435734,0.0173181597435734,0.0187398462838653,0.0196704716807066,0.0196704716807066,0.0196704716807066,0.0196704716807066,0.0196704716807066,0.0196704716807066,0.0234380658360855,0.0256328305966202,0.0279567855415904,0.0303586234543444,0.0370127952817134,0.0370127952817134,0.0387148570655236,0.0387148570655236,0.0387148570655236,0.041070844808561,0.0431723533574806,0.0471389526492672,0.0471389526492672,0.0471389526492672,0.0471389526492672,0.0493385914865635],[[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]],["TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL","TOTAL"],[5,8,7,25,35,14,43,2,29,5,14,25,17,334,40,40,27,98,56,2,30,14,4,6,449,13,350,78,80,48,2,259,60,289,4,320,42,27,6,24,17,8,15,11,12,10,6,15,3,7,7,10,26,113,13,15,3,2,1,1,5,15,210,28,85,85,33],[3,4,1,3,17,7,7,2,12,5,7,19,15,12,32,32,6,64,27,2,12,2,2,6,183,3,156,49,51,8,2,172,18,46,2,46,27,10,4,11,10,2,15,7,5,6,5,7,3,4,4,10,23,48,5,3,2,2,1,1,2,5,91,2,5,5,2],[0.000372411738417995,0.000595858781468792,0.000521376433785193,0.00186205869208997,0.00260688216892596,0.00104275286757039,0.00320274095039476,0.000148964695367198,0.00215998808282437,0.000372411738417995,0.00104275286757039,0.00186205869208997,0.00126619991062118,0.0248771041263221,0.00297929390734396,0.00297929390734396,0.00201102338745717,0.0072992700729927,0.00417101147028154,0.000148964695367198,0.00223447043050797,0.00104275286757039,0.000297929390734396,0.000446894086101594,0.0334425741099359,0.000968270519886787,0.0260688216892596,0.00580962311932072,0.00595858781468792,0.00357515268881275,0.000148964695367198,0.0192909280500521,0.00446894086101594,0.0215253984805601,0.000297929390734396,0.0238343512587517,0.00312825860271116,0.00201102338745717,0.000446894086101594,0.00178757634440638,0.00126619991062118,0.000595858781468792,0.00111723521525398,0.000819305824519589,0.000893788172203188,0.00074482347683599,0.000446894086101594,0.00111723521525398,0.000223447043050797,0.000521376433785193,0.000521376433785193,0.00074482347683599,0.00193654103977357,0.00841650528824668,0.000968270519886787,0.00111723521525398,0.000223447043050797,0.000148964695367198,7.4482347683599e-05,7.4482347683599e-05,0.000372411738417995,0.00111723521525398,0.0156412930135558,0.00208550573514077,0.00633099955310591,0.00633099955310591,0.00245791747355877]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>stId<\/th>\n      <th>dbId<\/th>\n      <th>name<\/th>\n      <th>llp<\/th>\n      <th>inDisease<\/th>\n      <th>species.dbId<\/th>\n      <th>species.taxId<\/th>\n      <th>species.name<\/th>\n      <th>entities.resource<\/th>\n      <th>entities.total<\/th>\n      <th>entities.found<\/th>\n      <th>entities.ratio<\/th>\n      <th>entities.pValue<\/th>\n      <th>entities.fdr<\/th>\n      <th>entities.exp<\/th>\n      <th>reactions.resource<\/th>\n      <th>reactions.total<\/th>\n      <th>reactions.found<\/th>\n      <th>reactions.ratio<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":5,"columnDefs":[{"className":"dt-right","targets":[2,6,10,11,12,13,14,17,18,19]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]}},"evals":[],"jsHooks":[]}</script><p>As mentioned, some of <code><a href="../reference/rba_reactome_analysis.html">rba_reactome_analysis()</a></code>’s arguments correspond to the wizard of <a href="https://reactome.org/PathwayBrowser/#TOOL=AT">analyse gene list</a> tool; Other arguments corresponds to the contents of “Filter your results” tab in the results page.</p>
<p>Having the analysis’s token, you can retrieve the analysis results in many formats using <code><a href="../reference/rba_reactome_analysis_pdf.html">rba_reactome_analysis_pdf()</a></code> and <code><a href="../reference/rba_reactome_analysis_download.html">rba_reactome_analysis_download()</a></code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># download a full pdf report</span>
<span class="fu"><a href="../reference/rba_reactome_analysis_pdf.html">rba_reactome_analysis_pdf</a></span><span class="op">(</span>token <span class="op">=</span> <span class="va">analyzed</span><span class="op">$</span><span class="va">summary</span><span class="op">$</span><span class="va">token</span>,
                          species <span class="op">=</span> <span class="fl">9606</span><span class="op">)</span>
<span class="co"># download the result in compressed json.gz format</span>
<span class="fu"><a href="../reference/rba_reactome_analysis_download.html">rba_reactome_analysis_download</a></span><span class="op">(</span>token <span class="op">=</span> <span class="va">analyzed</span><span class="op">$</span><span class="va">summary</span><span class="op">$</span><span class="va">token</span>,
                               request <span class="op">=</span> <span class="st">"results"</span>,
                               save_to <span class="op">=</span> <span class="st">"reactome_results.json"</span><span class="op">)</span></code></pre></div>
<p>Your token is only guaranteed to be stored for 7 days. After that, you can upload the JSON file you have downloaded using <code>rba_reactome_analysis_download</code> and get a token for that:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">re_uploaded</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_analysis_import.html">rba_reactome_analysis_import</a></span><span class="op">(</span>input <span class="op">=</span> <span class="st">"reactome_results.json"</span><span class="op">)</span></code></pre></div>
<div id="analysis_see-also-in-functions-manuals" class="section level2 heading3">
<h2 class="hasAnchor">
<a href="#analysis_see-also-in-functions-manuals" class="anchor"></a>See also in Functions’ manuals</h2>
<p>Some rbioapi Reactome analysis functions were not covered in this vignette, be sure to check their manuals:</p>
<ul>
<li><p><code><a href="../reference/rba_reactome_analysis_mapping.html">rba_reactome_analysis_mapping()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_analysis_species.html">rba_reactome_analysis_species()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_analysis_token.html">rba_reactome_analysis_token()</a></code></p></li>
</ul>
<hr>
</div>
</div>
<div id="reactome-contents-services" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#reactome-contents-services" class="anchor"></a>Reactome contents services</h1>
<p>rbioapi functions that correspond to Reactome content services are those starting with rba_reactome_* but without “_analysis” infix. These functions cover what you can do with objects in Reactome knowledge-base. In simpler terms, most -but not all of them- correspond to what you can find in <a href="https://reactome.org/PathwayBrowser/" title="Reactome Pathway Browser">Reactome Pathway Browser</a> and <a href="https://reactome.org/content/query?q=apoptosis&amp;species=Homo+sapiens" title="Search results for apoptosis">search results</a>. (e.g. a <a href="https://reactome.org/content/detail/R-HSA-109581" title="Apoptosis Homo sapiens">pathway</a>, a <a href="https://reactome.org/content/detail/R-HSA-202939" title="Caspase-mediated cleavage of E-Cadherin">reaction</a>, a <a href="https://reactome.org/content/detail/R-HSA-350870" title="Caspase-3 [cytosol]">physical Entity</a>, etc.)</p>
<div id="retrieve-any-object-from-reactome-knowledge-base" class="section level2">
<h2 class="hasAnchor">
<a href="#retrieve-any-object-from-reactome-knowledge-base" class="anchor"></a>Retrieve any object from Reactome knowledge-base</h2>
<p>Using <code><a href="../reference/rba_reactome_query.html">rba_reactome_query()</a></code>, you can retrieve any object from Reactome knowledge-base. In simpler terms, what I mean by the object is roughly anything that Reactome associated an ID to it. This can range from a person’s entry to proteins, reactions, pathways, species, and many more! You can explore <a href="https://reactome.org/content/schema/DatabaseObject" title="Graph Database :: Data Schema">Reactome’s data schema</a> to learn about Reactome knowledge-base objects and their organization. Here are some examples, note that you are not limited to only one ID per query. You can use a vector of inputs, the only limitation is that when you provide more than one ID, you cannot have <code>enhanced = TRUE</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## 1 query a pathway Entry</span>
<span class="va">pathway</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_query.html">rba_reactome_query</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="st">"R-HSA-109581"</span>, enhanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">## 2 As always we use str() to inspect the output's structure</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">pathway</span>, <span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; List of 25</span>
<span class="co">#&gt;  $ dbId               : int 109581</span>
<span class="co">#&gt;  $ displayName        : chr "Apoptosis"</span>
<span class="co">#&gt;  $ stId               : chr "R-HSA-109581"</span>
<span class="co">#&gt;  $ stIdVersion        : chr "R-HSA-109581.3"</span>
<span class="co">#&gt;  $ created            :List of 5</span>
<span class="co">#&gt;   ..$ dbId       : int 109608</span>
<span class="co">#&gt;   ..$ displayName: chr "Alnemri, E, Hengartner, Michael, Tschopp, Jürg, Tsujimoto, Yoshihide, Hardwick, JM, 2004-01-16"</span>
<span class="co">#&gt;   ..$ dateTime   : chr "2004-01-16 21:01:51.0"</span>
<span class="co">#&gt;   ..$ schemaClass: chr "InstanceEdit"</span>
<span class="co">#&gt;   ..$ className  : chr "InstanceEdit"</span>
<span class="co">#&gt;  $ modified           :List of 6</span>
<span class="co">#&gt;   ..$ dbId       : int 10686365</span>
<span class="co">#&gt;   ..$ displayName: chr "Shorser, Solomon, 2021-03-06"</span>
<span class="co">#&gt;   ..$ dateTime   : chr "2021-03-06 05:25:32.0"</span>
<span class="co">#&gt;   ..$ note       : chr "Inserted by org.reactome.orthoinference"</span>
<span class="co">#&gt;   ..$ schemaClass: chr "InstanceEdit"</span>
<span class="co">#&gt;   ..$ className  : chr "InstanceEdit"</span>
<span class="co">#&gt;  $ isInDisease        : logi FALSE</span>
<span class="co">#&gt;  $ isInferred         : logi FALSE</span>
<span class="co">#&gt;  $ name               :List of 1</span>
<span class="co">#&gt;   ..$ : chr "Apoptosis"</span>
<span class="co">#&gt;  $ releaseDate        : chr "2004-09-20"</span>
<span class="co">#&gt;  $ speciesName        : chr "Homo sapiens"</span>
<span class="co">#&gt;  $ authored           :List of 1</span>
<span class="co">#&gt;   ..$ : int 109608</span>
<span class="co">#&gt;  $ edited             :List of 1</span>
<span class="co">#&gt;   ..$ :List of 5</span>
<span class="co">#&gt;  $ figure             :List of 1</span>
<span class="co">#&gt;   ..$ :List of 5</span>
<span class="co">#&gt;  $ goBiologicalProcess:List of 9</span>
<span class="co">#&gt;   ..$ dbId        : int 2273</span>
<span class="co">#&gt;   ..$ displayName : chr "apoptotic process"</span>
<span class="co">#&gt;   ..$ accession   : chr "0006915"</span>
<span class="co">#&gt;   ..$ databaseName: chr "GO"</span>
<span class="co">#&gt;   ..$ definition  : chr "A programmed cell death process which begins when a cell receives an internal (e.g. DNA damage) or external sig"| __truncated__</span>
<span class="co">#&gt;   ..$ name        : chr "apoptotic process"</span>
<span class="co">#&gt;   ..$ url         : chr "https://www.ebi.ac.uk/QuickGO/term/GO:0006915"</span>
<span class="co">#&gt;   ..$ schemaClass : chr "GO_BiologicalProcess"</span>
<span class="co">#&gt;   ..$ className   : chr "GO_BiologicalProcess"</span>
<span class="co">#&gt;  $ literatureReference:List of 7</span>
<span class="co">#&gt;   ..$ :List of 11</span>
<span class="co">#&gt;   ..$ :List of 11</span>
<span class="co">#&gt;   ..$ :List of 11</span>
<span class="co">#&gt;   ..$ :List of 11</span>
<span class="co">#&gt;   ..$ :List of 11</span>
<span class="co">#&gt;   ..$ :List of 11</span>
<span class="co">#&gt;   ..$ :List of 11</span>
<span class="co">#&gt;  $ orthologousEvent   :List of 14</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;  $ reviewed           :List of 1</span>
<span class="co">#&gt;   ..$ :List of 5</span>
<span class="co">#&gt;  $ species            :List of 1</span>
<span class="co">#&gt;   ..$ :List of 8</span>
<span class="co">#&gt;  $ summation          :List of 1</span>
<span class="co">#&gt;   ..$ :List of 5</span>
<span class="co">#&gt;  $ hasDiagram         : logi TRUE</span>
<span class="co">#&gt;  $ hasEHLD            : logi TRUE</span>
<span class="co">#&gt;  $ hasEvent           :List of 4</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;   ..$ :List of 16</span>
<span class="co">#&gt;   ..$ :List of 16</span>
<span class="co">#&gt;   ..$ :List of 15</span>
<span class="co">#&gt;  $ schemaClass        : chr "Pathway"</span>
<span class="co">#&gt;  $ className          : chr "Pathway"</span>



<span class="co">## 3 You can compare it with the webpage of R-HSA-202939 entry:</span>
<span class="co"># https://reactome.org/content/detail/R-HSA-202939</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## 1 query a protein Entry</span>
<span class="va">protein</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_query.html">rba_reactome_query</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="fl">66247</span>, enhanced <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">## 2 As always we use str() to inspect the output's structure</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">protein</span>, <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; List of 27</span>
<span class="co">#&gt;  $ dbId               : int 66247</span>
<span class="co">#&gt;  $ displayName        : chr "UniProt:P25942-1 CD40"</span>
<span class="co">#&gt;  $ modified           :List of 6</span>
<span class="co">#&gt;  $ databaseName       : chr "UniProt"</span>
<span class="co">#&gt;  $ identifier         : chr "P25942"</span>
<span class="co">#&gt;  $ name               :List of 1</span>
<span class="co">#&gt;  $ otherIdentifier    :List of 117</span>
<span class="co">#&gt;  $ url                : chr "http://purl.uniprot.org/uniprot/P25942-1"</span>
<span class="co">#&gt;  $ crossReference     :List of 28</span>
<span class="co">#&gt;  $ referenceDatabase  :List of 7</span>
<span class="co">#&gt;  $ physicalEntity     :List of 1</span>
<span class="co">#&gt;  $ checksum           : chr "BC8776EC2C4A5680"</span>
<span class="co">#&gt;  $ comment            :List of 1</span>
<span class="co">#&gt;  $ description        :List of 1</span>
<span class="co">#&gt;  $ geneName           :List of 2</span>
<span class="co">#&gt;  $ isSequenceChanged  : logi FALSE</span>
<span class="co">#&gt;  $ keyword            :List of 16</span>
<span class="co">#&gt;  $ secondaryIdentifier:List of 8</span>
<span class="co">#&gt;  $ sequenceLength     : int 277</span>
<span class="co">#&gt;  $ species            : int 48887</span>
<span class="co">#&gt;  $ chain              :List of 2</span>
<span class="co">#&gt;  $ referenceGene      :List of 11</span>
<span class="co">#&gt;  $ referenceTranscript:List of 4</span>
<span class="co">#&gt;  $ variantIdentifier  : chr "P25942-1"</span>
<span class="co">#&gt;  $ isoformParent      :List of 1</span>
<span class="co">#&gt;  $ schemaClass        : chr "ReferenceIsoform"</span>
<span class="co">#&gt;  $ className          : chr "ReferenceIsoform"</span>



<span class="co">## 3 You can compare it with the webpage of R-HSA-202939 entry:</span>
<span class="co"># https://reactome.org/content/detail/R-HSA-202939</span></code></pre></div>
</div>
<div id="find-cross-reference-ids-in-reactome" class="section level2">
<h2 class="hasAnchor">
<a href="#find-cross-reference-ids-in-reactome" class="anchor"></a>Find Cross-Reference IDs in Reactome</h2>
<p>As you can see in the second example usage of we used Reactome’s dbID <code>66247</code> to query CD40 protein. How did we obtain that in the first place? You can use <code>rba_reactome_xref</code> to map any cross-reference (external) IDs to Reactome IDs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## 1 We provide HGNC ID to find what is the corresponding database ID in Reactome</span>
<span class="va">xref_protein</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_xref.html">rba_reactome_xref</a></span><span class="op">(</span><span class="st">"CD40"</span><span class="op">)</span>
<span class="co">## 2 As always use str() to inspect the output's structure</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">xref_protein</span>, <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; List of 19</span>
<span class="co">#&gt;  $ dbId               : int 66247</span>
<span class="co">#&gt;  $ displayName        : chr "UniProt:P25942-1 CD40"</span>
<span class="co">#&gt;  $ databaseName       : chr "UniProt"</span>
<span class="co">#&gt;  $ identifier         : chr "P25942"</span>
<span class="co">#&gt;  $ name               :List of 1</span>
<span class="co">#&gt;  $ otherIdentifier    :List of 1</span>
<span class="co">#&gt;  $ url                : chr "http://purl.uniprot.org/uniprot/P25942-1"</span>
<span class="co">#&gt;  $ checksum           : chr "BC8776EC2C4A5680"</span>
<span class="co">#&gt;  $ comment            :List of 1</span>
<span class="co">#&gt;  $ description        :List of 1</span>
<span class="co">#&gt;  $ geneName           :List of 1</span>
<span class="co">#&gt;  $ isSequenceChanged  : logi FALSE</span>
<span class="co">#&gt;  $ keyword            :List of 1</span>
<span class="co">#&gt;  $ secondaryIdentifier:List of 1</span>
<span class="co">#&gt;  $ sequenceLength     : int 277</span>
<span class="co">#&gt;  $ chain              :List of 1</span>
<span class="co">#&gt;  $ variantIdentifier  : chr "P25942-1"</span>
<span class="co">#&gt;  $ schemaClass        : chr "ReferenceIsoform"</span>
<span class="co">#&gt;  $ className          : chr "ReferenceIsoform"</span></code></pre></div>
</div>
<div id="map-cross-reference-ids-to-reactome" class="section level2">
<h2 class="hasAnchor">
<a href="#map-cross-reference-ids-to-reactome" class="anchor"></a>Map Cross-Reference IDs to Reactome</h2>
<p>While we are at the cross-reference topic, here is another useful resource. Using <code>rba_reactome_mapping</code> you can find the Reactome pathways or reactions which include your external ID:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## 1 Again, consider CD40 protein:</span>
<span class="va">xref_mapping</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_reactome_mapping.html">rba_reactome_mapping</a></span><span class="op">(</span>id <span class="op">=</span> <span class="st">"CD40"</span>,
                                    resource <span class="op">=</span> <span class="st">"hgnc"</span>,
                                    map_to <span class="op">=</span> <span class="st">"pathways"</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-9abbc2be978fac344de9" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-9abbc2be978fac344de9">{"x":{"filter":"none","data":[["1","2","3"],[198933,5668541,5676594],["Immunoregulatory interactions between a Lymphoid and a non-Lymphoid cell","TNFR2 non-canonical NF-kB pathway","TNF receptor superfamily (TNFSF) members mediating non-canonical NF-kB pathway"],["R-HSA-198933","R-HSA-5668541","R-HSA-5676594"],["R-HSA-198933.6","R-HSA-5668541.3","R-HSA-5676594.1"],[false,false,false],[false,false,false],["Immunoregulatory interactions between a Lymphoid and a non-Lymphoid cell",["TNFR2 non-canonical NF-kB pathway","TNFRSF mediated non-canonical NF-kB pathway"],"TNF receptor superfamily (TNFSF) members mediating non-canonical NF-kB pathway"],["2007-08-21","2015-06-17","2015-06-17"],["Homo sapiens","Homo sapiens","Homo sapiens"],["10.3180/r-hsa-198933.2","10.3180/r-hsa-5668541.1",null],[true,true,false],[false,false,false],["Pathway","Pathway","Pathway"],["Pathway","Pathway","Pathway"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>dbId<\/th>\n      <th>displayName<\/th>\n      <th>stId<\/th>\n      <th>stIdVersion<\/th>\n      <th>isInDisease<\/th>\n      <th>isInferred<\/th>\n      <th>name<\/th>\n      <th>releaseDate<\/th>\n      <th>speciesName<\/th>\n      <th>doi<\/th>\n      <th>hasDiagram<\/th>\n      <th>hasEHLD<\/th>\n      <th>schemaClass<\/th>\n      <th>className<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":1},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><hr>
</div>
<div id="see-also-in-functions-manuals" class="section level2 heading2">
<h2 class="hasAnchor">
<a href="#see-also-in-functions-manuals" class="anchor"></a>See also in Functions’ manuals</h2>
<p>There are still more rbioapi f Reactome content functions that were not covered in this vignette. Here is a brief overview, refer to the functions’ manual for detailed guides and examples.</p>
<div id="retrieve-reactome-database-information" class="section level3">
<h3 class="hasAnchor">
<a href="#retrieve-reactome-database-information" class="anchor"></a>Retrieve Reactome Database information</h3>
<ul>
<li><p><code><a href="../reference/rba_reactome_version.html">rba_reactome_version()</a></code>: Return current Reactome version</p></li>
<li><p><code><a href="../reference/rba_reactome_diseases.html">rba_reactome_diseases()</a></code>: Retrieve a list of disease annotated in Reactome.</p></li>
<li><p><code><a href="../reference/rba_reactome_species.html">rba_reactome_species()</a></code>: Retrieve a list of species annotated in Reactome.</p></li>
</ul>
</div>
<div id="general-mappingquerying" class="section level3">
<h3 class="hasAnchor">
<a href="#general-mappingquerying" class="anchor"></a>General Mapping/Querying</h3>
<ul>
<li><p><code><a href="../reference/rba_reactome_query.html">rba_reactome_query()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_mapping.html">rba_reactome_mapping()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_xref.html">rba_reactome_xref()</a></code></p></li>
</ul>
</div>
<div id="things-you-can-do-with-a-entities" class="section level3">
<h3 class="hasAnchor">
<a href="#things-you-can-do-with-a-entities" class="anchor"></a>Things you can do with a Entities</h3>
<ul>
<li><p><code>reactome_complex_list()</code>: Get a list of complexes that have your molecule in them.</p></li>
<li><p><code><a href="../reference/rba_reactome_complex_subunits.html">rba_reactome_complex_subunits()</a></code>: Get the list of subunits in your complex</p></li>
<li><p><code><a href="../reference/rba_reactome_participant_of.html">rba_reactome_participant_of()</a></code>: Get a list of Reactome sets and complexes that your entity (event, molecule, reaction, pathway etc.) is a participant in them.</p></li>
<li><p><code><a href="../reference/rba_reactome_entity_other_forms.html">rba_reactome_entity_other_forms()</a></code></p></li>
</ul>
</div>
<div id="things-you-can-do-with-events" class="section level3">
<h3 class="hasAnchor">
<a href="#things-you-can-do-with-events" class="anchor"></a>Things you can do with Events</h3>
<ul>
<li><p><code><a href="../reference/rba_reactome_event_ancestors.html">rba_reactome_event_ancestors()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_participants.html">rba_reactome_participants()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_pathways_events.html">rba_reactome_pathways_events()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_event_ancestors.html">rba_reactome_event_ancestors()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_orthology.html">rba_reactome_orthology()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_event_hierarchy.html">rba_reactome_event_hierarchy()</a></code>: Retrieve full event hierarchy of an species.</p></li>
</ul>
</div>
<div id="pathways" class="section level3">
<h3 class="hasAnchor">
<a href="#pathways" class="anchor"></a>Pathways</h3>
<ul>
<li><p><code><a href="../reference/rba_reactome_pathways_low.html">rba_reactome_pathways_low()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_pathways_events.html">rba_reactome_pathways_events()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_pathways_top.html">rba_reactome_pathways_top()</a></code></p></li>
</ul>
</div>
<div id="interactors" class="section level3">
<h3 class="hasAnchor">
<a href="#interactors" class="anchor"></a>Interactors</h3>
<ul>
<li><p><code><a href="../reference/rba_reactome_interactors_psicquic.html">rba_reactome_interactors_psicquic()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_interactors_static.html">rba_reactome_interactors_static()</a></code></p></li>
</ul>
</div>
<div id="people" class="section level3">
<h3 class="hasAnchor">
<a href="#people" class="anchor"></a>People</h3>
<ul>
<li><p><code><a href="../reference/rba_reactome_people_name.html">rba_reactome_people_name()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_people_id.html">rba_reactome_people_id()</a></code></p></li>
</ul>
</div>
<div id="export-diagrams-and-events" class="section level3">
<h3 class="hasAnchor">
<a href="#export-diagrams-and-events" class="anchor"></a>Export diagrams and events</h3>
<ul>
<li><p><code><a href="../reference/rba_reactome_exporter_diagram.html">rba_reactome_exporter_diagram()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_exporter_overview.html">rba_reactome_exporter_overview()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_exporter_reaction.html">rba_reactome_exporter_reaction()</a></code></p></li>
<li><p><code><a href="../reference/rba_reactome_exporter_event.html">rba_reactome_exporter_event()</a></code></p></li>
</ul>
<hr>
</div>
</div>
</div>
<div id="session-info" class="section level1 heading2">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<pre><code>#&gt; R version 4.1.0 (2021-05-18)
#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)
#&gt; Running under: macOS Catalina 10.15.7
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] rbioapi_0.7.3
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] knitr_1.33        magrittr_2.0.1    R6_2.5.0          ragg_1.1.2       
#&gt;  [5] rlang_0.4.11      fastmap_1.1.0     stringr_1.4.0     httr_1.4.2       
#&gt;  [9] tools_4.1.0       DT_0.18           xfun_0.23         crosstalk_1.1.1  
#&gt; [13] htmltools_0.5.1.1 systemfonts_1.0.2 yaml_2.2.1        digest_0.6.27    
#&gt; [17] rprojroot_2.0.2   pkgdown_1.6.1     crayon_1.4.1      textshaping_0.3.4
#&gt; [21] htmlwidgets_1.5.3 fs_1.5.0          curl_4.3.1        memoise_2.0.0    
#&gt; [25] cachem_1.0.5      evaluate_0.14     rmarkdown_2.8     stringi_1.6.2    
#&gt; [29] compiler_4.1.0    desc_1.3.0        jsonlite_1.7.2</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
