<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rbioapi">
<title>2.D: PANTHER &amp; rbioapi • rbioapi</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="2.D: PANTHER &amp; rbioapi">
<meta property="og:description" content="rbioapi">
<meta property="og:image" content="https://rbioapi.moosa-r.com/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BF9VGFNZ4C"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BF9VGFNZ4C');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-secondary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbioapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/rbioapi.html">Get Started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles-amp-supported-services">Articles &amp; Supported Services</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles-amp-supported-services">
    <a class="dropdown-item" href="../articles/rbioapi.html">rbioapi: Get Started</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_enrichr.html">Enrichr &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_jaspar.html">JASPAR &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_mieaa.html">miEEA &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_panther.html">PANTHER &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_reactome.html">Reactome &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_string.html">STRING &amp; rbioapi</a>
    <a class="dropdown-item" href="../articles/rbioapi_uniprot.html">UniProt &amp; rbioapi</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">News</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/moosa-r/rbioapi/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="rbioapi_panther_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="rbioapi_panther_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="rbioapi_panther_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rbioapi_panther_files/datatables-binding-0.22/datatables.js"></script><link href="rbioapi_panther_files/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rbioapi_panther_files/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rbioapi_panther_files/dt-core-1.11.3/js/jquery.dataTables.min.js"></script><link href="rbioapi_panther_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="rbioapi_panther_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <main id="main"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>2.D: PANTHER &amp; rbioapi</h1>
                        <h4 data-toc-skip class="author">Moosa Rezwani</h4>
            
            <h4 data-toc-skip class="date">2022-04-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/moosa-r/rbioapi/blob/HEAD/vignettes/rbioapi_panther.Rmd" class="external-link"><code>vignettes/rbioapi_panther.Rmd</code></a></small>
      <div class="d-none name"><code>rbioapi_panther.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Directly quoting the the paper published by <a href="http://www.pantherdb.org" title="Protein Analysis THrough Evolutionary Relationships (PANTHER)" class="external-link">PANTHER</a> (Protein Analysis THrough Evolutionary Relationships) authors:</p>
<blockquote>
<p>The PANTHER classification system (<a href="http://www.pantherdb.org" class="external-link uri">http://www.pantherdb.org</a>) a comprehensive system that combines genomes, gene function , pathways and statistical analysis tools to enable to analyze large-scale genome-wide experimental data. The system (PANTHER v.14.0) covers 131 complete genomes organized gene families and subfamilies; evolutionary relationships between are represented in phylogenetic trees, multiple sequence and statistical models (hidden Markov models (HMMs)). The families and subfamilies are annotated with Gene Ontology (GO) terms, sequences are assigned to PANTHER pathways. A suite of tools has built to allow users to browse and query gene functions and analyze-scale experimental data with a number of statistical tests. is widely used by bench scientists, bioinformaticians, computer and systems biologists.</p>
<p>(source: Mi, Huaiyu, et al. “Protocol Update for large-scale genome and gene function analysis with the PANTHER classification system (v. 14.0).” <em>Nature protocols</em> 14.3 (2019): 703-721)</p>
</blockquote>
<p>The available tools in PANTHER’s <strong>RESTful API services</strong> can be divided into 3 broad categories: Mapping genes, retrieving information, and research tools. Herein, we provide a very short introduction; you can always check functions’ manuals for detailed guides and examples.</p>
<hr>
</div>
<div class="section level2">
<h2 id="map-genes">Map genes<a class="anchor" aria-label="anchor" href="#map-genes"></a>
</h2>
<ul>
<li><p><code><a href="../reference/rba_panther_mapping.html">rba_panther_mapping()</a></code>: map your gene-set to PANTHER database and retrieve attributes and annotations associated with your genes</p></li>
<li><p><code><a href="../reference/rba_panther_ortholog.html">rba_panther_ortholog()</a></code>: Retrieve Orthologs of your genes</p></li>
<li><p><code><a href="../reference/rba_panther_homolog.html">rba_panther_homolog()</a></code>: Retrieve Homologs of your genes</p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="get-information">Get information<a class="anchor" aria-label="anchor" href="#get-information"></a>
</h2>
<ul>
<li><p><code>rba_panther_info</code>(): Retrieve a list of PANTHER’s supported organisms, datasets, families, or pathways</p></li>
<li><p><code>rba_panther_family</code>(): Retrieve Orthologs, MSA, or Tree topology of a given PANTHER family.</p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="research-tools">Research tools<a class="anchor" aria-label="anchor" href="#research-tools"></a>
</h2>
<div class="section level3">
<h3 id="gene-list-analysis">Gene List Analysis<a class="anchor" aria-label="anchor" href="#gene-list-analysis"></a>
</h3>
<p><code><a href="../reference/rba_panther_enrich.html">rba_panther_enrich()</a></code> is an equivalent to <a href="http://www.pantherdb.org/index.jsp" title="PANTHER Gene List Analysis" class="external-link">Gene List analysis tool’s webpage.</a>. Here is a usage example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## 1 We get the available annotation datasets in PANTHER (we need to select one of them to submit an enrichment request)</span>
<span class="va">annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_panther_info.html">rba_panther_info</a></span><span class="op">(</span>what <span class="op">=</span> <span class="st">"datasets"</span><span class="op">)</span>
<span class="co">#&gt; Retrieving available annotation datasets.</span>
<span class="co">## 2 We create a variable with our genes' IDs</span>
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"cdk2"</span>, <span class="st">"Q99835"</span>, <span class="st">"CDC42"</span>,<span class="st">"CDK1"</span>,<span class="st">"KIF23"</span>,<span class="st">"PLK1"</span>,
           <span class="st">"RAC2"</span>,<span class="st">"RACGAP1"</span>,<span class="st">"RHOA"</span>,<span class="st">"RHOB"</span>, <span class="st">"PHF14"</span>, <span class="st">"RBM3"</span>, <span class="st">"MSL1"</span><span class="op">)</span>
<span class="co">## 3 Now we can submit the enrichment request.</span>
<span class="va">enriched</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rba_panther_enrich.html">rba_panther_enrich</a></span><span class="op">(</span>genes <span class="op">=</span> <span class="va">genes</span>,
                               organism <span class="op">=</span> <span class="fl">9606</span>,
                               annot_dataset <span class="op">=</span> <span class="st">"ANNOT_TYPE_ID_PANTHER_PATHWAY"</span>,
                               cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co">#&gt; Performing over-representation enrichment analysis of 15 input genes of organism 9606 against ANNOT_TYPE_ID_PANTHER_PATHWAY datasets.</span></code></pre></div>
<div id="htmlwidget-0a597852efdcb06fdb2c" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-0a597852efdcb06fdb2c">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14"],[4,3,4,3,3,2,3,7,2,2,2,3,2,2],[72.2421052631579,164.712,27.452,47.8813953488372,46.2674157303371,130.72380952381,27.0907894736842,0.53465026987925,78.4342857142857,66.9560975609756,53.8274509803922,15.7770114942529,29.8391304347826,21.9616],[4.45410626460892e-05,8.09208204082397e-05,0.000623633565611843,0.00136566718442141,0.00120654874506168,0.00330818481428414,0.00406368327874385,0.00385164488733462,0.00577158496309621,0.00702514668808674,0.00970505969622764,0.0112571820922484,0.0256032965462438,0.0429195734756196],[0.0553693719947545,0.0182136092088008,0.145708873670407,0.0626548156782748,0.0648404487833309,0.0152994317353927,0.110738743989509,13.0926708436544,0.0254990528923211,0.0298703191024333,0.0371557627859537,0.190150080139881,0.067026081888387,0.0910680460440041],[76,25,200,86,89,21,152,17971,35,41,51,261,92,125],[2.78381641538058e-07,1.011510255103e-06,1.16931293552221e-05,3.41416796105352e-05,3.77046482831776e-05,0.000124056930535655,0.000177786143445043,0.000192582244366731,0.000324651654174162,0.000439071668005421,0.00066722285411565,0.000844288656918627,0.00208026784438231,0.00375546267911671],["+","+","+","+","+","+","+","-","+","+","+","+","+","+"],["P04393","P00008","P00034","P00016","P00059","P00007","P00029",null,"P00009","P05918","P04398","P00031","P00053","P00027"],["Ras Pathway","Axon guidance mediated by Slit/Robo","Integrin signalling pathway","Cytoskeletal regulation by Rho GTPase","p53 pathway","Axon guidance mediated by semaphorins","Huntington disease","UNCLASSIFIED","Axon guidance mediated by netrin","p38 MAPK pathway","p53 pathway feedback loops 2","Inflammation mediated by chemokine and cytokine signaling pathway","T cell activation","Heterotrimeric G-protein signaling pathway-Gq alpha and Go alpha mediated pathway"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>number_in_list<\/th>\n      <th>fold_enrichment<\/th>\n      <th>fdr<\/th>\n      <th>expected<\/th>\n      <th>number_in_reference<\/th>\n      <th>pValue<\/th>\n      <th>plus_minus<\/th>\n      <th>term.id<\/th>\n      <th>term.label<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"paging":true,"fixedHeader":true,"keys":true,"pageLength":10,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p><strong>Please Note:</strong> Other services supported by rbioapi also provide Over-representation analysis tools. Please see the vignette article <a href="rbioapi_do_enrich.html">Do with rbioapi: Over-Representation (Enrichment) Analysis in R</a> (<a href="https://rbioapi.moosa-r.com/articles/rbioapi_do_enrich.html">link to the documentation site</a>) for an in-depth review.</p>
</div>
<div class="section level3">
<h3 id="tree-grafter">Tree grafter<a class="anchor" aria-label="anchor" href="#tree-grafter"></a>
</h3>
<p><code><a href="../reference/rba_panther_tree_grafter.html">rba_panther_tree_grafter()</a></code> is an equivalent to the “<a href="http://www.pantherdb.org/tools/sequenceSearchForm.jsp" class="external-link">Graft sequence into PANTHER library of trees</a>” tool.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="Citations">How to Cite?<a class="anchor" aria-label="anchor" href="#Citations"></a>
</h2>
<p>To cite PANTHER (Please see <a href="http://www.pantherdb.org/publications.jsp#HowToCitePANTHER" class="external-link uri">http://www.pantherdb.org/publications.jsp#HowToCitePANTHER</a>):</p>
<ul>
<li>Huaiyu Mi, Dustin Ebert, Anushya Muruganujan, Caitlin Mills, Laurent-Philippe Albou, Tremayne Mushayamaha, Paul D Thomas, PANTHER version 16: a revised family classification, tree-based classification tool, enhancer regions and extensive API, <em>Nucleic Acids Research</em>, Volume 49, Issue D1, 8 January 2021, Pages D394–D403, <a href="https://doi.org/10.1093/nar/gkaa1106" class="external-link uri">https://doi.org/10.1093/nar/gkaa1106</a>
</li>
</ul>
<p>To cite rbioapi:</p>
<ul>
<li>Moosa Rezwani, Ali Akbar Pourfathollah, Farshid Noorbakhsh, rbioapi: User-Friendly R Interface to Biologic Web Services’ API, <em>Bioinformatics</em>, 2022;, btac172, <a href="https://doi.org/10.1093/bioinformatics/btac172" class="external-link uri">https://doi.org/10.1093/bioinformatics/btac172</a>
</li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="links">Links<a class="anchor" aria-label="anchor" href="#links"></a>
</h2>
<ul>
<li><p><a href="https://rbioapi.moosa-r.com/articles/rbioapi_panther.html" title="2.C: PANTHER &amp; rbioapi">This article in rbioapi documentation site</a></p></li>
<li><p><a href="https://rbioapi.moosa-r.com/reference/index.html#section-enrichr-rba-enrichr-" title="rbioapi reference">Functions references in rbioapi documentation site</a></p></li>
<li><p><a href="rbioapi.html" title="rbioapi: User-Friendly R Interface to Biologic Web Services' API">rbioapi vignette index</a></p></li>
</ul>
<hr>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span class="co">#&gt; R version 4.1.3 (2022-03-10)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Big Sur/Monterey 10.16</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] rbioapi_0.7.6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] knitr_1.38        magrittr_2.0.3    R6_2.5.1          ragg_1.2.2       </span>
<span class="co">#&gt;  [5] rlang_1.0.2       fastmap_1.1.0     httr_1.4.2        stringr_1.4.0    </span>
<span class="co">#&gt;  [9] tools_4.1.3       DT_0.22           xfun_0.30         cli_3.2.0        </span>
<span class="co">#&gt; [13] jquerylib_0.1.4   crosstalk_1.2.0   htmltools_0.5.2   systemfonts_1.0.4</span>
<span class="co">#&gt; [17] yaml_2.3.5        digest_0.6.29     rprojroot_2.0.3   pkgdown_2.0.2    </span>
<span class="co">#&gt; [21] textshaping_0.3.6 purrr_0.3.4       htmlwidgets_1.5.4 sass_0.4.1       </span>
<span class="co">#&gt; [25] fs_1.5.2          curl_4.3.2        memoise_2.0.1     cachem_1.0.6     </span>
<span class="co">#&gt; [29] evaluate_0.15     rmarkdown_2.13    stringi_1.7.6     compiler_4.1.3   </span>
<span class="co">#&gt; [33] bslib_0.3.1       desc_1.4.1        jsonlite_1.8.0</span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Moosa Rezwani.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
